---
title: "Sheep miRNA"
author: "Tim Vigers"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 2
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(tidyverse)
```

# Notes
- Following the workflow in this vignette:
  - http://master.bioconductor.org/packages/release/workflows/html/rnaseqGene.html
- Use Salmon to quantify RNA:
  - https://combine-lab.github.io/salmon/
  - Tutorial: https://combine-lab.github.io/salmon/getting_started/
  - rnaseqGene recommends using the --gcBias flag which estimates a correction factor for systematic biases commonly present in RNA-seq data
- Sheep transcriptome can be downloaded from:
  - http://ftp.ensembl.org/pub/release-106/fasta/ovis_aries_rambouillet/cdna/
  - Browse other downloads with: http://uswest.ensembl.org/info/data/ftp/index.html/
- Copy of data stored on shared drive: `/RI Biostatistics Core/Shared/Shared Projects/Vigers/sheep_miRNA`
  
```{bash eval=FALSE}
# bash script
./GitHub/CHCO-Code/quant_sheep_miRNA.sh
```

```{r}
# Combine all quantified files from Salmon
files = list.files("/mnt/HD1/Tim/sheep_miRNA/data_clean/quants")
paths = paste0("/mnt/HD1/Tim/sheep_miRNA/data_clean/quants/",files,"/","quant.sf")
df = lapply(paths, function(p){
  name = sub("_quant","",strsplit(p,"/")[[1]][8])
  d = read.table(p,header = T)
  d = d[,c("Name","TPM")]
  colnames(d)[2] = name
  return(d)
})
df = df %>% purrr::reduce(full_join,by = "Name") %>%
  column_to_rownames(.,var = "Name")
# Sample info
samples = read.csv("/mnt/HD1/Tim/sheep_miRNA/data_clean/sheep_samples.csv",na.strings = "")
samples$Sample = toupper(samples$Sample)
samples$Sample = gsub("-","_",samples$Sample)
samples$Sample = gsub(" ","_",samples$Sample)
# Match count data names
colnames(df) = toupper(colnames(df))
colnames(df) = sub("_MIRNA","",colnames(df))
colnames(df) = sub("_S\\d.*","",colnames(df))
colnames(df) = sub("^\\d{,2}_","",colnames(df))
colnames(df) = sub("_\\d{,2}$","",colnames(df))
colnames(df) = sub("_SHEEP$","_SHEEP_52",colnames(df))
# Remove missing samples
samples = samples[samples$Sample %in% colnames(df),]
```

```{r}
txi <- tximport(paths, type="salmon",txOut=TRUE)
ddsTxi <- DESeqDataSetFromTximport(txi,
                                   colData = samples,
                                   design = ~ t)
dds <- DESeq(ddsTxi)
res <- results(dds)
res
```
