---
title: "OAT scRNA & PAH Clearance"
author: "Hailey Hampson"
date: "2024-09-24"
output: html_document
--- 

```{r libraries, echo=F, include = F}
library(tidyverse)
library(BiocManager)
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)

#Increase Memory
mem.maxVSize(64000000000)

#Set up directories
dir.dat <- c("/Users/hhampson/Dropbox/Bjornstad data")
dir.home <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/OATs and PAH/OATs and PAH Clearance")

#Load functions
source(fs::path(dir.home,"OAT_PAH_Functions.R"))
```

# scRNA
```{r echo = F}
# Kidney scRNA data processing
so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))

# Add missing covariate meta data to Seurat object
meta_raw <- read.csv(fs::path(dir.dat,"Clinical Data","renal_clearance_biopsy.csv"))

#Filter out meta data participants without sc 
ids <- unique(so_kidney_sc@meta.data$cryostor_id)
meta_raw <- meta_raw %>% 
  filter(cryostor_id %in% ids) 

#Merge new metadata with 
meta_kidney_sc <-  tidylog::left_join(so_kidney_sc@meta.data, meta_raw,by="cryostor_id")

#Check there are no duplicates
length(which(duplicated(meta_kidney_sc$barcode)))

#Add rownames back to metadata to merge back into saurat object 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)

#Create study visit variale to filter to RH, CROCCODILE and IMPROVE
so_kidney_sc@meta.data$study= ifelse(grepl("RH",so_kidney_sc@meta.data$record_id) | grepl("CRC",so_kidney_sc@meta.data$record_id) | grepl("IT",so_kidney_sc@meta.data$record_id),"subset","non-subset")

# Add new meta data to so 
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)

#Filter to baseline visits only
so_kidney_sc <- subset(so_kidney_sc, visit == "baseline")

#Filter to RH/RH2, CROCODILE, and IMPROVE
so_kidney_sc <- subset(so_kidney_sc, study == "subset")

so_kidney_sc <- NormalizeData(so_kidney_sc)
so_kidney_sc <- ScaleData(so_kidney_sc)
# # PCA
# so_kidney_sc <- RunPCA(so_kidney_sc, features = VariableFeatures(object = so_kidney_sc))
# ElbowPlot(so_kidney_sc)
# # Cluster cells
# so_kidney_sc <- FindNeighbors(so_kidney_sc)
# so_kidney_sc <- FindClusters(so_kidney_sc)
# # Perform UMAP and tSNE
# so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
# DimPlot(so_kidney_sc, reduction = "umap") 
```

## Obese vs. Obese + T2D
### Descriptive table
```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_diabetes", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## MASLD Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
meta_oat_sc <- so_oat_sc@meta.data %>%
  filter(cryostor_id!="")
form <- paste("diagnosis_of_MASLD", paste(colnames(meta_oat_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = meta_oat_sc, test = F))

```

## OSA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(cryostor_id!="")
form <- paste("diagnosis_of_osa", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## GLP-1RA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- liver_meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("glp1agonist", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Insulin Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("insulin", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Kidney disease Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_kidney_disease", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## Stratified UMAP scRNA 
### Obese vs. Obese + T2D
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_diabetes") 

```

### MASLD Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_MASLD") 

```

### OSA Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_osa") 

```

### GLP-1RA Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "glp1agonist") 

```

### Insulin Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "insulin") 

```

### Kidney disease Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_kidney_disease") 

```

## Differential Expression by Group
### Obese vs. Obese + T2D
```{r}
so_oat_sc@meta.data$hepatocyte <- ifelse(grepl("Hepatocyte",meta_liver_sc$celltype),"Hepatocyte","Non-Hepatocyte")
genes <- unique(rownames(so_oat_sc))
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### MASLD Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### OSA Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### GLP-1RA Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### Insulin Y/N
```{r}
# Idents(so_oat_sc) <- so_oat_sc$celltype
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_oat_sc) <- so_oat_sc$hepatocyte
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

### Kidney disease Y/N
```{r}
# Idents(so_oat_sc) <- so_oat_sc$celltype
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_oat_sc) <- so_oat_sc$hepatocyte
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

## Figures
### Obese vs. Obese + T2D
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### MASLD Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### OSA Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

## Obese vs. Obese + T2D
### Descriptive table
```{r}
# Descriptive stats
meta_kidney_raw <- so_kidney_sc@meta.data
meta_kidney_raw <- meta_kidney_raw %>%
  filter(cryostor_id!="")
form <- paste("diagnosis_of_diabetes", paste(colnames(meta_kidney_raw)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = meta_kidney_raw, test = F))

```


## MASLD Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
meta_oat_sc <- so_kidney_sc@meta.data %>%
  filter(cryostor_id!="")
form <- paste("diagnosis_of_MASLD", paste(colnames(meta_oat_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = meta_oat_sc, test = F))

```

## OSA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(cryostor_id!="")
form <- paste("diagnosis_of_osa", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## GLP-1RA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- liver_meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("glp1agonist", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Insulin Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("insulin", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Kidney disease Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_kidney_disease", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## Stratified UMAP scRNA 
### Obese vs. Obese + T2D
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "diagnosis_of_diabetes") 

```

### MASLD Y/N
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "diagnosis_of_MASLD") 

```

### OSA Y/N
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "diagnosis_of_osa") 

```

### GLP-1RA Y/N
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "glp1agonist") 

```

### Insulin Y/N
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "insulin") 

```

### Kidney disease Y/N
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "diagnosis_of_kidney_disease") 

```

## Differential Expression by Group
### Obese vs. Obese + T2D
```{r}
so_kidney_sc@meta.data$hepatocyte <- ifelse(grepl("Hepatocyte",meta_liver_sc$celltype),"Hepatocyte","Non-Hepatocyte")
genes <- unique(rownames(so_kidney_sc))
Idents(so_kidney_sc) <- so_kidney_sc$celltype
de.markers(so_kidney_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
de.markers(so_kidney_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### MASLD Y/N
```{r}
Idents(so_kidney_sc) <- so_kidney_sc$celltype
de.markers(so_kidney_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
de.markers(so_kidney_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### OSA Y/N
```{r}
Idents(so_kidney_sc) <- so_kidney_sc$celltype
de.markers(so_kidney_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_kidney_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
de.markers(so_kidney_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### GLP-1RA Y/N
```{r}
Idents(so_kidney_sc) <- so_kidney_sc$celltype
de.markers(so_kidney_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_kidney_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_kidney_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
de.markers(so_kidney_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### Insulin Y/N
```{r}
# Idents(so_kidney_sc) <- so_kidney_sc$celltype
# de.markers(so_kidney_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_kidney_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_kidney_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
# de.markers(so_kidney_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

### Kidney disease Y/N
```{r}
# Idents(so_kidney_sc) <- so_kidney_sc$celltype
# de.markers(so_kidney_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_kidney_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_kidney_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
# de.markers(so_kidney_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

## Figures
### Obese vs. Obese + T2D
```{r}
Idents(so_kidney_sc) <- so_kidney_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### MASLD Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_kidney_sc) <- so_kidney_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### OSA Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_kidney_sc) <- so_kidney_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_kidney_sc) <- so_kidney_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```