---
title: "OAT scRNA & PAH Clearance"
author: "Hailey Hampson"
date: "2024-09-24"
output: html_document
--- 

#1. Load Libraries & Directories
# 1. Set up Libraries & Directores
```{r, include=F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
library(GSEABase)
library(limma)
library(reshape2)
library(data.table)
library(knitr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
library(NMF)
library(rsvd)
library(RColorBrewer)
library(MAST)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
#library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
registerDoParallel(cores = 6)
library(VennDiagram)
library(janitor)


#Local file path
#dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
#dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
#dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
#dir.results <- c("/Volumes/Peds Endo/Petter Bjornstad/Kidney Project/Results")

#Lambda file path
dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Kidney scRNA/Kidney scRNA")
dir.results <- c(fs::path(dir.dat,"Kidney Project/Results"))

#Load functions
source("Kidney_functions_sc.R")

```

#2. Load Data & Format
## a. Format Data
```{r, include=F}
#Local
# dir.dat <- c("/Users/hhampson/Dropbox/Bjornstad data")
# so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
# load("/Volumes/Peds Endo/Petter Bjornstad/Data Harmonization/Data Exports/PB90_clinical_metadata.RData")
# so_kidney_sc <- AddMetaData(so_kidney_sc, so_meta_combined)
# rm(so_meta_combined)
# gc()

#Lambda
so_kidney_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="KI-0014643")] <- "KL-0014643"
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="kl-0023998")] <- "KL-0023998"
#check2 <- so_kidney_sc@meta.data

#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv")) #small meta set
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata.csv")) #all metadata vars
#"KL-0014634" "KL-0014629" "KL-0014628" coenrolled IDs
#"KL-0014634" "KL-0014629" "KL-0014628" having missing meta data 

#KL-0014628: IT_10/RH-66-T KL-0014628 KL-0019101
#KL-0014629: IT_09/RH-65-T
#KL-0014634: IT_07/RH-59-T
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group")))

#Partipant 30051 (kit id KL-0027474 at 4m has sc, but at baseline no sc has kit id KL-0027474 metadata)
#missing <- c("KL-0014628", "KL-0014629", "KL-0014634", "KL-0019059", "KL-0019092", "KL-0019101", "KL-0022159", "KL-0024005", "KL-0027470", #"KL-0027478")
#Check these ids in metadat
#check <- harm_meta_data %>%
  #filter(kit_id %in% missing)

#IDs missing group information  
#Filter out attempt participants
harm_meta_data <- harm_meta_data %>% 
  dplyr::select(-X)
#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>% 
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>% 
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>% 
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
  #mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups: 
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
gc()
harm_meta_data <- harm_meta_data %>% 
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                            sglt2 == "Yes" ~ "sglt2",
                            glp1 == "Yes" ~ "glp1",
                            no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
gc()
meta_kidney_sc <-  so_kidney_sc@meta.data 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
meta_kidney_sc <- meta_kidney_sc %>%
  #dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(harm_meta_data,by="kit_id")
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
#Check to see if all ids still have metadata
#check4 <- meta_kidney_sc %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check4 <- check4 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","group2")))

#rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
ids <- harm_meta_data$kit_id
#Filter seurat object to only IDs that have the metadata (83 individuals of interest)
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)
#check5 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check5 <- check5 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","group2")))
so_kidney_sc <- subset(so_kidney_sc, kit_id %in% ids)
#check6 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check6 <- check6 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","group2")))
#unique(so_kidney_sc$group)
#unique(so_kidney_sc$group2)
rm(meta_kidney_sc,harm_meta_data)

#Switch default assay in seurat object to RNA
DefaultAssay(so_kidney_sc) <- "RNA"
gc()

check3 <- so_kidney_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","group2","medication","pah_clear_bsa","pah_clear_abs")))
#saveRDS(so_kidney_sc,fs::path(dir.dat,"scRNA","data_clean","Formatted_so_kidney_sc_01_26_25.rds"))
#Create senesence so in all cell types
#sens_genes <- c(sens_genes,"CDKN1A")
#sens_gene_in_data <- intersect(sens_genes, rownames(so_kidney_sc))
# Subset the Seurat object to include only genes in sens_gene_in_data
#so_sens_all <- subset(so_kidney_sc, features = sens_gene_in_data)
#rm(so_kidney_sc)
# # saveRDS(so_sens_all,fs::path(dir.dat,"so_sens_genes_all_cells_KIDNEY.RDS"))
# # rm(so_sens_all)
# # 
# # #Load sens all cells
# # so_sens <- readRDS(fs::path(dir.dat,"so_sens_genes_all_cells_KIDNEY.RDS"))
# # dat <- so_sens@meta.data
# # so_diab <- subset(so_sens, diagnosis_of_diabetes == "Yes")
# # 
# #PT Cells
# so_sens_all@meta.data$PT <- ifelse(grepl("PT-",so_sens_all@meta.data$celltype_rpca),"PT","Non-PT")
# Idents(so_sens_all) <- so_sens_all$PT
# so_sens <- subset(so_sens_all, PT=="PT")
# # saveRDS(so_sens,fs::path(dir.dat,"so_sens_genes_PT_Cells_KIDNEY.RDS"))
# rm(so_sens_all,so_sens)
# so_sens_PT <- readRDS(fs::path(dir.dat,"Liver Project","so_sens_genes_PT_Cells_KIDNEY.RDS"))
# so_sens_PT@meta.data$group1 <- ifelse(so_sens_PT@meta.data$group=="Type 2 Diabetes" | so_sens_PT@meta.data$group=="Lean Control" ,"Yes","No")
# so_sens_PT1 <- subset(so_sens_PT, group1=="Yes")
# so_sens_PT@meta.data$group2 <- ifelse(so_sens_PT@meta.data$group=="Type 2 Diabetes","Yes","No")
# so_sens_PT2 <- subset(so_sens_PT, group2=="Yes")
# so_sens_PT3 <- subset(so_sens_PT, group=="Type 2 Diabetes")
# 
# 
# 
# #TAL Cells
# so_sens_all@meta.data$TAL <- ifelse(grepl("TAL",so_sens_all@meta.data$celltype_rpca),"TAL","Non-TAL")
# Idents(so_sens_all) <- so_sens_all$TAL
# so_sens <- subset(so_sens_all, TAL=="TAL")
# # saveRDS(so_sens,fs::path(dir.dat,"so_sens_genes_TAL_Cells_KIDNEY.RDS"))
# rm(so_sens_all,so_sens,so_sens_TAL,so_sens_TAL1)
# so_sens_TAL <- readRDS(fs::path(dir.dat,"Liver Project","so_sens_genes_TAL_Cells_KIDNEY.RDS"))
# so_sens_TAL@meta.data$group1 <- ifelse(so_sens_TAL@meta.data$group=="Type 2 Diabetes"| so_sens_TAL@meta.data$group=="Lean Control","Yes","No")
# so_sens_TAL1 <- subset(so_sens_TAL, group1=="Yes")
# so_sens_TAL@meta.data$group2 <- ifelse(so_sens_TAL@meta.data$group=="Type 2 Diabetes","Yes","No")
# so_sens_TAL2 <- subset(so_sens_TAL, group2=="Yes")
# so_sens_TAL3 <- subset(so_sens_TAL, group=="Type 2 Diabetes")
```
```{r echo = F}
##a. Kidney scRNA----
so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))

##b. Metadata ----
meta_raw <- read.csv(fs::path(dir.dat,"Clinical Data","renal_clearance_biopsy.csv"))
harm_data <- read.csv("/Users/hhampson/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Biostatistics Core Shared Drive/Data Harmonization/Data Clean/harmonized_dataset.csv") %>% 
  dplyr::select(record_id,cryostor_id,visit,pah_clear_bsa,pah_clear_abs,group)

##c. Process Data----
#Filter out meta data participants without sc 
ids <- unique(so_kidney_sc@meta.data$cryostor_id)
meta_raw <- meta_raw %>% 
  filter(cryostor_id %in% ids) 
ids2 <- unique(so_kidney_sc@meta.data$record_id)
harm_data <- harm_data %>% 
  filter(record_id %in% ids2) %>% 
  filter(!is.na(pah_clear_bsa)) %>% 
  # filter(visit=="baseline") %>% 
  mutate(diabetes=ifelse(group=="Type 1 Diabetes"|group=="Type 2 Diabetes","Yes","No")) %>% 
  dplyr::select(record_id,visit,pah_clear_bsa,diabetes,pah_clear_abs,pah_bsa,
                pah_bsa_plasma_urine)
meta_raw <- tidylog::left_join(meta_raw,harm_data,by=c("record_id","visit"))

#Merge new metadata with 
meta_kidney_sc <-  tidylog::left_join(so_kidney_sc@meta.data, meta_raw,by="cryostor_id")

#Check there are no duplicates
length(which(duplicated(meta_kidney_sc$barcode)))

#Add rownames back to metadata to merge back into saurat object 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)

#Create study visit variale to filter to RH, CROCCODILE and IMPROVE
so_kidney_sc@meta.data$study= ifelse(grepl("RH",so_kidney_sc@meta.data$record_id) | grepl("CRC",so_kidney_sc@meta.data$record_id) | grepl("IT",so_kidney_sc@meta.data$record_id),"subset","non-subset")

# Add new meta data to so 
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)

gc()
#Filter to baseline visits only
so_kidney_sc <- subset(so_kidney_sc, visit != "12_months_post_surgery")

gc()
#Filter to RH/RH2, CROCODILE, and IMPROVE
so_kidney_sc <- subset(so_kidney_sc, study == "subset")


# #Normalize & Scale Data
# so_kidney_sc <- NormalizeData(so_kidney_sc)
# so_kidney_sc <- ScaleData(so_kidney_sc)

```

#3. Visualize Data 
```{r Visualize Data}
#Perform PCA
so_kidney_sc <- RunPCA(so_kidney_sc, features = VariableFeatures(object = so_kidney_sc))
ElbowPlot(so_kidney_sc)
# Cluster cells
so_kidney_sc <- FindNeighbors(so_kidney_sc)
so_kidney_sc <- FindClusters(so_kidney_sc)
# Perform UMAP and tSNE
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap")
DimPlot(so_kidney_sc, reduction = "umap", group.by = "sglt2i_ever")
DimPlot(so_kidney_sc, reduction = "umap", group.by = "diabetes")
DimPlot(so_kidney_sc, reduction = "umap", group.by = "group")
```

#4. Linear Regression 
```{r}
agg_expression <- AggregateExpression(
  object = so_kidney_sc,
  features = NULL,
  group.by = c("kit_id", "celltype_rpca"),  # Aggregate by cell_type
  fun = "mean",
  return.seurat = FALSE
)

# Check the structure of the result
agg_expression_matrix <- agg_expression$RNA  # Assuming RNA is the name of your assay

# Ensure 'pah' values are aligned with the conditions in agg_expression_matrix
# Assuming that the 'kit_id' and 'celltype_rpca' are the columns that the data was aggregated by.

# Extract metadata for 'pah' and align it with the columns of aggregated expression matrix
agg_metadata <- so_kidney_sc@meta.data[, c("kit_id", "celltype_rpca", "pah_clear_bsa")]

# Create a combined factor for 'kit_id' and 'celltype_rpca' to match columns in agg_expression_matrix
agg_metadata$condition <- interaction(agg_metadata$kit_id, agg_metadata$celltype_rpca, sep = "_")

# Check if the names match the column names of the aggregated expression matrix
colnames(agg_expression_matrix) <- as.character(agg_metadata$condition)

# Now align the pah values to the column names (conditions) of agg_expression_matrix
pah_values <- agg_metadata$pah

# Make sure pah_values is of the correct length to match the columns of the aggregated expression matrix
if(length(pah_values) != ncol(agg_expression_matrix)) {
  stop("The number of pah values does not match the number of columns in the aggregated expression matrix.")
}

# Now you can run linear regression for each gene with pah as the predictor
regression_results <- data.frame(Gene = character(),
                                 Coefficient = numeric(),
                                 PValue = numeric(),
                                 stringsAsFactors = FALSE)

# Loop over each gene and perform linear regression
for (gene in rownames(agg_expression_matrix)) {
  gene_expression <- agg_expression_matrix[gene, ]
  
  # Perform linear regression with pah as the predictor
  lm_model <- lm(gene_expression ~ pah_values)
  
  # Extract the coefficient and p-value from the model
  model_summary <- summary(lm_model)
  coef <- model_summary$coefficients[2, 1]  # Coefficient for pah
  p_value <- model_summary$coefficients[2, 4]  # P-value for pah
  
  # Store the results
  regression_results <- rbind(regression_results, data.frame(Gene = gene, 
                                                           Coefficient = coef, 
                                                           PValue = p_value))
}

# View the results
head(regression_results)


# Extract 'pah' variable from Seurat metadata
pah_values <- so_kidney_sc@meta.data$pah_clear_bsa
length(pah_values)
# Create an empty data frame to store results
regression_results <- data.frame(Gene = character(), 
                                 Coefficient = numeric(), 
                                 PValue = numeric(), 
                                 stringsAsFactors = FALSE)

# Loop over each gene and perform linear regression
for (gene in rownames(agg_expression_matrix)) {
  gene_expression <- agg_expression_matrix[gene, ]
  
  # Perform linear regression with pah as the predictor
  lm_model <- lm(gene_expression ~ pah_values)
  
  # Extract the coefficient and p-value from the model
  model_summary <- summary(lm_model)
  coef <- model_summary$coefficients[2, 1]  # Coefficient for pah
  p_value <- model_summary$coefficients[2, 4]  # P-value for pah
  
  # Store the results
  regression_results <- rbind(regression_results, data.frame(Gene = gene, 
                                                           Coefficient = coef, 
                                                           PValue = p_value))
}

# View the results
head(regression_results)


```

#5. Differential Expression Analysis
```{r}
#Define Gene Sets
all.genes = rownames(so_kidney_sc)
oat.genes <- all.genes[which(grepl("SLC22",all.genes))]

#Differential Expression of SLC22 genes in PT by Group
#Define Proximal Tubule Cells (PT)
so_kidney_sc@meta.data$PT <- ifelse(grepl("PT",so_kidney_sc$celltype_rpca),"PT","")
Idents(so_kidney_sc) <- so_kidney_sc$PT

#PAH Clearance Normal vs. Low 
so_kidney_sc@meta.data$pah_norm <- ifelse(so_kidney_sc@meta.data$pah_clear_bsa>=625,"Normal","Low")
## Top 2000 genes
de.markers(so_kidney_sc, oat.genes, "pah_norm", id2 = "Normal", id1 = "Low", "PT", "_top")
m_top <- m_top %>% head(2000)

# significant_genes <- m_top %>% filter(p_val_adj < 0.05)
# 
# # Select the top 10 positive and top 10 negative log2FC genes that are significant
# top_genes <- rbind(
#   significant_genes %>% arrange(desc(avg_log2FC)) %>% head(10),  # Top 10 positive log2FC
#   significant_genes %>% arrange(avg_log2FC) %>% head(10)         # Top 10 negative log2FC
# )
# 
# labels <- ifelse(rownames(m_top) %in% rownames(top_genes), rownames(m_top), NA)
# p <- EnhancedVolcano(m_top,
#                      lab = labels,
#                      x = 'avg_log2FC',
#                      y = 'p_val_adj',
#                      title = paste0("Differentially Expressed OAT Genes by PAH Clearance (Pseudobulk PT)"),
#                      subtitle = paste0("Positive Log2 FC = Greater Expression in PAH Clearance Normal vs. Low\n",
#                                        "(Significant at FDR-P<0.05, FC Threshold = 0.5)"),
#                      pCutoff = 0.05,
#                      FCcutoff = 0.1,
#                      labFace = 'bold',
#                      pointSize = 4,
#                      labSize = 5,
#                      drawConnectors = TRUE,
#                      widthConnectors = 1.0,
#                      colConnectors = 'black',
#                      legendPosition=NULL,
#                      boxedLabels = TRUE,
#                      max.overlaps=20)
# plot(p)

```

#6. Model-based Analysis of Single-cell Transcriptomics 
```{r}
# Get expression matrix (counts)
expr_matrix <- as.matrix(GetAssayData(so_kidney_sc, layer = "counts"))

# Extract metadata containing your continuous variable, e.g., PAH clearance
metadata <- so_kidney_sc@meta.data

# Assuming the rownames of your metadata or expression matrix correspond to unique cell IDs
metadata$wellKey <- rownames(metadata)

# Create SingleCellAssay object with the updated metadata containing wellKey
sca <- FromMatrix(exprsArray = expr_matrix, cData = metadata)

#Specify the Regression Model
zlm_model <- zlm(~ pah_clear_bsa, sca)

#Extract results
# Test for the effect of PAH clearance
summary_res <- summary(zlm_model, doLRT='pah_clearance')

# Extract the summary DataFrame
summary_dt <- summary_res$datatable

# Filtering for significant genes based on a specific threshold (e.g., p < 0.05)
significant_genes <- summary_dt[summary_dt$component == "H" & summary_dt$`Pr(>Chisq)` < 0.05,]

# Plot the effect of PAH clearance on specific genes
# Example: Plot gene expression of gene X against PAH clearance
gene_of_interest <- "GeneX"

ggplot(metadata, aes(x = pah_clearance, y = expr_matrix[gene_of_interest, ])) +
  geom_point() +
  geom_smooth(method = "lm") +
  labs(title = paste("Expression of", gene_of_interest, "vs PAH clearance"))

# Include additional covariates like batch or condition
zlm_model <- zlm(~ pah_clearance + batch + condition, sca)


```

#7. Pseudotime Analysis
```{r pseudotime analysis}
## Make SO into SCE object for pseudotime analysis
sce_PT <- as.SingleCellExperiment(subset(so_kidney_sc, PT == "PT"), assay = "RNA")
rm(harm_data,meta_kidney_sc,meta_raw,so_kidney_sc)

# Totem clustering for trajectory analysis
gc()
sce_PT <- PrepareTotem(sce_PT)
sce_PT <- RunDimRed(object = sce_PT,
                    dim.red.method = "pca",
                    dim.red.features = row.names(m_top),
                    dim.reduction.par.list = list(ndim=5))

## where so@assays$RNA@counts is the normalized expression count
# gc()
# dim_red <- dimred_pca(t(subset(so, generaltype == "PT" & celltype != "PT_lowQuality")@assays$RNA@counts), ndim=2)
dim_red <- reducedDim(sce_PT, type = "pca")

sce_PT <- RunClustering(sce_PT,
                        k.range = 3:20,
                        min.cluster.size = 5,
                        N.clusterings=10000)
gc()
viz_cell <- VizCellConnectivity(sce_PT,viz.dim.red = dim_red)

pushover(message = "done w/ VizCellConnectivity")

sce_PT <- SelectClusterings(sce_PT,selection.method = 1,
                       selection.N.models = 10,
                       selection.stratified=FALSE,
                       prior.clustering = NULL)
VizMST(sce_PT,clustering.names = ReturnTrajNames(sce_PT),viz.dim.red = dim_red)
```
