---
title: "OAT scRNA & PAH Clearance"
author: "Hailey Hampson"
date: "2024-09-24"
output: html_document
--- 
  
```{r libraries, echo=F, include = F}
library(tidyverse)
library(BiocManager)
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
```

# scRNA
```{r echo = F}
# Liver scRNA data processing
# so_liver_sc <- readRDS("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_raw/PB_Liver_7SingleCellDatasets.RDS")
so_liver_sc <- readRDS("/Users/hhampson/Dropbox/Bjornstad data/Liver/PB_Liver_7SingleCellDatasets.RDS")

# Add missing liver meta data to Seurat object
# meta_liver_raw <- read.csv("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean/liver_biopsy_metadata_PN.csv")
meta_liver_raw <- read.csv("/Users/hhampson/Dropbox/Bjornstad data/Liver/liver_biopsy_metadata_PN.csv")

meta_liver_sc <-  so_liver_sc@meta.data[,1:9] %>%
  dplyr::mutate(Cryostor_ID = gsub("-Liv", "", orig.ident)) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sc) <- rownames(so_liver_sc@meta.data)

so_liver_sc <- AddMetaData(so_liver_sc, meta_liver_sc)
so_liver_sc <- NormalizeData(so_liver_sc)
so_liver_sc <- ScaleData(so_liver_sc)
# PCA
so_liver_sc <- RunPCA(so_liver_sc, features = VariableFeatures(object = so_liver_sc))
ElbowPlot(so_liver_sc)
# Cluster cells
so_liver_sc <- FindNeighbors(so_liver_sc)
so_liver_sc <- FindClusters(so_liver_sc)
# Perform UMAP and tSNE
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap") 
```

# snRNA
```{r echo = F}
# Liver snRNA data processing
# so_liver_sn <- readRDS("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_raw/NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS")
so_liver_sn <- readRDS("/Users/hhampson/Dropbox/Bjornstad data/Liver/NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS")

meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
```

## snRNA UMAP

```{r echo=F}
# Set a higher memory limit
mem.maxVSize(64000000000)
so_liver_sn_rna <- NormalizeData(so_liver_sn@assays$RNA)
so_liver_sn_rna <- ScaleData(so_liver_sn@assays$RNA)

so_liver_sn_int <- NormalizeData(so_liver_sn@assays$integrated)
so_liver_sn_int <- ScaleData(so_liver_sn@assays$integrated)

# PCA
so_liver_sn_int <- RunPCA(so_liver_sn_int, features = VariableFeatures(object = so_liver_sn_int))
ElbowPlot(so_liver_sn_int)
# Cluster cells
so_liver_sn_int <- FindNeighbors(so_liver_sn_int)
so_liver_sn_int <- FindClusters(so_liver_sn_int)
# Perform UMAP and tSNE
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap") 

```

## Obese vs. Obese + T2D
### Descriptive table
```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_diabetes", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## MASLD Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_MASLD", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## OSA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_osa", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## GLP-1RA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- liver_meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("glp1agonist", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Insulin Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("insulin", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Kidney disease Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_kidney_disease", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## Stratified UMAP scRNA 
### Obese vs. Obese + T2D
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_diabetes") 

```

### MASLD Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_MASLD") 

```

### OSA Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_osa") 

```

### GLP-1RA Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "glp1agonist") 

```

### Insulin Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "insulin") 

```

### Kidney disease Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_kidney_disease") 

```

## Differential Expression by Group
### Obese vs. Obese + T2D
```{r}
so_liver_sc@meta.data$hepatocyte <- ifelse(grepl("Hepatocyte",meta_liver_sc$celltype),"Hepatocyte","Non-Hepatocyte")
genes <- unique(rownames(so_liver_sc))
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### MASLD Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### OSA Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### GLP-1RA Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### Insulin Y/N
```{r}
# Idents(so_liver_sc) <- so_liver_sc$celltype
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_liver_sc) <- so_liver_sc$hepatocyte
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

### Kidney disease Y/N
```{r}
# Idents(so_liver_sc) <- so_liver_sc$celltype
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_liver_sc) <- so_liver_sc$hepatocyte
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

## Figures
### Obese vs. Obese + T2D
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### MASLD Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### OSA Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```