---
title: "Proteomics and DKD"
author: "Laura Pyle & Tim Vigers"
date: "today"
date-format: long
format:
  html:
    toc: true
    toc-depth: 5
    code-fold: true
    self-contained: true
    fig-cap-location: top
    page-layout: full
editor: source
execute:
  echo: false
---

```{r libraries}
#| include: false
library(readxl)
library(tidyverse)
library(psych)
library(corrplot)
library(ggpmisc)
library(pander)
library(Seurat)
library(clusterProfiler)
library(ReactomePA)
library(enrichplot)
library(ggrepel)
```

```{r data cleaning}
#| include: false
# Import top proteins for MIC, MAC, etc.
top_mic_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "MIC CPH")
de_genes_mic = top_mic_df[top_mic_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_mic <- setNames(de_genes_mic$estimate, de_genes_mic$EntrezGeneID)
top_mic <- top_mic_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_mac_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "MAC CPH")
de_genes_mac = top_mac_df[top_mac_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_mac <- setNames(de_genes_mac$estimate, de_genes_mac$EntrezGeneID)
top_mac <- top_mac_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_hyp_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "HYP CPH")
de_genes_hyp = top_hyp_df[top_hyp_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_hyp <- setNames(de_genes_hyp$estimate, de_genes_hyp$EntrezGeneID)
top_hyp <- top_hyp_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_rapid_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "RAPID CPH")
de_genes_rapid = top_rapid_df[top_rapid_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_rapid <- setNames(de_genes_rapid$estimate, de_genes_rapid$EntrezGeneID)
top_rapid <- top_rapid_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_htn_sbp_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "HTN with SBP CPH")
de_genes_htn_sbp = top_htn_sbp_df[top_htn_sbp_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_htn_sbp <- setNames(de_genes_htn_sbp$estimate, de_genes_htn_sbp$EntrezGeneID)
top_htn_sbp <- top_htn_sbp_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_htn_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "HTN CPH")
de_genes_htn = top_htn_df[top_htn_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_htn <- setNames(de_genes_htn$estimate, de_genes_htn$EntrezGeneID)
top_htn <- top_htn_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_neuro_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "NEURO CPH")
de_genes_neuro = top_neuro_df[top_neuro_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_neuro <- setNames(de_genes_neuro$estimate, de_genes_neuro$EntrezGeneID)
top_neuro <- top_neuro_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_retino_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "RETINO CPH")
de_genes_retino = top_retino_df[top_retino_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_retino <- setNames(de_genes_retino$estimate, de_genes_retino$EntrezGeneID)
top_retino <- top_retino_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_glyc_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "GLYC CPH")
de_genes_glyc = top_glyc_df[top_glyc_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_glyc <- setNames(de_genes_glyc$estimate, de_genes_glyc$EntrezGeneID)
top_glyc <- top_glyc_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
top_glyc_a1c_df <- read_excel("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/TODAY subaward/Results/Linear and Cox models/TODAY somalogic Cox models scaled baseline adjusted.xlsx", sheet = "GLYC with A1c CPH")
de_genes_glyc_a1c = top_glyc_a1c_df[top_glyc_a1c_df$p.value <= 0.05,c("EntrezGeneID","estimate")]
de_genes_glyc_a1c <- setNames(de_genes_glyc_a1c$estimate, de_genes_glyc_a1c$EntrezGeneID)
top_glyc_a1c <- top_glyc_a1c_df %>%
  filter(adj.p.value <= 0.05) %>%
  slice_max(abs(estimate), n = 5) %>%
  pull(AptName)
# Import and clean data
df <- read.csv("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Data Harmonization/Data Clean/harmonized_dataset.csv",na.strings = "")
df = df %>% filter(study %in% c("CROCODILE","IMPROVE","RENAL-HEIR"),
                   !grepl("IT2D",co_enroll_id),participation_status=="Participated") %>%
  select(record_id,co_enroll_id,visit,group, age, sex,race_ethnicity, 
         diabetes_duration,albuminuria_cat,
         bmi, hba1c, gfr_bsa_plasma, gfr_raw_plasma, gfr_bsa_plasma_urine, 
         gfr_raw_plasma_urine,acr_u,map) %>%
  group_by(record_id,visit) %>%
  summarise(across(where(is.character),~last(na.omit(.x))),
            across(where(is.factor),~last(na.omit(.x))),
            across(where(is.numeric),~mean(.x,na.rm = T)),.groups = "drop") %>%
  mutate_all(~ifelse(is.nan(.), NA, .))
# Import proteomics data for RH and IMPROVE
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Renal HERITAGE/Somalogic data/analytes.Rdata")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Renal HERITAGE/Somalogic data/rh_soma.Rdata")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/IMPROVE T2D/Somalogic data/improve_soma.Rdata")
# Format and combine
improve_soma <- improve_soma %>% select(SampleDescription, TimePoint, contains("seq."))
rh_soma <- rh_soma %>% select(SampleDescription, TimePoint, contains("seq."))
soma <- rbind(improve_soma, rh_soma)
# Transform
soma[,3:ncol(soma)] = lapply(soma[,3:ncol(soma)],log)
# Merge clinical and SOMA data
soma <- soma %>%
  rename(record_id = "SampleDescription", visit = "TimePoint") %>%
  mutate(
    record_id = sub("IT2D-", "IT_", record_id),
    visit = case_when(
      visit == "BL" ~ "baseline",
      visit == "3M" ~ "3_months_post_surgery",
      visit == "12M" ~ "12_months_post_surgery"
    )
  )
df <- full_join(df, soma, by = c("record_id", "visit"))
# Import Olink data and combine
olink_map <- read.csv("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Olink Data/Data_Clean/olink_id_map.csv")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/IMPROVE T2D/Olink Data/improve_olink_plasma.Rdata")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/IMPROVE T2D/Olink Data/improve_olink_urine.Rdata")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Renal HERITAGE/Olink Data/rh_olink_plasma.Rdata")
load("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Renal HERITAGE/Olink Data/rh_olink_urine.Rdata")
olink_plasma <- rbind(improve_olink_plasma, rh_olink_plasma)
olink_urine <- rbind(improve_olink_urine, rh_olink_urine)
# Log transform
olink_plasma[,3:ncol(olink_plasma)]
olink_urine[,3:ncol(olink_urine)]
# Add visit column, format IDs
olink_plasma$visit <- sapply(str_split(olink_plasma$record_id, "_"), "[", 3)
olink_plasma <- olink_plasma %>%
  mutate(
    visit = case_when(
      visit == "BL" ~ "baseline",
      visit == "3M" ~ "3_months_post_surgery",
      visit == "12M" ~ "12_months_post_surgery",
      is.na(visit) ~ "baseline"
    ),
    record_id = sub("_BL|_3M|_12M", "", record_id)
  )
olink_urine$visit <- sapply(str_split(olink_urine$record_id, "_"), "[", 3)
olink_urine <- olink_urine %>%
  mutate(
    visit = case_when(
      visit == "BL" ~ "baseline",
      visit == "3M" ~ "3_months_post_surgery",
      visit == "12M" ~ "12_months_post_surgery",
      is.na(visit) ~ "baseline"
    ),
    record_id = sub("_BL|_3M|_12M", "", record_id)
  )
# Baseline visit only for IMPROVE
df <- df %>% filter(visit == "baseline")
# Combine
plasma <- left_join(df, olink_plasma, by = c("record_id", "visit"))
urine <- left_join(df, olink_urine, by = c("record_id", "visit"))
# Read in scRNA object
so <- readRDS("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/scRNA/Data_Raw/PB_68samples_harmonyIntegration_092322_Fadhl.RDS")
# Strip computations
so <- DietSeurat(so)
# Limit to RH/IMPROVE baseline visit
so = so[,!grepl("_12M",so$michigan_id)]
# Remove HC with igA
so = so[,so$T2D_HC_Phil %in% c("T2Di","T2D","HC")]
# Combined groups
so$diabetes = sub("i","",so$T2D_HC_Phil)
so$SGLT2i = factor(so$T2D_HC_Phil,levels = c("HC","T2D","T2Di"),
                   labels = c("No","No","Yes"))
so$acr_u_cat = cut_number(so$acr_u,n = 2,labels=c("Low UACR","High UACR"))
# Normalize and scale
so <- NormalizeData(so)
so <- ScaleData(so, features = rownames(so))
# PCA
so <- RunPCA(so, features = VariableFeatures(object = so))
# Cluster cells
so <- FindNeighbors(so)
so <- FindClusters(so)
# Perform UMAP
so <- RunUMAP(so, dims = 1:20)
# General cell types as identifiers
so$generaltype = sub("_.*","",so$LR_clusters)
Idents(so) <- so$generaltype
```

```{r functions}
# Define correlation function
cor_and_plot_soma <- function(data = df, vars) {
  # Format matrix
  d <- data[data$group=="Type 2 Diabetes", 
            c(vars, "acr_u","map","gfr_raw_plasma","gfr_bsa_plasma")]
  n <- c(analytes$Target[match(vars, analytes$AptName)], 
         "acr_u","map","gfr_raw_plasma","gfr_bsa_plasma")
  n[duplicated(n)] <- paste0(n[duplicated(n)], "(2)")
  names(d) <- n
  # Correlations
  M <- corr.test(data.matrix(d), normal = F, method = "spearman")
  corrplot(M$r,
           p.mat = M$p, method = "circle", type = "lower", insig = "blank",
           addCoef.col = "white", diag = FALSE
  )
  # Print a nice looking table
  r <- data.frame(M$r)
  emphasize.strong.cells(which(M$p <= 0.05, arr.ind = TRUE))
  pander(r)
  # Print p values
  p = data.frame(M$p)
  p[]=sapply(p, format.pval,eps=0.001)
  pander(p)
}
# Function for converting SOMAScan to Olink
soma_to_olink <- function(soma_ids) {
  ids <- list()
  length(ids) <- length(soma_ids)
  ids <- lapply(soma_ids, function(n) {
    uni <- analytes$UniProt[analytes$AptName == n]
    olink <- olink_map$OlinkID[olink_map$Uniprot.ID == uni]
    return(as.character(na.omit(olink)))
  })
  names(ids) <- soma_ids
  ids <- ids[lapply(ids, length) > 0]
  return(ids)
}
# Plot agreement
soma_olink_plots <- function(top_olink, data) {
  invisible(lapply(names(top_olink), function(n) {
    title <- analytes$TargetFullName[analytes$AptName == n]
    plots <- lapply(top_olink[n][[1]], function(o) {
      panel <- olink_map$Panel[olink_map$OlinkID == o]
      p <- ggplot(data[data$group=="Type 2 Diabetes",], aes_string(x = n, y = o)) +
        geom_point(alpha = 0.5) +
        stat_poly_line(se = F) +
        stat_poly_eq(
          aes(label = paste(after_stat(eq.label),
                            after_stat(rr.label),
                            sep = "*\", \"*"
          )),
          label.y = 0.2, label.x = 0.9
        ) +
        theme_bw() +
        ylab(paste0(o, " [", panel, "]")) +
        ggtitle(title)
      print(p)
    })
  }))
}
# Correlate Olink with UACR
cor_and_plot_olink <- function(plasma_df = plasma, urine_df = urine, vars) {
  # Get Olink names
  vars <- as.character(vars)
  # Format matrices
  plasma_df <- plasma_df[plasma_df$group=="Type 2 Diabetes", c(vars, "acr_u")]
  urine_df <- urine_df[urine_df$group=="Type 2 Diabetes", c(vars, "acr_u")]
  # Gene names
  n <- c(olink_map$Assay[match(vars, olink_map$OlinkID)], "UACR")
  n[duplicated(n)] <- paste0(n[duplicated(n)], "(2)")
  names(plasma_df) <- n
  names(urine_df) <- n
  # Plasma correlations
  M_plasma <- corr.test(data.matrix(plasma_df), normal = F, method = "spearman")
  corrplot(M_plasma$r,
           p.mat = M_plasma$p, method = "circle", type = "lower",
           insig = "blank", addCoef.col = "white", diag = FALSE, title = "Plasma"
  )
  # Print nice looking table
  r <- data.frame(M_plasma$r)
  emphasize.strong.cells(which(M_plasma$p <= 0.05, arr.ind = TRUE))
  pander(r)
  # Print p values
  p = data.frame(M_plasma$p)
  p[]=sapply(p, format.pval,eps=0.001)
  pander(p)
  # Urine correlations
  M_urine <- corr.test(data.matrix(urine_df), normal = F, method = "spearman")
  corrplot(M_urine$r,
           p.mat = M_urine$p, method = "circle", type = "lower",
           insig = "blank", addCoef.col = "white", diag = FALSE, title = "Urine"
  )
  # Print nice looking table
  r <- data.frame(M_urine$r)
  emphasize.strong.cells(which(M_urine$p <= 0.05, arr.ind = TRUE))
  pander(r)
  # Print p values
  p = data.frame(M_urine$p)
  p[]=sapply(p, format.pval,eps=0.001)
  pander(p)
}
# UMAP function
umap_plot <- function(seurat_object = so, vars) {
  # Get Entrez Gene IDs
  entrez <- analytes$EntrezGeneSymbol[match(vars, analytes$AptName)]
  entrez <- entrez[entrez %in% rownames(so)]
  # Plot each gene
  invisible(lapply(entrez, function(e) {
    p <- FeaturePlot(seurat_object, features = e)
    p <- LabelClusters(plot = p, id = "ident")
    print(p)
  }))
}
# DE function
de = function(seurat_object = so,outcome,ref_group,vars){
  # Get Entrez Gene IDs
  entrez <- analytes$EntrezGeneSymbol[match(vars, analytes$AptName)]
  entrez <- entrez[entrez %in% rownames(seurat_object)]
  entrez= unique(entrez)
  # Find cells where most highly expressed
  cells = FetchData(seurat_object,vars = entrez)
  cells$type = seurat_object$generaltype
  cells = cells %>% group_by(type) %>%
    summarise(across(where(is.numeric), ~ mean(.x>0, na.rm = TRUE))) %>%
    column_to_rownames("type")
  max = lapply(cells,function(c){
    rownames(cells)[which.max(c)]
  })
  # DE testing
  de_tables = lapply(names(max),function(n){
    m = FindMarkers(seurat_object,features = n,ident.1 = ref_group,
                    group.by = outcome,subset.ident = max[[n]],
                    min.pct = 0,logfc.threshold = 0,verbose = F)
    m = data.frame(m)
    m$celltype = max[[n]]
    return(m)
  })
  # Combine rows and print
  de_table = data.frame(do.call(rbind,de_tables))
  de_table = de_table[order(de_table$p_val),]
  # Plots
  invisible(lapply(names(max),function(n){
    p = VlnPlot(seurat_object,features = n,group.by = outcome,
                idents = max[[n]],pt.size=0.01,log = F)
    print(p)
  }))
  pander(de_table)
}
# Volcano plot
volcano = function(data,lab = "Target",xcol = "estimate",ycol = "p.value",top = 5,
                   xlimit = c(-1, 1),ylimit = c(0, -log10(10e-7)),
                   xlabel = "HR",pCutoff = 0.05,overlaps=10){
  data = as.data.frame(data)
  t = data[data[,"adj.p.value"]<=pCutoff,]
  t = t[order(abs(t[,xcol]),decreasing = T),]
  data$top = data[,"AptName"] %in% t[1:top,"AptName"]
  data$logp = -log10(data[,ycol])
  data$fc = data[,xcol]
  data$sig = data[,ycol] <= pCutoff
  p = ggplot(data = data,aes(x = fc,y = logp,color = sig)) + 
    geom_hline(yintercept = -log10(pCutoff),linetype = "dashed") +
    geom_point(size=2) + 
    geom_label_repel(data=data[data$top,],aes(label=Target),color = "black",
                     max.overlaps = overlaps,size=6) +
    scale_color_manual(values = c("grey","#3e6dbf")) +
    xlab(xlabel) + ylab(bquote(~-Log[10] ~ italic(P))) +
    theme_bw() +
    theme(legend.position = "none",
          axis.text=element_text(size=15),
          axis.title=element_text(size=20)) 
  return(p)
}
# GSEA
pathway_analysis = function(de_genes,log_t=T){
  # log transform if necessary
  if(log_t){
    de_genes=log(de_genes)
  }
  # Sort
  de_genes = sort(de_genes,decreasing = T)
  # Remove duplicates and fix names
  de_genes = de_genes[!is.na(names(de_genes))]
  names(de_genes) = sub("\\|.*","",names(de_genes))
  de_genes = de_genes[which(!duplicated(names(de_genes)))]
  # GSEA with KEGG
  gsea = gsePathway(geneList = de_genes,pvalueCutoff = 0.1)
  dot= dotplot(gsea)
  ridge = ridgeplot(gsea)
  sig = which(gsea@result$p.adjust<=0.05)
  if(length(sig)>0){
    gseap = gseaplot2(gsea,geneSetID = sig,
                    pvalue_table = TRUE,ES_geom = "dot")
  } else {
   gseap = NULL 
  }
  # Convert to gene symbols
  gseax = setReadable(gsea, 'org.Hs.eg.db', 'ENTREZID')
  # More plots!
  cnet = cnetplot(gseax, foldChange=de_genes,colorEdge = T,node_label="category")
  # Return those beautiful plots
  return(list("ridge" = ridge,"gseap" = gseap,"dot" = dot,"cnet" = cnet))
}
# Function for comparing two gene lists
# de_gene_list is a list of multiple gene sets
pathway_comparison = function(de_gene_list){
  # Format names
  de_gene_list = lapply(de_gene_list, function(l){
    n = names(l)
    n = sub("\\|.*","",n)
    eg = bitr(n, fromType="UNIPROT", toType="ENTREZID", OrgDb="org.Hs.eg.db")
    n = eg$ENTREZID[match(n,eg$UNIPROT)]
    n = n[!is.na(n)]
    dups = n[duplicated(n)]
    n = n[!n %in% dups]
    return(n)
  })
  # Compare with GO enrichment
  ck = compareCluster(geneClusters = de_gene_list,
                      fun='enrichGO', OrgDb='org.Hs.eg.db')
  # Gene names
  ckx = setReadable(ck, OrgDb = "org.Hs.eg.db", keyType="ENTREZID")
  # Plots
  dot = dotplot(ckx) + ggtitle("Cluster Comparison Dot Plot")
  net = cnetplot(ckx) + ggtitle("Cluster Comparison Network Plot")
  xx = pairwise_termsim(ckx)
  emap = emapplot(xx) + ggtitle("Cluster Comparison Enrichment Map")
  # Return
  return(list("dot" = dot,"net" = net,"emap" = emap))
}
```

# MAC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_mac,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_mac,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_mac)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_mac_olink <- soma_to_olink(top_mac)
soma_olink_plots(top_olink = top_mac_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_mac_olink, data = urine)
```

## Olink associations with UACR

```{r}
cor_and_plot_olink(vars = top_mac_olink)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_mac)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_mac)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_mac)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_mac)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_mac_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MAC_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
mac_p = pathway_analysis(de_genes_mac)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MAC_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mac_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MAC_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mac_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MAC_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mac_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MAC_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mac_p$cnet,device = "jpeg")
```

# MIC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_mic,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_mic,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_mic)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_mic_olink <- soma_to_olink(top_mic)
soma_olink_plots(top_olink = top_mic_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_mic_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_mic)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_mic)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_mic)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_mic)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_mic_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MIC_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
mic_p = pathway_analysis(de_genes_mic)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MIC_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mic_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MIC_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mic_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MIC_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mic_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/MIC_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = mic_p$cnet,device = "jpeg")
```

# HYP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_hyp,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_hyp,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_hyp)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_hyp_olink <- soma_to_olink(top_hyp)
soma_olink_plots(top_olink = top_hyp_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_hyp_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_hyp)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_hyp)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_hyp)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_hyp)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_hyp_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HYP_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
hyp_p = pathway_analysis(de_genes_hyp)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HYP_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = hyp_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HYP_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = hyp_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HYP_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = hyp_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HYP_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = hyp_p$cnet,device = "jpeg")
```

# RAPID

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_rapid,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_rapid,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_rapid)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_rapid_olink <- soma_to_olink(top_rapid)
soma_olink_plots(top_olink = top_rapid_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_rapid_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_rapid)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_rapid)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_rapid)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_rapid)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_rapid_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RAPID_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
rapid_p = pathway_analysis(de_genes_rapid)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RAPID_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = rapid_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RAPID_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = rapid_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RAPID_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = rapid_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RAPID_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = rapid_p$cnet,device = "jpeg")
```

# HTN with SBP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_htn_sbp,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_htn_sbp,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_htn_sbp)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_htn_sbp_olink <- soma_to_olink(top_htn_sbp)
soma_olink_plots(top_olink = top_htn_sbp_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_htn_sbp_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_htn_sbp)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_htn_sbp)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_htn_sbp)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_htn_sbp)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_htn_sbp_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_SBP_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
htn_sbp_p = pathway_analysis(de_genes_htn_sbp)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_SBP_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_sbp_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_SBP_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_sbp_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_SBP_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_sbp_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_SBP_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_sbp_p$cnet,device = "jpeg")
```

# HTN

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_htn,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_htn,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_htn)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_htn_olink <- soma_to_olink(top_htn)
soma_olink_plots(top_olink = top_htn_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_htn_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_htn)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_htn)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_htn)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_htn)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_htn_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
htn_p = pathway_analysis(de_genes_htn)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/HTN_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = htn_p$cnet,device = "jpeg")
```

# NEURO

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_neuro,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_neuro,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_neuro)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_neuro_olink <- soma_to_olink(top_neuro)
soma_olink_plots(top_olink = top_neuro_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_neuro_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_neuro)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_neuro)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_neuro)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_neuro)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_neuro_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/NEURO_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
neuro_p = pathway_analysis(de_genes_neuro)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/NEURO_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = neuro_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/NEURO_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = neuro_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/NEURO_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = neuro_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/NEURO_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = neuro_p$cnet,device = "jpeg")
```

# RETINO

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_retino,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_retino,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_retino)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_retino_olink <- soma_to_olink(top_retino)
soma_olink_plots(top_olink = top_retino_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_retino_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_retino)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_retino)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_retino)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_retino)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_retino_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RETINO_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
retino_p = pathway_analysis(de_genes_retino)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RETINO_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = retino_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RETINO_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = retino_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RETINO_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = retino_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/RETINO_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = retino_p$cnet,device = "jpeg")
```

# GLYC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_glyc,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_glyc,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_glyc)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_glyc_olink <- soma_to_olink(top_glyc)
soma_olink_plots(top_olink = top_glyc_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_glyc_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_glyc)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_glyc)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_glyc)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_glyc)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_glyc_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
glyc_p = pathway_analysis(de_genes_glyc)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_p$cnet,device = "jpeg")
```

# GLYC with A1c

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_glyc_a1c,analytes$AptName)])` (associated with genes `r toString(analytes$Target[match(top_glyc_a1c,analytes$AptName)])`, respectively).

## SOMAScan correlation matrix (with UACR)

```{r results='asis'}
cor_and_plot_soma(vars = top_glyc_a1c)
```

## Olink associations with SOMAScan

### Plasma

```{r warning=FALSE}
top_glyc_a1c_olink <- soma_to_olink(top_glyc_a1c)
soma_olink_plots(top_olink = top_glyc_a1c_olink, data = plasma)
```

### Urine

```{r warning=FALSE}
soma_olink_plots(top_olink = top_glyc_a1c_olink, data = urine)
```

## scRNA-seq

### Overall expression

```{r message=FALSE,warning=FALSE}
umap_plot(so, top_glyc_a1c)
```

### Differential expression in the highest expressing cell type

#### Diabetes status

```{r}
de(outcome = "diabetes",ref_group="HC",vars=top_glyc_a1c)
```

#### SGLT2i status

```{r}
de(outcome = "SGLT2i",ref_group="No",vars=top_glyc_a1c)
```

#### UACR

There are only 3 participants with UACR >= 30, so participants were divided into two groups: below ("Low") and above ("High") the median. Median UACR was 6.21.

```{r}
de(outcome = "acr_u_cat",ref_group="Low UACR",vars=top_glyc_a1c)
```

## TODAY pathway analysis

```{r warning=FALSE}
v = volcano(top_glyc_a1c_df,overlaps = 5)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_A1C_TODAY_cox_scaled_baseline_adjusted_volcano_plot_1512x916.jpeg",
       units = "px",width = 1512,height = 916,scale = 2,plot = v,device = "jpeg")
v
```

```{r}
glyc_a1c_p = pathway_analysis(de_genes_glyc_a1c)
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_A1C_TODAY_cox_scaled_baseline_adjusted_dot_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_a1c_p$dot,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_A1C_TODAY_cox_scaled_baseline_adjusted_gsea_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_a1c_p$gseap,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_A1C_TODAY_cox_scaled_baseline_adjusted_ridge_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = glyc_a1c_p$ridge,device = "jpeg")
ggsave("/home/tim/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/Proteomics and DKD/Figures/GLYC_A1C_TODAY_cox_scaled_baseline_adjusted_cnet_plot_1512x916.jpeg",units = "px",
       width = 1512,height = 916,scale = 2,
       plot = retino_p$cnet,device = "jpeg")
```
