---
title: "Proteomics and DKD"
author: "Laura Pyle, Ye Ji Choi & Tim Vigers"
date: "today"
date-format: long
format:
  html:
    toc: true
    toc-depth: 5
    toc-float: true
    code-fold: true
    self-contained: true
    fig-cap-location: top
    page-layout: full
editor: source
execute:
  echo: false
  message: false
---

```{r libraries}
#| include: false
library(readxl)
library(tidyverse)
library(psych)
library(corrplot)
library(ggpmisc)
library(pander)
library(Seurat)
library(clusterProfiler)
library(ReactomePA)
library(enrichplot)
library(ggrepel)
library(patchwork)
library(gtsummary)
library(arsenal)
library(pedbp)
# Print all pander tables
panderOptions("knitr.auto.asis", FALSE)
```

```{r data import}
#| include: false
# source("~/GitHub/CHCO-Code/Petter Bjornstad/Proteomics and DKD/create_today_dkd_analysis_dataset.R")
load("/Volumes/Peds Endo/Petter Bjornstad/Proteomics and DKD/Data_Cleaned/analysis_dataset.RData")
```

```{r functions}
# UMAP function
umap_plot <- function(seurat_object = so, vars) {
  # Get Entrez Gene IDs
  entrez <- analytes$EntrezGeneSymbol[match(vars, analytes$AptName)]
  entrez <- entrez[entrez %in% rownames(so)]
  # Plot each gene
  invisible(lapply(entrez, function(e) {
    p <- FeaturePlot(seurat_object, features = e)
    p <- LabelClusters(plot = p, id = "ident")
    cat("\n")
    print(p)
    cat("\n")
  }))
}
# DE function
de <- function(seurat_object = so, outcome, ref_group, vars) {
  # Get Entrez Gene IDs
  entrez <- analytes$EntrezGeneSymbol[match(vars, analytes$AptName)]
  entrez <- entrez[entrez %in% rownames(seurat_object)]
  entrez <- unique(entrez)
  # Find cells where most highly expressed
  cells <- FetchData(seurat_object, vars = entrez,slot = "counts")
  cells$type <- seurat_object$generaltype
  cell_count = data.frame(table(cells$type))
  colnames(cell_count) = c("Cell.Type","n Cells")
  cells <- cells %>%
    group_by(type) %>%
    summarise(across(where(is.numeric), ~ sum(.x, na.rm = TRUE))) %>%
    column_to_rownames("type")
  max <- lapply(cells, function(c) {
    rownames(cells)[order(c,decreasing = T)[1:2]]
  })
  # DE testing
  de_tables <- lapply(names(max), function(n) {
    # Compare expression in two highest expressing cell types
    c1 = tryCatch(FindMarkers(seurat_object,features = n, ident.1 = ref_group,
                              group.by = outcome, subset.ident = max[[n]][1],
                              logfc.threshold = 0, verbose = F),
                  error=function(err) NULL,warning=function(war) NULL)
    if(!is.null(c1)){
      c1$`Cell Type` = max[[n]][1]
      c1$`Gene` = n
    }
    c2 = tryCatch(FindMarkers(seurat_object,features = n, ident.1 = ref_group,
                              group.by = outcome, subset.ident = max[[n]][2],
                              logfc.threshold = 0, verbose = F),
                  error=function(err) NULL,warning=function(war) NULL)
    if(!is.null(c2)){
      c2$`Cell Type` = max[[n]][2]
      c2$`Gene` = n
    }
    # Merge tables if necessary
    t = do.call(rbind,list(c1,c2))
    return(t)
  })
  # Combine rows and print
  de_table <- data.frame(do.call(rbind, de_tables))
  de_table <- de_table[order(de_table$Gene), ]
  # Plots
  invisible(apply(de_table,1, function(r) {
    g = r["Gene"]
    i = r["Cell.Type"]
    adjp = format.pval(as.numeric(r["p_val_adj"]),eps = 0.001,digits = 3)
    fc = round(as.numeric(r["avg_log2FC"]),3)
    t = bquote(log[2]~FC==~.(fc)~(q==~.(adjp)))
    p <- VlnPlot(seurat_object,
                 features = g, group.by = outcome,
                 idents = i, pt.size = 0.01, log = F
    ) 
    p = p + 
      annotate(geom = 'text', label = deparse(t), 
               x = 1.5, y = max(p$data[,1]), parse=TRUE) + 
      ggtitle(paste(g,"in",i)) + 
      theme(legend.position = "none",axis.title.x = element_blank())
    cat("\n")
    print(p)
    cat("\n")
  }))
  de_table = de_table %>% select(Gene,Cell.Type,avg_log2FC,pct.1,pct.2,p_val,p_val_adj)
  de_table=left_join(de_table,cell_count,by = join_by(Cell.Type))
  rownames(de_table)=NULL
  cat("\n")
  pander(de_table)
  cat("\n")
}
# scRNA correlation with proteomics
scRNA_corr <- function(seurat_object = so, vars) {
  # Get Entrez Gene IDs
  entrez <-
    as.list(analytes$EntrezGeneSymbol[match(vars, analytes$AptName)])
  names(entrez) <- vars
  entrez <- entrez[entrez %in% rownames(seurat_object)]
  entrez <- entrez[!duplicated(entrez)]
  olink <- soma_to_olink(vars)
  # Find cells where most highly expressed
  cells <- FetchData(seurat_object, vars = as.character(entrez))
  cells$type <- seurat_object$generaltype
  cell_counts = data.frame(table(cells$type))
  cell_perc <- cells %>%
    group_by(type) %>%
    summarise(across(where(is.numeric), ~ mean(.x>0, na.rm = TRUE))) %>%
    column_to_rownames("type")
  cells <- cells %>%
    group_by(type) %>%
    summarise(across(where(is.numeric), ~ sum(.x, na.rm = TRUE))) %>%
    column_to_rownames("type")
  max <- lapply(cells, function(c) {
    rownames(cells)[order(c,decreasing = T)[1:2]]
  })
  w = which(unlist(lapply(cell_perc,function(c){any(c>=0.1)})))
  # Plot correlations in first cell type
  invisible(lapply(names(max)[w], function(n) {
    # Expression by cell type and ID
    avg_exp <- subset(seurat_object, idents = max[[n]][1])
    avg_exp <- AverageExpression(avg_exp, group.by = "michigan_id")
    avg_exp <- as.data.frame(avg_exp$RNA)
    avg_exp <- as.data.frame(t(avg_exp[rownames(avg_exp) == n, ]))
    avg_exp$record_id <- sub("_BL", "", rownames(avg_exp))
    # Get IDs
    prot_soma <- names(entrez[which(entrez == n)])
    prot_olink <- olink[[prot_soma]]
    # Plot linear model
    t <- left_join(df[, c("record_id", prot_soma)], avg_exp,
                   by = join_by(record_id)
    )
    p_soma <- ggplot(t, aes_string(x = n, y = prot_soma)) +
      geom_point(alpha = 0.5) +
      stat_poly_line(se = F) +
      stat_poly_eq(
        aes(label = paste(after_stat(eq.label),
                          after_stat(rr.label),
                          after_stat(p.value.label),
                          sep = "*\", \"*"
        ))
      ) +
      xlab(paste0("Mean ", n, " expression in ", max[[n]][1])) +
      theme_bw()
    cat("\n")
    cat("### SOMAScan")
    cat("\n")
    print(p_soma)
    cat("\n")
    if (!is.null(prot_olink)) {
      # Plasma
      t <- left_join(plasma[, c("record_id", prot_olink)], avg_exp,
                     by = join_by(record_id)
      )
      p_plasma <- ggplot(t, aes_string(x = n, y = prot_olink)) +
        geom_point(alpha = 0.5) +
        stat_poly_line(se = F) +
        stat_poly_eq(
          aes(label = paste(after_stat(eq.label),
                            after_stat(rr.label),
                            after_stat(p.value.label),
                            sep = "*\", \"*"
          ))
        ) +
        xlab(paste0("Mean ", n, " expression in ", max[[n]][1])) +
        theme_bw()
      cat("\n")
      cat("### Plasma Olink")
      cat("\n")
      print(p_plasma)
      cat("\n")
      # Urine
      t <- left_join(urine[, c("record_id", prot_olink)], avg_exp,
                     by = join_by(record_id)
      )
      p_urine <- ggplot(t, aes_string(x = n, y = prot_olink)) +
        geom_point(alpha = 0.5) +
        stat_poly_line(se = F) +
        stat_poly_eq(
          aes(label = paste(after_stat(eq.label),
                            after_stat(rr.label),
                            after_stat(p.value.label),
                            sep = "*\", \"*"
          ))
        ) +
        xlab(paste0("Mean ", n, " expression in ", max[[n]][1])) +
        theme_bw()
      cat("\n")
      cat("### Urine Olink")
      cat("\n")
      print(p_urine)
      cat("\n")
    }
  }))
  # Plot correlations in second cell type
  invisible(lapply(names(max)[w], function(n) {
    # Expression by cell type and ID
    avg_exp <- subset(seurat_object, idents = max[[n]][2])
    avg_exp <- AverageExpression(avg_exp, group.by = "michigan_id")
    avg_exp <- as.data.frame(avg_exp$RNA)
    avg_exp <- as.data.frame(t(avg_exp[rownames(avg_exp) == n, ]))
    avg_exp$record_id <- sub("_BL", "", rownames(avg_exp))
    # Get IDs
    prot_soma <- names(entrez[which(entrez == n)])
    prot_olink <- olink[[prot_soma]]
    # Plot linear model
    t <- left_join(df[, c("record_id", prot_soma)], avg_exp,
                   by = join_by(record_id)
    )
    p_soma <- ggplot(t, aes_string(x = n, y = prot_soma)) +
      geom_point(alpha = 0.5) +
      stat_poly_line(se = F) +
      stat_poly_eq(
        aes(label = paste(after_stat(eq.label),
                          after_stat(rr.label),
                          after_stat(p.value.label),
                          sep = "*\", \"*"
        ))
      ) +
      xlab(paste0("Mean ", n, " expression in ", max[[n]][2])) +
      theme_bw()
    cat("\n")
    cat("### SOMAScan")
    cat("\n")
    print(p_soma)
    cat("\n")
    if (!is.null(prot_olink)) {
      # Plasma
      t <- left_join(plasma[, c("record_id", prot_olink)], avg_exp,
                     by = join_by(record_id)
      )
      p_plasma <- ggplot(t, aes_string(x = n, y = prot_olink)) +
        geom_point(alpha = 0.5) +
        stat_poly_line(se = F) +
        stat_poly_eq(
          aes(label = paste(after_stat(eq.label),
                            after_stat(rr.label),
                            after_stat(p.value.label),
                            sep = "*\", \"*"
          ))
        ) +
        xlab(paste0("Mean ", n, " expression in ", max[[n]][2])) +
        theme_bw()
      cat("\n")
      cat("### Plasma Olink")
      cat("\n")
      print(p_plasma)
      cat("\n")
      # Urine
      t <- left_join(urine[, c("record_id", prot_olink)], avg_exp,
                     by = join_by(record_id)
      )
      p_urine <- ggplot(t, aes_string(x = n, y = prot_olink)) +
        geom_point(alpha = 0.5) +
        stat_poly_line(se = F) +
        stat_poly_eq(
          aes(label = paste(after_stat(eq.label),
                            after_stat(rr.label),
                            after_stat(p.value.label),
                            sep = "*\", \"*"
          ))
        ) +
        xlab(paste0("Mean ", n, " expression in ", max[[n]][2])) +
        theme_bw()
      cat("\n")
      cat("### Urine Olink")
      cat("\n")
      print(p_urine)
      cat("\n")
    }
  }))
}
# scRNA correlation with outcomes
scRNA_corr_out <- function(seurat_object = so, vars,outcome) {
  # Get Entrez Gene IDs
  entrez <-
    as.list(analytes$EntrezGeneSymbol[match(vars, analytes$AptName)])
  names(entrez) <- vars
  entrez <- entrez[entrez %in% rownames(seurat_object)]
  entrez <- entrez[!duplicated(entrez)]
  # Find cells where most highly expressed
  cells <- FetchData(seurat_object, vars = as.character(entrez))
  cells$type <- seurat_object$generaltype
  cell_counts = data.frame(table(cells$type))
  cell_perc <- cells %>%
    group_by(type) %>%
    summarise(across(where(is.numeric), ~ mean(.x>0, na.rm = TRUE))) %>%
    column_to_rownames("type")
  cells <- cells %>%
    group_by(type) %>%
    summarise(across(where(is.numeric), ~ sum(.x, na.rm = TRUE))) %>%
    column_to_rownames("type")
  max <- lapply(cells, function(c) {
    rownames(cells)[order(c,decreasing = T)[1:2]]
  })
  w = which(unlist(lapply(cell_perc,function(c){any(c>=0.1)})))
  # Get average expression
  # First cells
  first = sapply(max[w],"[[",1)
  f = lapply(names(first),function(n){
    ctype = as.character(first[n])
    avg_exp <- subset(seurat_object, idents = ctype)
    avg_exp <- AverageExpression(avg_exp, group.by = "michigan_id")
    avg_exp <- as.data.frame(avg_exp$RNA)
    avg_exp <- as.data.frame(t(avg_exp[rownames(avg_exp) == n, ]))
    avg_exp$record_id <- sub("_BL", "", rownames(avg_exp))
    colnames(avg_exp) = c(paste(n,"in",ctype),"record_id")
    data.frame(avg_exp,check.names = F)
  })
  f = purrr::reduce(f,full_join,by = "record_id")
  f = left_join(f,df[,c("record_id",outcome)],by = join_by(record_id))
  x = f %>% dplyr::select(-record_id,-all_of(outcome))
  y = f %>% dplyr::select(all_of(outcome))
  M = corr.test(x,y,method = "spearman")
  # Print nice looking table
  r <- data.frame(M$r)
  emphasize.strong.cells(which(M$p <= 0.05, arr.ind = TRUE))
  cat("\n")
  pander(r, caption = "Correlation in Top Cell Type")
  cat("\n")
  # Second cells
  second = sapply(max[w],"[[",2)
  s = lapply(names(second),function(n){
    ctype = as.character(second[n])
    avg_exp <- subset(seurat_object, idents = ctype)
    avg_exp <- AverageExpression(avg_exp, group.by = "michigan_id")
    avg_exp <- as.data.frame(avg_exp$RNA)
    avg_exp <- as.data.frame(t(avg_exp[rownames(avg_exp) == n, ]))
    avg_exp$record_id <- sub("_BL", "", rownames(avg_exp))
    colnames(avg_exp) = c(paste(n,"in",ctype),"record_id")
    data.frame(avg_exp,check.names = F)
  })
  s = purrr::reduce(s,full_join,by = "record_id")
  s = left_join(s,df[,c("record_id",outcome)],by = join_by(record_id))
  x = s %>% dplyr::select(-record_id,-all_of(outcome))
  y = s %>% dplyr::select(all_of(outcome))
  M = corr.test(x,y,method = "spearman")
  # Print nice looking table
  r <- data.frame(M$r)
  emphasize.strong.cells(which(M$p <= 0.05, arr.ind = TRUE))
  cat("\n")
  pander(r, caption = "Correlation in Second Cell Type")
  cat("\n")
}
# Volcano plot
volcano <- function(data, lab = "EntrezGeneSymbol", xcol = "estimate", ycol = "p.value", 
                    top = 3,
                    xlimit = c(-1, 1), ylimit = c(0, -log10(10e-7)),
                    xlabel = "HR", pCutoff = 0.05, overlaps = 10,
                    log_t = T) {
  data <- as.data.frame(data)
  t <- data[data[, "adj.p.value"] <= pCutoff, ]
  if (log_t) {
    t <- t[order(abs(log(t[, xcol])), decreasing = T), ]
  } else {
    t <- t[order(abs(t[, xcol]), decreasing = T), ]
  }
  data$top <- data[, "AptName"] %in% t[1:top, "AptName"]
  data$logp <- -log10(data[, ycol])
  data$fc <- data[, xcol]
  data$sig <- data[, ycol] <= pCutoff
  p <- ggplot(data = data, aes(x = fc, y = logp, color = sig)) +
    geom_hline(yintercept = -log10(pCutoff), linetype = "dashed") +
    geom_point(size = 2) +
    geom_label_repel(
      data = data[data$top, ], aes(label = EntrezGeneSymbol), color = "black",
      max.overlaps = overlaps
    ) +
    scale_color_manual(values = c("grey", "#3e6dbf")) +
    xlab(xlabel) +
    ylab(bquote(~ -Log[10] ~ italic(P))) +
    theme_bw() +
    theme(legend.position = "none")
  return(p)
}
# Compare soluble omics by SGLT2i status
soluble_comp = function(data = df, vars){
  data = data.frame(data)
  f = as.formula(paste0("sglt2i_ever~",paste0(vars,collapse = "+")))
  labs = analytes$EntrezGeneSymbol[match(vars,analytes$AptName)]
  names(labs) = vars
  t = tableby(f,data = data,numeric.test="kwt",
              numeric.stats=c("Nmiss", "median", "q1q3"))
  summary(t,pfootnote = T,labelTranslations = labs)
}
# GSEA
pathway_analysis <- function(de_genes) {
  # Log transform if necessary
  if(!any(de_genes < 0)){
    de_genes = log(de_genes)
  }
  # Upregulated and downregulated
  up = names(de_genes)[de_genes > 0]
  up = up[!is.na(up)]
  down = names(de_genes)[de_genes < 0]
  down = down[!is.na(down)]
  # Reactome enrichment
  upreact = enrichPathway(up)
  upreact = setReadable(upreact, "org.Hs.eg.db", "ENTREZID")
  downreact = enrichPathway(down)
  downreact = setReadable(downreact, "org.Hs.eg.db", "ENTREZID")
  # Plots
  updot <- dotplot(upreact) + ggtitle("Upregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  upnet <- cnetplot(upreact) + ggtitle("Upregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  downdot <- dotplot(downreact) + ggtitle("Downregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  downnet <- cnetplot(downreact) + ggtitle("Downregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  # Return
  return(list("updot" = updot, "upnet" = upnet,
              "downdot" = downdot, "downnet" = downnet))
}
# Function for comparing two gene lists
# de_gene_list is a named list of multiple gene sets
pathway_comparison <- function(de_gene_list) {
  # Log transform if necessary
  de_gene_list = lapply(de_gene_list,function(l){
    if(!any(l < 0)){
      l = log(l)
    }
    return(l)
  })
  # Upregulated and downregulated
  up = lapply(de_gene_list,function(l){
    return(names(l)[which(l > 0)])
  })
  down = lapply(de_gene_list,function(l){
    return(names(l)[which(l < 0)])
  })
  # Fix names
  up = lapply(up,function(l){
    sub("\\|.*","",l)
  })
  down = lapply(down,function(l){
    sub("\\|.*","",l)
  })
  # Compare with GO enrichment
  upck <- compareCluster(up, fun='enrichPathway')
  downck <- compareCluster(down, fun='enrichPathway')
  # Plots
  updot <- dotplot(upck) + ggtitle("Upregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  upnet <- cnetplot(upck) + ggtitle("Upregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  downdot <- dotplot(downck) + ggtitle("Downregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  downnet <- cnetplot(downck) + ggtitle("Downregulated") + 
    theme(plot.title = element_text(hjust = 0.5))
  # Return
  return(list("updot" = updot, "upnet" = upnet,
              "downdot" = downdot, "downnet" = downnet))
}
```

# Participant characteristics

```{r message = FALSE}
df %>%
  mutate(
    race_ethnicity_condensed = case_when(
      race == "White" & startsWith(ethnicity, "Not") ~ "Non-Hispanic White",
      race == "Black or African American" & startsWith(ethnicity, "Not") ~ 
        "Non-Hispanic Black",
      ethnicity == "Hispanic or Latino" ~ "Hispanic",
      T ~ "Other"
    ),
    soma = case_when(record_id %in% unique(sub("_BL|_12M", "", soma$record_id)) ~ 
                       "Yes", T ~ "No"),
    olink_plasma = case_when(record_id %in% unique(sub("_BL|_12M", "", 
                                                       olink_plasma$record_id)) ~ 
                               "Yes", T ~ "No"),
    olink_urine = case_when(record_id %in% unique(sub("_BL|_12M", "", 
                                                      olink_urine$record_id)) ~ 
                              "Yes", T ~ "No"),
    scRNA = case_when(record_id %in% unique(sub("_BL|_12M", "", 
                                                so@meta.data[["michigan_id"]])) ~ 
                        "Yes", T ~ "No")
  ) %>%
  select(
    group, age, sex, race_ethnicity_condensed, diabetes_duration, hba1c, bmi, acr_u, sbp, dbp, map, eGFR_fas_cr,gfr_bsa_plasma,gfr_raw_plasma,elevated_albuminuria, htn,soma, olink_plasma, olink_urine, scRNA
  ) %>%
  tbl_summary(by = group,
              statistic = list(all_continuous() ~ "{mean} ({sd})",
                               acr_u ~ "{median}[{p25},{p75}]"),
              missing_text = "Missing")
```

# scRNA-seq

## MAC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MAC were: `r toString(analytes$TargetFullName[match(top_mac,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_mac,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_mac)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_mac)
```

#### UACR

```{r results='asis', dpi=1200}
de(outcome = "elevated_uacr", ref_group = "UACR < 30", vars = top_mac)
```

## MIC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_mic,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_mic,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_mic)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_mic)
```

#### UACR

```{r results='asis', dpi=1200}
de(outcome = "elevated_uacr", ref_group = "UACR < 30", vars = top_mic)
```

## HYP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension were: `r toString(analytes$TargetFullName[match(top_hyp,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_hyp,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_hyp)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_hyp)
```

#### eGFR

```{r results='asis', dpi=1200}
de(outcome = "hyp", ref_group = "eGFR < 135", vars = top_hyp)
```

## RAPID

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for rapid eGFR decline were: `r toString(analytes$TargetFullName[match(top_rapid,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_rapid,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_rapid)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_rapid)
```

## HTN

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension were: `r toString(analytes$TargetFullName[match(top_htn,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_htn,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_htn)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_htn)
```

#### Hypertension

```{r results='asis', dpi=1200}
de(outcome = "htn", ref_group = "HTN-", vars = top_htn)
```

## HTN with SBP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension (with SBP) were: `r toString(analytes$TargetFullName[match(top_htn_sbp,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_htn_sbp,analytes$AptName)])`, respectively).

### Overall expression

```{r message=FALSE,warning=FALSE,dpi=1200}
umap_plot(so, top_htn_sbp)
```

### Differential expression in the highest expressing cell type

#### SGLT2i status

```{r results='asis', dpi=1200}
de(outcome = "SGLT2i", ref_group = "SGLT2i-", vars = top_htn_sbp)
```

#### Hypertension

```{r results='asis', dpi=1200}
de(outcome = "htn", ref_group = "HTN-", vars = top_htn_sbp)
```

# Soluble omics in SGLT2i+ vs. SGLT2i-

## MAC

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_mac)
```

## MIC

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_mic)
```

## HYP

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_hyp)
```

## RAPID

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_rapid)
```

## HTN

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_htn)
```

## HTN with SBP

```{r message=FALSE,warning=FALSE,dpi=1200,results='asis'}
soluble_comp(vars = top_htn_sbp)
```

# TODAY pathway analysis

## MAC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MAC were: `r toString(analytes$TargetFullName[match(top_mac,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_mac,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_mac <- volcano(top_mac_df)
v_mac <- v_mac +
  ggtitle("Severe Albuminuria") +
  theme(plot.title = element_text(hjust = 0.5))
v_mac
```

```{r message=FALSE,warning=FALSE,dpi=1200}
mac_p <- pathway_analysis(de_genes_mac)
mac_p$updot
mac_p$downdot

mac_plot = mac_p$updot + mac_p$downdot + 
  plot_annotation(title = "Severe Albuminuria",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/MAC_ORA_1600x900.jpeg",plot = mac_plot,
       width = 1600,height = 900,units = "px",scale = 2.5)

mac_p_comp <- pathway_comparison(list(
  Baseline = de_genes_mac,
  Followup = de_genes_mac_10
))
mac_p_comp$updot
mac_p_comp$downdot

mac_comp_plot = mac_p_comp$updot + mac_p_comp$downdot + 
  plot_annotation(title = "Severe Albuminuria",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/MAC_ORA_comparison_1600x900.jpeg",
       plot = mac_comp_plot,width = 1600,height = 900,units = "px",scale = 2.5)
```

## MIC

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for MIC were: `r toString(analytes$TargetFullName[match(top_mic,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_mic,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_mic <- volcano(top_mic_df)
v_mic <- v_mic +
  ggtitle("Elevated Albuminuria") +
  theme(plot.title = element_text(hjust = 0.5))
v_mic
```

```{r message=FALSE,warning=FALSE,dpi=1200}
mic_p <- pathway_analysis(de_genes_mic)
mic_p$updot
mic_p$downdot

mic_plot = mic_p$updot + mic_p$downdot + 
  plot_annotation(title = "Elevated Albuminuria",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/MIC_ORA_1600x900.jpeg",plot = mic_plot,
       width = 1600,height = 900,units = "px",scale = 2.5)

mic_p_comp <- pathway_comparison(list(
  Baseline = de_genes_mic,
  Followup = de_genes_mic_10
))
mic_p_comp$updot
mic_p_comp$downdot

mic_comp_plot = mic_p_comp$updot + mic_p_comp$downdot + 
  plot_annotation(title = "Elevated Albuminuria",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/MIC_ORA_comparison_1600x900.jpeg",
       plot = mic_comp_plot,width = 1600,height = 900,units = "px",scale = 2.5)
```

## HYP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension were: `r toString(analytes$TargetFullName[match(top_hyp,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_hyp,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_hyp <- volcano(top_hyp_df)
v_hyp <- v_hyp +
  ggtitle("Hyperfiltration") +
  theme(plot.title = element_text(hjust = 0.5))
v_hyp
```

```{r message=FALSE,warning=FALSE,dpi=1200}
hyp_p <- pathway_analysis(de_genes_hyp)
hyp_p$updot
hyp_p$downdot

hyp_plot = hyp_p$updot + hyp_p$downdot + 
  plot_annotation(title = "Hyperfiltration",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/HYP_ORA_1600x900.jpeg",plot = hyp_plot,
       width = 1600,height = 900,units = "px",scale = 3)

hyp_p_comp <- pathway_comparison(list(
  Baseline = de_genes_hyp,
  Followup = de_genes_hyp_10
))
hyp_p_comp$updot
hyp_p_comp$downdot

hyp_comp_plot = hyp_p_comp$updot + hyp_p_comp$downdot + 
  plot_annotation(title = "Hyperfiltration",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/HYP_ORA_comparison_1600x900.jpeg",
       plot = hyp_comp_plot,width = 1600,height = 900,units = "px",scale = 2.5)
```

## RAPID

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for rapid eGFR decline were: `r toString(analytes$TargetFullName[match(top_rapid,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_rapid,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_rapid <- volcano(top_rapid_df)
v_rapid <- v_rapid +
  ggtitle("Rapid eGFR Decline") +
  theme(plot.title = element_text(hjust = 0.5))
v_rapid
```

```{r message=FALSE,warning=FALSE,dpi=1200}
rapid_p <- pathway_analysis(de_genes_rapid)
rapid_p$updot
rapid_p$downdot

rapid_plot = rapid_p$updot + rapid_p$downdot + 
  plot_annotation(title = "Rapid eGFR decline",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/RAPID_ORA_1600x900.jpeg",plot = rapid_plot,
       width = 1600,height = 900,units = "px",scale = 2.8)

rapid_p_comp <- pathway_comparison(list(
  Baseline = de_genes_rapid,
  Followup = de_genes_rapid_10
))
rapid_p_comp$updot
rapid_p_comp$downdot

rapid_comp_plot = rapid_p_comp$updot + rapid_p_comp$downdot + 
  plot_annotation(title = "Rapid eGFR decline",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/RAPID_ORA_comparison_1600x900.jpeg",
       plot = rapid_comp_plot,width = 1600,height = 900,units = "px",scale = 2.5)
```

## HTN

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension were: `r toString(analytes$TargetFullName[match(top_htn,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_htn,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_htn <- volcano(top_htn_df)
v_htn <- v_htn +
  ggtitle("Hypertension") +
  theme(plot.title = element_text(hjust = 0.5))
v_htn
```

```{r message=FALSE,warning=FALSE,dpi=1200}
htn_p <- pathway_analysis(de_genes_htn)
htn_p$updot
htn_p$downdot

htn_plot = htn_p$updot + htn_p$downdot + 
  plot_annotation(title = "Hypertension",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/HTN_ORA_1600x900.jpeg",plot = htn_plot,
       width = 1600,height = 900,units = "px",scale = 2.8)

htn_p_comp <- pathway_comparison(list(
  Baseline = de_genes_htn,
  Followup = de_genes_htn_10
))
htn_p_comp$updot
htn_p_comp$downdot

htn_comp_plot = htn_p_comp$updot + htn_p_comp$downdot + 
  plot_annotation(title = "Hypertension",
                  theme = theme(plot.title = element_text(hjust = 0.5)))
ggsave(filename = "/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/HTN_ORA_comparison_1600x900.jpeg",
       plot = htn_comp_plot,width = 1600,height = 900,units = "px",scale = 2.5)
```

## HTN with SBP

The proteins with the largest absolute effect size and adjusted p value $\leq 0.05$ for hypertension (with SBP) were: `r toString(analytes$TargetFullName[match(top_htn_sbp,analytes$AptName)])` (associated with genes `r toString(analytes$EntrezGeneSymbol[match(top_htn_sbp,analytes$AptName)])`, respectively).

```{r message=FALSE,warning=FALSE,dpi=1200}
v_htn_sbp <- volcano(top_htn_sbp_df)
v_htn_sbp <- v_htn_sbp +
  ggtitle("Hypertension (with SBP)") +
  theme(plot.title = element_text(hjust = 0.5))
v_htn_sbp
```

```{r message=FALSE,warning=FALSE,dpi=1200}
htn_sbp_p <- pathway_analysis(de_genes_htn_sbp)
htn_sbp_p$updot
htn_sbp_p$downdot

htn_sbp_p_comp <- pathway_comparison(list(
  Baseline = de_genes_htn_sbp,
  Followup = de_genes_htn_sbp_10
))
htn_sbp_p_comp$updot
htn_sbp_p_comp$downdot
```

# scRNA-seq correlations with circulating proteins

## MAC

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_mac)
```

## MIC

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_mic)
```

## HYP

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_hyp)
```

## RAPID

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_rapid)
```

## HTN

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_htn)
```

## HTN with SBP

```{r message=FALSE,warning=FALSE,dpi=1200}
scRNA_corr(vars = top_htn_sbp)
```

# UACR correlation with TAL NELL-1 transcript.

```{r message=FALSE,warning=FALSE,dpi=1200}
# Pull data
tal <- subset(so, idents = "TAL")
avg_exp <- AverageExpression(tal,features = "NELL1" , group.by = "michigan_id")
avg_exp = as.data.frame(t(avg_exp$RNA))
avg_exp$record_id = sub("*_BL","",rownames(avg_exp))
avg_exp = left_join(avg_exp,df[,c("record_id","acr_u")])
colnames(avg_exp) = c("NELL1","record_id","UACR")
# Correlation
M = corr.test(avg_exp$NELL1,avg_exp$UACR,method = "spearman")
# Plot
ggplot(avg_exp,aes(x=NELL1,y=log(UACR))) + 
  geom_point() + theme_bw() + 
  geom_smooth(method = "lm",se=F)
```

Spearman correlation between UACR and mean NELL1 expression in TAL was `r round(M$r,3)` (p = `r round(M$p,3)`).

# scRNA correlations with continuous outcomes

## MAC correlation with UACR

```{r results='asis'}
scRNA_corr_out(vars = top_mac,outcome = "acr_u")
```

## MIC correlation with UACR

```{r results='asis'}
scRNA_corr_out(vars = top_mic,outcome = "acr_u")
```

## HYP correlation with BSA-indexed GFR

```{r results='asis'}
scRNA_corr_out(vars = top_hyp,outcome = "gfr_bsa_plasma")
```

## HTN correlation with MAP

```{r results='asis'}
scRNA_corr_out(vars = top_htn,outcome = "map")
```

## HTN with SBP correlation with MAP

```{r results='asis'}
scRNA_corr_out(vars = top_htn_sbp,outcome = "map")
```

# Figure 1

```{r message=FALSE,warning=FALSE,dpi=1200}
fig1 <- (v_mic + v_mac + v_rapid) / (v_hyp + v_htn + plot_spacer())
ggsave("/Users/timvigers/Library/CloudStorage/Dropbox/Work/Shared/TODAY/TODAY DKD manuscript [shared]/Figures/Figure_1_1600x900.jpeg",
  units = "px",
  width = 1600, height = 900, scale = 2, plot = fig1, device = "jpeg"
)
```
