---
title: "Sheep miRNA"
author: "Tim Vigers"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE}
library(DESeq2)
library(tximport)
library(tidyverse)
library(ggrepel)
library(knitr)
knitr::opts_chunk$set(echo = TRUE,message = FALSE)
if(Sys.info()["sysname"] == "Windows"){
  home_dir = "B:/Projects"
} else if (Sys.info()["sysname"] == "Linux"){
  home_dir = "~/UCD/PEDS/RI Biostatistics Core/Shared/Shared Projects/Vigers/sheep_miRNA"
} else if (Sys.info()["sysname"] == "Darwin"){
  home_dir = "/Volumes/PEDS/RI Biostatistics Core/Shared/Shared Projects/Vigers/sheep_miRNA"
}
knitr::opts_knit$set(root.dir = home_dir)
rm(home_dir)
```

# Notes
- Following the workflow in this vignette:
  - http://master.bioconductor.org/packages/release/workflows/html/rnaseqGene.html
- Use Salmon to quantify RNA:
  - https://combine-lab.github.io/salmon/
  - Tutorial: https://combine-lab.github.io/salmon/getting_started/
  - rnaseqGene recommends using the --gcBias flag which estimates a correction factor for systematic biases commonly present in RNA-seq data
- Sheep transcriptome can be downloaded from:
  - http://ftp.ensembl.org/pub/release-106/fasta/ovis_aries/cdna/Ovis_aries.Oar_v3.1.cdna.all.fa.gz
  - Browse other downloads with: http://uswest.ensembl.org/info/data/ftp/index.html/
- Sheep genome downloaded from:
  - http://ftp.ensembl.org/pub/release-106/fasta/ovis_aries/dna_index/Ovis_aries.Oar_v3.1.dna.toplevel.fa.gz
- Selective alignment performed according to: https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/
- Copy of data stored on shared drive: `/RI Biostatistics Core/Shared/Shared Projects/Vigers/sheep_miRNA`
  
```{bash eval=FALSE}
# bash script
# Move to base folder
cd /home/tim/UCD/PEDS/RI\ Biostatistics\ Core/Shared/Shared\ Projects/Vigers/sheep_miRNA
# List of genome targets (decoys)
grep "^>" <(gunzip -c data_clean/Ovis_aries.Oar_v3.1.dna.toplevel.fa.gz) | cut -d " " -f 1 >data_clean/decoys.txt
sed -i.bak -e 's/>//g' data_clean/decoys.txt
cat data_clean/Ovis_aries.Oar_v3.1.cdna.all.fa.gz data_clean/Ovis_aries.Oar_v3.1.dna.toplevel.fa.gz >data_clean/gentrome.fa.gz
# Index the sheep transcriptome
salmon index -t data_clean/gentrome.fa.gz -d data_clean/decoys.txt -p 24 -i data_clean/ovis_index
# Cut adapters
for file in data_clean/FASTQ/*; do
    samp=$(basename ${file})
    echo "Cutting adapters from sample ${samp}"
    cutadapt -j 24 -a TGGAATTCTCGGGTGCCAAGG data_clean/FASTQ/"$samp" -o data_clean/FASTQ_Trimmed/"$samp"
done
# Quantify reads
for fn in $(find data_clean/FASTQ_Trimmed -name "*_R1_001.fastq.gz" -type f | sed 's/_R._001\..*//' | sort | uniq); do
    samp=$(basename ${fn})
    echo "Processing sample ${samp}"
    salmon quant -i data_clean/ovis_index -l A \
        -1 ${fn}_R1_001.fastq.gz \
        -2 ${fn}_R2_001.fastq.gz \
        -p 24 --validateMappings --gcBias -o data_clean/quants/${samp}_quant
done
```

```{r}
# Sample info
samples = read.csv("./data_clean/sheep_samples.csv",na.strings = c("","NA"))
# Remove those with no data
missing = samples$Sample[is.na(samples$Path)]
samples = samples[!is.na(samples$Path),]
# Data files
samples$Path = paste0("./data_clean/quants/",samples$Path,"/quant.sf")
```

# Sample Information

```{r echo=FALSE}
kable(samples[,1:4])
```

# Analysis

```{r}
# Import to DESeq2
txi <- tximport(samples$Path, type="salmon", txOut=TRUE)
ddsTxi <- DESeqDataSetFromTximport(txi,colData = samples,design = ~ 1)
# Variance stabilizing transformation
vsd <- varianceStabilizingTransformation(ddsTxi)
# Analysis
dds = DESeq(ddsTxi)
```

## PCA by treatment

```{r warning=FALSE,fig.width=8}
plotPCA(vsd,intgroup = "Sample",ntop = nrow(vsd)) + 
  theme_bw() + theme(legend.position = "none")
```

Sample 3D-RFS-8W1 appears to be very different from the other samples based on PCA.

## Comparisons between groups

### 30 Day Samples

#### 8mm wounds treated w PBS vs. 8mm wounds treated w CNP-miR146a

```{r}
# New model design, exclude 2mm wounds
samples$Days = factor(samples$Days)
samples$Treatment = factor(samples$Treatment)
txi <- tximport(samples$Path[samples$Days == "30"], type="salmon", txOut=TRUE)
# New model design matrix
ddsTxi <- DESeqDataSetFromTximport(txi,colData = samples[samples$Days == "30",],
                                   design = ~ Treatment)
dds = DESeq(ddsTxi)
# Contrast
con = c("Treatment","PBS","CNP_miR146a")
res = results(dds,contrast = con,tidy = T)
res = res[res$pvalue < 0.05 & !is.na(res$pvalue),]
kable(res,row.names = F,digits = 3)
```

No transcripts were significantly different between groups after adjusting for multiple comparisons. `r nrow(res)` transcripts were significant at a nominal p < 0.05 level.

### 8mm wounds 

#### 0 day wounds vs. 3 day wounds

```{r}
# New model design, exclude 2mm wounds
txi <- tximport(samples$Path[samples$Size == "8mm"], type="salmon", txOut=TRUE)
ddsTxi <- DESeqDataSetFromTximport(txi,colData = samples[samples$Size == "8mm",],
                                   design = ~ Days)
dds = DESeq(ddsTxi)
# Contrast
con = c("Days","0","3")
res = results(dds,tidy = T)
res = res[res$pvalue < 0.05 & !is.na(res$pvalue),]
kable(res,row.names = F,digits = 3)
```

No transcripts were significantly different between groups after adjusting for multiple comparisons. `r nrow(res)` transcripts were significant at a nominal p < 0.05 level.

#### 3 day wounds vs 30 day wounds

```{r}
con = c("Days","3","30")
res = results(dds,contrast = con,tidy = T)
res = res[res$pvalue < 0.05 & !is.na(res$pvalue),]
kable(res,row.names = F,digits = 3)
```

No transcripts were significantly different between groups after adjusting for multiple comparisons. `r nrow(res)` transcripts were significant at a nominal p < 0.05 level.

#### 0 day wounds vs 30 day wounds

```{r}
con = c("Days","0","30")
res = results(dds,contrast = con,tidy = T)
res = res[res$pvalue < 0.05 & !is.na(res$pvalue),]
kable(res,row.names = F,digits = 3)
```

No transcripts were significantly different between groups after adjusting for multiple comparisons. `r nrow(res)` transcripts were significant at a nominal p < 0.05 level.

