---
title: "Kidney scRNA for Jenny Kanter"
author: "Hailey Hampson"
date: "2024-10-09"
output: html_document
  word_document: default
  html_document: default
  pdf_document: default
---
# 1. Set up Libraries & Directores
```{r, include=F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
library(GSEABase)
library(limma)
library(reshape2)
library(data.table)
library(knitr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
#library(NMF)
library(rsvd)
library(RColorBrewer)
library(MAST)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
#library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
registerDoParallel(cores = 6)
library(VennDiagram)
library(janitor)
devtools::install_github('immunogenomics/presto')
library(presto)
library(knitr)

#Local file path
#dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
#dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
#dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
#dir.results <- c("/Volumes/Peds Endo/Petter Bjornstad/Kidney Project/Results")

# #Lambda file path
#  dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
#  dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Kidney scRNA/Kidney scRNA")
#  dir.results <- c(fs::path(dir.dat,"Kidney Project/Results"))

#Mac Studio File Path
dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Kidney scRNAseq Project/Results")
dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Load functions
source("Kidney_functions_sc.R")
```

# 2. Load Full Data & Clean: scRNA & MetaData
## a. Format Data
```{r, include=F}
#Local
# dir.dat <- c("/Users/hhampson/Dropbox/Bjornstad data")
# so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
# load("/Volumes/Peds Endo/Petter Bjornstad/Data Harmonization/Data Exports/PB90_clinical_metadata.RData")
# so_kidney_sc <- AddMetaData(so_kidney_sc, so_meta_combined)
# rm(so_meta_combined)
# gc()

#Lambda
so_kidney_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="KI-0014643")] <- "KL-0014643"
so_kidney_sc$kit_id[which(so_kidney_sc$kit_id=="kl-0023998")] <- "KL-0023998"
#check2 <- so_kidney_sc@meta.data

#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv")) #small meta set
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata.csv")) #all metadata vars
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))
#Get coenrolled IDs
coenroll <- read.csv(fs::path(dir.dat,"Renal HERITAGE/Data_Cleaned/coenrolled_ids.csv"))

#Which IDs are coenrolled 
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")]
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")]

#"KL-0014634" "KL-0014629" "KL-0014628" coenrolled IDs
#"KL-0014634" "KL-0014629" "KL-0014628" having missing meta data 

#KL-0014628: IT_10/RH-66-T KL-0014628 KL-0019101
#KL-0014629: IT_09/RH-65-T
#KL-0014634: IT_07/RH-59-T
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group")))

#Partipant 30051 (kit id KL-0027474 at 4m has sc, but at baseline no sc has kit id KL-0027474 metadata)
#missing <- c("KL-0014628", "KL-0014629", "KL-0014634", "KL-0019059", "KL-0019092", "KL-0019101", "KL-0022159", "KL-0024005", "KL-0027470", #"KL-0027478")
#Check these ids in metadat
#check <- harm_meta_data %>%
#filter(kit_id %in% missing)

#IDs missing group information  
#Filter out attempt participants
harm_meta_data <- harm_meta_data %>% 
  dplyr::select(-X)
#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>% 
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>% 
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>% 
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
#mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups: 
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
gc()
harm_meta_data <- harm_meta_data %>% 
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
gc()
meta_kidney_sc <-  so_kidney_sc@meta.data 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
meta_kidney_sc <- meta_kidney_sc %>%
  #dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(harm_meta_data,by="kit_id")
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
#Check to see if all ids still have metadata
#check4 <- meta_kidney_sc %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check4 <- check4 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))

#rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
ids <- harm_meta_data$kit_id
#Filter seurat object to only IDs that have the metadata (83 individuals of interest)
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)
#check5 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check5 <- check5 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
so_kidney_sc <- subset(so_kidney_sc, kit_id %in% ids)
#check6 <- so_kidney_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check6 <- check6 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
#unique(so_kidney_sc$group)
#unique(so_kidney_sc$KPMP_celltype)
rm(meta_kidney_sc,harm_meta_data)

#Switch default assay in seurat object to RNA
DefaultAssay(so_kidney_sc) <- "RNA"
gc()

check3 <- so_kidney_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","celltype_rpca","medication")))
# rm(check3)
unique(so_kidney_sc$co_enroll_id) 
#"RH-66-T" "RH-65-T" "RH-59-T" "IT2D-08" "RH-98-T" "RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T"
#"RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T" IDs coenrolled in RH and RH2 as a baseline and followup - filter out their RH2 study visit?
#RH-67-T/RH2-19 kit id for second visit to filter out: KL-0030621
#RH-23/RH2-14 kit id for second visit to filter out: KL-0029535

# #Check if coenrolled IDs have multiple biopsies
# co1 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")] #Coenrolled in RH&RH2
# co2 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH/RH2
# co3 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH2
# co4 <- coenroll$rh_id[which(coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenroleld in Improve/RH
# co5 <- coenroll$crc_id[which(coenroll$crc_id!="" & coenroll$panda_id!="")] #Coenrolled in Crocodile and panda
# co6 <- coenroll$panda_id[which(coenroll$panda_id!="" & coenroll$crc_id!="")]
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co1)]) #"RH-68-T" "RH-23-T" "RH-60-T" "RH-67-T" "RH-93-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co2)]) #"RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co3)]) #"RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co4)]) #"RH-66-T" "RH-65-T" "RH-59-T" "RH-60-T"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co5)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# unique(so_kidney_sc$record_id[which(so_kidney_sc$record_id %in% co6)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# 
# rm(coenroll,check3)


so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0030621")
so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0029535")
check3 <- so_kidney_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","celltype_rpca","medication")))
length(unique(check3$kit_id)) #81
rm(check3)
rm(coenroll)

#Check the total number of cells & genes in the full dataset
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #31332 genes

dim(so_kidney_sc@assays$RNA@counts) #31332 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #101

# #Filter out mitochondrial DNA
# gene_expression_matrix <- so_kidney_sc@assays$RNA@counts
# 
# # Filter genes expressed in more than 10% of cells
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# so_kidney_sc@assays$RNA@counts <- gene_expression_matrix[expressed_genes,]
# 
# dim(so_kidney_sc@assays$RNA@counts) #5759 186125 
# sum(grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))) #45
# 
# ncol(so_kidney_sc) #186125 cells
# nrow(so_kidney_sc) #31332 genes
# 
# # Remove mitochondrial genes (prefix "MT")
# so_kidney_sc@assays$RNA@counts <- so_kidney_sc@assays$RNA@counts[grep("^MT", rownames(so_kidney_sc@assays$RNA@counts), invert = TRUE),]
# ncol(so_kidney_sc) #186125 cells
# nrow(so_kidney_sc) #31332 genes
# dim(so_kidney_sc@assays$RNA@counts) #5714 186125
# sum(grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))) #0

# Step 1: Filter to genes expressed in more than 5% of cells
gene_expression_matrix <- so_kidney_sc@assays$RNA@counts
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05

so_kidney_sc <- subset(so_kidney_sc, features = rownames(gene_expression_matrix)[expressed_genes])
dim(so_kidney_sc@assays$RNA@counts) #9289 186125
dim(so_kidney_sc@assays$RNA) #9289 186125
dim(so_kidney_sc) #9289 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #70
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #9289 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_kidney_sc), value = TRUE)
#keep_ids <- unique(rownames(so_kidney_sc)[which(!rownames(so_kidney_sc) %in% mito_genes)])
# so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc), mito_genes))
# so_kidney_sc <- subset(so_kidney_sc,kit_id!="KL-0029535")
#so_kidney_sc$Gene <- rownames(so_kidney_sc)
#so_kidney_sc <- subset(so_kidney_sc, Gene %in% keep_ids)
so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc@assays$RNA@counts), mito_genes))
#so_kidney_sc <- subset(so_kidney_sc, features = setdiff(rownames(so_kidney_sc@assays$RNA@counts), mito_genes))
# so_kidney_sc <- subset(so_kidney_sc, !rownames(so_kidney_sc) %in% mito_genes)
# grep("^MT-", rownames(so_kidney_sc@assays$RNA@counts), value = TRUE)
dim(so_kidney_sc@assays$RNA@counts) #9276 186125
dim(so_kidney_sc@assays$RNA@data) #9276 186125
dim(so_kidney_sc@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts))) #0
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #9276 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_kidney_sc@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_kidney_sc@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_kidney_sc@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_kidney_sc <- so_kidney_sc[, cells_to_keep]

# Check dimensions after filtering
dim(so_kidney_sc@assays$RNA@counts) #9276 81882
dim(so_kidney_sc@assays$RNA)#9276 81882
dim(so_kidney_sc)#9276 81882

# Optionally, you can also check how many MT genes remain in the filtered dataset
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@counts)))
ncol(so_kidney_sc)  # Number of cells remaining


# Step 3: Renormalize the Seurat object after filtering
so_kidney_sc <- NormalizeData(so_kidney_sc)

#Check normalized slot
dim(so_kidney_sc@assays$RNA@data) #5714 186125
sum(grepl("^MT-", rownames(so_kidney_sc@assays$RNA@data))) #70
ncol(so_kidney_sc) #186125 cells
nrow(so_kidney_sc) #5714 genes

#Load in med data to update medication information
med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))
med <- med %>% 
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
meta_kidney_sc <-  so_kidney_sc@meta.data
med <- med %>% 
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain
med <- med %>%  
  filter(record_id %in% meta_kidney_sc$record_id) #81 remain
# med$mrn <- as.numeric(med$mrn)
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)
med$mrn <- as.numeric(med$mrn)
meta_kidney_sc <- meta_kidney_sc %>%
  left_join(med,by=c("mrn","record_id"))
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)

#Add Med Meta Data to Seurat object
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)
rm(gene_expression_matrix,med,meta_kidney_sc)
# saveRDS(so_kidney_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds"))
```


## b. Format KPMP Data
```{r, include=F}
#Load KPMP Cell Types 2 (new)
#Mac Studio
so_kpmp_sc <- readRDS("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/scRNA/data_raw/PB_90_RPCAFix_Metadata_LR_RM_kpmpV1labelled.rds")

so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="KI-0014643")] <- "KL-0014643"
so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="kl-0023998")] <- "KL-0023998"
#check2 <- so_kpmp_sc@meta.data

#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv")) #small meta set
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata.csv")) #all metadata vars
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))
#Get coenrolled IDs
coenroll <- read.csv(fs::path(dir.dat,"Renal HERITAGE/Data_Cleaned/coenrolled_ids.csv"))

#Which IDs are coenrolled 
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")]
coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")]

#"KL-0014634" "KL-0014629" "KL-0014628" coenrolled IDs
#"KL-0014634" "KL-0014629" "KL-0014628" having missing meta data 

#KL-0014628: IT_10/RH-66-T KL-0014628 KL-0019101
#KL-0014629: IT_09/RH-65-T
#KL-0014634: IT_07/RH-59-T
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group")))

#Partipant 30051 (kit id KL-0027474 at 4m has sc, but at baseline no sc has kit id KL-0027474 metadata)
#missing <- c("KL-0014628", "KL-0014629", "KL-0014634", "KL-0019059", "KL-0019092", "KL-0019101", "KL-0022159", "KL-0024005", "KL-0027470", #"KL-0027478")
#Check these ids in metadat
#check <- harm_meta_data %>%
#filter(kit_id %in% missing)

#IDs missing group information  
#Filter out attempt participants
harm_meta_data <- harm_meta_data %>% 
  dplyr::select(-X)
#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>% 
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>% 
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>% 
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
#mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups: 
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
gc()
harm_meta_data <- harm_meta_data %>% 
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
gc()
meta_kidney_sc <-  so_kpmp_sc@meta.data 
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
meta_kidney_sc <- meta_kidney_sc %>%
  #dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(harm_meta_data,by="kit_id")
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
#Check to see if all ids still have metadata
#check4 <- meta_kidney_sc %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check4 <- check4 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))

#rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
ids <- harm_meta_data$kit_id
#Filter seurat object to only IDs that have the metadata (83 individuals of interest)
so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc)
#check5 <- so_kpmp_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check5 <- check5 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
so_kpmp_sc <- subset(so_kpmp_sc, kit_id %in% ids)
#check6 <- so_kpmp_sc@meta.data %>%
#  filter(kit_id=="KL-0014628" | kit_id=="KL-0014629" | kit_id=="KL-0014634")
#check6 <- check6 %>%
#  dplyr::select(all_of(c("kit_id","record_id.x","record_id.y","co_enroll_id","group","KPMP_celltype")))
#unique(so_kpmp_sc$group)
#unique(so_kpmp_sc$KPMP_celltype)
rm(meta_kidney_sc,harm_meta_data)

#Switch default assay in seurat object to RNA
DefaultAssay(so_kpmp_sc) <- "RNA"
gc()

check3 <- so_kpmp_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","KPMP_celltype","medication")))
# rm(check3)
unique(so_kpmp_sc$co_enroll_id) 
length(unique(check3$kit_id))
#"RH-66-T" "RH-65-T" "RH-59-T" "IT2D-08" "RH-98-T" "RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T"
#"RH-99-T" "RH-16-T" "RH-23-T" "RH-19-O" "RH-96-T" "RH-67-T" IDs coenrolled in RH and RH2 as a baseline and followup - filter out their RH2 study visit?
#RH-67-T/RH2-19 kit id for second visit to filter out: KL-0030621
#RH-23/RH2-14 kit id for second visit to filter out: KL-0029535

# #Check if coenrolled IDs have multiple biopsies
# co1 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="")] #Coenrolled in RH&RH2
# co2 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH/RH2
# co3 <- coenroll$rh_id[which(coenroll$rh2_id!="" & coenroll$improve_id!="")] #Coenrolled in Improve/RH2
# co4 <- coenroll$rh_id[which(coenroll$rh_id!="" & coenroll$improve_id!="")] #Coenroleld in Improve/RH
# co5 <- coenroll$crc_id[which(coenroll$crc_id!="" & coenroll$panda_id!="")] #Coenrolled in Crocodile and panda
# co6 <- coenroll$panda_id[which(coenroll$panda_id!="" & coenroll$crc_id!="")]
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co1)]) #"RH-68-T" "RH-23-T" "RH-60-T" "RH-67-T" "RH-93-T"
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co2)]) #"RH-60-T"
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co3)]) #"RH-60-T"
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co4)]) #"RH-66-T" "RH-65-T" "RH-59-T" "RH-60-T"
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co5)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# unique(so_kpmp_sc$record_id[which(so_kpmp_sc$record_id %in% co6)]) #"CRC-21" "CRC-04" "CRC-09" "CRC-24" "CRC-18" "CRC-07" "CRC-38" "CRC-32" "CRC-36" "CRC-44" "CRC-43" "CRC-16" "CRC-42" "CRC-31" "CRC-48" "CRC-59"
# 
# rm(coenroll,check3)


so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0030621")
so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0029535")
check3 <- so_kpmp_sc@meta.data
check3 <- check3 %>%
  dplyr::select(all_of(c("kit_id","record_id","co_enroll_id","group","KPMP_celltype","medication")))
length(unique(check3$kit_id)) #81
rm(check3)
rm(coenroll)

#Check the total number of cells & genes in the full dataset
ncol(so_kpmp_sc) #186125 cells
nrow(so_kpmp_sc) #31332 genes

dim(so_kpmp_sc@assays$RNA@layers$counts) #31332 186125
sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA))) #101

# #Filter out mitochondrial DNA
# gene_expression_matrix <- so_kpmp_sc@assays$RNA@counts
# 
# # Filter genes expressed in more than 10% of cells
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# so_kpmp_sc@assays$RNA@counts <- gene_expression_matrix[expressed_genes,]
# 
# dim(so_kpmp_sc@assays$RNA@counts) #5759 186125 
# sum(grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts))) #45
# 
# ncol(so_kpmp_sc) #186125 cells
# nrow(so_kpmp_sc) #31332 genes
# 
# # Remove mitochondrial genes (prefix "MT")
# so_kpmp_sc@assays$RNA@counts <- so_kpmp_sc@assays$RNA@counts[grep("^MT", rownames(so_kpmp_sc@assays$RNA@counts), invert = TRUE),]
# ncol(so_kpmp_sc) #186125 cells
# nrow(so_kpmp_sc) #31332 genes
# dim(so_kpmp_sc@assays$RNA@counts) #5714 186125
# sum(grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts))) #0

# Step 1: Filter to genes expressed in more than 5% of cells
gene_expression_matrix <- so_kpmp_sc@assays$RNA@layers$counts
rownames(gene_expression_matrix) <- rownames(so_kpmp_sc@assays$RNA)
colnames(gene_expression_matrix) <- colnames(so_kpmp_sc@assays$RNA)
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05
names(expressed_genes)
rownames(gene_expression_matrix)[expressed_genes]

so_kpmp_sc <- subset(so_kpmp_sc, features = rownames(gene_expression_matrix)[expressed_genes])
dim(so_kpmp_sc@assays$RNA@layers$counts) #9289 186125
dim(so_kpmp_sc@assays$RNA) #9289 186125
dim(so_kpmp_sc) #9289 186125
sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA))) #70
ncol(so_kpmp_sc) #186125 cells
nrow(so_kpmp_sc) #9289 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_kpmp_sc@assays$RNA), value = TRUE)
so_kpmp_sc <- subset(so_kpmp_sc, features = setdiff(rownames(so_kpmp_sc), mito_genes))
dim(so_kpmp_sc@assays$RNA@layers$counts) #9276 186125
dim(so_kpmp_sc@assays$RNA@layers$data) #9276 186125
dim(so_kpmp_sc@assays$RNA)#5714 186125
sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$counts))) #0
ncol(so_kpmp_sc) #186125 cells
nrow(so_kpmp_sc) #9276 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_kpmp_sc@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_kpmp_sc@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_kpmp_sc@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_kpmp_sc <- so_kpmp_sc[, cells_to_keep]

# Check dimensions after filtering
dim(so_kpmp_sc@assays$RNA@layers$counts) #9289 81882
dim(so_kpmp_sc@assays$RNA)#9289 81882
dim(so_kpmp_sc)#9289 81882

# Optionally, you can also check how many MT genes remain in the filtered dataset
sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$counts)))
ncol(so_kpmp_sc)  # Number of cells remaining


# Step 3: Renormalize the Seurat object after filtering
so_kpmp_sc <- NormalizeData(so_kpmp_sc)

#Check normalized slot
dim(so_kpmp_sc@assays$RNA@layers$data) #5714 186125
sum(grepl("^MT-", rownames(so_kpmp_sc@assays$RNA@layers$data))) #70
ncol(so_kpmp_sc) #186125 cells
nrow(so_kpmp_sc) #5714 genes

#Load in med data to update medication information
med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))
med <- med %>% 
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
meta_kidney_sc <-  so_kpmp_sc@meta.data
med <- med %>% 
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain
med <- med %>%  
  filter(record_id %in% meta_kidney_sc$record_id) #81 remain
# med$mrn <- as.numeric(med$mrn)
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
med$mrn <- as.numeric(med$mrn)
meta_kidney_sc <- meta_kidney_sc %>%
  left_join(med,by=c("mrn","record_id"))
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)

#Add Med Meta Data to Seurat object
so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc)
rm(gene_expression_matrix,med,meta_kidney_sc)
# saveRDS(so_kpmp_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds"))
```

# 3. Outline 
### a. Comparison Groups of Interest
1. T2D vs. OB
2. T2D vs. T1D
3. T2D vs. LC
4. T2D/OB vs. LC
5. T2D/OB vs. T1D

### b. Disease Categories of Interest
1. Albuminuria (Y/N) - UACR >= 30 mg/g
a. T2D
b. T2D/OB
2. UACR >= 100mg/g
3. UACR >= 300mg/g

### c. Medication Groups of Interest 
1. SLGT2i+/GLP-1ra+
2. SGLT2i-/GLP-1ra+
3. SGLT2i+/GLP-1ra-
4. SGLT2i-/GLP-1ra-

### d. Additional Medications of Interest  
Metformin, Insulin, ACEi/ARB

### e. Characteristics of Interest
UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV

### f. Gene Sets of Interest
1. All Genes 
2. Senescence 
3. Metabolism
4. Inflammation

### g. Analysis Plan
1. All Genes
a. DEGS + GSEA + IPA
b. MAST + GSEA + IPA
c. Pseuodbulk all cell types
d. Cell Specific (PT, TAL, EC, etc.)
2. Pathway Gene Sets
a. DEGS
b. MAST
c. Pseudobulk all cell types
d. Cell Specific (PT, TAL, EC, etc. )

# 3. Descriptive Statistics
```{r, echo = F,warning=F}
invisible(gc())


#Load in med data to update medication information
med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))
med <- med %>% 
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
meta_kidney_sc <-  so_kidney_sc@meta.data
med <- med %>% 
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) #95 remain
med <- med %>%  
  filter(record_id %in% meta_kidney_sc$record_id) #81 remain
med$mrn <- as.numeric(med$mrn)
rm(meta_kidney_sc)

# saveRDS(so_kidney_sc,fs::path(dir.dat,"Kidney scRNAseq Project","Data","so_kidney_sc_10_mito.rds"))

#Load metadata for descriptive stats
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc.csv"))
#harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney Project","Data","harmonized_data_kidney_sc_all_metadata2.csv")) #all metadata vars
harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))
harm_meta_data <- harm_meta_data %>%
  dplyr::select(-X)
harm_meta_data <- harm_meta_data %>% 
  filter(kit_id %in% unique(so_kidney_sc$kit_id))
harm_meta_data <- tidylog::full_join(med,harm_meta_data,by=c("mrn"))
rm(med)

#Create Medication Groups
harm_meta_data <- harm_meta_data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>%
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>%
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>%
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))
#mutate(=ifelse(group!="Type 2 Diabetes" &  epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups:
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)
harm_meta_data <- harm_meta_data %>%
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
harm_meta_data$medication <- factor(harm_meta_data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))

#Define a T2D+OB group
harm_meta_data <- harm_meta_data %>%
  mutate(group2 = ifelse((group=="Type 2 Diabetes" | group=="Obese Control"),"T2D + OB",group))
#check <- harm_meta_data %>%
#  dplyr::select(all_of(c("kit_id","group","group2")))

# #Number of Individuals Per Study
# harm_summary1 <- harm_meta_data %>%
#   group_by(study) %>%
#   summarize(n=n())
# 
# #Number of Individuals Per Disease Group
# harm_summary2 <- harm_meta_data %>%
#   group_by(group) %>%
#   summarize(n=n())
# 
# #Number of Individuals Per Medication Group
# harm_summary3 <- harm_meta_data %>%
#   group_by(medication) %>%
#   summarize(n=n())
# 
# #Number of Individuals per med group & disease group
# harm_summary4 <- harm_meta_data %>%
#   group_by(group,medication) %>%
#   summarize(n=n())

#Note: 83 participants have sc and metadata
#Covars of interest:
#UACR, SBP/DBP (MAP), GFR, RPF, BMI, Fat Mass, Lean Mass, HbA1C, TKV/htTKV
dat <- harm_meta_data
rm(harm_meta_data)
label(dat$age)      <- "Age (y)"
label(dat$sex)      <- "Sex"
label(dat$hba1c) <- "HbA1c (%)"
dat$medication <- case_when(dat$medication=="no_med"~"No Medication",
                            dat$medication=="sglt2"~"Exclusive SLGT2i",
                            dat$medication=="glp1"~"Exclusive GLP-1ra",
                            dat$medication=="glp1_sglt2"~"GLP-1ra + SGLT2i")
#dat$medication <- factor(dat$medication)
label(dat$medication) <- "Medication Status"
label(dat$ace_inhibitor) <- "ACE Inhibitor (Yes vs. No)"
label(dat$angiotensin_receptor_blocker) <- "Angiotensin Receptor Blocker (Yes vs. No)"
label(dat$raasi_1) <- "ACEi/ARB (Yes vs. No)"
label(dat$sbp) <- "Systolic Blood Pressure (mmHg)"
label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$map) <- "Mean Arterial Pressure (mmHg)"
label(dat$erpf_raw_plasma) <- "Estimated Renal Plasma Flow (mL/min)"
label(dat$bmi) <- "Body Mass Index (kg/m²)"
label(dat$dexa_fat_kg) <- "DEXA Fat Mass (kg)"
label(dat$dexa_lean_kg) <- "DEXA Lean Mass (kg)"
label(dat$total_kidney_volume_ml) <- "Total Kidney Volume (mL)"
label(dat$ht_adj_tkv) <- "Height-Adjusted Total Kidney Volume (mL/m)"
label(dat$gfr_raw_plasma) <- "Glomerular Filtration Rate (mL/min)"
label(dat$eGFR_CKD_epi) <- "Estimated Glomerular Filtration Rate (mL/min/1.73 m²)"
label(dat$race_ethnicity)    <- "Race/Ethnicity"
label(dat$triglycerides) <-  "Triglycerides (mg/dL)"
label(dat$group) <- "Disease Status"
label(dat$group2) <- "Disease Status"
label(dat$metformin_timepoint) <- "Metformin (Yes vs. No)" 
label(dat$mfm_1) <- "Metformin (Yes vs. No)"
label(dat$insulin_med_timepoint) <- NULL
label(dat$insulin_injections_timepoint) <- NULL
label(dat$insulin_1) <- NULL

#race_ethnicity
## Table 1. Descriptive Statistics by Study  ----
table1(~ age + sex + bmi + map + group + medication | study, data=dat,overall=F)
# t1 <- table1(~ age + sex + bmi + map + group + medication | study, data=dat,overall=F)
# kable(t1,caption="Table 1. Descriptive Statistics by Study")
# table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
#          sbp + dbp + map +
#          triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
#          erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
#          ace_inhibitor + angiotensin_receptor_blocker+
#          + group + medication | study, data=dat, overall=F)
# table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
#          sbp + dbp + map +
#          triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
#          erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
#          ace_inhibitor + angiotensin_receptor_blocker+
#          + group + medication | study, data=dat, overall=F)
## Table 2. Table 2. Descriptive Statistics by Disease Status ----
table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
         sbp + dbp + map +
         triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
         erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
         raasi_1+mfm_1+insulin_1+medication | group2, data=dat)
#Med table
# table1(~insulin_injections_timepoint + insulin_med_timepoint+insulin_1| group, data=dat)
# kable(t2,caption="Table 2. Descriptive Statistics by Disease Status")

#Table 3. Table 3. Descriptive Statistics by Disease Status (T2D & OB Combined) ----
table1(~ age + sex + bmi + dexa_fat_kg + dexa_lean_kg + hba1c +
         sbp + dbp + map +
         triglycerides + gfr_raw_plasma + eGFR_CKD_epi +
         erpf_raw_plasma+ total_kidney_volume_ml + ht_adj_tkv+
         ace_inhibitor + angiotensin_receptor_blocker+metformin_timepoint+insulin_med_timepoint+
         + medication | group2, data=dat)
# kable(t3,caption="Table 3. Descriptive Statistics by Disease Status (T2D & OB Combined)")

#Compare meds and groups
# table1(~ medication + insulin_injections_timepoint +insulin_med_timepoint+ ace_inhibitor+angiotensin_receptor_blocker |group,data=dat,overall=F)
# table1(~ medication + insulin_injections_timepoint+insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker  |group2,data=dat,overall=F)
# table1(~ medication +insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker + metformin_timepoint  |group2,data=dat,overall=F)
# table1(~ medication + insulin_injections_timepoint+insulin_med_timepoint+ ace_inhibitor +angiotensin_receptor_blocker + metformin_timepoint  |group,data=dat,overall=F)
rm(dat)
```

# 4. Analysis
## A. All Genes
### i. DEGS + IPA
#### a. Pseuodbulk all cell types
```{r,echo=F,warning=F,fig.width=15, fig.height=9}
#Select full gene set
all_genes <- rownames(so_kidney_sc)

#Disease Groups
#1. T2D vs. OB
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30)
# ipa <- read_delim()

#2. T2D vs. T1D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)


#3. T2D vs. LC
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30)


#4. T2D/OB vs. LC
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="Yes",top_gsea = 30)

#5. T2D/OB vs. T1D
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,additional_group = NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="Yes",top_gsea = 30)

#1. Albuminuria (Y/N) - UACR >= 30 mg/g
so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)

#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)


#2. UACR >= 100mg/g
so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)

#"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)

#3. UACR >= 300mg/g
so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)

#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)

#Medications (AMong Type 2s)
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)


#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)


#Medications (AMong T2D + OB)
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group=="T2D + OB")
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)


#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)


#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)

##RAASI - Type 2 Diabetes
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)

##RAASI - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)

#Metformin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
#
#Metformin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)

#Insulin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)

#
#Insulin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
result <- degs_fxn(so=so_subgroup,cell=NULL,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
```

#### b. PT & TAL Pseudobulk
```{r,echo=F,warning=F,fig.width=15, fig.height=9}
#Pseuodbulk Tall and PT cells
so_kidney_sc$cell_type2 <- ifelse(grepl("PT-",so_kidney_sc$celltype_rpca),"PT",
                                  ifelse(grepl("TAL-",so_kidney_sc$celltype_rpca),"TAL",NA))
so_kidney_sc$cell_type2 <- as.character(so_kidney_sc$cell_type2)
all_cell_types <- unique(so_kidney_sc$cell_type2)
all_cell_types <- all_cell_types[-1]
Idents(so_kidney_sc) <- so_kidney_sc$cell_type2

#1. T2D vs. OB
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
all_genes <- rownames(so_subgroup)
#Diabetes yes vs. no
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30)
  print(result)
}

#2. T2D vs. T1D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}

#3. T2D vs. LC
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}

#4. T2D/OB vs. LC
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}

#5. T2D/OB vs. T1D
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}

#1. Albuminuria (Y/N) - UACR >= 30 mg/g
so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}

#2. UACR >= 100mg/g
so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}

#"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}
#3. UACR >= 300mg/g
so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}

#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}

#Medications (AMong Type 2s)
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")

#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#T2D + OB
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Other med groups
#RAASI - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}
#
##RAASI - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

```



#### c. KPMP Cell Type 1 (PT, TAL, EC, etc.)
```{r,echo=F,warning=F,fig.width=15, fig.height=9}

so_kidney_sc$celltype_rpca <- as.character(so_kidney_sc$celltype_rpca)
all_cell_types <- unique(so_kidney_sc$celltype_rpca)
Idents(so_kidney_sc) <- so_kidney_sc$celltype_rpca
all_cell_types <- all_cell_types[-29]
#Select full gene set
all_genes <- rownames(so_kidney_sc)

#1. T2D vs. OB
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
#Diabetes yes vs. no
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30)
  print(result)
}

#2. T2D vs. T1D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}
#3. T2D vs. LC
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}
#4. T2D/OB vs. LC
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL, exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}
#5. T2D/OB vs. T1D
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}

#1. Albuminuria (Y/N) - UACR >= 30 mg/g
so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}

#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}

#2. UACR >= 100mg/g
so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
# all_cell_types2 <- all_cell_types[-29]
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}

#"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}

#3. UACR >= 300mg/g
so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}

#Medications (Among Type 2 Diabetes)
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group2=="Type 2 Diabetes")

#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}
#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Medications (Among T2D + OB)
#Filter to T2D + OB only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")

#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}
#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Other med groups
#RAASI - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

##RAASI - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - T2D + OB
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

```

#### d. KPMP Cell Type 2 (PT, TAL, EC, etc.)
```{r,echo=F,warning=F,fig.width=15, fig.height=9}
so_kidney_sc$celltype_rpca <- as.character(so_kidney_sc$celltype_rpca)
all_cell_types <- unique(so_kidney_sc$celltype_rpca)
Idents(so_kidney_sc) <- so_kidney_sc$celltype_rpca
all_cell_types <- all_cell_types[-29]
#Select full gene set
all_genes <- rownames(so_kidney_sc)

#1. T2D vs. OB
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Obese Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
#Diabetes yes vs. no
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Obese Control",enrichment="No",top_gsea = 30)
  print(result)
}

#2. T2D vs. T1D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}
#3. T2D vs. LC
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="Type 2 Diabetes",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}
#4. T2D/OB vs. LC
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Lean Control")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL, exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Lean Control",enrichment="No",top_gsea = 30)
  print(result)
}
#5. T2D/OB vs. T1D
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB" | group=="Type 1 Diabetes")
so_subgroup$DiseaseGroup <- so_subgroup$group2
so_subgroup$DiseaseGroup <- factor(so_subgroup$DiseaseGroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,additional_group=NULL,exposure="DiseaseGroup",exp_group="T2D + OB",ref_group="Type 1 Diabetes",enrichment="No",top_gsea = 30)
  print(result)
}

#1. Albuminuria (Y/N) - UACR >= 30 mg/g
so_kidney_sc$albuminuria_30 <- ifelse(so_kidney_sc$acr_u>=30,"Above_30","Below_30")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="Type 2 Diabetes",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}

#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_30 <- factor(so_subgroup$albuminuria_30)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_30",additional_group="T2D + OB",exp_group="Above_30",ref_group="Below_30",enrichment="No",top_gsea = 30)
  print(result)
}

#2. UACR >= 100mg/g
so_kidney_sc$albuminuria_100 <- ifelse(so_kidney_sc$acr_u>=100,"Above_100","Below_100")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
# all_cell_types2 <- all_cell_types[-29]
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="Type 2 Diabetes",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}

#"PT_lowQuality" has fewer than 3 cells in group 1, remove from analyses
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_100 <- factor(so_subgroup$albuminuria_100)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_100",additional_group="T2D + OB",exp_group="Above_100",ref_group="Below_100",enrichment="No",top_gsea = 30)
  print(result)
}

#3. UACR >= 300mg/g
so_kidney_sc$albuminuria_300 <- ifelse(so_kidney_sc$acr_u>=300,"Above_300","Below_300")
#  a. T2D
so_subgroup <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="Type 2 Diabetes",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}
#  b. T2D/OB
so_subgroup <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup$albuminuria_300 <- factor(so_subgroup$albuminuria_300)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="albuminuria_300",additional_group="T2D + OB",exp_group="Above_300",ref_group="Below_300",enrichment="No",top_gsea = 30)
  print(result)
}

#Medications (Among Type 2 Diabetes)
#Filter to type 2 diabetes only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group2=="Type 2 Diabetes")

#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}
#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="Type 2 Diabetes",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Medications (Among T2D + OB)
#Filter to T2D + OB only since no other participants are on these main meds
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")

#levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))
#1. SLGT2i+/GLP-1ra+
#Combo vs. No Med
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}
#Combo vs. GLP1
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="glp1")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="glp1",enrichment="No",top_gsea = 30)
  print(result)
}

#Combo vs. SGLT2
so_subgroup <- subset(so_diab,medication=="glp1_sglt2" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1_sglt2",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#2. SGLT2i-/GLP-1ra+
#GLP-1 exclusive vs. no med
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Glp1 vs. sglt2
so_subgroup <- subset(so_diab,medication=="glp1" | medication=="sglt2")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="glp1",ref_group="sglt2",enrichment="No",top_gsea = 30)
  print(result)
}

#3. SGLT2i+/GLP-1ra-
#sglt2 vs. no med
so_subgroup <- subset(so_diab,medication=="sglt2" | medication=="no_med")
so_subgroup$medication <- factor(so_subgroup$medication)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
  result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="medication",additional_group="T2D + OB",exp_group="sglt2",ref_group="no_med",enrichment="No",top_gsea = 30)
  print(result)
}

#Other med groups
#RAASI - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

##RAASI - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,raasi_1=="Yes" | raasi_1=="No")
so_subgroup$raasi_1 <- factor(so_subgroup$raasi_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="raasi_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Metformin - T2D + OB
# #Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,mfm_1=="Yes" | mfm_1=="No")
so_subgroup$mfm_1 <- factor(so_subgroup$mfm_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="mfm_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - type 2s only
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group=="Type 2 Diabetes")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="Type 2 Diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}

#Insulin - T2D + OB
#Filter out NAs for ace inhibitors
so_diab <- subset(so_kidney_sc,group2=="T2D + OB")
so_subgroup <- subset(so_diab,insulin_1=="Yes" | insulin_1=="No")
so_subgroup$insulin_1 <- factor(so_subgroup$insulin_1)
all_genes <- rownames(so_subgroup)
for (cell_type in all_cell_types[-c(26,32)]) {
result <- degs_fxn(so=so_subgroup,cell=cell_type,gene_set=all_genes,exposure="insulin_1",additional_group="T2D + OB",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
print (result)
}
```















#1. Set Up Libraries & Directories
```{r Set up, echo=F, include = F}
# Function to install and load packages
install_and_load <- function(package_name) {
  if (!requireNamespace(package_name, quietly = TRUE)) {
    install.packages(package_name)
  }
  library(package_name, character.only = TRUE)
}

# Load required packages
#library(tidyverse)
install_and_load("tidyr")
install_and_load("BiocManager")
install_and_load("arsenal")
install_and_load("dplyr")
install_and_load("ggplot2")
install_and_load("ggrepel")
install_and_load("Seurat")
install_and_load("future")
install_and_load("colorspace")
install_and_load("patchwork")
install_and_load("ggdendro")
install_and_load("cowplot")
install_and_load("ggpubr")
install_and_load("rstatix")
install_and_load("table1")
install_and_load("Biobase")
install_and_load("ReactomeGSA")
install_and_load("GSEABase")
install_and_load("msigdbr")
install_and_load("kableExtra")
install_and_load("knitr")
if (!requireNamespace('BiocManager', quietly = TRUE))
  install.packages('BiocManager')
library(table1)

# BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
#install.packages("openxlsx")
#library(openxlsx)

#Increase Memory
mem.maxVSize(64000000000)

#Set up directories
dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
#dir.dat <- c("/Users/hhampson/Dropbox/Bjornstad data")
#dir.dat <- c("/Desktop/Peds/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/scRNA/data_raw")
dir.meta.dat <- c("/Volumes/Peds Endo/Petter Bjornstad/Data Harmonization/Data Clean")
# dir.dat2 <- c("/Users/hhampson/Dropbox/Bjornstad data/Clinical data")
# "smb://ucdenver.pvt/som/PEDS/RI%20Biostatistics%20Core/Shared/Shared%20Projects/Laura/Peds%20Endo/Petter%20Bjornstad/Data%20Harmonization/"
# smb://ucdenver.pvt/som/PEDS/RI%20Biostatistics%20Core/Shared/Shared%20Projects/Laura/Peds%20Endo/Petter%20Bjornstad/scRNA
dir.home <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Kidney scRNA/Kidney scRNA")

#Load functions
# source(fs::path(dir.home,"Kidney_Functions.R"))
source("Kidney_Functions.R")
```


#2. Load & Format scRNAseq Data 
```{r echo = F}
# Kidney scRNA data processing
#so_kidney_sc <- readRDS(fs::path(dir.dat,"Kidney scRNA","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
so_kidney_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS"))
# so_kidney_sc <- readRDS("/Desktop/Peds/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad/scRNA/data_raw/PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS")

# Add missing covariate meta data to Seurat object
meta_raw <- read.csv(fs::path(dir.dat,"Data Harmonization","Data Exports","renal_clearance_biopsy.csv"))
# meta_raw <- read.csv(fs::path(dir.dat2,"renal_clearance_biopsy.csv"))
# meta_raw <- read.csv("/Users/hhampson/Dropbox/Bjornstad data/Clinical data/renal_clearance_biopsy.csv")
# harm_data <- read.csv(fs::path(dir.meta.dat,"harmonized_dataset.csv"))
# harm_data2 <- harm_data %>% 
#   filter(group=="Lean Control"|group=="Obese Control"|group=="Type 2 Diabetes" |group=="Type 1 Diabetes")
# sglt2s <- harm_data2 %>% 
#   filter(procedure=="medications") %>% 
#   filter(sglt2i_ever=="Yes" | sglt2i_ever=="No" | sglti_timepoint=="No") %>% 
#   filter(visit=="baseline")
# clinical <- harm_data %>% 
#   filter(procedure=="physical_exam")
# hba1c <- harm_data %>% 
#   filter(procedure=="clamp")
# sex <- harm_data %>% 
#   filter(procedure=="screening")


# meta.vars <- c("procedure",colnames(meta_raw))
# harm_data <- harm_data %>% 
#   dplyr::select(all_of(meta.vars))
# meta_data <- harm_data[which(harm_data$sglt2i_ever!=""),]
# meta_data <- meta_data[which(harm_data$sex=="Male" | harm_data$sex=="Female"),]


#Filter out meta data participants without sc 
record_ids <- unique(so_kidney_sc@meta.data$record_id)
meta_raw <- meta_raw %>%
  filter(record_id %in% record_ids) %>%
  dplyr::select(-record_id)

# #Create study id to filter to in so
# so_kidney_sc@meta.data <- so_kidney_sc@meta.data %>%
#   mutate(study = case_when(grepl("CRC",record_id) ~ "CRC",
#                            grepl("RH",record_id) ~ "RH",
#                            grepl("IT",record_id) ~ "IT"))

#Merge new metadata with 
meta_kidney_sc <-  tidylog::left_join(so_kidney_sc@meta.data, meta_raw,by="cryostor_id")
# meta_kidney_sc <-  so_kidney_sc@meta.data

#Check there are no duplicates
length(which(duplicated(meta_kidney_sc$barcode)))

#Add rownames back to metadata to merge back into saurat object 
rownames(meta_kidney_sc) <- rownames(so_kidney_sc@meta.data)

# Add new meta data to so 
so_kidney_sc <- AddMetaData(so_kidney_sc, meta_kidney_sc)

#Filter to baseline visits only
so_kidney_sc <- subset(so_kidney_sc, visit == "baseline")
# so_kidney_sc <- subset(so_kidney_sc)

# so_kidney_sc <- subset(so_kidney_sc, record_id=="CRC" | grepl("CRC",record_id) | grepl("RH",record_id) | grepl("IT",record_id))
# meta_kidney_sc <-  meta_kidney_sc %>%
#   filter(grepl("CRC",record_id) | grepl("RH",record_id) | grepl("IT",record_id))

#so_kidney_sc <- NormalizeData(so_kidney_sc)
#so_kidney_sc <- ScaleData(so_kidney_sc)

#Filter to only MAC cells 
so_kidney_sc$celltype

# # # PCA
# so_kidney_sc <- RunPCA(so_kidney_sc, features = VariableFeatures(object = so_kidney_sc))
# ElbowPlot(so_kidney_sc)
# # Cluster cells
# so_kidney_sc <- FindNeighbors(so_kidney_sc)
# so_kidney_sc <- FindClusters(so_kidney_sc)
# # Perform UMAP and tSNE
# so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
# DimPlot(so_kidney_sc, reduction = "umap")
# 
# #UMAP by disease status
# DimPlot(so_kidney_sc, reduction = "umap", group.by = "group") 

```

## Descriptive Statistics
```{r Table 1}
#Number of cells and MON/MAC among each type 2 participant
#Number of MON and MAC from each of the subjects in the T2 group & their corresponding total cell count with gender identified.
meta_full <- so_kidney_sc@meta.data 
meta_full <- meta_full %>% 
  filter(group=="Type 2 Diabetes") %>% 
  group_by(record_id,sex) %>% 
  summarize(MON_count = sum(celltype_rpca=="MON"),
            MAC_count = sum(celltype_rpca=="MAC"),
            total_cell_count = n(),
            .groups = "drop")
write.csv(meta_full,"/Users/hhampson/Documents/UW/1_Ongoing Projects/Kidney scRNA/cell_counts_by_participant.csv")

#Descriptive stats
dat <- read.csv(fs::path(dir.dat,"Clinical Data","renal_clearance_biopsy.csv"))

dat$sex <- as.numeric(as.factor(dat$sex))
dat$sex <- as.factor(dat$sex)
# dat$sex    <- factor(dat$sex, levels=c(0, 1))
label(dat$sex) <- "Sex"
dat$ethnicity   <- factor(dat$ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
label(dat$ethnicity) <- "Ethnicity"
dat$race   <- factor(dat$race, levels=c("White","Black or African American"),
                      labels=c("White","Black"))
label(dat$race) <- "Race"
label(dat$age) <- "Age (y)"
label(dat$bmi) <- "BMI (kg/m2)"
label(dat$hba1c) <- "HbA1c (mg/dL)"
label(dat$acr_u) <- "uACR (mg/g)"

dat_T2 <- dat %>% 
  filter(group=="Type 2 Diabetes")
dat_HC <- dat %>% 
  filter(group=="Lean Control")

caption1  <- "Descriptive Statistics by Sex among youth with Type 2 Diabetes"
# footnote <- "ᵃ Also known as Breslow thickness"
caption2  <- "Descriptives by Sex among HC"

#Type 2
table1(~ age + race + ethnicity + bmi + hba1c + acr_u | sex, data=dat_T2,caption = caption1)
#LC
table1(~ age + race + ethnicity + bmi + hba1c + acr_u | sex, data=dat_HC)

pvalue <- function(x, ...) {
    # Construct vectors of data y, and groups (strata) g
    y <- unlist(x)
    g <- factor(rep(1:length(x), times=sapply(x, length)))
    if (is.numeric(y)) {
        # For numeric variables, perform a standard 2-sample t-test
        p <- t.test(y ~ g)$p.value
    } else {
        # For categorical variables, perform a chi-squared test of independence
        p <- chisq.test(table(y, g))$p.value
    }
    # Format the p-value, using an HTML entity for the less-than sign.
    # The initial empty string places the output on the line below the variable label.
    c("", sub("<", "&lt;", format.pval(p, digits=3, eps=0.001)))
}
dat_T2$sex
table1(~ age + race + ethnicity + bmi + hba1c + acr_u | sex, data=dat_T2,caption = caption1)

t.test(age~sex,data=dat_T2)
t.test(bmi~sex,data=dat_T2)
t.test(hba1c~sex,data=dat_T2)
t.test(acr_u~sex,data=dat_T2)

table1(~ age + race + ethnicity + bmi + hba1c + acr_u | sex, data=dat_HC,caption = caption1)

t.test(age~sex,data=dat_T2)
t.test(bmi~sex,data=dat_T2)
t.test(hba1c~sex,data=dat_T2)
t.test(acr_u~sex,data=dat_T2)
chisq.test(table(dat_T2$race, dat_T2$sex))$p.value
chisq.test(table(dat_T2$ethnicity, dat_T2$sex))$p.value

t.test(age~sex,data=dat_HC)
t.test(bmi~sex,data=dat_HC)
t.test(hba1c~sex,data=dat_HC)
t.test(acr_u~sex,data=dat_HC)
chisq.test(table(dat_HC$race, dat_HC$sex))$p.value
chisq.test(table(dat_HC$ethnicity, dat_HC$sex))$p.value
```


#3. Examine Descriptive Statistics by Characteristics of Interest (Sex & SGLT2i Status)
## Sex (M/F)
### Descriptive table
```{r echo = F}
# Descriptive stats
kidney_meta_raw_sc <- so_kidney_sc@meta.data %>%
  filter(cryostor_id!="")
form <- paste("sex", paste(colnames(kidney_meta_raw_sc)[c(15,18:27)], collapse = " + "), sep = " ~ ")
summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = F))



#By group
kidney_meta_raw_sc <- so_HC@meta.data %>%
  filter(cryostor_id!="")
form <- paste("sex", paste(colnames(kidney_meta_raw_sc)[c(18:27)], collapse = " + "), sep = " ~ ")
summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = T))


t.test(kidney_meta_raw_sc$age ~ kidney_meta_raw_sc$sex)
t.test(kidney_meta_raw_sc$hba1c ~ kidney_meta_raw_sc$sex)

#cell numbers for the HC and T2 for MON, MAC, and GC-EC
#HC
so_HC_mon<- subset(so_HC, celltype_rpca == "MON")
ncol(so_HC_mon)
so_HC_mac<- subset(so_HC, celltype_rpca == "MAC")
ncol(so_HC_mac)
so_HC_ecgc<- subset(so_HC, celltype_rpca == "EC-GC")
ncol(so_HC_ecgc)
#T2
so_T2_mon<- subset(so_T2, celltype_rpca == "MON")
ncol(so_T2_mon)
so_T2_mon_male<- subset(so_T2_mon, sex == "Male")
ncol(so_T2_mon_male)
so_T2_mon_fem<- subset(so_T2_mon, sex == "Female")
ncol(so_T2_mon_fem)

so_T2_mac<- subset(so_T2, celltype_rpca == "MAC")
ncol(so_T2_mac)
so_T2_mac_male<- subset(so_T2_mac, sex == "Male")
ncol(so_T2_mac_male)
so_T2_mac_fem<- subset(so_T2_mac, sex == "Female")
ncol(so_T2_mac_fem)

so_T2_ecgc<- subset(so_T2, celltype_rpca == "EC-GC")
ncol(so_T2_ecgc)

ncol(so_T2)

```
## Group 
### Descriptive table
```{r echo = F}
# Descriptive stats
kidney_meta_raw_sc <- meta_kidney_sc %>%
  filter(cryostor_id!="")
form <- paste("group", paste(colnames(kidney_meta_raw_sc)[c(15,17:27)], collapse = " + "), sep = " ~ ")
summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = T))

```


## SGLT2i Y/N
### Descriptive table
```{r echo = F}
# Descriptive stats
# kidney_meta_raw_sc <- meta_raw %>%
#   filter(cryostor_id!="")
form <- paste("sglt2i_ever", paste(colnames(kidney_meta_raw_sc)[c(15,18:27)], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = F))

```

#4. Visualize using UMAPs Stratified by Characteristics of Interest (Sex & SGLT2i Status)
## Stratified UMAP scRNA 
### a. Sex
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "sex") 

```

### b. SGLT2is
```{r echo = F}
so_kidney_sc <- RunUMAP(so_kidney_sc, dims = 1:15)
DimPlot(so_kidney_sc, reduction = "umap", group.by = "sglt2i_ever") 

```


#5. Differential Expression of Genes by Group for All Disease Types
## Top 2000 genes by Sex
```{r}
#Sex
de.markers(so_kidney_sc, NULL, "sex", id1 = "Female", id2 = "Male", NULL, "_top")
m_top <- m_top %>% head(2000)
```

#Top 2000 genes by SGLT2is
```{r}
#SGLT2is
de.markers(so_kidney_sc, NULL, "sglt2i_ever", id1 = "No", id2 = "Yes", NULL, "_top")
m_top <- m_top %>% head(2000)

```


### By Sex & SGLT2is
```{r}
# so_kidney_sc@meta.data$hepatocyte <- ifelse(grepl("Hepatocyte",meta_liver_sc$celltype),"Hepatocyte","Non-Hepatocyte")
genes <- unique(rownames(so_kidney_sc))
Idents(so_kidney_sc) <- so_kidney_sc$celltype_rpca

#Sex
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "MON", "")
m_top <- m_top %>% head(2000)
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "MAC", "")
m_top <- m_top %>% head(2000)
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "EC-GC", "")
m_top <- m_top %>% head(2000)
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "EC-LYM", "")
m_top <- m_top %>% head(2000)
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "EC-AEA", "")
m_top <- m_top %>% head(2000)
de.markers(so_kidney_sc, genes, "sex", id2 = "Female", id1 = "Male", "EC-PTC", "")
m_top <- m_top %>% head(2000)

#SGLT2is
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-GC", "")
m_top <- m_top %>% head(20)
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-LYM", "")
m_top <- m_top %>% head(20)
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_kidney_sc, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-PTC", "")
m_top <- m_top %>% head(20)
```

## Figures
### By Sex & SGLT2is
```{r}
# Idents(so_kidney_sc) <- so_kidney_sc$celltype_rpca
# #Sex
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = NULL, m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "MON", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "MAC", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "EC-GC", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "EC_LYM", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "EC-AEA", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sex",
#                               celltype = "EC-PTC", m = m)
# 
# #SGLT2is
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = NULL, m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "MON", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "MAC", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "EC-GC", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "EC_LYM", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "EC-AEA", m = m)
# pt.plot_table <- dp.formatted(seurat_object = so_kidney_sc, genes = genes, group.by = "sglt2i_ever",
#                               celltype = "EC-PTC", m = m)

```

#6. Differential Expression of Genes by Group Stretified by Disease Status (HC/OB/T1/T2)
##a. Healthy Controls

```{r subset by disease status}
#Filter to HC only
so_HC <- subset(so_kidney_sc, group == "Lean Control")

#Filter to OB only
so_OB <- subset(so_kidney_sc, group == "Obese Control")

#Filter to T1 only
so_T1 <- subset(so_kidney_sc, group == "Type 1 Diabetes")

#Filter to T2 only
so_T2 <- subset(so_kidney_sc, group == "Type 2 Diabetes")

length(which(meta_raw$group=="Lean Control"))
length(which(meta_raw$group=="Obese Control"))
length(which(meta_raw$group=="Type 1 Diabetes"))
length(which(meta_raw$group=="Type 2 Diabetes" & meta_raw$sex=="Female"))
length(which(meta_raw$group=="Type 2 Diabetes" & meta_raw$sex=="Male"))
```

### Sex
```{r echo = F}
#HC
so_HC <- RunUMAP(so_HC, dims = 1:15)
DimPlot(so_HC, reduction = "umap", group.by = "sex") 

#OB
so_OB <- RunUMAP(so_OB, dims = 1:15)
DimPlot(so_OB, reduction = "umap", group.by = "sex") 

#T1
so_T1 <- RunUMAP(so_T1, dims = 1:15)
DimPlot(so_T1, reduction = "umap", group.by = "sex") 

#T2
so_T2 <- RunUMAP(so_T2, dims = 1:15)
DimPlot(so_T2, reduction = "umap", group.by = "sex") 

```

### SGLT2is
```{r echo = F}
#HC
so_HC <- RunUMAP(so_HC, dims = 1:15)
DimPlot(so_HC, reduction = "umap", group.by = "sglt2i_ever") 

#OB
so_OB <- RunUMAP(so_OB, dims = 1:15)
DimPlot(so_OB, reduction = "umap", group.by = "sglt2i_ever") 

#T1
so_T1 <- RunUMAP(so_T1, dims = 1:15)
DimPlot(so_T1, reduction = "umap", group.by = "sglt2i_ever") 

#T2
so_T2 <- RunUMAP(so_T2, dims = 1:15)
DimPlot(so_T2, reduction = "umap", group.by = "sglt2i_ever") 

```


## Differential Expression by Group

## Top 2000 genes
```{r}
#Sex
#HC
de.markers(so_HC, NULL, "sex", id1 = "Female", id2 = "Male", NULL, "_top")
m_top <- m_top %>% head(20)
#OB
de.markers(so_OB, NULL, "sex", id1 = "Female", id2 = "Male", NULL, "_top")
m_top <- m_top %>% head(20)
#T1
de.markers(so_T1, NULL, "sex", id1 = "Female", id2 = "Male", NULL, "_top")
m_top <- m_top %>% head(20)
#T2
de.markers(so_T2, NULL, "sex", id1 = "Female", id2 = "Male", NULL, "_top")
m_top <- m_top %>% head(20)

#SGLT2is
#T2
de.markers(so_T2, NULL, "sglt2i_ever", id1 = "No", id2 = "Yes", NULL, "_top")
m_top <- m_top %>% head(20)

```


### By Sex & SGLT2is
###HC
```{r}
genes <- unique(rownames(so_HC))
Idents(so_HC) <- so_HC$celltype_rpca

#Sex
de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "EC-GC", "")
m_top <- m_top %>% head(20)
# de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "EC-LYM", "")
de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_HC, genes, "sex", id2 = "Female", id1 = "Male", "EC-PTC", "")
m_top <- m_top %>% head(20)
```

###OB
```{r}
genes <- unique(rownames(so_OB))
Idents(so_OB) <- so_OB$celltype_rpca

#Sex
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "EC-GC", "")
m_top <- m_top %>% head(20)
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "EC-LYM", "")
m_top <- m_top %>% head(20)
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_OB, genes, "sex", id2 = "Female", id1 = "Male", "EC-PTC", "")
m_top <- m_top %>% head(20)
```

###T1
```{r}
genes <- unique(rownames(so_T1))
Idents(so_T1) <- so_T1$celltype_rpca

#Sex
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "EC-GC", "")
m_top <- m_top %>% head(20)
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "EC-LYM", "")
m_top <- m_top %>% head(20)
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_T1, genes, "sex", id2 = "Female", id1 = "Male", "EC-PTC", "")
m_top <- m_top %>% head(20)
```

###T2
```{r}
genes <- unique(rownames(so_T2))
Idents(so_T2) <- so_T2$celltype_rpca

#Sex
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "EC-GC", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "EC-LYM", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sex", id2 = "Female", id1 = "Male", "EC-PTC", "")
m_top <- m_top %>% head(20)

#SGLT2is
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "MON", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "MAC", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-GC", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-LYM", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-AEA", "")
m_top <- m_top %>% head(20)
de.markers(so_T2, genes, "sglt2i_ever", id2 = "No", id1 = "Yes", "EC-PTC", "")
m_top <- m_top %>% head(20)
```
