---
title: "Kidney scRNA for Jenny Kanter"
author: "Hailey Hampson"
date: "2024-10-09"
output: html_document
---

```{r libraries, echo=F, include = F}
library(tidyverse)
library(BiocManager)
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
```

# scRNA
```{r echo = F}
# Kidney OAT scRNA data processing
so_oat_sc <- readRDS("/Users/hhampson/Dropbox/Bjornstad data/Kidney scRNA/PB_90samples_Harmony_rpca_Fadhl_PhilApproved_091024.RDS")

# Add missing covariate meta data to Seurat object
meta_raw <- read.csv("/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/PAH Clearance_OATS/0_Data/renal_clearance_biopsy (1).csv")

#Filter meta data participants without sc 
record_ids <- unique(so_oat_sc@meta.data$record_id)
meta_raw <- meta_raw %>% 
  filter(record_id %in% record_ids) %>% 
  dplyr::select(-record_id)
meta_kidney_sc <-  tidylog::left_join(so_oat_sc@meta.data, meta_raw,by="cryostor_id")
#FIlter to baseline visits only
# meta_kidney_sc <- meta_kidney_sc %>% 
#   filter(visit!="12_months_post_surgery") 
#Check there are no duplicates
length(which(duplicated(meta_kidney_sc$barcode)))

#Add rownames back to metadata to merge back into saurat object 
rownames(meta_kidney_sc) <- rownames(so_oat_sc@meta.data)

#Filter to only RH/RH2, IMPROVE & CROCODILE
# meta_kidney_sc <-  meta_kidney_sc %>% 
#   filter(grepl("CRC",record_id) | grepl("RH",record_id) | grepl("IT",record_id))

# Add new meta data to so 
so_oat_sc <- AddMetaData(so_oat_sc, meta_kidney_sc)
so_oat_sc <- NormalizeData(so_oat_sc)
so_oat_sc <- ScaleData(so_oat_sc)
# PCA
so_oat_sc <- RunPCA(so_oat_sc, features = VariableFeatures(object = so_oat_sc))
ElbowPlot(so_oat_sc)
# Cluster cells
so_oat_sc <- FindNeighbors(so_oat_sc)
so_oat_sc <- FindClusters(so_oat_sc)
# Perform UMAP and tSNE
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap") 
```

## Obese vs. Obese + T2D
### Descriptive table
```{r echo = F}
# Descriptive stats
kidney_meta_raw_sc <- meta_kidney_sc %>%
  filter(cryostor_id!="")
form <- paste("sex", paste(colnames(kidney_meta_raw_sc)[c(15,18:25)], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = F))

```


## SGLT2i Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
kidney_meta_raw_sc <- meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("sglt2", paste(colnames(kidney_meta_raw_sc)[c(15,18:25)], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = kidney_meta_raw_sc, test = F))

```


## Stratified UMAP scRNA 
### Obese vs. Obese + T2D
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "sex") 

```

### MASLD Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_MASLD") 

```

### OSA Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_osa") 

```

### GLP-1RA Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "glp1agonist") 

```

### Insulin Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "insulin") 

```

### Kidney disease Y/N
```{r echo = F}
so_oat_sc <- RunUMAP(so_oat_sc, dims = 1:15)
DimPlot(so_oat_sc, reduction = "umap", group.by = "diagnosis_of_kidney_disease") 

```

## Differential Expression by Group

## Top 2000 genes
```{r}
de.markers(so, NULL, "Group", id1 = "T1D", id2 = "HC", "PT", "_top")
m_top <- m_top %>% head(2000)
```


### Obese vs. Obese + T2D
```{r}
so_oat_sc@meta.data$hepatocyte <- ifelse(grepl("Hepatocyte",meta_liver_sc$celltype),"Hepatocyte","Non-Hepatocyte")
genes <- unique(rownames(so_oat_sc))
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### MASLD Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### OSA Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### GLP-1RA Y/N
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_oat_sc) <- so_oat_sc$hepatocyte
de.markers(so_oat_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### Insulin Y/N
```{r}
# Idents(so_oat_sc) <- so_oat_sc$celltype
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_oat_sc) <- so_oat_sc$hepatocyte
# de.markers(so_oat_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

### Kidney disease Y/N
```{r}
# Idents(so_oat_sc) <- so_oat_sc$celltype
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_oat_sc) <- so_oat_sc$hepatocyte
# de.markers(so_oat_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

## Figures
### Obese vs. Obese + T2D
```{r}
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### MASLD Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### OSA Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_oat_sc) <- so_oat_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_oat_sc) <- so_oat_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_oat_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```


