---
title: "ROCKIES_RH_RH2_Project"
author: "Hailey Hampson"
date: "2025-04-15"
output: html_document
---

#Analysis Plan
# Within T2D and T2D/OB: TCA and oxphos transcripts SGLT2i y/n
# Ye Ji and Hailey you have done most of this, we just need to run the new mixed models and also determine if we should adjust for any covariates
# Evaluate associations between TCA and oxphos transcripts and K2 and K2/F in a mixed models. We will likely have better power looking at this in a linear model with continuous K2 and K2/F in lieu of groups (i.e., above and below median / tertiles), but can try both.

# 1. Set up 
## a. Libraries & Directores
```{r, include=F}
library(reprex)
# library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
# BiocManager::install("Biobase")
library(Biobase)
# BiocManager::install("ReactomeGSA")
library(ReactomeGSA)
# BiocManager::install("GSEABase")
library(GSEABase)
# install.packages('msigdbdf', repos = 'https://igordot.r-universe.dev')
library(msigdbr)
# library(kableExtra)
library(knitr)
# BiocManager::install("SingleCellExperiment")
library(SingleCellExperiment)
# BiocManager::install("fgsea")
library(fgsea)
# BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
# library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
# library(limma)
library(reshape2)
library(data.table)
# BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
# library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
# library(NMF)
library(rsvd)
library(RColorBrewer)
# BiocManager::install("devtools")
# library(devtools)
# install_github("Sun-lab/ideas",force=T)
# library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
library(future)
# registerDoParallel(cores = 6)
# library(VennDiagram)
#install.packages("survival")
library(survival)
#install.packages("lme4")  # If not already installed
library(lme4)
#install.packages("lmerTest")
library(lmerTest)
# install.packages("emmeans")
library(emmeans)
# install.packages("glmmTMB")
library(glmmTMB)
library(ggrepel)
# library(qpcR)
library(ggpubr)
#install.packages("glmmTMB")
# Reinstall glmmTMB from source
#install.packages("glmmTMB", type = "source")
# library(glmmTMB)
# Install DoubletFinder (if not already installed)
# devtools::install_github("chris-mcginnis-ucsf/DoubletFinder",force=T)
# Load the package
# Install DoubletFinder from GitHub (use devtools to install)
# if (!requireNamespace("devtools", quietly = TRUE)) install.packages("devtools")
# devtools::install_github("chris-mcginnis-ucsf/DoubletFinder",force=T)
# library(DoubletFinder)
# install.packages("emmeans")
library(emmeans)
library(ggrepel)
# library(qpcR)
# BiocManager::install("enrichplot")
library(enrichplot)
library(enrichR)
dbs <- c("GO_Biological_Process_2023",
         "KEGG_2021_Human",
         "Reactome_2022",
         "Reactome_Pathways_2024")
library(doFuture)


#Set number of cores for parallellization
#maxCores <- detectCores()
#numCores <- maxCores-1
#cl <- makeCluster(numCores)  # Create a cluster with the desired number of cores
#registerDoParallel(cl) 

#Local file path
#dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
#dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
#dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
#dir.results <- c("/Volumes/Peds Endo/Petter Bjornstad/Kidney Project/Results")

# #Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Kidney scRNA/Kidney scRNA")
# dir.results <- c(fs::path(dir.dat,"Kidney Project/Results"))

# # #Mac Studio File Path
# dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
# dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Kidney scRNAseq Project/Results")
# dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Load functions
# source("Kidney_functions_sc.R")
# source("/Users/hhampson/CHCO-Code/Petter Bjornstad/Data Processing and Analysis/Standard_Functions.R")
# dir.results <- c("/home/hailey/Documents/Results")

# # write file
# bucket <- "scrna" # bucket name in Kopah
# temp_file <- tempfile(fileext = ".csv") # need to create a temporary file
# write.csv(df, temp_file, row.names = FALSE) # save the temp file locally (will delete after R session)
# s3$upload_file(temp_file, bucket, 'Kidney transcriptomics/ROCKIES Results/filename.csv')

```

## b. Cyberduck 
```{r, Cyberduck setup}
# knitr::opts_knit$set(root.dir = dir.results)
# install.packages("reticulate")
# Sys.setenv(RETICULATE_PYTHON = "/home/hailey/miniconda3/envs/py310_env/bin/python")
library(reticulate)
reticulate::py_config()
# library(reticulate)
# reticulate::use_condaenv("py310_env", required = TRUE)
# reticulate::py_config()
# Sys.setenv(LD_LIBRARY_PATH = "/usr/lib/x86_64-linux-gnu")
# reticulate::use_python("/home/hailey/miniconda3/bin/python")
# reticulate::use_condaenv("base", required = TRUE)

## Load boto3 and pandas
boto3 <- reticulate::import("boto3")
pd <- reticulate::import("pandas")

## Create an S3 client
# install.packages("jsonlite")  # Install if not already installed
library(jsonlite)  # Load the package

keys <- fromJSON("/mmfs1/home/hhampson/keys.json") # replace with your Lambda username
session <- boto3$session$Session(
  aws_access_key_id = keys$MY_ACCESS_KEY,
  aws_secret_access_key = keys$MY_SECRET_KEY
)

## Create an S3 client with the session
s3 <- session$client("s3", endpoint_url = "https://s3.kopah.uw.edu")
```

# 2. Quality Control & Preprocessing
## a. Load Kidney scRNA seq Data
```{r, include=F}
#Load PB90 with new KPMP Cell Types
gc()
bucket <- "scrna" # bucket name in Kopah
temp_file <- tempfile(fileext = ".rds") # need to create a temporary file
s3$download_file(bucket, "Kidney transcriptomics/Single cell RNA seq/PB_90_RPCAFix_Metadata_LR_RM_kpmpV1labelled.rds", temp_file)
so_kpmp_sc <- readRDS(temp_file)
```

## b. Data Cleaning, QC & Preprocessing
```{r, echo=F, warning=F, fig.width=15, fig.height=15}
#Fix Typos in kit ids in PB90
so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="KI-0014643")] <- "KL-0014643"
so_kpmp_sc$kit_id[which(so_kpmp_sc$kit_id=="kl-0023998")] <- "KL-0023998"

#Load harmonized data that has been filtered from 90 to the 83 participants that have baseline single cell data
# harm_meta_data <- read.csv(fs::path(dir.dat,"Kidney scRNAseq Project","Data","harmonized_data_kidney_sc_all_metadata2.csv"))

bucket <- "scrna" # bucket name in Kopah
temp_file <- tempfile(fileext = ".csv") # need to create a temporary file
s3$download_file(bucket, "data_clean/harmonized_data_kidney_sc_all_metadata2.csv", temp_file)
harm_meta_data <- read.csv(temp_file)


harm_meta_data <- harm_meta_data %>%
  dplyr::select(-X)

#Select metadata from seurat object to facilitate merge of new metadata into seurat object
meta_kidney_sc <-  so_kpmp_sc@meta.data
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)

#Merge metadata from 83 participants at baseline into seurat object metadata
meta_kidney_sc <- meta_kidney_sc %>%
  left_join(harm_meta_data,by="kit_id")
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)

#Pull ids from 83 participants at baseline to filter seurat object to these participants only 
ids <- harm_meta_data$kit_id

#Merge metadata back into seurat object
so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc)

#Check number of unique ids
length(unique(so_kpmp_sc$kit_id)) #should be 90

#Filter seurat object to only IDs that have the metadata & sc data (83 individuals at baseline)
so_kpmp_sc <- subset(so_kpmp_sc, kit_id %in% ids)

#Check number of unique ids after filtering
length(unique(so_kpmp_sc$kit_id)) #should be 83

#Remove metadatasets
rm(meta_kidney_sc,harm_meta_data)

#Remove two coenrolled IDs
so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0030621")
so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0029535")
#Check unique number of kit ids now (should be 81)
length(unique(so_kpmp_sc$kit_id)) #Should be 81

#Load in most up to date medication data to update medication information
# med <- read.xlsx(fs::path(dir.dat,"Kidney scRNAseq Project/Data/Biopsies_w_mrn_Oct3.xlsx"))

bucket <- "scrna" # bucket name in Kopah
temp_file <- tempfile(fileext = ".xlsx") # need to create a temporary file
s3$download_file(bucket, "data_clean/Biopsies_w_mrn_Oct3.xlsx", temp_file)
med <- read.xlsx(temp_file)

# rm(s3,session,keys)

#Select Metformin, RASSI, Insulin data
med <- med %>%
  dplyr::select(all_of(c("record_id","mrn","raasi_1","insulin_1","mfm_1")))
#Pull seurat object metadata to help harmoinize in new metadata
meta_kidney_sc <-  so_kpmp_sc@meta.data
#Filter to only those with a unique identifier id in the seurat object metadata
med <- med %>%
  filter(mrn %in% as.character(meta_kidney_sc$mrn)) 
length(unique(med$mrn)) #95 total rows remain, 81 unique individual ids
#Filter to only those that have a unique record id in the seurat object
med <- med %>%
  filter(record_id %in% meta_kidney_sc$record_id) 
length(unique(med$mrn)) #81 remain
length(unique(med$record_id)) #81
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
med$mrn <- as.numeric(med$mrn) #Make numeric to merge
#Merge med data with seurat metadata
meta_kidney_sc <- meta_kidney_sc %>%
  left_join(med,by=c("mrn","record_id"))
rownames(meta_kidney_sc) <- rownames(so_kpmp_sc@meta.data)
length(unique(meta_kidney_sc$mrn)) #81 remain
length(unique(meta_kidney_sc$record_id)) #81

#Add Med Meta Data to Seurat object
so_kpmp_sc <- AddMetaData(so_kpmp_sc, meta_kidney_sc)
#Remove med metadatset
rm(med,meta_kidney_sc)

#Create medication & disease status groups of interest
so_kpmp_sc@meta.data <- so_kpmp_sc@meta.data %>%
  mutate(glp1_sglt2=ifelse(epic_glp1ra_1=="Yes" & epic_sglti2_1=="Yes","Yes","No")) %>%
  mutate(sglt2=ifelse(epic_sglti2_1=="Yes" & epic_glp1ra_1=="No","Yes","No")) %>%
  mutate(glp1=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="Yes","Yes","No")) %>%
  mutate(no_med=ifelse(epic_sglti2_1=="No" & epic_glp1ra_1=="No","Yes","No"))

#Define 4 exposure groups:
#SGLT2i(+)/GLP1RA(+), SGLT2i(+)/GLP1RA(-), SGLT2i(-)/GLPRA(+), SGLT2i(-)/GLP1RA(-)

so_kpmp_sc@meta.data <- so_kpmp_sc@meta.data %>%
  mutate(medication = case_when(glp1_sglt2 == "Yes" ~ "glp1_sglt2",
                                sglt2 == "Yes" ~ "sglt2",
                                glp1 == "Yes" ~ "glp1",
                                no_med == "Yes" ~ "no_med"))
so_kpmp_sc@meta.data$medication <- factor(so_kpmp_sc@meta.data$medication, levels = c("no_med", "sglt2", "glp1","glp1_sglt2"))


#Filter to RH/RH2 at baseline only
#RENAL-HEIR or RENAL-HEIRitage
so_kpmp_sc <- subset(so_kpmp_sc,study=="RENAL-HEIR" | study=="RENAL-HEIRitage")

#Ensure default assay in seurat object to RNA
DefaultAssay(so_kpmp_sc) <- "RNA"


#Before filtering
ncol(so_kpmp_sc)
#65062 cells
nrow(so_kpmp_sc)
#31332 genes
length(unique(so_kpmp_sc$kit_id))
#28 participants
length(unique(so_kpmp_sc$kit_id[which(so_kpmp_sc$study=="RENAL-HEIR")]))
#16,"RH-68-T" "RH-49-T" "RH-63-T" "RH-23-T" "RH-71-T" "RH-62-T" "RH-50-T" "RH-60-T" "RH-74-T" "RH-72-T" "RH-76-T" "RH-77-T" "RH-75-T" "RH-67-T" "RH-93-T" "RH-91-T"
length(unique(so_kpmp_sc$kit_id[which(so_kpmp_sc$study=="RENAL-HEIRitage")]))
#12, "RH2-51-T" "RH2-53-T" "RH2-07-O" "RH2-11-O" "RH2-23-T" "RH2-22-T" "RH2-38-T" "RH2-21-T" "RH2-55-T" "RH2-36-O" "RH2-43-T" "RH2-42-T"

#YE JI's filtering code for percent expression 
#Filter out rare genes expressed in less than "gene_pct" of cells
# expr_matrix <- as.matrix(GetAssayData(so_kpmp_sc, layer = "counts"))
expr_matrix <- as.matrix(GetAssayData(so_kpmp_sc, assay = "RNA", layer = "counts"))
# expr_matrix <- so_kpmp_sc@assays$RNA@counts
# Calculate the proportion of cells expressing each gene
num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
total_cells <- ncol(expr_matrix)  # Total number of cells
gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
remove(expr_matrix)
# Keep genes expressed in at least "gene_pct" of cells
genes_to_keep <- names(gene_proportion[gene_proportion >= 0.01])
so_kpmp_sc <- subset(so_kpmp_sc, features = genes_to_keep)
#After filtering
ncol(so_kpmp_sc) #65062 cells
nrow(so_kpmp_sc) # 15193 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_kpmp_sc), value = TRUE)
# mito_genes2 <- grep("^MT.", rownames(so_kpmp_sc), value = TRUE)
#Filter out all mitochondrial genes

#keep_ids <- unique(rownames(so_kpmp_sc)[which(!rownames(so_kpmp_sc) %in% mito_genes)])
# so_kpmp_sc <- subset(so_kpmp_sc, features = setdiff(rownames(so_kpmp_sc), mito_genes))
# so_kpmp_sc <- subset(so_kpmp_sc,kit_id!="KL-0029535")
#so_kpmp_sc$Gene <- rownames(so_kpmp_sc)
#so_kpmp_sc <- subset(so_kpmp_sc, Gene %in% keep_ids)
so_kpmp_sc <- subset(so_kpmp_sc, features = setdiff(rownames(so_kpmp_sc), mito_genes))
#so_kpmp_sc <- subset(so_kpmp_sc, features = setdiff(rownames(so_kpmp_sc@assays$RNA@counts), mito_genes))
# so_kpmp_sc <- subset(so_kpmp_sc, !rownames(so_kpmp_sc) %in% mito_genes)
# grep("^MT-", rownames(so_kpmp_sc@assays$RNA@counts), value = TRUE)
# dim(so_kpmp_sc@assays$RNA@counts) #9276 186125
# dim(so_kpmp_sc@assays$RNA@data) #9276 186125
# dim(so_kpmp_sc@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_kpmp_sc))) #0
rownames(so_kpmp_sc)[(grepl("^MT", rownames(so_kpmp_sc)))] #0
ncol(so_kpmp_sc) #186125 cells
nrow(so_kpmp_sc) #15180 genes

#Renormalize & Scale after filtering
so_kpmp_sc <- NormalizeData(so_kpmp_sc)
so_kpmp_sc <- ScaleData(so_kpmp_sc, features = VariableFeatures(so_kpmp_sc))

#Create general hepatocyte cell type variable
#Create PT and TAL pseudobulk cell type variable
so_kpmp_sc$celltype1 <- case_when(grepl("PT-",so_kpmp_sc$celltype_rpca)~"PT",
                                  grepl("TAL-",so_kpmp_sc$celltype_rpca)~"TAL",
                                  grepl("EC-",so_kpmp_sc$celltype_rpca)~"EC",
                                  grepl("POD",so_kpmp_sc$celltype_rpca)~"POD",
                                  grepl("MAC",so_kpmp_sc$celltype_rpca)~"MAC",
                                  grepl("MON",so_kpmp_sc$celltype_rpca)~"MON",
                                  grepl("PC-",so_kpmp_sc$celltype_rpca)~"PC",
                                  grepl("FIB",so_kpmp_sc$celltype_rpca)~"FIB_MC_VSMC",
                                  grepl("DTL",so_kpmp_sc$celltype_rpca)~"DTL",
                                  so_kpmp_sc$celltype_rpca=="DCT"~"DCT",
                                  so_kpmp_sc$celltype_rpca=="ATL"~"ATL",
                                  so_kpmp_sc$celltype_rpca=="B"~"B",
                                  so_kpmp_sc$celltype_rpca=="T"~"T")
so_kpmp_sc$celltype1 <- as.character(so_kpmp_sc$celltype1)

so_kpmp_sc$KPMP_celltype2 <- as.character(so_kpmp_sc$KPMP_celltype)
so_kpmp_sc$celltype2 <- ifelse(so_kpmp_sc$KPMP_celltype=="aPT" | 
                                 so_kpmp_sc$KPMP_celltype=="PT-S1/S2" | 
                                 so_kpmp_sc$KPMP_celltype == "PT-S3","PT",
                               ifelse(grepl("TAL",so_kpmp_sc$KPMP_celltype),"TAL",
                                      ifelse(grepl("EC-",so_kpmp_sc$KPMP_celltype),"EC",so_kpmp_sc$KPMP_celltype2)))
# so_kpmp_sc$celltype2 <- factor(so_kpmp_sc$celltype2)
```

## c. Visualizations 
```{r}
# PCA
so_kpmp_sc <- FindVariableFeatures(object = so_kpmp_sc)
so_kpmp_sc <- RunPCA(so_kpmp_sc, features = VariableFeatures(object = so_kpmp_sc),assay="RNA")
ElbowPlot(so_kpmp_sc)

# # Find neighbors and clusters (again using integrated data)
so_kpmp_sc <- FindNeighbors(so_kpmp_sc, assay = "RNA", dims = 1:20)
so_kpmp_sc <- FindClusters(so_kpmp_sc, resolution = 0.5)

# Perform UMAP and tSNE
# so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
# so_kpmp_sc@reductions
# DimPlot(so_kpmp_sc, reduction = "umap.harmony", raster = F)
# DimPlot(so_kpmp_sc, reduction = "umap.harmony", group.by = "KPMP_celltype", raster = FALSE)
p <- DimPlot(so_kpmp_sc, reduction = "umap.harmony", group.by = "KPMP_celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot with Cell Type Labels")

# # write file
# bucket <- "scrna" # bucket name in Kopah
# temp_file <- tempfile(fileext = ".jpeg") # need to create a temporary file
# write.csv(df, temp_file, row.names = FALSE) # save the temp file locally (will delete after R session)
# s3$upload_file(temp_file, bucket, 'Kidney transcriptomics/ROCKIES Results/filename.csv')
# Open the JPEG file in binary read mode
 DimPlot(so_kpmp_sc, reduction = "umap.harmony", group.by = "KPMP_celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot with Cell Type Labels")
temp_file <- tempfile(fileext = ".jpeg")
ggsave(temp_file, width = 10)
s3$upload_file(temp_file, "scrna", "Kidney transcriptomics/ROCKIES Results/umap_by_celltype.jpeg")

# # Create a temporary file with .jpeg extension
# temp_file <- tempfile(fileext = ".jpeg")
# # Save the plot as a JPEG
# ggsave(filename = temp_file, plot = p, device = "jpeg", width = 6, height = 4, dpi = 300)
# # Connect to S3
# s3 <- paws::s3()
# # Define the S3 bucket and path
# bucket <- "scrna"
# s3_key <- "Kidney transcriptomics/ROCKIES Results/filename.jpeg"
# # Upload the JPEG to the bucket
# s3$upload_file(temp_file, bucket, s3_key)


so_kpmp_sc$group <- factor(so_kpmp_sc$group)
DimPlot(so_kpmp_sc, reduction = "umap.harmony",group.by = "group",label=F,raster=F) +
  ggtitle(paste0("UMAP by Diabetes Category"))

so_kpmp_sc$medication <- factor(so_kpmp_sc$medication)
DimPlot(so_kpmp_sc, reduction = "umap.harmony",group.by = "medication",label=F,raster=F) +
  ggtitle(paste0("UMAP by Medication Category"))


#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(celltype2 == "PT")

label(cellcount$group) <- "Disease Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>% 
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Disease Status") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(celltype2 == "TAL")

label(cellcount$group) <- "Disease Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
# cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>%
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Disease Status") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(grepl("EC-",KPMP_celltype) | grepl("EC/VSMC",KPMP_celltype))

label(cellcount$group) <- "Disease Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
# cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>%
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Disease Status") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1

#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(celltype2 == "PT")

label(cellcount$medication) <- "Medication Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
# cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>% 
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(medication, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Medication Status (T2D + OB)") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

cellcount <- cellcount %>%
  filter(group!="Obese Controls")
prop_plot1 <- ggplot(data=cellcount,aes(medication, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Medication Status (T2D)") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(celltype2 == "TAL")

label(cellcount$medication) <- "Medication Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
# cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
# cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>%
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Medication Status (T2D + OB)") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

cellcount <- cellcount %>%
  filter(group!="Obese Controls")
prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Medication Status (T2D + OB)") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1 

#Barcharts of proportions
# By PT subtypes
cellcount <- so_kpmp_sc@meta.data %>% 
  filter(grepl("EC-",KPMP_celltype) | grepl("EC/VSMC",KPMP_celltype))

label(cellcount$group) <- "Disease Status"
# label(cellcount$steatosis_cat) <- "Steatosis Category"
# cellcount$group <- ifelse(cellcount$group=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$group <- factor(cellcount$group,levels=c("Lean_Control","Obese_Control","Type_1_Diabetes","Type_2_Diabetes"),labels=c("Lean Controls","Obese Controls","Type 1 Diabetes","Type 2 Diabetes"))
# cellcount$KPMP_celltype2 <- factor(cellcount$KPMP_celltype2,levels=c("PT-S1/S2","PT-S3","aPT"),labels=c("PT-S1/S2","PT-S3","aPT"))
# cellcount <- cellcount %>%
#   filter(group!="Obese Controls")

prop_plot1 <- ggplot(data=cellcount,aes(group, fill = KPMP_celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Disease Status") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

prop_plot1
```
## d. Descriptive Stats
```{r, echo = F,warning=F}
#Define disease and drug groups
# dat <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#dat <- read.csv("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean/liver_biopsy_metadata_PN.csv")

dat <- so_liver_sn@meta.data %>%
  dplyr::select(all_of(c("StudyID","RNAlater_ID","Cryostor_ID","nih_sex","nih_ethnicity","nih_race","bmi","diagnosis_of_diabetes",
                         "diagnosis_of_MASLD",
                         "steatosis_grade","fibrosis_stage",
                         "lobular_inflammation_percent","tg","creatinine",
                         "alt","ast","ggt","age_surgery","glp1agonist","sglt2","group2",
                         "group3","meds","insulin","metformin","StudyID","steatosis_cat","fibrosis_cat","a1c"
  )))  %>%
  group_by(RNAlater_ID) %>%
  summarise(across(everything(), first)) %>%
  ungroup() 
dat$study <- case_when(grepl("IT2D",dat$StudyID) ~ "IMPROVE",
                       grepl("MANGO",dat$StudyID) ~ "MANGO",
                       grepl("BSA",dat$StudyID) ~ "BIOBANK")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting
# dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (2+3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD + diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ a1c+ diagnosis_of_MASLD+diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent+  
         group2| steatosis_cat, data=dat)
table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent  
       | fibrosis_cat, data=dat)
rm(dat)
```

# 3. Comparitive Data Analysis
## A. Medication Status - Type 2 Diabetes Only
### i. All PT Cells - Pseudobulk 
```{r}
so_celltype <- subset(so_kpmp_sc,celltype1=="PT")
ncol(so_celltype) #12312 cells
nrow(so_celltype) #15180 genes
so_celltype <- subset(so_celltype,group=="Type_2_Diabetes")
ncol(so_celltype) #11107 cells
nrow(so_celltype) #15180 genes

# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)
data_subset <- data
data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
# num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
# cl <- makeCluster(num_cores)
# registerDoParallel(cl)
Sys.setenv(LD_LIBRARY_PATH = "/usr/lib/x86_64-linux-gnu")  # Or your system default lib path
plan(multisession, workers = 40)  # Already done
registerDoFuture()
options(future.globals.maxSize = 3 * 1024^3)  # 3 GiB, adjust as needed


# List of genes to process
genes <- gene_list_total[1:100]  # Assuming gene_hold is your full list of genes
# genes <- gene_list_total[which(!gene_list_total %in% results2$Gene)]
# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
#results2 <- final_results
# final_results <- rbind(final_results,results2)
# rm(results2)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"All_PT_Cells_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

# Stop the parallel cluster
stopCluster(cl)
Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")
rm(data_subset,cl,all_results)

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S1/S2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S1_S2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()
```
#### ii. PT-S1/2
```{r}
#PT-S1/S2
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="PT-S1/S2")
ncol(so_celltype) #37948 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total
# genes <- gene_list_total[9000:14627]  # Assuming gene_hold is your full list of genes
# genes <- gene_list_total[which(!gene_list_total %in% results2$Gene)]
# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# chunks19 <- final_results
# results2 <- rbind(results2,final_results)
# final_results <- results2
# rm(results2)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"PT_S1_S2_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))

# Stop the parallel cluster
stopCluster(cl)

# full_results <- read.csv(fs::path(dir.results,"PT_S1_S2_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")
rm(data_subset,cl,all_results)

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S1/S2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S1_S2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
#### iii. PT-S3
```{r}
#PT-S3
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="PT-S3")
ncol(so_celltype) #37948 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"PT_S3_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"PT_S3_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S3, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S3_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()
```
#### iv. aPT
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="aPT")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"aPT_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"aPT_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "aPT, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_aPT_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
###ii. TAL
####i. C-TAL-1
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="C-TAL-1")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"C-TAL-1_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"C-TAL-1_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-(length(unique(final_results$Gene))/length(gene_list_total)))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "C-TAL-1, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_C-TAL-1_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
####i. C-TAL-2
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="C-TAL-2")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"C-TAL-2_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"C-TAL-2_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "C-TAL-2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_C-TAL-2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```

## B. Medication Status - Type 2 Diabetes + OB Controls
### i. All PT Cells - Pseudobulk 
```{r}
so_celltype <- subset(so_kpmp_sc,celltype1=="PT")
ncol(so_celltype) #37948 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
# num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
# cl <- makeCluster(num_cores)
# registerDoParallel(cl)
plan(multisession, workers = 40)

# List of genes to process
genes <- gene_list_total[1:100]  # Assuming gene_hold is your full list of genes
# genes <- gene_list_total[which(!gene_list_total %in% results2$Gene)]
# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
#results2 <- final_results
# final_results <- rbind(final_results,results2)
rm(results2)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"All_PT_Cells_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

# Stop the parallel cluster
stopCluster(cl)
Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")
rm(data_subset,cl,all_results)

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S1/S2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S1_S2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()
```
#### ii. PT-S1/2
```{r}
#PT-S1/S2
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="PT-S1/S2")
ncol(so_celltype) #37948 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total
# genes <- gene_list_total[9000:14627]  # Assuming gene_hold is your full list of genes
# genes <- gene_list_total[which(!gene_list_total %in% results2$Gene)]
# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# chunks19 <- final_results
# results2 <- rbind(results2,final_results)
# final_results <- results2
# rm(results2)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"PT_S1_S2_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))

# Stop the parallel cluster
stopCluster(cl)

# full_results <- read.csv(fs::path(dir.results,"PT_S1_S2_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")
rm(data_subset,cl,all_results)

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S1/S2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S1_S2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
#### iii. PT-S3
```{r}
#PT-S3
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="PT-S3")
ncol(so_celltype) #37948 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"PT_S3_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"PT_S3_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "PT-S3, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_PT_S3_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()
```
#### iv. aPT
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="aPT")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"aPT_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"aPT_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "aPT, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_aPT_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
###ii. TAL
####i. C-TAL-1
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="C-TAL-1")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"C-TAL-1_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"C-TAL-1_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-(length(unique(final_results$Gene))/length(gene_list_total)))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "C-TAL-1, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_C-TAL-1_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```
####i. C-TAL-2
```{r}
so_celltype <- subset(so_kpmp_sc,KPMP_celltype2=="C-TAL-2")
ncol(so_celltype) #9052 cells
nrow(so_celltype) #14627 genes
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list_total <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

#Compare Health Controls to Type 2 Diabetes on no meds
data_subset <- data %>%
  filter(group=="Type_2_Diabetes") 
# filter(medication=="no_med")

data_subset$medication <- factor(data_subset$medication)
data_subset <- data_subset %>% 
  dplyr::select(c("kit_id","medication","group","age","sex","bmi",all_of(gene_list_total)))
rm(data)
gc()
# rm(so_celltype)

# Define your function for processing a gene
process_gene <- function(gene) {
  # Use tryCatch to handle errors
  result <- tryCatch({
    # Define the formula for the linear mixed-effects model
    formula <- as.formula(paste0(gene, " ~ medication + age + sex + bmi + (1 | kit_id)"))
    
    # Fit the model
    model <- suppressWarnings(lmer(formula, data = data_subset))
    
    # Compute emmeans
    emm <- emmeans(model, ~ medication, lmerTest.limit = 19037, pbkrtest.limit = 19037)
    
    # Define the contrasts to test
    contrast_results <- contrast(emm, list(
      "GLP1 vs No Med"        = c(0, 1, 0, -1),
      "SGT2 vs No Med"        = c(0, 0, 1, -1),
      "GLP1+SGT2 vs No Med"   = c(0, 0, 0, 1),
      "GLP1+SGT2 vs GLP1"     = c(0, -1, 1, 0),
      "GLP1+SGT2 vs SGT2"     = c(0, 0, -1, 1)
    ))
    
    # Convert the contrast results to a data frame and add PValue and Gene columns
    contrast_df <- as.data.frame(confint(contrast_results))
    contrast_df$PValue <- summary(contrast_results)$p.value
    contrast_df$Gene <- gene
    contrast_df <- as.data.frame(contrast_df)
    
    return(contrast_df)
  }, error = function(e) {
    # If there's an error, return NULL and print the error message
    message(paste("Error processing gene:", gene, "\n", e))
    return(NULL)  # Return NULL if an error occurs
  })
  
  return(result)
}

# Set up parallel backend (use maximum available cores)
num_cores <- detectCores() - 1  # Reserve one core for the system, adjust as needed
cl <- makeCluster(num_cores)
registerDoParallel(cl)

# List of genes to process
genes <- gene_list_total  # Assuming gene_hold is your full list of genes

# Process genes in chunks of 1000
chunk_size <- 1000
num_chunks <- ceiling(length(genes) / chunk_size)

# Initialize an empty list to store all results
all_results <- list()

# Loop over each chunk
for (i in 1:num_chunks) {
  # Determine the start and end indices for the current chunk
  start_idx <- ((i - 1) * chunk_size) + 1
  end_idx <- min(i * chunk_size, length(genes))
  
  # Get the current chunk of genes
  current_genes <- genes[start_idx:end_idx]
  
  # Run the analysis for this chunk in parallel
  chunk_results <- foreach(gene = current_genes, .combine = rbind, .packages = c("lme4", "emmeans", "dplyr")) %dopar% {
    process_gene(gene)
  }
  
  # Remove any NULL results (in case of errors)
  chunk_results <- chunk_results[!sapply(chunk_results, is.null), ]
  
  # Add the results of the current chunk to the list of all results
  all_results[[i]] <- chunk_results
  
  # Save the current chunk's results to a CSV file
  # write.csv(chunk_results, file = fs::path(dir.results,paste0("Full_results_chunk_", i, "_04_07_2025.csv")))
  
  # Optionally, print progress
  cat("Finished chunk", i, "out of", num_chunks, "\n")
}

# Combine all chunks into a single data frame
final_results <- do.call(rbind, all_results)
# Stop the parallel cluster
stopCluster(cl)
# Save the final combined results
write.csv(final_results, file = fs::path(dir.results,"C-TAL-2_T2D_Medication_Full_results_all_genes_04_14_2025.csv"))
# final_results <- read.csv(fs::path(dir.results,"C-TAL-2_T2D_Medication_Full_results_all_genes_04_07_2025.csv"))

Nonconvergence_Rate <- paste0(round((1-length(unique(final_results$Gene))/length(gene_list_total))*100,3),"%")

#Visulaize the results
# #Make volcano plot of all gene results for group
full_results <- final_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# group_by(contrast) %>% 
# mutate(fdr=p.adjust(PValue,method="fdr"))  %>% 
# ungroup()
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)

# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$estimate > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$estimate < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

Genes <- length(unique(full_results$Gene))
Cells <- ncol(so_celltype)

#Figure Range
# max <- max(significant_df$Beta,na.rm=T)
# min <- min(significant_df$Beta,na.rm=T)

max <- max(full_results$estimate,na.rm=T)
min <- min(full_results$estimate,na.rm=T)

full_results$contrast <- factor(full_results$contrast)
# Reorder the contrast variable
full_results$contrast <- factor(full_results$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))
significant_df$contrast <- factor(significant_df$contrast, levels = c("GLP1 vs No Med", "SGT2 vs No Med", "GLP1+SGT2 vs No Med","GLP1+SGT2 vs GLP1","GLP1+SGT2 vs SGT2"))

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = estimate, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
facet_wrap(~ contrast, nrow = 2, ncol = 3)+
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Comparison of Medication Status among Youth with T2D",
    subtitle = "C-TAL-2, Adjusted for Age, Sex & BMI",
    x = "Fold Change",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Cells,", Non-Convergence Rate: ",Nonconvergence_Rate)
  ) +
  # xlim(min,max)+
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot
pdf(fs::path(dir.results,"Plot_C-TAL-2_T2D_Medications_lmm_no_zi.pdf"),width=12,height=10)
print(volcano_plot)
dev.off()

```