---
title: "Liver_snRNAseq"
output: html_document
date: "2025-04-23"
---

# 1. Set up Libraries & Directores
```{r libraries, echo=F, include = F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
# library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
# library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(ggplot2)
library(GGally)
# library(limma)
# library(reshape2)
# library(data.table)
# library(knitr)
# library(TxDb.Hsapiens.UCSC.hg19.knownGene)
# library(stringr)
# library(NMF)
# library(rsvd)
# library(RColorBrewer)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
# library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
library(future)
# registerDoParallel(cores = 6)
library(VennDiagram)
#install.packages("survival")
library(survival)
#install.packages("lme4")  # If not already installed
library(lme4)
#install.packages("lmerTest")
library(lmerTest)
# install.packages("emmeans")
library(emmeans)
# install.packages("glmmTMB")
library(glmmTMB)
library(ggrepel)
# library(qpcR)
library(ggpubr)
# BiocManager::install("enrichplot")
# library(enrichplot)
library(enrichR)
dbs <- c("GO_Biological_Process_2023", 
         "KEGG_2021_Human",
         # "Reactome_2022", 
         "Reactome_Pathways_2024",
         # "MSigDB_Oncogenic_Signatures",
         # "MSigDB_Computational",
         "MSigDB_Hallmark_2020")

#Local file path
# dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
# dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
# dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c("/Users/hhampson/Documents/UW/1_Ongoing Projects/Liver scRNAseq/2_Results")

#Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c(fs::path(dir.dat,"Liver project/Results"))

# #Mac studio
# #Mac Studio File Path
# dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
# dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/Results")
# dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Laura's Mac Studio File Path
dir.dat <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/Results")

#Load functions
source("Liver_functions.R")
# source("/Users/hhampson/CHCO-Code/Petter Bjornstad/Data Processing and Analysis/Standard_Functions.R")


knitr::opts_knit$set(root.dir = dir.results)
```

# 2. Pre-Processing & Qualtiy Control
## a. Load Full Data & Clean:## a. snRNA & MetaData
```{r, include = F, fig.height=7,fig.width=10}
# Liver snRNA data processing
invisible(gc())
#Local
# dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Biostatistics Core Shared Drive")
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
# #Lambda
# so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
DefaultAssay(so_liver_sn) <- "RNA"
dim(so_liver_sn)
nrow(so_liver_sn) #36,601 genes
ncol(so_liver_sn) # 130,124 nuceli
invisible(gc())
#Local
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#Lambda
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
unique(meta_liver_raw$Kit_Lot)
meta_liver_raw <- meta_liver_raw %>%
  filter(StudyID!="IT2D-18") %>%
  filter(StudyID!="IT2D-16")
meta_liver_raw <- meta_liver_raw %>%
  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "glp1_or_sglt2",
                          neither == "Yes" ~ "no_med",
                          hc == "Yes" ~ "obese_control"))
# meta_liver_raw <- meta_liver_raw %>%
#   mutate(group3=ifelse(either=="Yes","either","neither"))

#Create Liver Status Variables
meta_liver_raw <- meta_liver_raw %>%
  mutate(steatosis_cat = ifelse((steatosis_grade==0 | steatosis_grade==1),"Low Steatosis (0+1)","High Steatosis (2+3)")) %>%
  mutate(fibrosis_cat = case_when(steatosis_grade==0 & fibrosis_stage==0 ~ "No Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage==0 ~ "Any Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage>0 ~ "Any Fibrosis & Any Steatosis"))

meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"
invisible(gc())

#Create liver disease and drug groups
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
# mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
# mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
#  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
#  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
#  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))

# Create a single grouping variable
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(group2=case_when(both == "Yes" ~ "both",
#                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
#                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
#                          either == "Yes" ~ "either",
#                          neither == "Yes" ~ "neither",
#                          hc == "Yes" ~ "obese_control"))

#so_liver_sn@meta.data$group2 <- factor(so_liver_sn@meta.data$group2, levels = c("obese_control","neither","sglt2_exclusive", #"glp1_exclusive","either","both"))

# #Create Diabetes only seurat object
# so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")

#Create senesence so in all cell types
#sens_gene_in_data <- intersect(sens_genes, rownames(so_liver_sn))
# Subset the Seurat object to include only genes in sens_gene_in_data
#so_sens_all <- subset(so_liver_sn, features = sens_gene_in_data)

# #Filter to hepatocytes only
# so_liver_sn$Hepatocyte <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Yes","No")
# so_hep <- subset(so_liver_sn, Hepatocyte=="Yes")
# rm(so_liver_sn)
# DefaultAssay(so_hep) <- "RNA"  

dim(so_liver_sn) #36601 130124
# so_object <- qc_function_sn(so=so_liver_sn,var_pct=0.75)
# so_filtered <- so_object$so
# print(so_object$elbow_plot)
# 
# #Number of Cells after filtering: 
# cell_total2 <- ncol(so_filtered) #14733
# #Number of Genes after filtering: 
# gene_total2 <- nrow(so_filtered) #8966
# print(paste0(cell_total1," cells before filtereing, ",cell_total2," cells after filtering (removed ",cell_total1-cell_total2," cells); ",gene_total1," genes before filtering, ",gene_total2," genes after filtering (removed ",gene_total1-gene_total2," genes)"))
# 
# #Visualize filtered seurat object on umaps by cell type and diabetes group
# umap_vis(so=so_filtered ,cell_type="KPMP_celltype",group_variable="group") 
# 
# #Remove unfiltered seurat object
# rm(so_object,so_kpmp_sc)

# # Step 1: Filter to genes expressed in more than 5% of cells
# gene_expression_matrix <- so_liver_sn@assays$RNA@counts
# # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05
# 
# so_liver_sn <- subset(so_liver_sn, features = rownames(gene_expression_matrix)[expressed_genes])
# dim(so_liver_sn@assays$RNA@counts) #4502 130124
# dim(so_liver_sn@assays$RNA) #4502 130124
# dim(so_liver_sn) #4502 130124
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #13
# ncol(so_liver_sn) #130124 cells
# nrow(so_liver_sn) # 36601 genes
ncol(so_liver_sn) #130124 nuclei 
nrow(so_liver_sn) # 36601 genes
# 
# #Code for filtering by count
# # Create a matrix of the expression data (e.g., counts or scaled data) from the Seurat object
# expression_data <- GetAssayData(so_liver_sn, assay = "RNA", layer = "counts")
# # Calculate the number of nuclei each gene is expressed in (i.e., count non-zero entries per gene)
# gene_expression_counts <- rowSums(expression_data > 0)
# # Filter out genes that are expressed in less than 10 nuclei
# filtered_genes <- names(which(gene_expression_counts >= 10))
# length(filtered_genes) #28977 genes remain of 36601
# rare_genes <- names(which(gene_expression_counts < 10))
# length(rare_genes) #7624

# # Subset the Seurat object to keep only genes that are expressed in at least 10 nuclei
# so_liver_sn <- subset(so_liver_sn, features = filtered_genes)

#YE JI's filtering code for percent expression 
#Filter out rare genes expressed in less than "gene_pct" of cells
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, layer = "counts"))
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, assay = "RNA", layer = "counts"))
expr_matrix <- so_liver_sn@assays$RNA@counts
# Calculate the proportion of cells expressing each gene
num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
total_cells <- ncol(expr_matrix)  # Total number of cells
gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
remove(expr_matrix)
# Keep genes expressed in at least "gene_pct" of cells
genes_to_keep <- names(gene_proportion[gene_proportion >= 0.01])
so_liver_sn <- subset(so_liver_sn, features = genes_to_keep)
ncol(so_liver_sn) #130,124 nuclei
nrow(so_liver_sn) # 11313 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_liver_sn), value = TRUE)
#keep_ids <- unique(rownames(so_liver_sn)[which(!rownames(so_liver_sn) %in% mito_genes)])
# so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn), mito_genes))
# so_liver_sn <- subset(so_liver_sn,kit_id!="KL-0029535")
#so_liver_sn$Gene <- rownames(so_liver_sn)
#so_liver_sn <- subset(so_liver_sn, Gene %in% keep_ids)
so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
#so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
# so_liver_sn <- subset(so_liver_sn, !rownames(so_liver_sn) %in% mito_genes)
# grep("^MT-", rownames(so_liver_sn@assays$RNA@counts), value = TRUE)
dim(so_liver_sn@assays$RNA@counts) #9276 186125
dim(so_liver_sn@assays$RNA@data) #9276 186125
dim(so_liver_sn@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #0
ncol(so_liver_sn) #130124 cells
nrow(so_liver_sn) #11300 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_liver_sn@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_liver_sn@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_liver_sn@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_liver_sn <- so_liver_sn[, cells_to_keep]



# # Check dimensions after filtering
# dim(so_liver_sn@assays$RNA@counts) #9276 81882
# dim(so_liver_sn@assays$RNA)#9276 81882
# dim(so_liver_sn)#9276 81882
# 
# # Optionally, you can also check how many MT genes remain in the filtered dataset
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts)))
# ncol(so_liver_sn)  # Number of cells remaining
# 
# # Step 3: Renormalize the Seurat object after filtering
# so_liver_sn <- NormalizeData(so_liver_sn)
# 
# #Check normalized slot
# dim(so_liver_sn@assays$RNA@data) #5714 186125
# sum(grepl("^MT", rownames(so_liver_sn@assays$RNA@data))) #70
# ncol(so_liver_sn) #186125 cells
# nrow(so_liver_sn) #5714 genes
# 
# rm(gene_expression_matrix)
so_liver_sn <- NormalizeData(so_liver_sn)
so_liver_sn <- ScaleData(so_liver_sn, features = VariableFeatures(so_liver_sn))

#Create general hepatocyte cell type variable
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype) | grepl("dHep",so_liver_sn$celltype),"Hep","Non-Hep")
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

# table(so_liver_sn$celltype,so_liver_sn$celltype2)

# rm(expression_data)
```

## b. Explore Distribution & Zero-Inflation
```{r}
#Check if data have been normalized 
so_liver_sn@assays  # Should show e.g., "RNA" with "counts" and "data"
head(GetAssayData(so_liver_sn, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_liver_sn, layer = "data")[, 1:5])    # Normalized data

#Check for normality
# Open a PDF device (all plots go here)
pdf(fs::path(dir.results,"Normalization_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesN <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Normalized Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

pdf(fs::path(dir.results,"Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesR <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Try log transforming
log_data <- log1p(GetAssayData(so_liver_sn, layer = "data"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Normalized_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesN
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

log_data <- log1p(GetAssayData(so_liver_sn, layer = "count"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesR
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Check for zero inflation
pdf(fs::path(dir.results,"Zero_Inflation_Visualization.pdf"), width = 8, height = 6)
counts <- GetAssayData(so_liver_sn, layer = "counts")
zero_prop <- rowMeans(counts == 0)
hist(zero_prop, main = "Proportion of Zeros for Genes", xlab = "Zero Proportion")
# Add a vertical line at 50%
abline(v = 0.5, col = "red", lwd = 2, lty = 2)
dev.off()

#Check for overdispersion
# Use raw counts
counts <- GetAssayData(so_liver_sn, layer = "counts")

# Pick a few genes or do this genome-wide
gene_means <- rowMeans(counts)
gene_vars <- apply(counts, 1, var)

# Dispersion estimate (Var / Mean)
dispersion <- gene_vars / gene_means

# Filter to only genes where mean > 0 (to avoid divide-by-zero)
valid_genes <- gene_means > 0
dispersion <- dispersion[valid_genes]

# View distribution
pdf(fs::path(dir.results,"Overdispersion_Visualization.pdf"), width = 8, height = 6)
hist(dispersion, breaks=200, main="Dispersion across genes", xlab="Dispersion (Var / Mean)",xlim = c(0, 2))
abline(v = 1, col="red")  # Poisson expectation
dev.off()

# Check how many genes have dispersion > 1
overdispersed_genes <- sum(dispersion > 1)
total_genes <- length(dispersion)
percent_overdispersed <- (overdispersed_genes / total_genes) * 100

# Print summary and recommendation
print(paste0("Overdispersed genes: ", 
             overdispersed_genes, 
             " out of ", 
             total_genes, 
             " (", 
             round(percent_overdispersed, 3), 
             "%)"))

if (percent_overdispersed > 10) {
  print("✅ Overdispersion detected. Consider using a Negative Binomial model.")
} else {
  print("ℹ️ Little evidence of overdispersion. Poisson model may be sufficient.")
}
rm(counts)
```

## c. Select Highly Variable Genes (HVGs)
```{r}
so_liver_sn <- FindVariableFeatures(so_liver_sn, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_liver_sn)

# Subset Seurat object to only HVGs
so_liver_sn_hvg <- subset(so_liver_sn, features = hvgs)
```

## d. Evaluate Covariates before Adjustment
```{r}
#Categorical Covariates
#Get metadata for everyone
# dat <- so_liver_sn@meta.data %>%
#   group_by(StudyID) %>%
#   summarise(across(everything(), first)) %>%
#   ungroup() 
dat <- so_liver_sn@meta.data %>%
  group_by(StudyID) %>%
  summarise(across(everything(), ~ .x[1])) %>%
  ungroup()

dat$study <- case_when(grepl("IT2D",dat$StudyID) ~ "IMPROVE",
                       grepl("MANGO",dat$StudyID) ~ "MANGO",
                       grepl("BSA",dat$StudyID) ~ "BIOBANK")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

dat <- dat %>% 
  mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "glp1_or_sglt2",
                          neither == "Yes" ~ "no_med",
                          hc == "Yes" ~ "obese_control"))

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting
# dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (2+3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD + diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither| steatosis_cat, data=dat)
table1(~ alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)
table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither + alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)

#Examine covariates
age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
  diagnosis_of_diabetes + diagnosis_of_MASLD+
  sglt2 + glp1agonist
cat_covariates <- c("nih_sex", "nih_race","nih_ethnicity", "diagnosis_of_diabetes", "diagnosis_of_MASLD", "both","glp1_exclusive","sglt2_exclusive","neither","fibrosis_cat")
plot_list <- list()

for (covariate in cat_covariates) {
  p <- ggplot(dat, aes_string(x = covariate, fill = "steatosis_cat")) +
    geom_bar(position = "fill") +  # use "dodge" for absolute counts
    labs(title = paste0("Distribution of ", covariate, " by Steatosis Category"),
         y = "Proportion", x = NULL) +
    scale_y_continuous(labels = scales::percent_format()) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust depending on how many you have)
wrap_plots(plot_list, ncol = 2)


#Continuous Covariates
con_covariates <- c("age", "bmi", "triglycerides", "hba1c", "pah_clear_bsa", "eGFR_CKD_epi", "acr_u")
plot_list <- list()

for (covariate in con_covariates) {
  p <- ggplot(dat, aes_string(x = "group", y = covariate, fill = "group")) +
    geom_boxplot(outlier.shape = NA, alpha = 0.7) +
    geom_jitter(width = 0.2, alpha = 0.4, color = "black") +  # optional: show points
    labs(title = paste0("Distribution of ", covariate, " by group"),
         x = "Group", y = covariate) +
    theme_minimal() +
    theme(legend.position = "none")
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust ncol/nrow as needed)
wrap_plots(plot_list, ncol = 4)
```
