---
  title: "Liver scRNA analysis HEH"
author: "Hailey Hampson"
date: "2024-09-24"
output: html_document
--- 
  
```{r libraries, echo=F, include = F}
# Load necessary package for Bioconductor management
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# List of CRAN packages to install
cran_packages <- c(
  "reprex", "tidyverse", "arsenal", "dplyr", "ggplot2", "ggrepel",
  "Seurat", "future", "colorspace", "patchwork", "ggdendro",
  "cowplot", "ggpubr", "rstatix", "table1", "kableExtra",
  "knitr", "msigdbr", "openxlsx"
)

# Install CRAN packages
install.packages(cran_packages)

# List of Bioconductor packages to install
bioc_packages <- c(
  "Biobase", "ReactomeGSA", "GSEABase", "SingleCellExperiment",
  "fgsea", "EnhancedVolcano"
)

# Install Bioconductor packages
BiocManager::install(bioc_packages)

# Install 'presto' package from GitHub
#install.packages("devtools")
#devtools::install_github("immunogenomics/presto")

library(reprex)
library(tidyverse)
library(BiocManager)
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
# install.packages('devtools')
# devtools::install_github('immunogenomics/presto')
# BiocManager::install('EnhancedVolcano')
library(EnhancedVolcano)
# install.packages("openxlsx")
library(openxlsx)
# install.packages("fgsea")
# library(fgsea)

#Lambda file path
dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
dir.code <- c("/home/Petter Bjornstad/Liver analysis/Liver scRNAseq")
dir.results <- c(fs::path(dir.code,"Results"))

plan()
future::plan("sequential")
options(future.globals.maxSize = 3e9)

```

# scRNA
```{r echo = F}
# Liver scRNA data processing
# so_liver_sc <- readRDS("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_raw/PB_Liver_7SingleCellDatasets.RDS")
# so_liver_sc <- readRDS("/Users/hhampson/Dropbox/Bjornstad data/Liver/PB_Liver_7SingleCellDatasets.RDS")
so_liver_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_Liver_7SingleCellDatasets.RDS"))

# Add missing liver meta data to Seurat object
# meta_liver_raw <- read.csv("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean/liver_biopsy_metadata_PN.csv")
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))

meta_liver_sc <-  so_liver_sc@meta.data[,1:9] %>%
  dplyr::mutate(Cryostor_ID = gsub("-Liv", "", orig.ident)) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sc) <- rownames(so_liver_sc@meta.data)

so_liver_sc <- AddMetaData(so_liver_sc, meta_liver_sc)
so_liver_sc <- NormalizeData(so_liver_sc)
so_liver_sc <- ScaleData(so_liver_sc)
# PCA
so_liver_sc <- RunPCA(so_liver_sc, features = VariableFeatures(object = so_liver_sc))
ElbowPlot(so_liver_sc)
# Cluster cells
so_liver_sc <- FindNeighbors(so_liver_sc)
so_liver_sc <- FindClusters(so_liver_sc)
# Perform UMAP and tSNE
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap") 
```

# snRNA
```{r echo = F}
# Liver snRNA data processing
# so_liver_sn <- readRDS("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_raw/NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS")
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))

meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))

meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"

gc()
# Normalize the RNA assay
so_liver_sn <- NormalizeData(so_liver_sn, assay = "RNA")
so_liver_sn <- ScaleData(so_liver_sn, assay = "RNA")
gc()
# Scale the integrated assay (no normalization needed)
so_liver_sn <- ScaleData(so_liver_sn, assay = "integrated")

so_liver_sn@meta.data$hepatocyte <- ifelse(grepl("Hep",so_liver_sn@meta.data$celltype),"Hepatocyte","Non-Hepatocyte")

```

## snRNA UMAP

```{r echo=F}
# Set a higher memory limit
# mem.maxVSize(64000000000)

# PCA
# so_liver_sn <- RunPCA(so_liver_sn, features = VariableFeatures(object = so_liver_sn),assay="integrated")
so_liver_sn <- FindVariableFeatures(object = so_liver_sn)
so_liver_sn <- RunPCA(so_liver_sn, features = VariableFeatures(object = so_liver_sn),assay="RNA")
ElbowPlot(so_liver_sn)
# Cluster cells
# so_liver_sn_int <- FindNeighbors(so_liver_sn_int)
# so_liver_sn_int <- FindClusters(so_liver_sn_int)

# Find neighbors and clusters (again using integrated data)
# so_liver_sn <- FindNeighbors(so_liver_sn, assay = "integrated", dims = 1:15)
so_liver_sn <- FindNeighbors(so_liver_sn, assay = "RNA", dims = 1:15)
so_liver_sn <- FindClusters(so_liver_sn, resolution = 0.5)

# UMAP for visualization
# so_liver_sn <- RunUMAP(so_liver_sn, assay = "integrated", dims = 1:15)
so_liver_sn <- RunUMAP(so_liver_sn, assay = "RNA", dims = 1:15)

# Perform UMAP and tSNE
# so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sn, reduction = "umap", raster = F) 

```

## Obese vs. Obese + T2D
### Descriptive table
```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_diabetes", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## MASLD Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_MASLD", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## OSA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_osa", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## GLP-1RA Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- liver_meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("glp1agonist", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Insulin Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("insulin", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

## Kidney disease Y/N
### Descriptive table

```{r echo = F}
# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_kidney_disease", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```


## Stratified UMAP scRNA 
### Obese vs. Obese + T2D
```{r echo = F}
so_liver_sn <- RunUMAP(so_liver_sn, dims = 1:15)
DimPlot(so_liver_sn, reduction = "umap", group.by = "diagnosis_of_diabetes") 

```

### MASLD Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_MASLD") 

```

### OSA Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_osa") 

```

### GLP-1RA Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "glp1agonist") 

```

### Insulin Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "insulin") 

```

### Kidney disease Y/N
```{r echo = F}
so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sc, reduction = "umap", group.by = "diagnosis_of_kidney_disease") 

```

## Differential Expression by Group
### Obese vs. Obese + T2D
```{r}
genes <- unique(rownames(so_liver_sn))
Idents(so_liver_sn) <- so_liver_sn$celltype

de.markers(so_liver_sn, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", NULL, "_top")
de.markers(so_liver_sn, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hep-1", "_top")
de.markers(so_liver_sn, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sn, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sn) <- so_liver_sn$hepatocyte
gc()
so_liver_sn_hep <- subset(so_liver_sn, hepatocyte == "Hepatocyte")
gc()
memory.limit(size = 64000)
de.markers(so_liver_sn_hep, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", "Hepatocyte", "")
```

### MASLD Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "diagnosis_of_MASLD", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### OSA Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "diagnosis_of_osa", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### GLP-1RA Y/N
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte3", "")

Idents(so_liver_sc) <- so_liver_sc$hepatocyte
de.markers(so_liver_sc, genes, "glp1agonist", id2 = "No", id1 = "Yes", "Hepatocyte", "")

```

### Insulin Y/N
```{r}
# Idents(so_liver_sc) <- so_liver_sc$celltype
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_liver_sc) <- so_liver_sc$hepatocyte
# de.markers(so_liver_sc, genes, "insulin", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

### Kidney disease Y/N
```{r}
# Idents(so_liver_sc) <- so_liver_sc$celltype
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte1", "")
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte2", "")
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte3", "")
# 
# Idents(so_liver_sc) <- so_liver_sc$hepatocyte
# de.markers(so_liver_sc, genes, "diagnosis_of_kidney_disease", id2 = "No", id1 = "Yes", "Hepatocyte", "")
#All no's for insulin
```

## Figures
### Obese vs. Obese + T2D
```{r}
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### MASLD Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

### OSA Y/N
```{r}
split.vp(seurat_object = so, genes = genes_subset, filepath = "/Users/hhampson/Library/Mobile Documents/com~apple~CloudDocs/UW/1_Ongoing Projects/Liver scRNAseq")
Idents(so_liver_sc) <- so_liver_sc$celltype
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte1", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte2", m = m)
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte3", m = m)
Idents(so_liver_sc) <- so_liver_sc$hepatocyte
pt.plot_table <- dp.formatted(seurat_object = so_liver_sc, genes = genes, group.by = "diagnosis_of_diabetes",
                              celltype = "Hepatocyte", m = m)
```

# Gene set enrichment analysis
```{r echo = F}
sce_sn_hep <- as.SingleCellExperiment(so_liver_sn_hep)
## C2 category is according to canonical pathways: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4707969/pdf/nihms-743907.pdf
geneSets <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG")
### filter background to only include genes that we assessed
geneSets$gene_symbol <- toupper(geneSets$gene_symbol)
geneSets <- geneSets[geneSets$gene_symbol %in% names(sce_sn_hep),]
m_list <- geneSets %>% split(x = .$gene_symbol, f = .$gs_name)
stats <- m$p_val_adj
names(stats) <- rownames(m)
eaRes <- fgsea(pathways = m_list, stats = na.omit(stats))
ooEA <- order(eaRes$pval, decreasing = FALSE)
kable(head(eaRes[ooEA, 1:7], n = 20))
```