---
title: "Liver Project"
author: "Hailey Hampson"
date: "2025-02-10"
output:
  word_document: default
  html_document: default
  pdf_document: default
---

# 1. Set up Libraries & Directores
```{r libraries, echo=F, include = F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
library(GSEABase)
library(limma)
library(reshape2)
library(data.table)
library(knitr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
library(NMF)
library(rsvd)
library(RColorBrewer)
library(MAST)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
# library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
library(future)
# registerDoParallel(cores = 6)
library(VennDiagram)
#install.packages("survival")
library(survival)
#install.packages("lme4")  # If not already installed
library(lme4)
#install.packages("lmerTest")
library(lmerTest)
# install.packages("emmeans")
library(emmeans)
# install.packages("glmmTMB")
library(glmmTMB)
library(ggrepel)
# library(qpcR)
library(ggpubr)
# BiocManager::install("enrichplot")
library(enrichplot)
library(enrichR)
dbs <- c("GO_Biological_Process_2023", 
         "KEGG_2021_Human",
         "Reactome_2022", 
         "Reactome_Pathways_2024")

#Local file path
# dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
# dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
# dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c("/Users/hhampson/Documents/UW/1_Ongoing Projects/Liver scRNAseq/2_Results")

#Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c(fs::path(dir.dat,"Liver project/Results"))

#Mac studio
#Mac Studio File Path
dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/Results")
dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Load functions
source("Liver_functions.R")
source("/Users/hhampson/CHCO-Code/Petter Bjornstad/Data Processing and Analysis/Standard_Functions.R")


knitr::opts_knit$set(root.dir = dir.results)
```

# 2. Load Full Data & Clean: 
## a. snRNA & MetaData
```{r, include = F, fig.height=7,fig.width=10}
# Liver snRNA data processing
invisible(gc())
#Local
# so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
#Lambda
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))

invisible(gc())
#Local
# meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#Lambda
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
unique(meta_liver_raw$Kit_Lot)
meta_liver_raw <- meta_liver_raw %>%
  filter(StudyID!="IT2D-18") %>%
  filter(StudyID!="IT2D-16")
meta_liver_raw <- meta_liver_raw %>%
  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group2=case_when(both == "Yes" ~ "both",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "either",
                          neither == "Yes" ~ "neither",
                          hc == "Yes" ~ "obese_control"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group3=ifelse(either=="Yes","either","neither"))

#Create Liver Status Variables
meta_liver_raw <- meta_liver_raw %>%
  mutate(steatosis_cat = ifelse((steatosis_grade==0 | steatosis_grade==1),"Low Steatosis (0+1)","High Steatosis (2+3)")) %>%
  mutate(fibrosis_cat = case_when(steatosis_grade==0 & fibrosis_stage==0 ~ "No Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage==0 ~ "Any Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage>0 ~ "Any Fibrosis & Any Steatosis"))

meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"
invisible(gc())

#Create liver disease and drug groups
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
# mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
# mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
#  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
#  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
#  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))

# Create a single grouping variable
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(group2=case_when(both == "Yes" ~ "both",
#                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
#                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
#                          either == "Yes" ~ "either",
#                          neither == "Yes" ~ "neither",
#                          hc == "Yes" ~ "obese_control"))

#so_liver_sn@meta.data$group2 <- factor(so_liver_sn@meta.data$group2, levels = c("obese_control","neither","sglt2_exclusive", #"glp1_exclusive","either","both"))

# #Create Diabetes only seurat object
# so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")

#Create senesence so in all cell types
#sens_gene_in_data <- intersect(sens_genes, rownames(so_liver_sn))
# Subset the Seurat object to include only genes in sens_gene_in_data
#so_sens_all <- subset(so_liver_sn, features = sens_gene_in_data)

# #Filter to hepatocytes only
# so_liver_sn$Hepatocyte <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Yes","No")
# so_hep <- subset(so_liver_sn, Hepatocyte=="Yes")
# rm(so_liver_sn)
# DefaultAssay(so_hep) <- "RNA"  

dim(so_liver_sn) #36601 130124
# so_object <- qc_function_sn(so=so_liver_sn,var_pct=0.75)
# so_filtered <- so_object$so
# print(so_object$elbow_plot)
# 
# #Number of Cells after filtering: 
# cell_total2 <- ncol(so_filtered) #14733
# #Number of Genes after filtering: 
# gene_total2 <- nrow(so_filtered) #8966
# print(paste0(cell_total1," cells before filtereing, ",cell_total2," cells after filtering (removed ",cell_total1-cell_total2," cells); ",gene_total1," genes before filtering, ",gene_total2," genes after filtering (removed ",gene_total1-gene_total2," genes)"))
# 
# #Visualize filtered seurat object on umaps by cell type and diabetes group
# umap_vis(so=so_filtered ,cell_type="KPMP_celltype",group_variable="group") 
# 
# #Remove unfiltered seurat object
# rm(so_object,so_kpmp_sc)

# # Step 1: Filter to genes expressed in more than 5% of cells
# gene_expression_matrix <- so_liver_sn@assays$RNA@counts
# # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05
# 
# so_liver_sn <- subset(so_liver_sn, features = rownames(gene_expression_matrix)[expressed_genes])
# dim(so_liver_sn@assays$RNA@counts) #4502 130124
# dim(so_liver_sn@assays$RNA) #4502 130124
# dim(so_liver_sn) #4502 130124
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #13
# ncol(so_liver_sn) #130124 cells
# nrow(so_liver_sn) # 36601 genes
ncol(so_liver_sn) #130124 nuclei 
nrow(so_liver_sn) # 36601 genes
# 
# #Code for filtering by count
# # Create a matrix of the expression data (e.g., counts or scaled data) from the Seurat object
# expression_data <- GetAssayData(so_liver_sn, assay = "RNA", layer = "counts")
# # Calculate the number of nuclei each gene is expressed in (i.e., count non-zero entries per gene)
# gene_expression_counts <- rowSums(expression_data > 0)
# # Filter out genes that are expressed in less than 10 nuclei
# filtered_genes <- names(which(gene_expression_counts >= 10))
# length(filtered_genes) #28977 genes remain of 36601
# rare_genes <- names(which(gene_expression_counts < 10))
# length(rare_genes) #7624

# # Subset the Seurat object to keep only genes that are expressed in at least 10 nuclei
# so_liver_sn <- subset(so_liver_sn, features = filtered_genes)

#YE JI's filtering code for percent expression 
#Filter out rare genes expressed in less than "gene_pct" of cells
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, layer = "counts"))
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, assay = "RNA", layer = "counts"))
expr_matrix <- so_liver_sn@assays$RNA@counts
# Calculate the proportion of cells expressing each gene
num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
total_cells <- ncol(expr_matrix)  # Total number of cells
gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
remove(expr_matrix)
# Keep genes expressed in at least "gene_pct" of cells
genes_to_keep <- names(gene_proportion[gene_proportion >= 0.01])
so_liver_sn <- subset(so_liver_sn, features = genes_to_keep)
ncol(so_liver_sn) #130,124 nuclei
nrow(so_liver_sn) # 11313 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_liver_sn), value = TRUE)
#keep_ids <- unique(rownames(so_liver_sn)[which(!rownames(so_liver_sn) %in% mito_genes)])
# so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn), mito_genes))
# so_liver_sn <- subset(so_liver_sn,kit_id!="KL-0029535")
#so_liver_sn$Gene <- rownames(so_liver_sn)
#so_liver_sn <- subset(so_liver_sn, Gene %in% keep_ids)
so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
#so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
# so_liver_sn <- subset(so_liver_sn, !rownames(so_liver_sn) %in% mito_genes)
# grep("^MT-", rownames(so_liver_sn@assays$RNA@counts), value = TRUE)
dim(so_liver_sn@assays$RNA@counts) #9276 186125
dim(so_liver_sn@assays$RNA@data) #9276 186125
dim(so_liver_sn@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #0
ncol(so_liver_sn) #130124 cells
nrow(so_liver_sn) #11300 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_liver_sn@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_liver_sn@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_liver_sn@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_liver_sn <- so_liver_sn[, cells_to_keep]



# # Check dimensions after filtering
# dim(so_liver_sn@assays$RNA@counts) #9276 81882
# dim(so_liver_sn@assays$RNA)#9276 81882
# dim(so_liver_sn)#9276 81882
# 
# # Optionally, you can also check how many MT genes remain in the filtered dataset
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts)))
# ncol(so_liver_sn)  # Number of cells remaining
# 
# # Step 3: Renormalize the Seurat object after filtering
# so_liver_sn <- NormalizeData(so_liver_sn)
# 
# #Check normalized slot
# dim(so_liver_sn@assays$RNA@data) #5714 186125
# sum(grepl("^MT", rownames(so_liver_sn@assays$RNA@data))) #70
# ncol(so_liver_sn) #186125 cells
# nrow(so_liver_sn) #5714 genes
# 
# rm(gene_expression_matrix)
so_liver_sn <- NormalizeData(so_liver_sn)
so_liver_sn <- ScaleData(so_liver_sn, features = VariableFeatures(so_liver_sn))

#Create general hepatocyte cell type variable
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype) | grepl("dHep",so_liver_sn$celltype),"Hep","Non-Hep")
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

# table(so_liver_sn$celltype,so_liver_sn$celltype2)

# rm(expression_data)
```

## b. scRNA & MetaData 
```{r, include=F}
invisible(gc())
so_liver_sc <- readRDS(fs::path(dir.dat,"scRNA","data_raw","PB_Liver_7SingleCellDatasets.RDS"))
invisible(gc())

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sc) <- "RNA"
invisible(gc())

#Get metadata for each individual with sc data
#Lambda
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
so_liver_sc$Cryostor_ID <- str_remove(unique(so_liver_sc@meta.data$orig.ident),"-Liv")
liver_ids <- str_remove(unique(so_liver_sc@meta.data$orig.ident),"-Liv")
liver_ids %in% meta_liver_raw$Cryostor_ID

meta_liver_raw <- meta_liver_raw %>%
  filter(Cryostor_ID %in% liver_ids)

meta_liver_raw <- meta_liver_raw %>%
  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group2=case_when(both == "Yes" ~ "both",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "either",
                          neither == "Yes" ~ "neither",
                          hc == "Yes" ~ "obese_control"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group3=ifelse(either=="Yes","either","neither"))

#Create Liver Status Variables
meta_liver_raw <- meta_liver_raw %>%
  mutate(steatosis_cat = ifelse((steatosis_grade==0 | steatosis_grade==1),"Low Steatosis (0+1)","High Steatosis (2+3)")) %>%
  mutate(fibrosis_cat = case_when(steatosis_grade==0 & fibrosis_stage==0 ~ "No Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage==0 ~ "Any Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage>0 ~ "Any Fibrosis & Any Steatosis"))



meta_liver_sc <-  so_liver_sc@meta.data[,c(9,123)] %>%
  #dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw,by=c("Cryostor_ID"))

rownames(meta_liver_sc) <- rownames(so_liver_sc@meta.data)
so_liver_sc <- AddMetaData(so_liver_sc, meta_liver_sc)
rm(meta_liver_sc,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sc) <- "RNA"
invisible(gc())


dim(so_liver_sc) #23697  7383

# # Step 1: Filter to genes expressed in more than 5% of cells
# gene_expression_matrix <- so_liver_sc@assays$RNA@counts
# # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05
# 
# so_liver_sc <- subset(so_liver_sc, features = rownames(gene_expression_matrix)[expressed_genes])
# dim(so_liver_sc@assays$RNA@counts) #4502 130124
# dim(so_liver_sc@assays$RNA) #4502 130124
# dim(so_liver_sc) #4502 130124
# sum(grepl("^MT-", rownames(so_liver_sc@assays$RNA@counts))) #13
ncol(so_liver_sc) #7383 cells
nrow(so_liver_sc) #23697 genes
# so_liver_sc <- t(so_liver_sc)

#Code for filtering by count
# Create a matrix of the expression data (e.g., counts or scaled data) from the Seurat object
expression_data <- GetAssayData(so_liver_sc, assay = "RNA", layer = "counts")
# Calculate the number of nuclei each gene is expressed in (i.e., count non-zero entries per gene)
gene_expression_counts <- rowSums(expression_data > 0)
# Filter out genes that are expressed in less than 10 nuclei
filtered_genes <- names(which(gene_expression_counts >= 5))
length(filtered_genes) #20194 genes remain of 36601
rare_genes <- names(which(gene_expression_counts < 5))
length(rare_genes) #7624

# Subset the Seurat object to keep only genes that are expressed in at least 10 nuclei
so_liver_sc <- subset(so_liver_sc, features = filtered_genes)

#YE JI's filtering code 
# expr_matrix <- so_liver_sc@assays$RNA@counts
# # Calculate the proportion of cells expressing each gene
# num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
# total_cells <- ncol(expr_matrix)  # Total number of cells
# gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
# remove(expr_matrix)
# # Keep genes expressed in at least "gene_pct" of cells
# genes_to_keep <- names(gene_proportion[gene_proportion >= 0.05])
# so_liver_sc <- subset(so_liver_sc, features = genes_to_keep)
# ncol(so_liver_sc) #130124 cells
# nrow(so_liver_sc) # 4502 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_liver_sc), value = TRUE)
#keep_ids <- unique(rownames(so_liver_sc)[which(!rownames(so_liver_sc) %in% mito_genes)])
# so_liver_sc <- subset(so_liver_sc, features = setdiff(rownames(so_liver_sc), mito_genes))
# so_liver_sc <- subset(so_liver_sc,kit_id!="KL-0029535")
#so_liver_sc$Gene <- rownames(so_liver_sc)
#so_liver_sc <- subset(so_liver_sc, Gene %in% keep_ids)
so_liver_sc <- subset(so_liver_sc, features = setdiff(rownames(so_liver_sc@assays$RNA@counts), mito_genes))
#so_liver_sc <- subset(so_liver_sc, features = setdiff(rownames(so_liver_sc@assays$RNA@counts), mito_genes))
# so_liver_sc <- subset(so_liver_sc, !rownames(so_liver_sc) %in% mito_genes)
# grep("^MT-", rownames(so_liver_sc@assays$RNA@counts), value = TRUE)
dim(so_liver_sc@assays$RNA@counts) #9276 186125
dim(so_liver_sc@assays$RNA@data) #9276 186125
dim(so_liver_sc@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_liver_sc@assays$RNA@counts))) #0
ncol(so_liver_sc) #7383 cells remain
nrow(so_liver_sc) #21,822 genes remain

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_liver_sc@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_liver_sc@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_liver_sc@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_liver_sc <- so_liver_sc[, cells_to_keep]

# Check dimensions after filtering
dim(so_liver_sc@assays$RNA@counts) #9276 81882
dim(so_liver_sc@assays$RNA)#9276 81882
dim(so_liver_sc)#9276 81882

# Optionally, you can also check how many MT genes remain in the filtered dataset
sum(grepl("^MT-", rownames(so_liver_sc@assays$RNA@counts)))
ncol(so_liver_sc)  # Number of cells remaining


# Step 3: Renormalize the Seurat object after filtering
so_liver_sc <- NormalizeData(so_liver_sc)
so_liver_sc <- ScaleData(so_liver_sc)

# #Check normalized slot
# dim(so_liver_sc@assays$RNA@data) #5714 186125
# sum(grepl("^MT", rownames(so_liver_sc@assays$RNA@data))) #70
# ncol(so_liver_sc) #186125 cells
# nrow(so_liver_sc) #5714 genes

rm(expression_data,filtered_genes,gene_expression_counts,liver_ids)
#Create general hepatocyte cell type variable
so_liver_sc$celltype2 <- ifelse(grepl("Hepatocyte",so_liver_sc$celltype),"Hep","Non-Hep")
table(so_liver_sc$celltype,so_liver_sc$celltype2)
```

## c. Integration of Single Nuc and Single Cell
```{r}
#Integration of sc and sn
gc()
num_cores <- detectCores()
num_cores
plan("multisession", workers = 24)
# Increase the maximum size for globals (e.g., 50 GB)
options(future.globals.maxSize = 100 * 1024^3)  # 50 GB
gc()
so_liver_sc <- FindVariableFeatures(so_liver_sc, selection.method = "vst", nfeatures = 2000)
gc()
so_liver_sn <- FindVariableFeatures(so_liver_sn, selection.method = "vst", nfeatures = 2000)
gc()

so_liver_sn$orig.ident <- "Single_Nuc"
so_liver_sc$orig.ident <- "Single_Cell"

#-----------------------
# 4. Integration
#-----------------------
# Identify integration anchors using the top 20 principal components (PCs)
gc()
anchors <- FindIntegrationAnchors(object.list = list(so_liver_sc, so_liver_sn), dims = 1:20,verbose=T)
gc()
# Computing 2000 integration features
# Scaling features for provided objects
# Finding all pairwise anchors
# Running CCA
# Merging objects
# Finding neighborhoods
# Finding anchors
# 	Found 19518 anchors
# Filtering anchors
# 	Retained 935 anchors
# Integrate data; this creates a new 'integrated' assay in the resulting object
combined <- IntegrateData(anchorset = anchors, dims = 1:20)
gc()
#-----------------------
# 5. Downstream Analysis
#-----------------------
DefaultAssay(combined) <- "integrated"

saveRDS(combined,fs::path(dir.dat,"scRNA","data_clean","integrated_sn_sc.rds"))

# Scale the integrated data
combined <- ScaleData(combined)
# Run PCA
combined <- RunPCA(combined, npcs = 30)
# Run UMAP for visualization
combined <- RunUMAP(combined, dims = 1:20)
# Optionally, cluster cells
combined <- FindNeighbors(combined, dims = 1:20)
combined <- FindClusters(combined, resolution = 0.5)
# Visualize UMAP results
DimPlot(combined, reduction = "umap", group.by = "orig.ident", repel = TRUE, label = T)

DefaultAssay(combined) <- "integrated"

DimPlot(combined, reduction = "umap", group.by = "Kit_Lot", label = TRUE, repel = TRUE)
DimPlot(combined, reduction = "umap", group.by = "celltype", label = TRUE)

# cluster_markers <- FindAllMarkers(combined, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
# head(cluster_markers)
Idents(combined) <- combined$celltype
all_genes <- rownames(combined)
FeaturePlot(combined, features = c("SAMD5","AC104041.1","MARCH1"), min.cutoff = "q10", max.cutoff = "q90")
VlnPlot(combined, features = c("SAMD5","AC104041.1","MARCH1"), group.by = "seurat_clusters", pt.size = 0)
DotPlot(combined, features = c("PNPLA3"), group.by = "celltype")

```

# 3. Outline 
### a. Comparison Groups of Interest
1. Steatosis (0+1 vs. 2+3)
2. No Liver Disease (No Steatosis, No Fibrosis) vs. Steatosis (1,2, or 3 Steatosis, No Fibrosis) vs. Fibrosis (Any Fibrosis, Any Steatosis)

### b. Disease Categories of Interest
1. OB (Obese Controls Only)
a. T2D + OB (Full Cohort: Type 2s and Obese Controls)

### e. Characteristics of Interest
HbA1c, Medication Status, TGs

### f. Gene Sets & Genes of Interest
1. All Genes 
2. Senescence 
3. Metabolism
4. Inflammation
5. PNLAP3
6. GCKR
7. TM6SF2
8. APOC3
9. MTOR
10. Growth Factors
11. HNF4alpha?

### g. Analysis Plan
1. All Genes
a. DEGS + IPA
b. LMEM + IPA, adjusting for HbA1c, Med Status, TGs 
c. Pseuodbulk all cell types
d. Hepatocytes
e. Cell Specific (All other cell types)
2. Pathway Gene Sets
a. DEGS + IPA
b. LMEM + IPA, adjusting for HbA1c, Med Status, TGs
c. Pseuodbulk all cell types
d. Hepatocytes
e. Cell Specific (All other cell types)

#4. Descriptive Statistics
## a. Participants with Single Nucleus Data
```{r, echo = F,warning=F}
#Define disease and drug groups
# dat <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#dat <- read.csv("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean/liver_biopsy_metadata_PN.csv")

dat <- so_liver_sn@meta.data %>%
  dplyr::select(all_of(c("StudyID","RNAlater_ID","Cryostor_ID","nih_sex","nih_ethnicity","nih_race","bmi","diagnosis_of_diabetes",
                         "diagnosis_of_MASLD",
                         "steatosis_grade","fibrosis_stage",
                         "lobular_inflammation_percent","tg","creatinine",
                         "alt","ast","ggt","age_surgery","glp1agonist","sglt2","group2",
                         "group3","meds","insulin","metformin","StudyID","steatosis_cat","fibrosis_cat","a1c"
  )))  %>%
  group_by(RNAlater_ID) %>%
  summarise(across(everything(), first)) %>%
  ungroup() 
dat$study <- case_when(grepl("IT2D",dat$StudyID) ~ "IMPROVE",
                       grepl("MANGO",dat$StudyID) ~ "MANGO",
                       grepl("BSA",dat$StudyID) ~ "BIOBANK")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting
dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (2+3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent  
       | steatosis_cat, data=dat)
table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent  
       | fibrosis_cat, data=dat)
rm(dat)
```

## b. Participants with Single Cell Data
```{r, echo = F,warning=F}
#Define disease and drug groups
# dat <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#dat <- read.csv("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean/liver_biopsy_metadata_PN.csv")

dat <- so_liver_sc@meta.data %>%
  dplyr::select(all_of(c("Cryostor_ID","nih_sex","nih_ethnicity","nih_race","bmi","diagnosis_of_diabetes",
                         "diagnosis_of_MASLD",
                         "steatosis_grade","fibrosis_stage",
                         "lobular_inflammation_percent","tg","creatinine",
                         "alt","ast","ggt","age_surgery","glp1agonist","sglt2","group2",
                         "group3","meds","insulin","metformin","StudyID","steatosis_cat","fibrosis_cat","a1c"
  )))  %>%
  group_by(Cryostor_ID) %>%
  summarise(across(everything(), first)) %>%
  ungroup() 
dat$study <- ifelse(grepl("IT2D",dat$StudyID),"IMPROVE","MANGO")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting

dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent  
       | steatosis_cat, data=dat)
table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + lobular_inflammation_percent  
       | fibrosis_cat, data=dat)

```

## c. UMAP snRNA & Barcharts
```{r}
# PCA
so_liver_sn <- FindVariableFeatures(object = so_liver_sn)
so_liver_sn <- RunPCA(so_liver_sn, features = VariableFeatures(object = so_liver_sn),assay="RNA")
ElbowPlot(so_liver_sn)

# # Find neighbors and clusters (again using integrated data)
so_liver_sn <- FindNeighbors(so_liver_sn, assay = "RNA", dims = 1:20)
so_liver_sn <- FindClusters(so_liver_sn, resolution = 0.5)

# Perform UMAP and tSNE
# so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
DimPlot(so_liver_sn, reduction = "umap", raster = F)
DimPlot(so_liver_sn, reduction = "umap", group.by = "celltype", raster = FALSE)
DimPlot(so_liver_sn, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot with Cell Type Labels")

so_liver_sn$steatosis_cat <- factor(so_liver_sn$steatosis_cat)
DimPlot(so_liver_sn, reduction = "umap",group.by = "steatosis_cat",label=F,raster=F) +
  ggtitle(paste0("UMAP by Steatosis Category"))

so_liver_sn$fibrosis_cat <- factor(so_liver_sn$fibrosis_cat)
DimPlot(so_liver_sn, reduction = "umap",group.by = "fibrosis_cat",label=F,raster=F) +
  ggtitle(paste0("UMAP by Fibrosis Category"))


#Barcharts of proportions
# By PT subtypes
cellcount<- so_liver_sn@meta.data %>% 
  filter(celltype2 == "Hep")

label(cellcount$fibrosis_cat) <- "Fibrosis Category"
label(cellcount$steatosis_cat) <- "Steatosis Category"
cellcount$steatosis_cat <- ifelse(cellcount$steatosis_cat=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$fibrosis_cat <- factor(cellcount$fibrosis_cat,levels=c("No Steatosis, No Fibrosis","Steatosis, No Fibrosis","Fibrosis & Steatosis"),labels=c("No Steatosis, No Fibrosis","Steatosis, No Fibrosis","Fibrosis & Steatosis"))

prop_plot1 <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Nuclei",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Steatosis Grade") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
pdf(fs::path(dir.results,"Hep_Proportions_Sn_Plot_Steatosis.pdf"),height=10,width=10)
plot(prop_plot1)
dev.off()
# ggsave(plot = cellcount, filename = "/home/yejichoi/Documents/lin_temp_folder/PT_cell_proportions.jpg", 
#        scale = 3,
#        width = 500,height = 1000,units = "px")

prop_plot2 <- ggplot(data=cellcount,aes(fibrosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Nuclei",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Fibrosis Stage & Steatosis Stage") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

pdf(fs::path(dir.results,"Hep_Proportions_Sn_Plot_Fibrosis.pdf"),height=10,width=10)
plot(prop_plot2)
dev.off()
```

## d. UMAP scRNA
```{r}
# PCA
so_liver_sc <- FindVariableFeatures(object = so_liver_sc)
so_liver_sc <- RunPCA(so_liver_sc, features = VariableFeatures(object = so_liver_sc),assay="RNA")
ElbowPlot(so_liver_sc)

# # Find neighbors and clusters (again using integrated data)
so_liver_sc <- FindNeighbors(so_liver_sc, assay = "RNA", dims = 1:20)
so_liver_sc <- FindClusters(so_liver_sc, resolution = 0.5)

# Perform UMAP and tSNE
# so_liver_sc <- RunUMAP(so_liver_sc, dims = 1:15)
# DimPlot(so_liver_sc, reduction = "umap", raster = F)
# DimPlot(so_liver_sc, reduction = "umap", group.by = "celltype", raster = FALSE)
DimPlot(so_liver_sc, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot with Cell Type Labels")

so_liver_sc$steatosis_cat <- factor(so_liver_sc$steatosis_cat)
DimPlot(so_liver_sc, reduction = "umap",group.by = "steatosis_cat",label=F,raster=F) +
  ggtitle(paste0("UMAP by Steatosis Category"))

so_liver_sc$fibrosis_cat <- factor(so_liver_sc$fibrosis_cat)
DimPlot(so_liver_sc, reduction = "umap",group.by = "fibrosis_cat",label=F,raster=F) +
  ggtitle(paste0("UMAP by Fibrosis Category"))

#Barcharts of proportions
# By PT subtypes
cellcount<- so_liver_sc@meta.data %>% 
  filter(celltype2 == "Hep")

label(cellcount$fibrosis_cat) <- "Fibrosis Category"
label(cellcount$steatosis_cat) <- "Steatosis Category"
cellcount$steatosis_cat <- ifelse(cellcount$steatosis_cat=="0+1", "Low Steatosis (0+1)","High Steatosis (2+3)")
cellcount$fibrosis_cat <- factor(cellcount$fibrosis_cat,levels=c("No Steatosis, No Fibrosis","Steatosis, No Fibrosis","Fibrosis & Steatosis"),labels=c("No Steatosis, No Fibrosis","Steatosis, No Fibrosis","Fibrosis & Steatosis"))

prop_plot1 <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Steatosis Grade") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))
pdf(fs::path(dir.results,"Hep_Proportions_Sn_Plot_Steatosis_single_cell.pdf"),height=10,width=10)
plot(prop_plot1)
dev.off()
# ggsave(plot = cellcount, filename = "/home/yejichoi/Documents/lin_temp_folder/PT_cell_proportions.jpg", 
#        scale = 3,
#        width = 500,height = 1000,units = "px")

prop_plot2 <- ggplot(data=cellcount,aes(fibrosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Fibrosis Stage & Steatosis Stage") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a"))

pdf(fs::path(dir.results,"Hep_Proportions_Sn_Plot_Fibrosis_single_cell.pdf"),height=10,width=10)
plot(prop_plot2)
dev.off()
```


# 5. Comparative Analyses
## A. Single Nuc
### i. All Genes
#### a. Pseudobulk all cell types
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
# so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="Low Steatosis (0+1)")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
gene_list <- gene_list_total[5001:11300]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results4 <- full_results
full_results <- rbind(full_results,full_results4)
# rm(full_results2,full_results3,full_results4,full_results5,full_results6,full_results7)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Pseudobulk_All_Celltypes_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","Pseudobulk_All_Celltypes_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "All Cell Types, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_All_Celltypes_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()
```

##### Fibrosis
```{r}
#Fibrosis
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
# so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
gene_list <- gene_list_total[3501:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    # SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,PValue2,Beta3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results7 <- full_results
full_results <- rbind(full_results,full_results2)
rm(full_results2,full_results3,full_results4,full_results5,full_results6)
# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))

#Fibrosis low vs. mid
full_results$color2 <- ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 > 0, "lightcoral",
                              ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df2 <- full_results[full_results$fdr2 < 0.2, ]

#Fibrosis low vs. high
full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
                              ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot2 <- ggplot(full_results, aes(x = Beta2, y = PValue10_2, color = color2)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "All Cell Types, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df2, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot3 <- ggplot(full_results, aes(x = Beta3, y = PValue10_3, color = color3)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "All Cell Types, Adjusted for HbA1c and Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df3, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot2+volcano_plot3

pdf(fs::path(dir.results,"Plot_All_Celltypes_Fibrosis_Steatosis.pdf"),width=20,height=7)
print(volcano_plot2+volcano_plot3)
dev.off()


```

#### b. Hepatocytes Pseudobulk Cells
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
Genes <- nrow(so_celltype) #11300 genes/4489
Nuclei <- ncol(so_celltype) #103101 nuclei
# so_celltype <- subset(so_celltype,features=rownames(so_celltype)[1:5650])
so_celltype <- subset(so_celltype,features=rownames(so_celltype)[5651:11300])
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
rm(data)
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="Low Steatosis (0+1)")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list <- gene_list_total[5650:36588]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
# full_results <- rbind(full_results,full_results2)
# full_results <- rbind(full_results,full_results3)
# rm(full_results2,full_results3,full_results4,full_results5)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]
#Accidentally ran which(duplicated(full_results$Gene)) muliple times
# full_results$Gene[which(duplicated(full_results$Gene))]
# full_results <- full_results %>% 
#   na.omit
rm(full_results2)

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Pseudobulk_Hepatocytes_Steatosis_03_20.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","Pseudobulk_Hepatocytes_Steatosis_03_20.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_Steatosis_03_19_no_dHep_10_count.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "All Hepatocytes, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_Pseudobulk_Hepatocytes_Steatosis_03_20.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```

##### Fibrosis 
```{r}
#Fibrosis
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
#Fibrosis
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
# so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
# so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[3501:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    # SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,PValue2,Beta3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
full_results7 <- full_results
full_results <- rbind(full_results,full_results2)
rm(full_results2,full_results3,full_results4,full_results5,full_results6)
# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))

#Fibrosis low vs. mid
full_results$color2 <- ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 > 0, "lightcoral",
                              ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df2 <- full_results[full_results$fdr2 < 0.2, ]

#Fibrosis low vs. high
full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
                              ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot2 <- ggplot(full_results, aes(x = Beta2, y = PValue10_2, color = color2)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "All Cell Types, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df2, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot3 <- ggplot(full_results, aes(x = Beta3, y = PValue10_3, color = color3)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "All Cell Types, Adjusted for HbA1c and Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df3, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot2+volcano_plot3

pdf(fs::path(dir.results,"Plot_Pseudobulk_Hepatocytes_Fibrosis_Steatosis.pdf"),width=20,height=7)
print(volcano_plot2+volcano_plot3)
dev.off()


```
#### c. Hepatocyte Types
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
so_celltype <- subset(so_liver_sn,celltype=="dHep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
#START HERE
# so_celltype <- subset(so_liver_sn,celltype=="dHep")
# so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #11300 genes
Nuclei <- ncol(so_celltype) #24808 (Hepatocytes 1),22983 (Hepatocytes 2), 19994 (Hepatocytes 3), 19,319 (Hepatocytes 4),4674 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="Low Steatosis (0+1)")
rm(data)
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

# First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[1:5650]
gene_list <- gene_list_total[5651:11300]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
full_results <- rbind(full_results,full_results2)
rm(full_results2)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"dHep_Steatosis_03_31.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","dHep_Steatosis_03_31.csv"))
# full_results <- read.csv(fs::path(dir.results,"dHep_Steatosis_03_31.csv")) %>%
# dplyr::select(-X)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "dHep, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_dHep_Steatosis_03_31.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()
```

###### GSEA
```{r echo = F}
#Hepatocytes
full_results <- read.csv(fs::path(dir.results,"Pseudobulk_All_Celltypes_Steatosis.csv")) %>%
dplyr::select(-X)
sig_pos <- full_results %>% 
  filter(Beta>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(Beta<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Positive Hepatocytes - GO")
ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Positive Hepatocytes - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hepatocytes - Kegg")
ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Positive Hepatocytes - Reactome '22")
# plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hepatocytes - Reactome '22")
ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Positive Hepatocytes - Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hepatocytes - Reactome '24")
ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Negative Hepatocytes - GO")
ggsave(fs::path(dir.results,"All_Cell_Types_neg_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Negative Hepatocytes - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hepatocytes - Kegg")
ggsave(fs::path(dir.results,"All_Cell_Types_neg_gsea_kegg.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Negative Hepatocytes - Reactome '22")
# plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hepatocytes - Reactome '22")
ggsave(fs::path(dir.results,"All_Cell_Types_neg_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
plotEnrich(filtered_results, title = "Model 1 Negative Hepatocytes - Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hepatocytes - Reactome '24")
ggsave(fs::path(dir.results,"All_Cell_Types_neg_gsea_reactome24.jpeg"),width=10,height=7)
```
#####Fibrosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    # SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,PValue2,Beta3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

#Fibrosis low vs. mid
full_results$color2 <- ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 > 0, "lightcoral",
                              ifelse(full_results$fdr2 < 0.2 & full_results$Beta2 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df2 <- full_results[full_results$fdr2 < 0.2, ]

#Fibrosis low vs. high
full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
                              ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot2 <- ggplot(full_results, aes(x = Beta2, y = PValue10_2, color = color2)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "Hep-5 Cells, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df2, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot3 <- ggplot(full_results, aes(x = Beta3, y = PValue10_3, color = color3)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis",
    subtitle = "Hep-5 Cells, Adjusted for HbA1c and Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene),
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df3, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
volcano_plot2+volcano_plot3

pdf(fs::path(dir.results,"Plot_Hep_5_Fibrosis_Steatosis.pdf"),width=20,height=7)
print(volcano_plot2+volcano_plot3)
dev.off()

```

#### d. All Cell Types
```{r, echo=F,warning=F,fig.width=15,fig.height=9}
#Major Cell Types
#unique(so_liver_sn$celltype)
so_liver_sn$celltype <- as.character(so_liver_sn$celltype)
all_genes <- rownames(so_liver_sn)
all_cell_types <- unique(so_liver_sn$celltype)
Idents(so_liver_sn) <- so_liver_sn$celltype

for (cell in all_cell_types[-1]) {
  result <- degs_fxn(so=so_liver_sn,cell=cell,gene_set=all_genes,exposure="steatosis_cat",exp_group="3",ref_group="0+1",enrichment="No",top_gsea = 30)
  print(result)
}

```

### ii. Specific Genes & Pathways
#### a. Pseudobulk All Cell Types
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_liver_sn, features = target_genes)

#START HERE
# so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
# gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[3500:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    SE <- summary(model1)$coef$cond[2,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    SE <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,SE=SE,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results7 <- full_results
# full_results <- rbind(full_results,full_results7)
# rm(full_results2,full_results3,full_results4,full_results5,full_results6,full_results7)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Target_Genes_Pseudobulk_All_Celltypes_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(PValue<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cells",
       x = "Log2FoldChange", y = "Gene",
       caption="Positive Expression = Higher expression in high steatosis vs. low steatosis") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
pdf(fs::path(dir.results,"Plot_Targeted_Genes_Pseudobulk_All_Celltypes_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()
```
##### Fibrosis
```{r}

#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_liver_sn, features = target_genes)
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,SE2,PValue2,Beta3,SE3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_Pseudobulk_All_Celltypes_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower2 = Beta2 - 1.96 * SE2,
    CI_Upper2 = Beta2 + 1.96 * SE2,
    CI_Lower3 = Beta3 - 1.96 * SE3,
    CI_Upper3 = Beta3 + 1.96 * SE3
  )
full_results <- full_results %>% 
  mutate(sig2=ifelse(PValue2<0.05,"*","")) %>% 
  mutate(sig3=ifelse(PValue3<0.05,"*",""))

# Create the coefficient plot
dot_plot2 <- ggplot(full_results, aes(x = Beta2, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower2, xmax = CI_Upper2), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis in All Cells",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot3 <- ggplot(full_results, aes(x = Beta3, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower3, xmax = CI_Upper3), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis in All Cells",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot2+dot_plot3

pdf(fs::path(dir.results,"Plot_Targeted_Genes_Pseudobulk_All_Celltypes_Fibrosis_Steatosis.pdf"),width=12,height=4)
print(dot_plot2+dot_plot3)
dev.off()
```

#### b. Pseudobulk Hepatocytes
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

#START HERE
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #98427 nuclei
#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_celltype, features = target_genes)

#START HERE
# so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
# gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[3500:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    SE <- summary(model1)$coef$cond[2,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    SE <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,SE=SE,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results7 <- full_results
# full_results <- rbind(full_results,full_results7)
# rm(full_results2,full_results3,full_results4,full_results5,full_results6,full_results7)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Target_Genes_Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(PValue<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Hepatocytes",
       x = "Log2FoldChange", y = "Gene",
       caption="Positive Expression = Higher expression in high steatosis vs. low steatosis") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
pdf(fs::path(dir.results,"Plot_Targeted_Genes_Pseudobulk_Hepatocytes_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()

```

##### Fibrosis
```{r}
#Fibrosis
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_celltype, features = target_genes)
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,SE2,PValue2,Beta3,SE3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_Pseudobulk_Hepatocytes_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower2 = Beta2 - 1.96 * SE2,
    CI_Upper2 = Beta2 + 1.96 * SE2,
    CI_Lower3 = Beta3 - 1.96 * SE3,
    CI_Upper3 = Beta3 + 1.96 * SE3
  )
full_results <- full_results %>% 
  mutate(sig2=ifelse(PValue2<0.05,"*","")) %>% 
  mutate(sig3=ifelse(PValue3<0.05,"*",""))

# Create the coefficient plot
dot_plot2 <- ggplot(full_results, aes(x = Beta2, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower2, xmax = CI_Upper2), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis in All Hepatocytes",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot3 <- ggplot(full_results, aes(x = Beta3, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower3, xmax = CI_Upper3), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis in All Hepatocytes",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot2+dot_plot3

pdf(fs::path(dir.results,"Plot_Targeted_Genes_Pseudobulk_Hepatocytes_Fibrosis_Steatosis.pdf"),width=12,height=4)
print(dot_plot2+dot_plot3)
dev.off()
```

#### c. Hepatocyte Types
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sn)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sn,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
#Filter to only target genes of interest
so_celltype <- subset(so_celltype,celltype=="Hep-5")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #98427 nuclei
#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_celltype, features = target_genes)

#START HERE
# so_celltype <- so_liver_sn
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
# gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[3500:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    SE <- summary(model1)$coef$cond[2,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    SE <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,SE=SE,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results7 <- full_results
# full_results <- rbind(full_results,full_results7)
# rm(full_results2,full_results3,full_results4,full_results5,full_results6,full_results7)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Target_Genes_Hep_5_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(PValue<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-5",
       x = "Log2FoldChange", y = "Gene",
       caption="Positive Expression = Higher expression in high steatosis vs. low steatosis") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_5_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()
```
##### Fibrosis
```{r}
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
gene_list <- target_genes
so_celltype <- subset(so_celltype, features = target_genes)
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)
data_subset$fibrosis_cat <- relevel(data_subset$fibrosis_cat,ref="No Steatosis, No Fibrosis")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    Beta2 <- summary(model1)$coef$cond[3,1]
    PValue2 <- summary(model1)$coef$cond[3,4]
    SE2 <- summary(model1)$coef$cond[3,2]
    Beta3 <- summary(model1)$coef$cond[2,1]
    PValue3 <- summary(model1)$coef$cond[2,4]
    SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    Beta2 <- NA
    PValue2 <- NA
    SE2 <- NA
    Beta3 <- NA
    PValue3 <- NA
    SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta2,SE2,PValue2,Beta3,SE3,PValue3,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr2=p.adjust(PValue2,method="fdr")) %>% 
  mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10_2 <- -log10(pmax(full_results$PValue2, 1e-10))  # Avoid log(0)
full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange2=2^Beta2-1) %>% 
  mutate(FoldChange3=2^Beta3-1)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_Hep_5_Fibrosis_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower2 = Beta2 - 1.96 * SE2,
    CI_Upper2 = Beta2 + 1.96 * SE2,
    CI_Lower3 = Beta3 - 1.96 * SE3,
    CI_Upper3 = Beta3 + 1.96 * SE3
  )
full_results <- full_results %>% 
  mutate(sig2=ifelse(PValue2<0.05,"*","")) %>% 
  mutate(sig3=ifelse(PValue3<0.05,"*",""))

# Create the coefficient plot
dot_plot2 <- ggplot(full_results, aes(x = Beta2, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower2, xmax = CI_Upper2), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & No Fibrosis vs. No Steatosis & No Fibrosis in Hep-5",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot3 <- ggplot(full_results, aes(x = Beta3, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower3, xmax = CI_Upper3), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis & Fibrosis vs. No Steatosis & No Fibrosis in Hep-5",
       x = "Log2FoldChange", y = "Gene",
       caption="Significant at P-Value < 0.05") +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot2+dot_plot3

pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_5_Fibrosis_Steatosis.pdf"),width=12,height=4)
print(dot_plot2+dot_plot3)
dev.off()


```

# Dot Plots
```{r}
key_transcripts <- c()

# Create a dot plot for key genes across clusters
DotPlot(object = seurat_object, 
        features = key_genes, 
        cols = c("lightgrey", "blue"),  # Adjust color scale
        dot.scale = 5) + 
  RotatedAxis() + 
  theme_minimal() +
  ggtitle("Dot Plot of Key Genes Across Cell Types")


```

## B. Single Cell
### i. All Genes
#### a. Pseudobulk All Cell Types
##### Steatosis
```{r}
so_celltype <- so_liver_sc
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")

#First half of genes
# gene_list_total <- gene_list
gene_list <- gene_list_total[3001:4489]
# gene_list2 <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results5 <- full_results
# full_results <- rbind(full_results,full_results5)
# rm(full_results2,full_results3,full_results4,full_results5)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"SC_Pseudobulk_All_Celltypes_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","SC_Pseudobulk_All_Celltypes_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.2 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.2 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.2, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Single Cell: High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "All Cell Types, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.2, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_SC_All_Celltypes_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```
#### b. Pseudobulk Hepatocytes Cells
##### Steatosis
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sc)
# length(all_genes)
#Hepatocytes Cells
so_liver_sc$celltype2 <- ifelse(grepl("Hepatocyte",so_liver_sc$celltype),"Hep","Non-Hep")
so_celltype <- subset(so_liver_sc,celltype2=="Hep")
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[1:2245]
gene_list <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
# full_results <- rbind(full_results,full_results2)
# rm(full_results2,full_results3,full_results4,full_results5)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"SC_Pseudobulk_Hepatocytes_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","SC_Pseudobulk_Hepatocytes_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Single Cell: High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hepatocytes, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(-4,4)+
  # ylim(0,12)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_SC_Hepatocytes_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```
#### c. Hepatocyte Types
```{r}
so_celltype <- subset(so_liver_sc,celltype=="Hepatocyte1")
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")

#First half of genes
gene_list_total <- gene_list
# gene_list <- gene_list_total[1:2245]
gene_list <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
# full_results <- rbind(full_results,full_results2)
# rm(full_results2,full_results3,full_results4,full_results5)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"SC_Hep_1_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","SC_Hep_1_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Single Cell: High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-1, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(-4,4)+
  # ylim(0,12)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_SC_Hep_1_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()
Genes <- nrow(so_celltype) #4489 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total[1:2245]
gene_list <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
# full_results <- rbind(full_results,full_results2)
# rm(full_results2,full_results3,full_results4,full_results5)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"SC_Pseudobulk_Hepatocytes_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","SC_Pseudobulk_Hepatocytes_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Single Cell: High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hepatocytes, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Cells = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(-4,4)+
  # ylim(0,12)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_SC_Hepatocytes_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()
```
### ii. Specific Genes & Pathways
#### a. Pseudobulk All Cell Types
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
all_genes <- rownames(so_liver_sc)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sc$celltype2 <- ifelse(grepl("Hep-",so_liver_sc$celltype),"Hep","Non-Hep")

#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)

#START HERE
# so_celltype <- subset(so_celltype,celltype2=="Hep")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #7383 cells
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), SE = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
print(full_results)

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Target_Genes_Pseudobulk_All_Celltypes_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cells",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 

pdf(fs::path(dir.results,"Plot_Single_Cell_Targeted_Genes_Pseudobulk_All_Celltypes_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()
```
#### b. Pseudobulk Hepatocytes
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
all_genes <- rownames(so_liver_sc)
# length(all_genes)
#Hepatocytes Cells
so_liver_sc$celltype2 <- ifelse(grepl("Hepatocyte",so_liver_sc$celltype),"Hep","Non-Hep")

#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)

#START HERE
so_celltype <- subset(so_celltype,celltype2=="Hep")
nrow(so_celltype) #5 genes
ncol(so_celltype) #cells
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), SE = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# print(full_results)

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Single_Cell_Target_Genes_Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hepatocytes",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 

pdf(fs::path(dir.results,"Plot_Single_Cell_Targeted_Genes_Pseudobulk_Hepatocytes_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()
# }
# Add a column to determine color based on the sign of the beta
# Create a new color column based on beta sign and FDR threshold
# full_results$color <- ifelse(full_results$fdr < 0.2 & full_results$Beta > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.2 & full_results$Beta < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df <- full_results[full_results$fdr < 0.2, ]
# # full_results$color <- ifelse(full_results$PValue < 0.05 & full_results$Beta > 0, "red", 
# #                              ifelse(full_results$PValue < 0.05 & full_results$Beta < 0, "blue", "gray"))
# # 
# # # Identify significant points (PValue < 0.05)
# # significant_df <- full_results[full_results$PValue < 0.05, ]
# 
# # Create the volcano plot using ggplot
# volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
#   geom_point(alpha = 0.7) +  # Plot points with transparency
#   scale_color_identity() +  # Use the color column directly
#   theme_minimal() +  # Minimal theme
#   labs(
#     title = "Pseudobulk Hepatocytes Cells: High vs. Low Steatosis",
#     subtitle = "Adjusted for HbA1c",
#     x = "Log2FC",
#     y = "-log10(P-Value)",
#     color = "Beta Direction",
#     caption = "Positive Beta = Upregulation in T2D vs. LC, Negative Beta = Downregulation in T2D vs. LC \n Signficance threshold: FDR < 0.2"
#   ) +
#   # Add a horizontal line at -log10(0.05) ≈ 1.3 (FDR threshold)
#   # geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "red", linewidth = 1) +
#   theme(
#     plot.title = element_text(hjust = 0),
#     axis.text.x = element_text(angle = 45, hjust = 1)
#   )+
#   # # Add labels for significant points
#   # geom_text(data = significant_df, aes(label = gene), 
#   #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
#   # Add labels for significant points with ggrepel
#   geom_text_repel(data = significant_df, aes(label = Gene),
#                   size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)


#Fibrosis
so_celltype <- subset(so_liver_sc,celltype2=="Hep")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$fibrosis_cat <- factor(data_subset$fibrosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]
gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    emm_fibrosis <- emmeans(model1, ~ fibrosis_cat,lmerTest.limit = 98427, pbkrtest.limit = 98427)
    
    # Perform pairwise contrasts to compare the levels of fibrosis_cat
    contrast_results <- contrast(emm_fibrosis, method = "pairwise")
    contrast_results <- contrast(emm_fibrosis, list(
      "Steatosis, No Fibrosis vs No Steatosis, No Fibrosis" = c(0, 1, 0),  # Steatosis, no fibrosis vs No steatosis, No fibrosis
      "Fibrosis & Steatosis vs No Steatosis, No Fibrosis" = c(0, 0, 1),  # Steatosis and fibrosis vs No steatosis, No fibrosis
      "Fibrosis & Steatosis vs Steatosis, No Fibrosis" = c(0, 0, -1)     # Steatosis and fibrosis vs Steatosis, no fibrosis
    ))
    
    
    contrast_p <- as.data.frame(summary(contrast_results))
    contrast_summary <- confint(contrast_results)
    contrast_df <- as.data.frame(contrast_summary)
    contrast_df$PValue <- contrast_p$p.value
    contrast_df$Gene <- gene
    
    # Estimate the overall beta and confidence intervals
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 2], 5)
    df <- round(summary(model1)$coef[2, 3], 5)
    lowerci <- round(summary(model1)$coef[2, 1] - 1.96 * se, 5)
    upperci <- round(summary(model1)$coef[2, 1] + 1.96 * se, 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
                       lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result2 <- rbind(result, contrast_df)
  return(result2)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Pseudobulk_Hepatocytes_High_vs_Low_steatosis.csv"))

# }
# Add a column to determine color based on the sign of the beta
# Create a new color column based on beta sign and FDR threshold
full_results$color <- ifelse(full_results$fdr < 0.2 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.2 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.2, ]
# full_results$color <- ifelse(full_results$PValue < 0.05 & full_results$Beta > 0, "red", 
#                              ifelse(full_results$PValue < 0.05 & full_results$Beta < 0, "blue", "gray"))
# 
# # Identify significant points (PValue < 0.05)
# significant_df <- full_results[full_results$PValue < 0.05, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "Pseudobulk Hepatocytes Cells: High vs. Low Steatosis",
    subtitle = "Adjusted for HbA1c",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "Beta Direction",
    caption = "Positive Beta = Upregulation in T2D vs. LC, Negative Beta = Downregulation in T2D vs. LC \n Signficance threshold: FDR < 0.2"
  ) +
  # Add a horizontal line at -log10(0.05) ≈ 1.3 (FDR threshold)
  # geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "red", linewidth = 1) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  # geom_text(data = significant_df, aes(label = gene), 
  #           vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  geom_text_repel(data = significant_df, aes(label = Gene),
                  size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

pdf(fs::path(dir.results,"Plot_Pseudobulk_Hepatocytes_High_vs_Low_steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```
#### c. Hepatocyte Types
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(so_liver_sc)
# length(all_genes)
#Hepatocytes Cells
# so_liver_sc$celltype2 <- ifelse(grepl("Hep-",so_liver_sc$celltype),"Hep","Non-Hep")
# so_celltype <- subset(so_liver_sc,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)
so_celltype <- subset(so_celltype,celltype=="Hepatocyte1")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #24808 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Single_Cell_Targeted_Genes_Hep_1_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-1",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 


pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_1_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()

# #Hep - 2 
#START HERE
#Filter to only target genes of interest
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)
so_celltype <- subset(so_celltype,celltype=="Hepatocyte2")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #24808 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Single_Cell_Targeted_Genes_Hep_2_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-2",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 


pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_2_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()

#Hep - 3 
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)
so_celltype <- subset(so_celltype,celltype=="Hepatocyte3")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #24808 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>%
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Single_Cell_Targeted_Genes_Hep_3_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-3",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10))


pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_3_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()

#Hep - 4
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)
so_celltype <- subset(so_celltype,celltype=="Hep-4")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #24808 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_Hep_4_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-4",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 


pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_4_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()

#Hep-5
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype <- subset(so_liver_sc, features = target_genes)
so_celltype <- subset(so_celltype,celltype=="Hep-5")
nrow(so_celltype) #4489 genes
ncol(so_celltype) #24808 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression) 
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)

#Log transform gene expression
data_subset <- data_subset %>% 
  mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes 
# gene_list_total <- gene_list
# gene_list <- gene_list_total
# gene_list2 <- gene_list_total[2246:4489]

# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    
    # Fit the model with key covariates
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + (1 | Kit_Lot)"))
    model1 <- lmer(m1, data = data_subset)
    
    # Extract beta and p-value
    beta1 <- round(summary(model1)$coef[2, 1], 3)
    pval1 <- round(summary(model1)$coef[2, 5], 5)
    se <- round(summary(model1)$coef[2, 3], 5)
    
  } else {
    beta1 <- NA
    pval1 <- NA
    se <- NA
  }
  
  # Return the result for the current gene
  result <- data.frame(Gene = gene, Beta = beta1, SE=se,PValue = pval1)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)


# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# view(full_results)

full_results <- full_results %>% 
  mutate(FoldChange=2^Beta-1)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_Hep_5_High_vs_Low_steatosis.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = Beta - 1.96 * SE,
    CI_Upper = Beta + 1.96 * SE
  )

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = Beta, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in Hep-5",
       x = "Log2FoldChange", y = "Gene") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 


pdf(fs::path(dir.results,"Plot_Targeted_Genes_Hep_5_High_vs_Low_steatosis.pdf"),width=6,height=4)
print(dot_plot)
dev.off()
```
#### d. All Cell Types
```{r, echo=F,warning=F,fig.width=15,fig.height=9}
#Major Cell Types
#unique(so_liver_sc$celltype)
so_liver_sc$celltype <- as.character(so_liver_sc$celltype)
all_genes <- rownames(so_liver_sc)
all_cell_types <- unique(so_liver_sc$celltype)
Idents(so_liver_sc) <- so_liver_sc$celltype

for (cell in all_cell_types[-1]) {
  result <- degs_fxn(so=so_liver_sc,cell=cell,gene_set=all_genes,exposure="steatosis_cat",exp_group="3",ref_group="0+1",enrichment="No",top_gsea = 30)
  print(result)
}

```

## C. Integrated 
```{r}
#Steatosis Low vs. Steatosis High - ="2+3",ref_group="0+1
# all_genes <- rownames(combined)
# length(all_genes)
#Hepatocytes Cells
combined$celltype2 <- ifelse(grepl("Hep-",combined$celltype) | grepl("dHep",combined$celltype) | grepl("Hepatocyte",combined$celltype),"Hep","Non-Hep")
# so_celltype <- subset(combined,celltype2=="Hep")
# 
# # Extract the gene expression data for all genes
# gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#START HERE
so_celltype <- subset(combined,celltype2=="Hep")
DefaultAssay(combined) <- "integrated"
#START HERE
# so_celltype <- subset(combined,celltype=="Hep-5")
# so_celltype <- combined
Genes <- nrow(so_celltype) #2000 genes
Nuclei <- ncol(so_celltype) #98427 nuclei
# Extract the gene expression data for all genes
gene_expression <- as.data.frame(GetAssayData(so_celltype, layer = "data"))

#Assign gene names as rownames, cell names as colnames
# rownames(gene_expression) <- rownames(so_celltype) #Gene Names
# colnames(gene_expression) <- colnames(so_celltype) #Cell Names
#Transpose gene expression dataset to merge with metadata
gene_expression <- t(gene_expression)
gene_expression <- data.frame(gene_expression) #Make a dataframe again after transposing
#Set gene list
gene_list <- colnames(gene_expression)
gene_expression$cellname <- rownames(gene_expression)
rownames(gene_expression) <- NULL

# Extract the metadata
metadata <- so_celltype@meta.data
# metadata <- metadata %>%
#   mutate(across(everything(),~ifelse(.==".",NA,.)))
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,gene_expression,by="cellname")
rm(metadata,gene_expression)

data_subset <- data
data_subset$steatosis_cat <- factor(data_subset$steatosis_cat)
data_subset$steatosis_cat <- relevel(data_subset$steatosis_cat,ref="0+1")
# #Log transform gene expression
# data_subset <- data_subset %>%
#   mutate(across(all_of(gene_list),~log2(.+1)))

#First half of genes
# gene_list_total <- rownames(combined)
gene_list <- gene_list_total[1]
# gene_list <- gene_list_total[2246:4489]
# gene <- gene_list[1]
# Define the function to process each gene
process_gene <- function(gene_idx) {
  
  gene <- gene_list[gene_idx]  # Get the gene name based on the index
  
  # Ensure that the gene exists in your data subset
  if (sum(data_subset[[gene]]) > 0) {
    # 
    # # Fit the model with key covariates
    # m1 <- as.formula(paste0(gene, " ~ fibrosis_cat + a1c + (1 | Kit_Lot)"))
    m1 <- as.formula(paste0(gene, " ~ steatosis_cat + a1c + nih_ethnicity + (1 | Kit_Lot)"))
    # model1 <- glmmTMB(Expression ~ Treatment * Visit + (1 | Subject / Visit),
    #         data = sc_dat_sub, family = gaussian, ziformula = ~1)
    model1 <- glmmTMB(m1,data = data_subset, family = gaussian, ziformula = ~1)
    # model1 <- lmer(m1, data = data_subset)
    # Beta2 <- summary(model1)$coef$cond[3,1]
    # PValue2 <- summary(model1)$coef$cond[3,4]
    # # SE2 <- summary(model1)$coef$cond[3,2]
    Beta <- summary(model1)$coef$cond[2,1]
    PValue <- summary(model1)$coef$cond[2,4]
    # SE3 <- summary(model1)$coef$cond[3,2]
    zi <- summary(model1)$coef$zi[1,1]
    zi_pval <- summary(model1)$coef$zi[1,4]
    
  } else {
    # Beta2 <- NA
    # PValue2 <- NA
    # SE2 <- NA
    Beta <- NA
    PValue <- NA
    #SE3 <- NA
    zi <- NA
    zi_pval <- NA
  }
  
  # Return the result for the current gene
  # result <- data.frame(contrast = "Overall", estimate = beta1, SE = se, df = df,
  #                      lower.CL = lowerci, upper.CL = upperci, PValue = pval1, Gene = gene)
  result <- data.frame(Gene=gene,Beta,PValue,ZI_Intercept=zi,ZI_PValue=zi_pval)
  return(result)
}

# Total number of genes
total_genes <- length(gene_list)

# Calculate the batch size
batch_size <- round(total_genes / 5)

# Pre-initialize an empty data frame for results
full_results <- data.frame(Gene = character(), Beta = numeric(), PValue = numeric())

# Split the gene list into batches for parallel processing
gene_batches <- split(seq(1, total_genes), ceiling(seq_along(seq(1, total_genes)) / batch_size))

# Use mclapply to process the batches in parallel
results_list <- mclapply(gene_batches, function(batch) {
  # Apply the process_gene function to each gene index in the current batch
  lapply(batch, process_gene)
}, mc.cores = detectCores() - 1)  # Use one less core than available to avoid overloading

# Flatten the list and combine the results into a single data frame
full_results <- do.call(rbind, unlist(results_list, recursive = FALSE))

# Print the final results
view(full_results)
# full_results2 <- full_results
# full_results <- rbind(full_results,full_results2)
# rm(full_results2,full_results3,full_results4,full_results5,full_results6,full_results7)
#Accidentally ran RORA twice, remove second entry (they are identical)
# full_results <- full_results[-3501,]

# #Make volcano plot of all gene results for group
full_results <- full_results %>%
  mutate(fdr=p.adjust(PValue,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$PValue, 1e-10))  # Avoid log(0)
# full_results$PValue10_3 <- -log10(pmax(full_results$PValue3, 1e-10))  # Avoid log(0)

# view(full_results)

full_results <- full_results %>%
  mutate(FoldChange=2^Beta-1)  
# mutate(FoldChange3=2^Beta3-1)
# full_results <- full_results %>% 
#   dplyr::rename(color=color2,
#                 FoldChange=FoldChange2) %>%
#   # dplyr::rename(fdr=fdr2,
#   #               PValue10=PValue10_2) %>%
#   # dplyr::select(-c("Beta3","PValue3")) %>% 
#   dplyr::select(-c("color3","FoldChange3"))

write.csv(full_results,fs::path(dir.results,"Pseudobulk_Hepatocytes_Steatosis.csv"))
write.csv(full_results,fs::path(dir.ipa %>% dirname(),"Liver Pathways","Pseudobulk_Hepatocytes_Steatosis.csv"))
# full_results <- read.csv(fs::path(dir.results,"Hep_5_Fibrosis_Steatosis.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$Beta > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$Beta < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]
# 
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$Beta3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = Beta, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "All Cell Types, Adjusted for HbA1c & Ethnicity",
    x = "Log2FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_Pseudobulk_Hepatocytes_Steatosis.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```
