---
title: "Liver_snRNAseq"
output: html_document
date: "2025-04-23"
---

# 1. Set up Libraries & Directores
```{r libraries, echo=F, include = F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(fgsea)
library(ReactomeGSA)
library(msigdbr)
library(kableExtra)
library(knitr)
# library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
# library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(ggplot2)
library(GGally)
# library(limma)
# library(reshape2)
# library(data.table)
# library(knitr)
# library(TxDb.Hsapiens.UCSC.hg19.knownGene)
# library(stringr)
# library(NMF)
# library(rsvd)
# library(RColorBrewer)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
# library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
library(future)
# registerDoParallel(cores = 6)
library(VennDiagram)
#install.packages("survival")
library(survival)
#install.packages("lme4")  # If not already installed
library(lme4)
#install.packages("lmerTest")
library(lmerTest)
# install.packages("emmeans")
library(emmeans)
# install.packages("glmmTMB")
library(glmmTMB)
library(ggrepel)
# library(qpcR)
library(ggpubr)
#install.packages("devtools")
library(devtools)
# install_github("lhe17/nebula",force=T)
library(nebula)
# BiocManager::install("enrichplot")
# library(enrichplot)
library(enrichR)
dbs <- c("GO_Biological_Process_2023", 
         "KEGG_2021_Human",
         # "Reactome_2022", 
         "Reactome_Pathways_2024",
         # "MSigDB_Oncogenic_Signatures",
         # "MSigDB_Computational",
         "MSigDB_Hallmark_2020")
library(pheatmap)
library(janitor)


#Local file path
# dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
# dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
# dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c("/Users/hhampson/Documents/UW/1_Ongoing Projects/Liver scRNAseq/2_Results")

#Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c(fs::path(dir.dat,"Liver project/Results"))

# #Mac studio
#Mac Studio File Path
dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/NEBULA Results")
# dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

#Laura's Mac Studio File Path
# dir.dat <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
# dir.results <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/NEBULA Results")

#Load functions
source("Liver_functions.R")
# source("/Users/hhampson/CHCO-Code/Petter Bjornstad/Data Processing and Analysis/Standard_Functions.R")

knitr::opts_knit$set(root.dir = dir.results)
```

# 2. Pre-Processing & Qualtiy Control
## a. Load Full Data & Clean:## a. snRNA & MetaData
```{r, include = F, fig.height=7,fig.width=10}
# Liver snRNA data processing
invisible(gc())
#Local
# dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Biostatistics Core Shared Drive")
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
# #Lambda
# so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
DefaultAssay(so_liver_sn) <- "RNA"
dim(so_liver_sn)
nrow(so_liver_sn) #36,601 genes
ncol(so_liver_sn) # 130,124 nuceli
invisible(gc())
#Local
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#Lambda
# meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
unique(meta_liver_raw$Kit_Lot)
meta_liver_raw <- meta_liver_raw %>%
  filter(StudyID!="IT2D-18") %>%
  filter(StudyID!="IT2D-16")
meta_liver_raw <- meta_liver_raw %>%
  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))
meta_liver_raw <- meta_liver_raw %>%
  mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "glp1_or_sglt2",
                          neither == "Yes" ~ "no_med",
                          hc == "Yes" ~ "obese_control"))
# meta_liver_raw <- meta_liver_raw %>%
#   mutate(group3=ifelse(either=="Yes","either","neither"))

#Create Liver Status Variables
meta_liver_raw <- meta_liver_raw %>%
  mutate(steatosis_cat = ifelse((steatosis_grade==0 | steatosis_grade==1),"Low Steatosis (0+1)","High Steatosis (2+3)")) %>%
  mutate(fibrosis_cat = case_when(steatosis_grade==0 & fibrosis_stage==0 ~ "No Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage==0 ~ "Any Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage>0 ~ "Any Fibrosis & Any Steatosis"))

meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"
invisible(gc())

#Create liver disease and drug groups
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
# mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
# mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
#  mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
#  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
#  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))

# Create a single grouping variable
#so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
#  mutate(group2=case_when(both == "Yes" ~ "both",
#                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
#                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
#                          either == "Yes" ~ "either",
#                          neither == "Yes" ~ "neither",
#                          hc == "Yes" ~ "obese_control"))

#so_liver_sn@meta.data$group2 <- factor(so_liver_sn@meta.data$group2, levels = c("obese_control","neither","sglt2_exclusive", #"glp1_exclusive","either","both"))

# #Create Diabetes only seurat object
# so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")

#Create senesence so in all cell types
#sens_gene_in_data <- intersect(sens_genes, rownames(so_liver_sn))
# Subset the Seurat object to include only genes in sens_gene_in_data
#so_sens_all <- subset(so_liver_sn, features = sens_gene_in_data)

# #Filter to hepatocytes only
# so_liver_sn$Hepatocyte <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Yes","No")
# so_hep <- subset(so_liver_sn, Hepatocyte=="Yes")
# rm(so_liver_sn)
# DefaultAssay(so_hep) <- "RNA"  

dim(so_liver_sn) #36601 130124
# so_object <- qc_function_sn(so=so_liver_sn,var_pct=0.75)
# so_filtered <- so_object$so
# print(so_object$elbow_plot)
# 
# #Number of Cells after filtering: 
# cell_total2 <- ncol(so_filtered) #14733
# #Number of Genes after filtering: 
# gene_total2 <- nrow(so_filtered) #8966
# print(paste0(cell_total1," cells before filtereing, ",cell_total2," cells after filtering (removed ",cell_total1-cell_total2," cells); ",gene_total1," genes before filtering, ",gene_total2," genes after filtering (removed ",gene_total1-gene_total2," genes)"))
# 
# #Visualize filtered seurat object on umaps by cell type and diabetes group
# umap_vis(so=so_filtered ,cell_type="KPMP_celltype",group_variable="group") 
# 
# #Remove unfiltered seurat object
# rm(so_object,so_kpmp_sc)

# # Step 1: Filter to genes expressed in more than 5% of cells
# gene_expression_matrix <- so_liver_sn@assays$RNA@counts
# # expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.1
# expressed_genes <- Matrix::rowSums(gene_expression_matrix != 0) > dim(gene_expression_matrix)[2] * 0.05
# 
# so_liver_sn <- subset(so_liver_sn, features = rownames(gene_expression_matrix)[expressed_genes])
# dim(so_liver_sn@assays$RNA@counts) #4502 130124
# dim(so_liver_sn@assays$RNA) #4502 130124
# dim(so_liver_sn) #4502 130124
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #13
# ncol(so_liver_sn) #130124 cells
# nrow(so_liver_sn) # 36601 genes
ncol(so_liver_sn) #130124 nuclei 
nrow(so_liver_sn) # 36601 genes
# 
# #Code for filtering by count
# # Create a matrix of the expression data (e.g., counts or scaled data) from the Seurat object
# expression_data <- GetAssayData(so_liver_sn, assay = "RNA", layer = "counts")
# # Calculate the number of nuclei each gene is expressed in (i.e., count non-zero entries per gene)
# gene_expression_counts <- rowSums(expression_data > 0)
# # Filter out genes that are expressed in less than 10 nuclei
# filtered_genes <- names(which(gene_expression_counts >= 10))
# length(filtered_genes) #28977 genes remain of 36601
# rare_genes <- names(which(gene_expression_counts < 10))
# length(rare_genes) #7624

# # Subset the Seurat object to keep only genes that are expressed in at least 10 nuclei
# so_liver_sn <- subset(so_liver_sn, features = filtered_genes)

#YE JI's filtering code for percent expression 
#Filter out rare genes expressed in less than "gene_pct" of cells
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, layer = "counts"))
# expr_matrix <- as.matrix(GetAssayData(so_liver_sn, assay = "RNA", layer = "counts"))
expr_matrix <- so_liver_sn@assays$RNA@counts
# Calculate the proportion of cells expressing each gene
num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
total_cells <- ncol(expr_matrix)  # Total number of cells
gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
remove(expr_matrix)
# Keep genes expressed in at least "gene_pct" of cells
genes_to_keep <- names(gene_proportion[gene_proportion >= 0.01])
so_liver_sn <- subset(so_liver_sn, features = genes_to_keep)
ncol(so_liver_sn) #130,124 nuclei
nrow(so_liver_sn) # 11313 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_liver_sn), value = TRUE)
#keep_ids <- unique(rownames(so_liver_sn)[which(!rownames(so_liver_sn) %in% mito_genes)])
# so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn), mito_genes))
# so_liver_sn <- subset(so_liver_sn,kit_id!="KL-0029535")
#so_liver_sn$Gene <- rownames(so_liver_sn)
#so_liver_sn <- subset(so_liver_sn, Gene %in% keep_ids)
so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
#so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
# so_liver_sn <- subset(so_liver_sn, !rownames(so_liver_sn) %in% mito_genes)
# grep("^MT-", rownames(so_liver_sn@assays$RNA@counts), value = TRUE)
dim(so_liver_sn@assays$RNA@counts) #9276 186125
dim(so_liver_sn@assays$RNA@data) #9276 186125
dim(so_liver_sn@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #0
ncol(so_liver_sn) #130124 cells
nrow(so_liver_sn) #11300 genes

# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_liver_sn@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_liver_sn@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_liver_sn@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_liver_sn <- so_liver_sn[, cells_to_keep]



# # Check dimensions after filtering
# dim(so_liver_sn@assays$RNA@counts) #9276 81882
# dim(so_liver_sn@assays$RNA)#9276 81882
# dim(so_liver_sn)#9276 81882
# 
# # Optionally, you can also check how many MT genes remain in the filtered dataset
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts)))
# ncol(so_liver_sn)  # Number of cells remaining
# 
# # Step 3: Renormalize the Seurat object after filtering
# so_liver_sn <- NormalizeData(so_liver_sn)
# 
# #Check normalized slot
# dim(so_liver_sn@assays$RNA@data) #5714 186125
# sum(grepl("^MT", rownames(so_liver_sn@assays$RNA@data))) #70
# ncol(so_liver_sn) #186125 cells
# nrow(so_liver_sn) #5714 genes
# 
# rm(gene_expression_matrix)
so_liver_sn <- NormalizeData(so_liver_sn)
so_liver_sn <- ScaleData(so_liver_sn, features = VariableFeatures(so_liver_sn))

#Create general hepatocyte cell type variable
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype) | grepl("dHep",so_liver_sn$celltype),"Hep","Non-Hep")
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

# table(so_liver_sn$celltype,so_liver_sn$celltype2)

# rm(expression_data)
DefaultAssay(so_liver_sn) <- "RNA" 
```

## b. Explore Distribution & Zero-Inflation
```{r}
#Check if data have been normalized 
so_liver_sn@assays  # Should show e.g., "RNA" with "counts" and "data"
head(GetAssayData(so_liver_sn, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_liver_sn, layer = "data")[, 1:5])    # Normalized data

#Check for normality
# Open a PDF device (all plots go here)
pdf(fs::path(dir.results,"Normalization_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesN <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Normalized Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

pdf(fs::path(dir.results,"Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesR <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Try log transforming
log_data <- log1p(GetAssayData(so_liver_sn, layer = "data"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Normalized_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesN
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

log_data <- log1p(GetAssayData(so_liver_sn, layer = "count"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesR
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Check for zero inflation
pdf(fs::path(dir.results,"Zero_Inflation_Visualization.pdf"), width = 8, height = 6)
counts <- GetAssayData(so_liver_sn, layer = "counts")
zero_prop <- rowMeans(counts == 0)
hist(zero_prop, main = "Proportion of Zeros for Genes", xlab = "Zero Proportion")
# Add a vertical line at 50%
abline(v = 0.5, col = "red", lwd = 2, lty = 2)
dev.off()

#Check for overdispersion
# Use raw counts
counts <- GetAssayData(so_liver_sn, layer = "counts")

# Pick a few genes or do this genome-wide
gene_means <- rowMeans(counts)
gene_vars <- apply(counts, 1, var)

# Dispersion estimate (Var / Mean)
dispersion <- gene_vars / gene_means

# Filter to only genes where mean > 0 (to avoid divide-by-zero)
valid_genes <- gene_means > 0
dispersion <- dispersion[valid_genes]

# View distribution
pdf(fs::path(dir.results,"Overdispersion_Visualization.pdf"), width = 8, height = 6)
hist(dispersion, breaks=200, main="Dispersion across genes", xlab="Dispersion (Var / Mean)",xlim = c(0, 2))
abline(v = 1, col="red")  # Poisson expectation
dev.off()

# Check how many genes have dispersion > 1
overdispersed_genes <- sum(dispersion > 1)
total_genes <- length(dispersion)
percent_overdispersed <- (overdispersed_genes / total_genes) * 100

# Print summary and recommendation
print(paste0("Overdispersed genes: ", 
             overdispersed_genes, 
             " out of ", 
             total_genes, 
             " (", 
             round(percent_overdispersed, 3), 
             "%)"))

if (percent_overdispersed > 10) {
  print("✅ Overdispersion detected. Consider using a Negative Binomial model.")
} else {
  print("ℹ️ Little evidence of overdispersion. Poisson model may be sufficient.")
}
rm(counts)
```

## d. Evaluate Covariates before Adjustment
```{r}
#Categorical Covariates
#Get metadata for everyone
# dat <- so_liver_sn@meta.data %>%
#   group_by(StudyID) %>%
#   summarise(across(everything(), first)) %>%
#   ungroup() 
dat <- so_liver_sn@meta.data %>%
  group_by(StudyID) %>%
  summarise(across(everything(), ~ .x[1])) %>%
  ungroup()

dat$study <- case_when(grepl("IT2D",dat$StudyID) ~ "IMPROVE",
                       grepl("MANGO",dat$StudyID) ~ "MANGO",
                       grepl("BSA",dat$StudyID) ~ "BIOBANK")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

dat <- dat %>% 
  mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "glp1_or_sglt2",
                          neither == "Yes" ~ "no_med",
                          hc == "Yes" ~ "obese_control"))

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting
# dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (2+3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD + diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither| steatosis_cat, data=dat)
table1(~ alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)
table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither + alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)

#Examine covariates
age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
  diagnosis_of_diabetes + diagnosis_of_MASLD+
  sglt2 + glp1agonist
cat_covariates <- c("nih_sex", "nih_race","nih_ethnicity", "diagnosis_of_diabetes", "diagnosis_of_MASLD", "both","glp1_exclusive","sglt2_exclusive","neither","fibrosis_cat")
plot_list <- list()

for (covariate in cat_covariates) {
  p <- ggplot(dat, aes_string(x = covariate, fill = "steatosis_cat")) +
    geom_bar(position = "fill") +  # use "dodge" for absolute counts
    labs(title = paste0("Distribution of ", covariate, " by Steatosis Category"),
         y = "Proportion", x = NULL) +
    scale_y_continuous(labels = scales::percent_format()) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust depending on how many you have)
wrap_plots(plot_list, ncol = 2)


#Continuous Covariates
con_covariates <- c("age", "bmi", "triglycerides", "hba1c", "pah_clear_bsa", "eGFR_CKD_epi", "acr_u")
plot_list <- list()

for (covariate in con_covariates) {
  p <- ggplot(dat, aes_string(x = "group", y = covariate, fill = "group")) +
    geom_boxplot(outlier.shape = NA, alpha = 0.7) +
    geom_jitter(width = 0.2, alpha = 0.4, color = "black") +  # optional: show points
    labs(title = paste0("Distribution of ", covariate, " by group"),
         x = "Group", y = covariate) +
    theme_minimal() +
    theme(legend.position = "none")
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust ncol/nrow as needed)
wrap_plots(plot_list, ncol = 4)
```

# 3. Steatosis Analysis 
## a. HEP Cells 
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
ncol(so_celltype) #130124 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

offset_hvg_CELLTYPE = Matrix::colSums(data_g_hvg_CELLTYPE$count) 
# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,reml=T,output_re = T,covariance=T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)
# Use lapply to pull all summaries and add gene names
# nebula_summaries <- do.call(rbind, lapply(names(nebula_results_list), function(gene) {
#   if (!is.null(nebula_results_list[[gene]]$summary)) {
#     df <- nebula_results_list[[gene]]$summary
#     df$gene <- gene  # Add gene name as a new column
#     return(df)
#   } else {
#     NULL
#   }
# }))
nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hepatocytes, Unadjusted (REML)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```

### ii. glmmTMB
```{r}
#Access the raw count matrix
raw_counts <- GetAssayData(so_celltype_hvg, layer = "counts")

# Step 2: Round manually for sparse matrix (dgCMatrix)
raw_counts_rounded <- raw_counts
raw_counts_rounded@x <- round(raw_counts@x)

# Step 3: Extract the full gene expression matrix (rounded counts)
expr_matrix <- as.matrix(raw_counts_rounded)

#View how rounded counts data looks
#1. Randomly select a few genes for visualization (e.g., top 50 or random)
selected_genes <- sample(rownames(raw_counts_rounded), 25)  # Random 50 genes

#2. Get the expression data for those genes (rounded counts)
selected_data <- as.matrix(raw_counts_rounded[selected_genes, ])

# Convert matrix to data.frame
df_selected_data <- as.data.frame(t(selected_data))  # Transpose so that genes are columns

# 3. **Boxplot** for raw counts of selected genes across cells (to see distribution)
df_selected_data_long <- reshape2::melt(df_selected_data)  # Convert to long format

ggplot(df_selected_data_long, aes(x = variable, y = value, fill = variable)) +
  geom_boxplot() +
  labs(title = "Boxplot of Selected Genes' Counts", x = "Genes", y = "Count") +
  theme_minimal()+
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
    # legend.position = "none"
  )

# 4. **Heatmap** of raw counts (scaled by genes) for a few cells (e.g., top 50 cells)
top_cells <- sample(colnames(raw_counts_rounded), 50)  # Random 50 cells
heatmap_data <- as.matrix(raw_counts_rounded[selected_genes, top_cells])

# Plot heatmap
pheatmap(heatmap_data, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE, 
         show_colnames = FALSE,
         main = "Heatmap of Gene Expression (Selected Genes & Cells)") 

#Check rownames and colnames: Rownames = genes, Colnames = cells
#view(expr_matrix)
#Transpose gene expression dataset to merge with metadata
expr_matrix <- t(expr_matrix)
expr_matrix <- data.frame(expr_matrix) #Make a dataframe again after transposing

#Set gene list
gene_list_total <- hvgs

#Create cell variable to merge metadata in
expr_matrix$cellname <- rownames(expr_matrix) 
rownames(expr_matrix) <- NULL

# Extract the metadata
metadata <- so_celltype_hvg@meta.data

#Create cell variable to merge metadata into gene expression data
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,expr_matrix,by="cellname")
rm(metadata,expr_matrix)

#Make sure exposure/independent/x variable or group variable is a factor variable
data$steatosis_cat <- factor(data$steatosis_cat)
#Make sure to set reference level
data$steatosis_cat <- relevel(data$steatosis_cat,ref="Low Steatosis (0+1)")

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hepatocytes,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                           stats = rankings_Hepatocytes,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hepatocytes <- fgsea(pathways = go,
                     stats = rankings_Hepatocytes,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
Hepatocytes_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 KEGG (REML)", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 REACTOME (REML)", xlimit = 5)

ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
##### GO
plot_fgsea_transpose(go_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 GO (REML)", xlimit = 8)

ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T,covariance=T,reml=T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hepatocytes_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hepatocytes, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hepatocytes_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)


#Examine distributions 
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
# Step 1: Pull expression for gene PNPLA3
gene_expr <- FetchData(so_celltype, vars = target_genes)

# Step 2: Pull metadata for steatosis_cat
metadata <- so_celltype@meta.data["steatosis_cat"]

# Step 3: Combine into one dataframe
df <- cbind(gene_expr, metadata)
# colnames(df) <- c("PNPLA3", "steatosis_cat")

# Step 4: Plot using ggplot
ggplot(df, aes(x = PNPLA3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "PNPLA3 Expression by Steatosis Category",
       x = "PNPLA3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = GCKR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "GCKR Expression by Steatosis Category",
       x = "GCKR Expression",
       y = "Cell Count")

ggplot(df, aes(x = TM6SF2, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "TM6SF2 Expression by Steatosis Category",
       x = "TM6SF2 Expression",
       y = "Cell Count")

ggplot(df, aes(x = APOC3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "APOC3 Expression by Steatosis Category",
       x = "APOC3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = MTOR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "MTOR Expression by Steatosis Category",
       x = "MTOR Expression",
       y = "Cell Count")
```

## b. Hep-1
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-1, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-1
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-1_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_1_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep1 <- full_results$LogFC
names(rankings_hep1) <- full_results$Gene
rankings_hep1 <- sort(rankings_hep1, decreasing = TRUE)
plot(rankings_hep1)
min(rankings_hep1)
max(rankings_hep1)

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep1, title = "Hep-1 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep1_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep1, title = "Hep-1 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep1_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep1, title = "Hep-1 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep1_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_1_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-1, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_1_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```


## c. Hep-2
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-2, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-2
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-2_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_2_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep2 <- full_results$LogFC
names(rankings_hep2) <- full_results$Gene
rankings_hep2 <- sort(rankings_hep2, decreasing = TRUE)
plot(rankings_hep2)
min(rankings_hep2)
max(rankings_hep2)


set.seed(1234)

kegg_legacy_res_hep2 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_hep2,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_hep2 <- fgsea(pathways = reactome,
                           stats = rankings_hep2,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_hep2 <- fgsea(pathways = go,
                     stats = rankings_hep2,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

hep2_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_hep2[, padj < 0.05]), sum(kegg_legacy_res_hep2[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_hep2[, padj < 0.05]), sum(reactome_res_hep2[, pval < 0.05])),
                         "GO"=c(sum(go_res_hep2[, padj < 0.05]), sum(go_res_hep2[, pval < 0.05])))
rownames(hep2_fgsea) <- c("adj.pval", "p.val")
hep2_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep2, title = "Hep-2 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"hep2_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep2, title = "Hep-2 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"hep2_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep2, title = "Hep-2 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"hep2_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_2_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-2, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_2_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## d. Hep-3
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-3, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep3 <- full_results$LogFC
names(rankings_Hep3) <- full_results$Gene
rankings_Hep3 <- sort(rankings_Hep3, decreasing = TRUE)
plot(rankings_Hep3)
min(rankings_Hep3)
max(rankings_Hep3)


set.seed(1234)

kegg_legacy_res_Hep3 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep3,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep3 <- fgsea(pathways = reactome,
                           stats = rankings_Hep3,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep3 <- fgsea(pathways = go,
                     stats = rankings_Hep3,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep3_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep3[, padj < 0.05]), sum(kegg_legacy_res_Hep3[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep3[, padj < 0.05]), sum(reactome_res_Hep3[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep3[, padj < 0.05]), sum(go_res_Hep3[, pval < 0.05])))
rownames(Hep3_fgsea) <- c("adj.pval", "p.val")
Hep3_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep3, title = "Hep-3 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep3_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep3, title = "Hep-3 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep3_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep3, title = "Hep-3 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep3_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_3_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-3, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_3_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## e. Hep-4
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-4, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep4 <- full_results$LogFC
names(rankings_Hep4) <- full_results$Gene
rankings_Hep4 <- sort(rankings_Hep4, decreasing = TRUE)
plot(rankings_Hep4)
min(rankings_Hep4)
max(rankings_Hep4)


set.seed(1234)

kegg_legacy_res_Hep4 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep4,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep4 <- fgsea(pathways = reactome,
                           stats = rankings_Hep4,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep4 <- fgsea(pathways = go,
                     stats = rankings_Hep4,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep4_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep4[, padj < 0.05]), sum(kegg_legacy_res_Hep4[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep4[, padj < 0.05]), sum(reactome_res_Hep4[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep4[, padj < 0.05]), sum(go_res_Hep4[, pval < 0.05])))
rownames(Hep4_fgsea) <- c("adj.pval", "p.val")
Hep4_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep4, title = "Hep-4 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep4_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep4, title = "Hep-4 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep4_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep4, title = "Hep-4 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep4_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_4_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-4, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_4_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## f. Hep-5
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-5, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep5 <- full_results$LogFC
names(rankings_Hep5) <- full_results$Gene
rankings_Hep5 <- sort(rankings_Hep5, decreasing = TRUE)
plot(rankings_Hep5)
min(rankings_Hep5)
max(rankings_Hep5)


set.seed(1234)

kegg_legacy_res_Hep5 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep5,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep5 <- fgsea(pathways = reactome,
                           stats = rankings_Hep5,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep5 <- fgsea(pathways = go,
                     stats = rankings_Hep5,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep5_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep5[, padj < 0.05]), sum(kegg_legacy_res_Hep5[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep5[, padj < 0.05]), sum(reactome_res_Hep5[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep5[, padj < 0.05]), sum(go_res_Hep5[, pval < 0.05])))
rownames(Hep5_fgsea) <- c("adj.pval", "p.val")
Hep5_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep5, title = "Hep-5 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep5_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep5, title = "Hep-5 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep5_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep5, title = "Hep-5 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep5_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_5_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-5, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_5_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## g. dHep
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "dHep, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_dHep_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_dHep <- full_results$LogFC
names(rankings_dHep) <- full_results$Gene
rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
plot(rankings_dHep)
min(rankings_dHep)
max(rankings_dHep)


set.seed(1234)

kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                              stats = rankings_dHep,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_dHep <- fgsea(pathways = reactome,
                           stats = rankings_dHep,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_dHep <- fgsea(pathways = go,
                     stats = rankings_dHep,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                         "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
rownames(dHep_fgsea) <- c("adj.pval", "p.val")
dHep_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"dHep_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_dHep, title = "dHep Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"dHep_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_dHep, title = "dHep Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"dHep_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_dHep_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "dHep, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_dHep_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

# 4. Oroboros Analysis 
## a. HEP Cells 
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
ncol(so_celltype) #130124 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
oro <- colnames(so_liver_sn@meta.data[,c(85:114)])

# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)
# Use lapply to pull all summaries and add gene names
# nebula_summaries <- do.call(rbind, lapply(names(nebula_results_list), function(gene) {
#   if (!is.null(nebula_results_list[[gene]]$summary)) {
#     df <- nebula_results_list[[gene]]$summary
#     df$gene <- gene  # Add gene name as a new column
#     return(df)
#   } else {
#     NULL
#   }
# }))
nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hepatocytes, Unadjusted (REML)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
print(volcano_plot)
dev.off()

```

### ii. glmmTMB
```{r}
#Access the raw count matrix
raw_counts <- GetAssayData(so_celltype_hvg, layer = "counts")

# Step 2: Round manually for sparse matrix (dgCMatrix)
raw_counts_rounded <- raw_counts
raw_counts_rounded@x <- round(raw_counts@x)

# Step 3: Extract the full gene expression matrix (rounded counts)
expr_matrix <- as.matrix(raw_counts_rounded)

#View how rounded counts data looks
#1. Randomly select a few genes for visualization (e.g., top 50 or random)
selected_genes <- sample(rownames(raw_counts_rounded), 25)  # Random 50 genes

#2. Get the expression data for those genes (rounded counts)
selected_data <- as.matrix(raw_counts_rounded[selected_genes, ])

# Convert matrix to data.frame
df_selected_data <- as.data.frame(t(selected_data))  # Transpose so that genes are columns

# 3. **Boxplot** for raw counts of selected genes across cells (to see distribution)
df_selected_data_long <- reshape2::melt(df_selected_data)  # Convert to long format

ggplot(df_selected_data_long, aes(x = variable, y = value, fill = variable)) +
  geom_boxplot() +
  labs(title = "Boxplot of Selected Genes' Counts", x = "Genes", y = "Count") +
  theme_minimal()+
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
    # legend.position = "none"
  )

# 4. **Heatmap** of raw counts (scaled by genes) for a few cells (e.g., top 50 cells)
top_cells <- sample(colnames(raw_counts_rounded), 50)  # Random 50 cells
heatmap_data <- as.matrix(raw_counts_rounded[selected_genes, top_cells])

# Plot heatmap
pheatmap(heatmap_data, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE, 
         show_colnames = FALSE,
         main = "Heatmap of Gene Expression (Selected Genes & Cells)") 

#Check rownames and colnames: Rownames = genes, Colnames = cells
#view(expr_matrix)
#Transpose gene expression dataset to merge with metadata
expr_matrix <- t(expr_matrix)
expr_matrix <- data.frame(expr_matrix) #Make a dataframe again after transposing

#Set gene list
gene_list_total <- hvgs

#Create cell variable to merge metadata in
expr_matrix$cellname <- rownames(expr_matrix) 
rownames(expr_matrix) <- NULL

# Extract the metadata
metadata <- so_celltype_hvg@meta.data

#Create cell variable to merge metadata into gene expression data
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,expr_matrix,by="cellname")
rm(metadata,expr_matrix)

#Make sure exposure/independent/x variable or group variable is a factor variable
data$steatosis_cat <- factor(data$steatosis_cat)
#Make sure to set reference level
data$steatosis_cat <- relevel(data$steatosis_cat,ref="Low Steatosis (0+1)")

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hepatocytes,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                           stats = rankings_Hepatocytes,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hepatocytes <- fgsea(pathways = go,
                     stats = rankings_Hepatocytes,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
Hepatocytes_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
##### GO
plot_fgsea_transpose(go_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```
### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)
# Use lapply to pull all summaries and add gene names
# nebula_summaries <- do.call(rbind, lapply(names(nebula_results_list), function(gene) {
#   if (!is.null(nebula_results_list[[gene]]$summary)) {
#     df <- nebula_results_list[[gene]]$summary
#     df$gene <- gene  # Add gene name as a new column
#     return(df)
#   } else {
#     NULL
#   }
# }))
nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hepatocytes_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hepatocytes, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hepatocytes_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## b. Hep-1
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-1, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-1
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-1_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_1_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep1 <- full_results$LogFC
names(rankings_hep1) <- full_results$Gene
rankings_hep1 <- sort(rankings_hep1, decreasing = TRUE)
plot(rankings_hep1)
min(rankings_hep1)
max(rankings_hep1)

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep1, title = "Hep-1 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep1_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep1, title = "Hep-1 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep1_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep1, title = "Hep-1 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep1_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_1_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-1, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_1_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```


## c. Hep-2
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-2, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-2
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-2_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_2_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep2 <- full_results$LogFC
names(rankings_hep2) <- full_results$Gene
rankings_hep2 <- sort(rankings_hep2, decreasing = TRUE)
plot(rankings_hep2)
min(rankings_hep2)
max(rankings_hep2)


set.seed(1234)

kegg_legacy_res_hep2 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_hep2,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_hep2 <- fgsea(pathways = reactome,
                           stats = rankings_hep2,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_hep2 <- fgsea(pathways = go,
                     stats = rankings_hep2,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

hep2_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_hep2[, padj < 0.05]), sum(kegg_legacy_res_hep2[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_hep2[, padj < 0.05]), sum(reactome_res_hep2[, pval < 0.05])),
                         "GO"=c(sum(go_res_hep2[, padj < 0.05]), sum(go_res_hep2[, pval < 0.05])))
rownames(hep2_fgsea) <- c("adj.pval", "p.val")
hep2_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep2, title = "Hep-2 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"hep2_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep2, title = "Hep-2 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"hep2_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep2, title = "Hep-2 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"hep2_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_2_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-2, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_2_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## d. Hep-3
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-3, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep3 <- full_results$LogFC
names(rankings_Hep3) <- full_results$Gene
rankings_Hep3 <- sort(rankings_Hep3, decreasing = TRUE)
plot(rankings_Hep3)
min(rankings_Hep3)
max(rankings_Hep3)


set.seed(1234)

kegg_legacy_res_Hep3 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep3,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep3 <- fgsea(pathways = reactome,
                           stats = rankings_Hep3,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep3 <- fgsea(pathways = go,
                     stats = rankings_Hep3,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep3_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep3[, padj < 0.05]), sum(kegg_legacy_res_Hep3[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep3[, padj < 0.05]), sum(reactome_res_Hep3[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep3[, padj < 0.05]), sum(go_res_Hep3[, pval < 0.05])))
rownames(Hep3_fgsea) <- c("adj.pval", "p.val")
Hep3_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep3, title = "Hep-3 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep3_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep3, title = "Hep-3 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep3_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep3, title = "Hep-3 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep3_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_3_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-3, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_3_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## e. Hep-4
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-4, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep4 <- full_results$LogFC
names(rankings_Hep4) <- full_results$Gene
rankings_Hep4 <- sort(rankings_Hep4, decreasing = TRUE)
plot(rankings_Hep4)
min(rankings_Hep4)
max(rankings_Hep4)


set.seed(1234)

kegg_legacy_res_Hep4 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep4,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep4 <- fgsea(pathways = reactome,
                           stats = rankings_Hep4,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep4 <- fgsea(pathways = go,
                     stats = rankings_Hep4,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep4_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep4[, padj < 0.05]), sum(kegg_legacy_res_Hep4[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep4[, padj < 0.05]), sum(reactome_res_Hep4[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep4[, padj < 0.05]), sum(go_res_Hep4[, pval < 0.05])))
rownames(Hep4_fgsea) <- c("adj.pval", "p.val")
Hep4_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep4, title = "Hep-4 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep4_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep4, title = "Hep-4 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep4_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep4, title = "Hep-4 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep4_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_4_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-4, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_4_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## f. Hep-5
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-5, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep5 <- full_results$LogFC
names(rankings_Hep5) <- full_results$Gene
rankings_Hep5 <- sort(rankings_Hep5, decreasing = TRUE)
plot(rankings_Hep5)
min(rankings_Hep5)
max(rankings_Hep5)


set.seed(1234)

kegg_legacy_res_Hep5 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep5,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep5 <- fgsea(pathways = reactome,
                           stats = rankings_Hep5,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep5 <- fgsea(pathways = go,
                     stats = rankings_Hep5,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep5_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep5[, padj < 0.05]), sum(kegg_legacy_res_Hep5[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep5[, padj < 0.05]), sum(reactome_res_Hep5[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep5[, padj < 0.05]), sum(go_res_Hep5[, pval < 0.05])))
rownames(Hep5_fgsea) <- c("adj.pval", "p.val")
Hep5_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep5, title = "Hep-5 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep5_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep5, title = "Hep-5 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep5_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep5, title = "Hep-5 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep5_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_5_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-5, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_5_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## g. dHep
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "dHep, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_dHep_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_dHep <- full_results$LogFC
names(rankings_dHep) <- full_results$Gene
rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
plot(rankings_dHep)
min(rankings_dHep)
max(rankings_dHep)


set.seed(1234)

kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                              stats = rankings_dHep,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_dHep <- fgsea(pathways = reactome,
                           stats = rankings_dHep,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_dHep <- fgsea(pathways = go,
                     stats = rankings_dHep,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                         "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
rownames(dHep_fgsea) <- c("adj.pval", "p.val")
dHep_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"dHep_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_dHep, title = "dHep Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"dHep_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_dHep, title = "dHep Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"dHep_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- 2000-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(2000-length(nonconverge_genes))/2000)*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_dHep_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "dHep, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_dHep_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```