---
title: "Liver_snRNAseq"
output: html_document
date: "2025-04-23"
---

# 1. Set up Libraries & Directores
```{r libraries, echo=F, include = F}
#Load Libraries
source("Libraries.R")

#Local file path
# dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
# dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
# dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c("/Users/hhampson/Documents/UW/1_Ongoing Projects/Liver scRNAseq/2_Results")

#Lambda file path
# dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
# dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c(fs::path(dir.dat,"Liver project/Results"))

#Mac Studio File Path
dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
dir.results <- c("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/NEBULA Results")
# # dir.ipa <- c("/Users/hhampson/Documents/IPA/Results")

# #Laura's Mac Studio File Path
# dir.dat <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive")
# dir.results <- c("/Users/haileyhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver Project/NEBULA Results")

#Load functions
source("Liver_functions.R")
# source("/Users/hhampson/CHCO-Code/Petter Bjornstad/Data Processing and Analysis/Standard_Functions.R")

knitr::opts_knit$set(root.dir = dir.results)
```

# 2. Pre-Processing & Qualtiy Control
## a. Load Seurat Object snRNA
```{r, include = F, fig.height=7,fig.width=10}
# Liver snRNA data processing
invisible(gc())
#Local
# dir.dat <- c("/Users/hhampson/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Biostatistics Core Shared Drive")
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
# #Lambda
# so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
DefaultAssay(so_liver_sn) <- "RNA"
dim(so_liver_sn)#36601 130124
nrow(so_liver_sn) #36,601 genes
ncol(so_liver_sn) # 130,124 nuceli
invisible(gc())
```

## b. Load MetaData, Clean and Merge
```{r}
#Local
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#Lambda
# meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
unique(meta_liver_raw$Kit_Lot)
unique(meta_liver_raw$StudyID)
meta_liver_raw <- meta_liver_raw %>%
  filter(StudyID!="IT2D-18") %>%
  filter(StudyID!="IT2D-16")

# meta_liver_raw <- meta_liver_raw %>%
#   mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
#   mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
#   mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
#   mutate(either=ifelse(glp1agonist=="Yes" | sglt2=="Yes","Yes","No")) %>%
#   mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
#   mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))
# meta_liver_raw <- meta_liver_raw %>%
#   mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
#                           sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
#                           glp1_exclusive == "Yes" ~ "glp1_exclusive",
#                           either == "Yes" ~ "glp1_or_sglt2",
#                           neither == "Yes" ~ "no_med",
#                           hc == "Yes" ~ "obese_control"))
# meta_liver_raw <- meta_liver_raw %>%
#   mutate(group3=ifelse(either=="Yes","either","neither"))

#Create Liver Status Variables
meta_liver_raw <- meta_liver_raw %>%
  mutate(steatosis_cat = ifelse((steatosis_grade==0 | steatosis_grade==1),"Low Steatosis (0+1)","High Steatosis (2+3)")) %>%
  mutate(fibrosis_cat = case_when(steatosis_grade==0 & fibrosis_stage==0 ~ "No Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage==0 ~ "Any Steatosis, No Fibrosis",
                                  steatosis_grade>0 & fibrosis_stage>0 ~ "Any Fibrosis & Any Steatosis"))

#Merge liver sn metadata with meta liver raw metadata
meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)
#Merge into seurat object
rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"
invisible(gc())

dim(so_liver_sn) #36601 130124
ncol(so_liver_sn) #130124 nuclei 
nrow(so_liver_sn) # 36601 genes

#YE JI's filtering code for percent expression 
#Filter out rare genes expressed in less than "gene_pct" of cells
expr_matrix <- as.matrix(GetAssayData(so_liver_sn, layer = "counts"))
# expr_matrix <- as.matrix(GetAssayData(so_liver_sn, assay = "RNA", layer = "counts"))
expr_matrix <- so_liver_sn@assays$RNA@counts
# Calculate the proportion of cells expressing each gene
num_cells_per_gene <- rowSums(expr_matrix > 0)  # Count nonzero cells per gene
total_cells <- ncol(expr_matrix)  # Total number of cells
gene_proportion <- num_cells_per_gene / total_cells  # Fraction of cells expressing each gene
remove(expr_matrix)
# Keep genes expressed in at least "gene_pct" of cells
genes_to_keep <- names(gene_proportion[gene_proportion >= 0.01])
so_liver_sn <- subset(so_liver_sn, features = genes_to_keep)
ncol(so_liver_sn) #130,124 nuclei
nrow(so_liver_sn) # 11313 genes

# Step 2: Remove mitochondrial genes (those starting with "MT")
mito_genes <- grep("^MT-", rownames(so_liver_sn), value = TRUE)
#keep_ids <- unique(rownames(so_liver_sn)[which(!rownames(so_liver_sn) %in% mito_genes)])
# so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn), mito_genes))
# so_liver_sn <- subset(so_liver_sn,kit_id!="KL-0029535")
#so_liver_sn$Gene <- rownames(so_liver_sn)
#so_liver_sn <- subset(so_liver_sn, Gene %in% keep_ids)
so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
#so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn@assays$RNA@counts), mito_genes))
# so_liver_sn <- subset(so_liver_sn, !rownames(so_liver_sn) %in% mito_genes)
# grep("^MT-", rownames(so_liver_sn@assays$RNA@counts), value = TRUE)
dim(so_liver_sn@assays$RNA@counts) #9276 186125
dim(so_liver_sn@assays$RNA@data) #9276 186125
dim(so_liver_sn@assays$RNA)#9276 186125
sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts))) #0
ncol(so_liver_sn) #130124 cells
nrow(so_liver_sn) #11300 genes


#Step 3: Remove ribosomal Genes
# Identify ribosomal genes
ribo_genes <- c(
  "RPL22", "RPL11", "RPS8", "RPL5", "RPS27", "RPS7", "RPS27A", "RPL31", "RPL37A", "RPL32", "RPL15", "RPL14", "RPL29",
  "RPL24", "RPL22L1", "RPL35A", "RPL9", "RPL34", "RPS3A", "RPL37", "RPS23", "RPS14", "RPS18", "RPS10", "RPL10A", 
  "RPS20", "RPL7", "RPL30", "RPL8", "RPS6", "RPL35", "RPL12", "RPL7A", "RPS24", "RPLP2", "RPL27A", "RPS13", "RPS3",
  "RPS25", "RPS26", "RPL41", "RPL6", "RPLP0", "RPL21", "RPS29", "RPL4", "RPLP1", "RPS17", "RPS2", "RPS15A", "RPL13",
  "RPL26", "RPL23A", "RPL23", "RPL19", "RPL27", "RPL38", "RPL17", "RPS15", "RPL36", "RPS28", "RPL18A", "RPS16", 
  "RPS19", "RPL18", "RPL13A", "RPS11", "RPS9", "RPL28", "RPS5", "RPS21", "RPL3", "RPS4X", "RPL36A", "RPL39", 
  "RPL10", "RPS4Y1"
)
so_liver_sn <- subset(so_liver_sn, features = setdiff(rownames(so_liver_sn), ribo_genes))
# sum(grepl("^MT-", rownames(so_liver_sn))) #0
length(which(rownames(so_liver_sn) %in% ribo_genes)) #0
ncol(so_liver_sn) #130,124 cells
nrow(so_liver_sn) #11224 genes


# # Identify mitochondrial genes
# mt_genes <- grepl("^MT", rownames(so_liver_sn@assays$RNA@counts))
# 
# # Calculate the percentage of mitochondrial genes per cell
# mt_percentage <- Matrix::colSums(so_liver_sn@assays$RNA@counts[mt_genes, ]) / Matrix::colSums(so_liver_sn@assays$RNA@counts)
# 
# # Filter out cells where the proportion of mitochondrial genes is > 15%
# cells_to_keep <- mt_percentage <= 0.15
# 
# # Subset the object to retain only cells with <= 15% mitochondrial genes
# so_liver_sn <- so_liver_sn[, cells_to_keep]



# # Check dimensions after filtering
# dim(so_liver_sn@assays$RNA@counts) #9276 81882
# dim(so_liver_sn@assays$RNA)#9276 81882
# dim(so_liver_sn)#9276 81882
# 
# # Optionally, you can also check how many MT genes remain in the filtered dataset
# sum(grepl("^MT-", rownames(so_liver_sn@assays$RNA@counts)))
# ncol(so_liver_sn)  # Number of cells remaining
# 
# # Step 3: Renormalize the Seurat object after filtering
# so_liver_sn <- NormalizeData(so_liver_sn)
# 
# #Check normalized slot
# dim(so_liver_sn@assays$RNA@data) #5714 186125
# sum(grepl("^MT", rownames(so_liver_sn@assays$RNA@data))) #70
# ncol(so_liver_sn) #186125 cells
# nrow(so_liver_sn) #5714 genes
# 
# rm(gene_expression_matrix)
so_liver_sn <- NormalizeData(so_liver_sn)
so_liver_sn <- ScaleData(so_liver_sn, features = VariableFeatures(so_liver_sn))

#Create general hepatocyte cell type variable
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype) | grepl("dHep",so_liver_sn$celltype),"Hep","Non-Hep")
# so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hep","Non-Hep")

# table(so_liver_sn$celltype,so_liver_sn$celltype2)

# rm(expression_data)
DefaultAssay(so_liver_sn) <- "RNA" 

# Calculate cell library size for offset in NEBULA --------------------------------------
counts_layer <- round(GetAssayData(so_liver_sn, layer = "counts"))
library_size <- Matrix::colSums(counts_layer)
so_liver_sn$library_size <- library_size
# View(so_liver_sn@meta.data)

# # TMM offset
# dge <- DGEList(counts = counts_layer)
# dge <- calcNormFactors(dge, method = "TMM")
# tmm_offset <- log(dge$samples$lib.size) + log(dge$samples$norm.factors)
# so_liver_sn$tmm_offset <- tmm_offset

# Pooled offset
bp <- MulticoreParam(workers = 63) 
sce <- SingleCellExperiment(assays = list(counts = counts_layer))
# sce <- computeSumFactors(sce)
sce <- computeSumFactors(sce, BPPARAM = bp)
# View size factors
# sizeFactors(sce)
# STEP 3: Calculate offset → log(size factors)
pooled_offset <- sizeFactors(sce)
so_liver_sn$pooled_offset <- pooled_offset
hist(so_liver_sn$pooled_offset)

# #Calculate cell library size for offset
# counts_layer <- round(GetAssayData(so_liver_sn, layer = "counts")) # load counts and round
# library_size <- Matrix::colSums(counts_layer) #calculate the column sums aka the sum of all genes in each cell
# so_liver_sn$library_size <- library_size #Put into the seurat object
# #Check that it is accessible in the metadata
# meta <- so_liver_sn@meta.data #It should be in the dataframe, delete meta data check df
# rm(meta,counts_layer)
# # hist(so_liver_sn$library_size)
```
## c. Visualize
###i. All Cells
```{r}
so_liver_sn <- FindVariableFeatures(object = so_liver_sn)
so_liver_sn <- RunPCA(so_liver_sn, features = VariableFeatures(object = so_liver_sn),assay="RNA")
ElbowPlot(so_liver_sn)

# # Find neighbors and clusters (again using integrated data)
so_liver_sn <- FindNeighbors(so_liver_sn, assay = "RNA", dims = 1:20)
so_liver_sn <- FindClusters(so_liver_sn, resolution = 0.5)

# Perform UMAP 
#Print plot
png(fs::path(dir.results, "UMAP_by_celltype.png"),
    width = 4500, height = 3000, res = 300)
p <- DimPlot(so_liver_sn, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")
print(p)
dev.off()


#Print plot
png(fs::path(dir.results, "UMAP_by_Steatosis.png"),
    width = 4500, height = 3000, res = 300)
p <- DimPlot(
  so_liver_sn,
  reduction = "umap",
  group.by = "steatosis_cat",
  label = FALSE,
  raster = FALSE
) +
  scale_color_manual(values = c("darkblue","lightgray")) +
  ggtitle("UMAP by Steatosis Status")
print(p)
dev.off()

#Barcharts of proportions
cellcount <- so_liver_sn@meta.data
prop_plot <- ggplot(data=cellcount,aes(StudyID, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) 
# ggtitle("PT Cells") +
# scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
png(fs::path(dir.results, "Barchart_All_Participants_All_Cells.png"),width = 4500, height = 3000, res = 300)
print(prop_plot)
dev.off()

prop_plot <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) 
png(fs::path(dir.results, "Barchart_Steatosis_Cat_All_Cells.png"),width = 3000, height = 3000, res = 300)
print(prop_plot)
dev.off()
```
### ii. Hepatocytes Only
```{r}
#Filter to hepatocytes only
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA"

so_celltype <- FindVariableFeatures(object = so_celltype)
so_celltype <- RunPCA(so_celltype, features = VariableFeatures(object = so_celltype),assay="RNA")
ElbowPlot(so_celltype)

# # Find neighbors and clusters (again using integrated data)
so_celltype <- FindNeighbors(so_celltype, assay = "RNA", dims = 1:20)
so_celltype <- FindClusters(so_celltype, resolution = 0.5)

# Perform UMAP 
#Print plot
png(fs::path(dir.results, "UMAP_by_celltype.png"),
    width = 4500, height = 3000, res = 300)
p <- DimPlot(so_celltype, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")
print(p)
dev.off()


#Print plot
png(fs::path(dir.results, "UMAP_by_Steatosis.png"),
    width = 4500, height = 3000, res = 300)
p <- DimPlot(
  so_celltype,
  reduction = "umap",
  group.by = "steatosis_cat",
  label = FALSE,
  raster = FALSE
) +
  scale_color_manual(values = c("darkblue","lightgray")) +
  ggtitle("UMAP by Steatosis Status")
print(p)
dev.off()

#Barcharts of proportions
cellcount <- so_celltype@meta.data
prop_plot <- ggplot(data=cellcount,aes(StudyID, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) 
# ggtitle("PT Cells") +
# scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
png(fs::path(dir.results, "Barchart_All_Participants_All_Cells.png"),width = 4500, height = 3000, res = 300)
print(prop_plot)
dev.off()

prop_plot <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) 
png(fs::path(dir.results, "Barchart_Steatosis_Cat_All_Cells.png"),width = 3000, height = 3000, res = 300)
print(prop_plot)
dev.off()


prop_plot <- ggplot(data=cellcount,aes(steatosis_grade, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
png(fs::path(dir.results, "Barchart_Steatosis_Grade_All_Cells.png"),width = 3500, height = 3000, res = 300)
print(prop_plot)
dev.off()


```


## d. Explore Distribution & Zero-Inflation
###i. All Celltypes
```{r}
#Check if data have been normalized 
so_liver_sn@assays  # Should show e.g., "RNA" with "counts" and "data"
head(GetAssayData(so_liver_sn, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_liver_sn, layer = "data")[, 1:5])    # Normalized data

#Check for normality
# Open a PDF device (all plots go here)
pdf(fs::path(dir.results,"Normalization_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesN <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Normalized Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

pdf(fs::path(dir.results,"Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_liver_sn), 100)
genesR <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_liver_sn, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Try log transforming
log_data <- log1p(GetAssayData(so_liver_sn, layer = "data"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Normalized_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesN
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

log_data <- log1p(GetAssayData(so_liver_sn, layer = "count"))  # log(x + 1)
pdf(fs::path(dir.results,"Log_Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
# genes <- sample(rownames(so_liver_sn), 100)
genes <- genesR
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(log_data[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

#Check for zero inflation
pdf(fs::path(dir.results,"Zero_Inflation_Visualization.pdf"), width = 8, height = 6)
counts <- GetAssayData(so_liver_sn, layer = "counts")
zero_prop <- rowMeans(counts == 0)
hist(zero_prop, main = "Proportion of Zeros for Genes", xlab = "Zero Proportion")
# Add a vertical line at 50%
abline(v = 0.5, col = "red", lwd = 2, lty = 2)
dev.off()

#Check for overdispersion
# Use raw counts
counts <- GetAssayData(so_liver_sn, layer = "counts")

# Pick a few genes or do this genome-wide
gene_means <- rowMeans(counts)
gene_vars <- apply(counts, 1, var)

# Dispersion estimate (Var / Mean)
dispersion <- gene_vars / gene_means

# Filter to only genes where mean > 0 (to avoid divide-by-zero)
valid_genes <- gene_means > 0
dispersion <- dispersion[valid_genes]

# View distribution
pdf(fs::path(dir.results,"Overdispersion_Visualization.pdf"), width = 8, height = 6)
hist(dispersion, breaks=200, main="Dispersion across genes", xlab="Dispersion (Var / Mean)",xlim = c(0, 2))
abline(v = 1, col="red")  # Poisson expectation
dev.off()

# Check how many genes have dispersion > 1
overdispersed_genes <- sum(dispersion > 1)
total_genes <- length(dispersion)
percent_overdispersed <- (overdispersed_genes / total_genes) * 100

# Print summary and recommendation
print(paste0("Overdispersed genes: ", 
             overdispersed_genes, 
             " out of ", 
             total_genes, 
             " (", 
             round(percent_overdispersed, 3), 
             "%)"))

if (percent_overdispersed > 10) {
  print("✅ Overdispersion detected. Consider using a Negative Binomial model.")
} else {
  print("ℹ️ Little evidence of overdispersion. Poisson model may be sufficient.")
}
rm(counts)
```

### ii. Hepatocytes Only
```{r}
so_celltype <- subset(so_celltype,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

#Check if data have been normalized 
so_celltype@assays  # Should show e.g., "RNA" with "counts" and "data"
head(GetAssayData(so_celltype, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_celltype, layer = "data")[, 1:5])    # Normalized data

#Check for normality
# Open a PDF device (all plots go here)
pdf(fs::path(dir.results,"Hep_Normalization_gene_expression_histograms_counts.pdf"), width = 8, height = 6)

# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_celltype), 100)
genesN <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_celltype, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Normalized Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

pdf(fs::path(dir.results,"Hep_Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# Randomly select 100 genes from the Seurat object
genes <- sample(rownames(so_celltype), 100)
genesR <- genes
# Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
par(mfrow = c(2, 3))

# Loop over each randomly selected gene
for (g in genes) {
  
  # Plot a histogram of the expression values for gene 'g'
  # using normalized expression values from the "data" slot
  hist(GetAssayData(so_celltype, layer = "counts")[g, ],
       main = g,                # Title of the plot = gene name
       xlab = "Raw Counts Expression")     # Label for x-axis
}

# Close the PDF device — this writes the file to disk
dev.off()

# #Try log transforming
# log_data <- log1p(GetAssayData(so_celltype, layer = "data"))  # log(x + 1)
# pdf(fs::path(dir.results,"Hep_Log_Normalized_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# 
# # Randomly select 100 genes from the Seurat object
# # genes <- sample(rownames(so_celltype), 100)
# genes <- genesN
# # Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
# par(mfrow = c(2, 3))
# 
# # Loop over each randomly selected gene
# for (g in genes) {
#   
#   # Plot a histogram of the expression values for gene 'g'
#   # using normalized expression values from the "data" slot
#   hist(log_data[g, ],
#        main = g,                # Title of the plot = gene name
#        xlab = "Log of Normalized Counts Expression")     # Label for x-axis
# }
# 
# # Close the PDF device — this writes the file to disk
# dev.off()
# 
# log_data <- log1p(GetAssayData(so_celltype, layer = "count"))  # log(x + 1)
# pdf(fs::path(dir.results,"Hep_Log_Raw_Count_gene_expression_histograms_counts.pdf"), width = 8, height = 6)
# # Randomly select 100 genes from the Seurat object
# # genes <- sample(rownames(so_celltype), 100)
# genes <- genesR
# # Set up a 2x3 plotting layout so you can plot multiple histograms in one figure
# par(mfrow = c(2, 3))
# 
# # Loop over each randomly selected gene
# for (g in genes) {
#   
#   # Plot a histogram of the expression values for gene 'g'
#   # using normalized expression values from the "data" slot
#   hist(log_data[g, ],
#        main = g,                # Title of the plot = gene name
#        xlab = "Log of Raw Counts Expression")     # Label for x-axis
# }
# 
# # Close the PDF device — this writes the file to disk
# dev.off()

#Check for zero inflation
pdf(fs::path(dir.results,"Hep_Zero_Inflation_Visualization.pdf"), width = 8, height = 6)
counts <- GetAssayData(so_celltype, layer = "counts")
zero_prop <- rowMeans(counts == 0)
hist(zero_prop, main = "Proportion of Zeros for Genes", xlab = "Zero Proportion")
# Add a vertical line at 50%
abline(v = 0.5, col = "red", lwd = 2, lty = 2)
dev.off()

#Check for overdispersion
# Use raw counts
counts <- GetAssayData(so_celltype, layer = "counts")

# Pick a few genes or do this genome-wide
gene_means <- rowMeans(counts)
gene_vars <- apply(counts, 1, var)

# Dispersion estimate (Var / Mean)
dispersion <- gene_vars / gene_means

# Filter to only genes where mean > 0 (to avoid divide-by-zero)
valid_genes <- gene_means > 0
dispersion <- dispersion[valid_genes]

# View distribution
pdf(fs::path(dir.results,"Hep_Overdispersion_Visualization.pdf"), width = 8, height = 6)
hist(dispersion, breaks=200, main="Dispersion across genes", xlab="Dispersion (Var / Mean)",xlim = c(0, 2))
abline(v = 1, col="red")  # Poisson expectation
dev.off()

# Check how many genes have dispersion > 1
overdispersed_genes <- sum(dispersion > 1)
total_genes <- length(dispersion)
percent_overdispersed <- (overdispersed_genes / total_genes) * 100

# Print summary and recommendation
print(paste0("Overdispersed genes: ", 
             overdispersed_genes, 
             " out of ", 
             total_genes, 
             " (", 
             round(percent_overdispersed, 3), 
             "%)"))

if (percent_overdispersed > 10) {
  print("✅ Overdispersion detected. Consider using a Negative Binomial model.")
} else {
  print("ℹ️ Little evidence of overdispersion. Poisson model may be sufficient.")
}
rm(counts)
```

## e. Evaluate Covariates before Adjustment
```{r}
#Categorical Covariates
#Get metadata for everyone
# dat <- so_liver_sn@meta.data %>%
#   group_by(StudyID) %>%
#   summarise(across(everything(), first)) %>%
#   ungroup() 
dat <- so_liver_sn@meta.data %>%
  group_by(StudyID) %>%
  summarise(across(everything(), ~ .x[1])) %>%
  ungroup()

dat$study <- case_when(grepl("IT2D",dat$StudyID) ~ "IMPROVE",
                       grepl("MANGO",dat$StudyID) ~ "MANGO",
                       grepl("BSA",dat$StudyID) ~ "BIOBANK")
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

dat <- dat %>% 
  mutate(group2=case_when(both == "Yes" ~ "glp1_and_sglt2",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          either == "Yes" ~ "glp1_or_sglt2",
                          neither == "Yes" ~ "no_med",
                          hc == "Yes" ~ "obese_control"))

# table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
# table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
# table(dat2$both) #+/+ #1 Both, 4 not both
# table(dat2$neither) #-/- #3 neither, 2 on someting
# dat$steatosis_cat <- factor(dat$steatosis_cat,levels=c("0+1","2+3"),labels=c("Low Steatosis (0+1)","High Steatosis (2+3)"))
dat$fibrosis_cat <- factor(dat$fibrosis_cat)
dat$diagnosis_of_MASLD <- factor(dat$diagnosis_of_MASLD, levels=c("Yes","No"), labels=c("MASLD", "No MASLD"))
dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age_surgery)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$sglt2) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status (Yes/No)"
label(dat$diagnosis_of_MASLD) <- "MASLD (Yes/No)"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$a1c) <- "HbA1c (%)"
#label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
#label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_grade) <- "Steatosis Grade"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
#label(dat$group2) <- "Disease Group"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No SGLT2i or GLP-1ra (Yes/No)"
label(dat$metformin) <- "Metformin (Yes/No)"
label(dat$insulin) <- "Insulin (Yes/No)"
label(dat$alt) <- "ALT (U/L)"
label(dat$ast) <- "AST (U/L)"
label(dat$ggt) <- "GGT (U/L)"
label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

label(dat$fibrosis_cat) <- "Fibrosis Category"
label(dat$steatosis_cat) <- "Steatosis Category"

#Table 1. 
table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
         diagnosis_of_diabetes + diagnosis_of_MASLD+
         sglt2 + glp1agonist  | study, data=dat)

table1(~ age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg +a1c+
         diagnosis_of_MASLD+
         sglt2 + glp1agonist +metformin + insulin | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_MASLD + diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_diabetes, data=dat)

table1(~ a1c+ diagnosis_of_diabetes+ sglt2 + glp1agonist + metformin + insulin + alt + ast + ggt + steatosis_cat + fibrosis_cat +steatosis_grade + fibrosis_stage + lobular_inflammation_percent  
       | diagnosis_of_MASLD, data=dat)

table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither| steatosis_cat, data=dat)
table1(~ alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)
table1(~ age_surgery + nih_sex + nih_ethnicity + bmi + tg +a1c
       +metformin + insulin + diagnosis_of_MASLD+ diagnosis_of_diabetes+both+glp1_exclusive+sglt2_exclusive+neither + alt + ast + ggt +lobular_inflammation_percent + fibrosis_cat| steatosis_cat, data=dat)

#Examine covariates
# age_surgery + nih_sex + nih_race + nih_ethnicity + bmi + tg + a1c+
#   diagnosis_of_diabetes + diagnosis_of_MASLD+
#   sglt2 + glp1agonist

cat_covariates <- c("nih_sex", "nih_race","nih_ethnicity", "diagnosis_of_diabetes", "diagnosis_of_MASLD", "group2","fibrosis_stage")
plot_list <- list()

for (covariate in cat_covariates) {
  p <- ggplot(dat, aes_string(x = "steatosis_cat", fill = covariate)) +
    geom_bar(position = "fill") +  # use "dodge" for absolute counts
    labs(y = "Proportion", x = NULL) +
    # title = paste0("Distribution of ", covariate, " by Steatosis Category")
    scale_y_continuous(labels = scales::percent_format()) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust depending on how many you have)
wrap_plots(plot_list, ncol = 4)


#Continuous Covariates
con_covariates <- c("age_surgery", "bmi", "tg", "a1c")
plot_list <- list()

for (covariate in con_covariates) {
  p <- ggplot(dat, aes_string(x = "steatosis_cat", y = covariate, fill = "steatosis_cat")) +
    geom_boxplot(outlier.shape = NA, alpha = 0.7) +
    geom_jitter(width = 0.2, alpha = 0.4, color = "black") +  # optional: show points
    labs(title = paste0("Distribution of ", covariate, " by group"),
         x = "Group", y = covariate) +
    theme_minimal() +
    theme(legend.position = "none")
  
  plot_list[[covariate]] <- p
}

# Display all plots together (adjust ncol/nrow as needed)
wrap_plots(plot_list, ncol = 4 )
```

# 3. Steatosis Analysis 
## a. HEP Cells 
```{r}
#Filter to Hep Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
ncol(so_celltype) #130124 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# # so_celltype_hvg <- subset(so_celltype, features = hvgs)
# subset_counts <- GetAssayData(so_celltype, layer = "counts")[hvgs, ]
# DefaultAssay(so_celltype_hvg) <- "RNA" 
# rm(so_celltype)
```

### i. NEBULA
#### Unadjusted
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)
# # count_gene <- counts_hvg[g, , drop = FALSE]
# meta_gene_hvg <- .subset(so_celltype_hvg,features=g)@meta.data
# pred_gene_hvg <- model.matrix(~steatosis_cat, data = meta_gene_hvg)
# data_g_hvg = list(count = counts_hvg, id = id, pred = pred_gene_hvg)
# offset_hvg = Matrix::colSums(data_g_hvg$count) 

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    # library <- meta_gene$pooled_offset
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_pooled_offset.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=logFC_steatosis_catHigh.Steatosis..2.3.)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
# volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
#   geom_point(alpha = 0.7) +  # Plot points with transparency
#   scale_color_identity() +  # Use the color column directly
#   theme_minimal() +  # Minimal theme
#   labs(
#     title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
#     subtitle = "Hepatocytes, Unadjusted (REML,Log Normal,Pooled Offset)",
#     x = "FC",
#     y = "-log10(P-Value)",
#     color = "FC Direction Direction",
#     caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
#   ) +
#   theme(
#     plot.title = element_text(hjust = 0),
#     axis.text.x = element_text(angle = 0, hjust = 1)
#   )+
#   xlim(min,max)+
#   # # Add labels for significant points
#   geom_text(data = significant_df, aes(label = gene),
#             vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# # Add labels for significant points with ggrepel
# # geom_text_repel(data = significant_df, aes(label = Gene),
# #                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
# 
# volcano_plot
# 
# # pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# # print(volcano_plot)
# # dev.off()
# png(fs::path(dir.results, "Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_pooled_offset.png"),
#     width = 3000, height = 2100, res = 300)
# print(volcano_plot)
# dev.off()


volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.6) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  # scale_color_identity(guide = "legend", 
  #                    name = "LogFC Direction", 
  #                    labels = c("Downregulated", "N.S.", "Upregulated"), 
  #                    breaks = c("#003366", "lightgrey", "#990000"))+
  theme_minimal() +  # Minimal theme
  labs(
    title = "Differentially Expressed Genes in High Steatosis vs. Low Steatosis",
    subtitle = "Hepatocytes",
    x = "LogFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0,face="bold"),
    plot.subtitle = element_text(face="bold"),
    axis.text.x = element_text(angle = 0, hjust = 1),
    legend.text = element_text(face="bold")
  )+
  # xlim(min,max)+
  xlim(-2,1.6)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black",fontface="bold")

volcano_plot
png(fs::path(dir.results, "Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000.png"),
    width = 2500, height = 2000, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                                     stats = rankings_Hepatocytes,
                                     scoreType = 'std', 
                                     minSize = 3,
                                     maxSize = 500,
                                     nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                                  stats = rankings_Hepatocytes,
                                  scoreType = 'std', 
                                  minSize = 3,
                                  maxSize = 500,
                                  nproc = 1)
go_res_Hepatocytes <- fgsea(pathways = go,
                            stats = rankings_Hepatocytes,
                            scoreType = 'std', 
                            minSize = 5,
                            maxSize = 500,
                            nproc = 1)

Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                                "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                                "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
Hepatocytes_fgsea

# ##### KEGG Legacy
# a <- plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

# combined_plot <- b + c + plot_layout(ncol = 3)
# print(combined_plot)
# ggsave(fs::path(dir.results,"Hepatocytes_top30_pathways_unadjusted.jpeg"),
#        width = 40, height = 10, scale = 1)
combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hepatocytes_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```
#### Adjusted
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)
# # count_gene <- counts_hvg[g, , drop = FALSE]
# meta_gene_hvg <- .subset(so_celltype_hvg,features=g)@meta.data
# id <- meta_gene_hvg$Kit_Lot
# pred_gene_hvg <- model.matrix(~steatosis_cat, data = meta_gene_hvg)
# data_g_hvg = list(count = counts_hvg, id = id, pred = pred_gene_hvg)
# offset_hvg = Matrix::colSums(data_g_hvg$count) 

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat+a1c, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_a1c.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hepatocytes, Adjusted for HbA1c (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_a1c.png"),
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### ii. glmmTMB
```{r}
#Access the raw count matrix
raw_counts <- GetAssayData(so_celltype_hvg, layer = "counts")

# Step 2: Round manually for sparse matrix (dgCMatrix)
raw_counts_rounded <- raw_counts
raw_counts_rounded@x <- round(raw_counts@x)

# Step 3: Extract the full gene expression matrix (rounded counts)
expr_matrix <- as.matrix(raw_counts_rounded)

#View how rounded counts data looks
#1. Randomly select a few genes for visualization (e.g., top 50 or random)
selected_genes <- sample(rownames(raw_counts_rounded), 25)  # Random 50 genes

#2. Get the expression data for those genes (rounded counts)
selected_data <- as.matrix(raw_counts_rounded[selected_genes, ])

# Convert matrix to data.frame
df_selected_data <- as.data.frame(t(selected_data))  # Transpose so that genes are columns

# 3. **Boxplot** for raw counts of selected genes across cells (to see distribution)
df_selected_data_long <- reshape2::melt(df_selected_data)  # Convert to long format

ggplot(df_selected_data_long, aes(x = variable, y = value, fill = variable)) +
  geom_boxplot() +
  labs(title = "Boxplot of Selected Genes' Counts", x = "Genes", y = "Count") +
  theme_minimal()+
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
    # legend.position = "none"
  )

# 4. **Heatmap** of raw counts (scaled by genes) for a few cells (e.g., top 50 cells)
top_cells <- sample(colnames(raw_counts_rounded), 50)  # Random 50 cells
heatmap_data <- as.matrix(raw_counts_rounded[selected_genes, top_cells])

# Plot heatmap
pheatmap(heatmap_data, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE, 
         show_colnames = FALSE,
         main = "Heatmap of Gene Expression (Selected Genes & Cells)") 

#Check rownames and colnames: Rownames = genes, Colnames = cells
#view(expr_matrix)
#Transpose gene expression dataset to merge with metadata
expr_matrix <- t(expr_matrix)
expr_matrix <- data.frame(expr_matrix) #Make a dataframe again after transposing

#Set gene list
gene_list_total <- hvgs

#Create cell variable to merge metadata in
expr_matrix$cellname <- rownames(expr_matrix) 
rownames(expr_matrix) <- NULL

# Extract the metadata
metadata <- .subset(so_celltype_hvg,features=g)@meta.data

#Create cell variable to merge metadata into gene expression data
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,expr_matrix,by="cellname")
rm(metadata,expr_matrix)

#Make sure exposure/independent/x variable or group variable is a factor variable
data$steatosis_cat <- factor(data$steatosis_cat)
#Make sure to set reference level
data$steatosis_cat <- relevel(data$steatosis_cat,ref="Low Steatosis (0+1)")

```


### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hepatocytes_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hepatocytes, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot


#Examine distributions 
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
# Step 1: Pull expression for gene PNPLA3
gene_expr <- FetchData(so_celltype, vars = target_genes)

# Step 2: Pull metadata for steatosis_cat
metadata <- so_celltype@meta.data["steatosis_cat"]

# Step 3: Combine into one dataframe
df <- cbind(gene_expr, metadata)
# colnames(df) <- c("PNPLA3", "steatosis_cat")

# Step 4: Plot using ggplot
ggplot(df, aes(x = PNPLA3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "PNPLA3 Expression by Steatosis Category",
       x = "PNPLA3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = GCKR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "GCKR Expression by Steatosis Category",
       x = "GCKR Expression",
       y = "Cell Count")

ggplot(df, aes(x = TM6SF2, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "TM6SF2 Expression by Steatosis Category",
       x = "TM6SF2 Expression",
       y = "Cell Count")

ggplot(df, aes(x = APOC3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "APOC3 Expression by Steatosis Category",
       x = "APOC3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = MTOR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "MTOR Expression by Steatosis Category",
       x = "MTOR Expression",
       y = "Cell Count")
```

## b. Hep-1
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 
so_celltype@assays
head(GetAssayData(so_celltype, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_celltype, layer = "data")[, 1:5])  

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

#Check for overdispersion
# Use raw counts
counts <- GetAssayData(so_celltype, layer = "counts")
counts <- counts[rownames(counts) %in% hvgs, ]


# Pick a few genes or do this genome-wide
gene_means <- rowMeans(counts)
gene_vars <- apply(counts, 1, var)

# Dispersion estimate (Var / Mean)
dispersion <- gene_vars / gene_means

# Filter to only genes where mean > 0 (to avoid divide-by-zero)
valid_genes <- gene_means > 0
dispersion <- dispersion[valid_genes]

# View distribution
# pdf(fs::path(dir.results,"Hep_Overdispersion_Visualization.pdf"), width = 8, height = 6)
hist(dispersion, breaks=200, main="Dispersion across genes", xlab="Dispersion (Var / Mean)",xlim = c(0, 2))
abline(v = 1, col="red")  # Poisson expectation
# dev.off()

# Check how many genes have dispersion > 1
overdispersed_genes <- sum(dispersion > 1)
total_genes <- length(dispersion)
percent_overdispersed <- (overdispersed_genes / total_genes) * 100

# Print summary and recommendation
print(paste0("Overdispersed genes: ", 
             overdispersed_genes, 
             " out of ", 
             total_genes, 
             " (", 
             round(percent_overdispersed, 3), 
             "%)"))

if (percent_overdispersed > 10) {
  print("✅ Overdispersion detected. Consider using a Negative Binomial model.")
} else {
  print("ℹ️ Little evidence of overdispersion. Poisson model may be sufficient.")
}
rm(counts)

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_pooled_offset.csv"))

# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-1, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hep_1_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                                     stats = rankings_Hepatocytes,
                                     scoreType = 'std', 
                                     minSize = 3,
                                     maxSize = 500,
                                     nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                                  stats = rankings_Hepatocytes,
                                  scoreType = 'std', 
                                  minSize = 3,
                                  maxSize = 500,
                                  nproc = 1)
reactome_res_Hepatocytes <- data.frame(reactome_res_Hepatocytes)
saveRDS(reactome_res_Hepatocytes,fs::path(dir.results,"Hep_1_Steatosis_Reactome.rds"))
go_res_Hepatocytes <- fgsea(pathways = go,
                            stats = rankings_Hepatocytes,
                            scoreType = 'std', 
                            minSize = 5,
                            maxSize = 500,
                            nproc = 1)
saveRDS(go_res_Hepatocytes,fs::path(dir.results,"Hep_1_Steatosis_GO.rds"))

Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                                "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                                "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
Hepatocytes_fgsea

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hep-1 Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hep-1 Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hepatocytes, title = "Hep-1 Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

# combined_plot <- b + c + plot_layout(ncol = 3)
# print(combined_plot)
# ggsave(fs::path(dir.results,"Hep_1_top30_pathways_unadjusted.jpeg"),
#        width = 40, height = 10, scale = 1)
combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hep_1_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_1_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-1, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot


#Examine distributions 
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
# Step 1: Pull expression for gene PNPLA3
gene_expr <- FetchData(so_celltype, vars = target_genes)

# Step 2: Pull metadata for steatosis_cat
metadata <- so_celltype@meta.data["steatosis_cat"]

# Step 3: Combine into one dataframe
df <- cbind(gene_expr, metadata)
# colnames(df) <- c("PNPLA3", "steatosis_cat")

# Step 4: Plot using ggplot
ggplot(df, aes(x = PNPLA3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "PNPLA3 Expression by Steatosis Category",
       x = "PNPLA3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = GCKR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "GCKR Expression by Steatosis Category",
       x = "GCKR Expression",
       y = "Cell Count")

ggplot(df, aes(x = TM6SF2, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "TM6SF2 Expression by Steatosis Category",
       x = "TM6SF2 Expression",
       y = "Cell Count")

ggplot(df, aes(x = APOC3, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "APOC3 Expression by Steatosis Category",
       x = "APOC3 Expression",
       y = "Cell Count")

ggplot(df, aes(x = MTOR, fill = steatosis_cat)) + 
  geom_histogram(bins = 30, position = "identity", alpha = 0.6) + 
  theme_minimal() +
  labs(title = "MTOR Expression by Steatosis Category",
       x = "MTOR Expression",
       y = "Cell Count")
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-1
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-1_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_1_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

## c. Hep-2
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 
# so_celltype@assays
# head(GetAssayData(so_celltype, layer = "counts")[, 1:5])  # Raw counts
# head(GetAssayData(so_celltype, layer = "data")[, 1:5])  

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_pooled_offset.csv"))

# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-2, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hep_2_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                                     stats = rankings_Hepatocytes,
                                     scoreType = 'std', 
                                     minSize = 3,
                                     maxSize = 500,
                                     nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                                  stats = rankings_Hepatocytes,
                                  scoreType = 'std', 
                                  minSize = 3,
                                  maxSize = 500,
                                  nproc = 1)
saveRDS(reactome_res_Hepatocytes,fs::path(dir.results,"Hep_2_Steatosis_Reactome.rds"))
go_res_Hepatocytes <- fgsea(pathways = go,
                            stats = rankings_Hepatocytes,
                            scoreType = 'std', 
                            minSize = 5,
                            maxSize = 500,
                            nproc = 1)
saveRDS(go_res_Hepatocytes,fs::path(dir.results,"Hep_2_Steatosis_GO.rds"))
Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                                "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                                "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
# Hepatocytes_fgsea

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hep-2 Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hep-2 Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hepatocytes, title = "Hep-2 Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hep_2_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>%
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_2_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-2, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_2_Steatosis_unadjusted.jpeg"), dot_plot, width = 7, height = 6)

```

## d. Hep-3
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_pooled_offset.csv"))

# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-3, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hep_3_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep3 <- full_results$LogFC
names(rankings_Hep3) <- full_results$Gene
rankings_Hep3 <- sort(rankings_Hep3, decreasing = TRUE)
plot(rankings_Hep3)
min(rankings_Hep3)
max(rankings_Hep3)


set.seed(1234)

kegg_legacy_res_Hep3 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep3,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep3 <- fgsea(pathways = reactome,
                           stats = rankings_Hep3,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
saveRDS(reactome_res_Hep3,fs::path(dir.results,"Hep_3_Steatosis_Reactome.rds"))
go_res_Hep3 <- fgsea(pathways = go,
                     stats = rankings_Hep3,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)
saveRDS(go_res_Hep3,fs::path(dir.results,"Hep_3_Steatosis_GO.rds"))
Hep3_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep3[, padj < 0.05]), sum(kegg_legacy_res_Hep3[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep3[, padj < 0.05]), sum(reactome_res_Hep3[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep3[, padj < 0.05]), sum(go_res_Hep3[, pval < 0.05])))
rownames(Hep3_fgsea) <- c("adj.pval", "p.val")

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hep-3 Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hep-3 Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hepatocytes, title = "Hep-3 Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hep_3_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>%
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_3_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-3, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_3_Steatosis_unadjusted.jpeg"), dot_plot, width = 7, height = 6)

```

## e. Hep-4
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_pooled_offset.csv"))

# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-4, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hep_4_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep4 <- full_results$LogFC
names(rankings_Hep4) <- full_results$Gene
rankings_Hep4 <- sort(rankings_Hep4, decreasing = TRUE)
plot(rankings_Hep4)
min(rankings_Hep4)
max(rankings_Hep4)


set.seed(1234)

kegg_legacy_res_Hep4 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep4,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep4 <- fgsea(pathways = reactome,
                           stats = rankings_Hep4,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
saveRDS(reactome_res_Hep4,fs::path(dir.results,"Hep_4_Steatosis_Reactome.rds"))
go_res_Hep4 <- fgsea(pathways = go,
                     stats = rankings_Hep4,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)
saveRDS(go_res_Hep4,fs::path(dir.results,"Hep_4_Steatosis_GO.rds"))
Hep4_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep4[, padj < 0.05]), sum(kegg_legacy_res_Hep4[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep4[, padj < 0.05]), sum(reactome_res_Hep4[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep4[, padj < 0.05]), sum(go_res_Hep4[, pval < 0.05])))
rownames(Hep4_fgsea) <- c("adj.pval", "p.val")

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_Hep4, title = "Hep-4 Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hep4, title = "Hep-4 Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hep4, title = "Hep-4 Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hep_4_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>%
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_4_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-4, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_4_Steatosis_unadjusted.jpeg"), dot_plot, width = 7, height = 6)

```

## f. Hep-5
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_pooled_offset.csv"))
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_pooled_offset.csv")) %>% 
  dplyr::select(-X) %>% 
  dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=logFC_steatosis_catHigh.Steatosis..2.3.)


# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-5, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-2.5,2.5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_Hep_5_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep5 <- full_results$LogFC
names(rankings_Hep5) <- full_results$Gene
rankings_Hep5 <- sort(rankings_Hep5, decreasing = TRUE)
plot(rankings_Hep5)
min(rankings_Hep5)
max(rankings_Hep5)


set.seed(1234)

kegg_legacy_res_Hep5 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep5,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep5 <- fgsea(pathways = reactome,
                           stats = rankings_Hep5,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
saveRDS(reactome_res_Hep5,fs::path(dir.results,"Hep_5_Steatosis_reactome.rds"))
go_res_Hep5 <- fgsea(pathways = go,
                     stats = rankings_Hep5,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)
saveRDS(go_res_Hep5,fs::path(dir.results,"Hep_5_Steatosis_GO.rds"))

Hep5_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep5[, padj < 0.05]), sum(kegg_legacy_res_Hep5[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep5[, padj < 0.05]), sum(reactome_res_Hep5[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep5[, padj < 0.05]), sum(go_res_Hep5[, pval < 0.05])))
rownames(Hep5_fgsea) <- c("adj.pval", "p.val")

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_Hep5, title = "Hep-5 Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_Hep5, title = "Hep-5 Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_Hep5, title = "Hep-5 Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"Hep_5_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>%
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_5_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-5, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_5_Steatosis_unadjusted.jpeg"), dot_plot, width = 7, height = 6)

```

## g. dHep
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Make sure exposure/independent/x variable or group variable is a factor variable
# so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
so_celltype$steatosis_cat <- factor(so_celltype$steatosis_cat)

#Make sure to set reference level
# so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")
so_celltype$steatosis_cat <- relevel(so_celltype$steatosis_cat,ref="Low Steatosis (0+1)")

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_pooled_offset.csv"))

# full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
#                              ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))
full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "#990000",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "#003366", "lightgray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "dHep, Unadjusted (REML,Log Normal,Pooled Offset)",
    x = "FC",
    y = "-log10(P-Value)",
    color = "FC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(-2.5,2.5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml_offset.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()
png(fs::path(dir.results, "Plot_NEBULA_dHep_Steatosis_unadjusted_2000_pooled_offset.png"),
    width = 2500, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_dHep <- full_results$LogFC
names(rankings_dHep) <- full_results$Gene
rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
plot(rankings_dHep)
min(rankings_dHep)
max(rankings_dHep)


set.seed(1234)

kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                              stats = rankings_dHep,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_dHep <- fgsea(pathways = reactome,
                           stats = rankings_dHep,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
saveRDS(reactome_res_dHep,fs::path(dir.results,"dHep_Steatosis_reactome.rds"))
go_res_dHep <- fgsea(pathways = go,
                     stats = rankings_dHep,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)
saveRDS(go_res_dHep,fs::path(dir.results,"dHep_Steatosis_GO.rds"))
dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                         "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
rownames(dHep_fgsea) <- c("adj.pval", "p.val")

##### KEGG Legacy
a <- plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)

##### REACTOME
b <- plot_fgsea_transpose(reactome_res_dHep, title = "dHep Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)", xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
#        width = 13, height = 10, scale = 1)
##### GO
c <- plot_fgsea_transpose(go_res_dHep, title = "dHep Unadjusted Top 30 GO (REML, Log Normal, Offset)", xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))

# ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
#        width = 13, height = 10, scale = 1)

combined_plot <- b + c + plot_layout(ncol = 1)
print(combined_plot)
ggsave(fs::path(dir.results,"dHep_top30_pathways_unadjusted.jpeg"),
       width = 15, height = 20, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

genes_list <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features =genes_list)
DefaultAssay(so_celltype_hvg) <- "RNA"

nrow(so_celltype_hvg) #28964 genes
Nuclei <- ncol(so_celltype_hvg) #130124 nuclei

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")

# With parallelization

cl <- makeCluster(2)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- .subset(so_celltype_hvg,features=g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>%
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_dHep_Steatosis_unadjusted.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "dHep, Unadjusted (REML, Log Normal, Offset)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred")
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_dHep_Steatosis_unadjusted.jpeg"), dot_plot, width = 7, height = 6)


dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color = "orange") +
  geom_point(aes(color = `p_steatosis_catHigh Steatosis (2+3)`), size = 3) +  # Map color to logFC
  labs(
    title = "Steatosis (High vs. Low) in All Cell Types",
    subtitle = "dHep, Unadjusted (REML, Log Normal, Offset)",
    x = "LogFC", y = "Gene",
    caption = paste0("Nuclei = ", Nuclei)
  ) +
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size = 10),
    axis.title.x = element_text(size = 10),
    axis.title.y = element_text(size = 10),
    text = element_text(size = 10)
  ) +
  scale_color_gradientn(
    colors = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51", "darkred"),
    name = "P-Value"
  )

dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "darkgray") +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color = "#3A3A3A") +
  geom_point(aes(color = `logFC_steatosis_catHigh Steatosis (2+3)`), size = 3) +
  labs(
    title = "Steatosis (High vs. Low) in All Cell Types",
    subtitle = "dHep, Unadjusted (REML, Log Normal, Offset)",
    x = "LogFC", y = "Gene",
    caption = paste0("Nuclei = ", Nuclei)
  ) +
  theme_minimal() +
  theme(
    axis.text.y = element_text(size = 10),
    axis.title.x = element_text(size = 10),
    axis.title.y = element_text(size = 10),
    text = element_text(size = 10)
  ) +
  scale_color_gradient(
    low = "blue",   # for low logFC
    high = "red",   # for high logFC
    name = "LogFC"
  )


dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_dHep_Steatosis_unadjusted_poster.jpeg"), dot_plot, width = 7, height = 6)



```

# 4. Oroboros Analysis 
##I. Visualize
```{r}
#Filter to hepatocytes only
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA"

# which(is.na(so_celltype$carbfccp))
meta <- so_celltype@meta.data
oro_vars <- colnames(meta)[85:114]

hist(meta$carbp_m)
```

##II. Analyze
## a. HEP Cells 
```{r}
#Filter to Hep Cells
so_Hep <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_Hep) <- "RNA" 

nrow(so_Hep) #28964 genes
ncol(so_Hep) #130124 nuclei
meta <- so_Hep@meta.data
oro_vars <- colnames(meta)[85:114]
# "carbp_m"             "carb_adp"            "carb_g"              "carb_cytoc"          "carb_s"              "carb_omy"
#  [7] "carbfccp"            "carbrcr_pm"          "carbrcr_pmgs"        "carbpe_pmgs"         "carb_ep"             "carbrcr_state2"
# [13] "carbrcr_state4_c1"   "carbrcr_state4_c12"  "carb_le"             "lipidpc_m"           "lipid_adp"           "lipid_cytoc"
# [19] "lipid_g"             "lipid_s"             "lipid_omy"           "lipidfccp"           "lipidrcr_pcm"        "lipidrcr_pcmgs"
# [25] "lipidpe_pcmgs"       "lipid_ep"            "lipidrcr_state2"     "lipidrcr_state4_c1"  "lipidrcr_state4_c12" "lipid_le"

## Select Highly Variable Genes (HVGs)
# results_total <- data.frame()
# for (oro in oro_vars) {
oro <- "lipid_g"
cells_to_keep <- !is.na(so_Hep@meta.data[[oro]])
so_celltype <- subset(so_Hep, cells = colnames(so_Hep)[cells_to_keep])

so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ])
genes_list <- hvgs

cl <- makeCluster(10)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    formula <- as.formula(paste("~", oro))
    pred_gene <- model.matrix(formula, data = meta_gene)
    # pred_gene <- model.matrix(~oro, data = meta_gene)
    # library <- meta_gene$pooled_offset
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
colnames(full_results) <- c("logFC_Intercept","logFC","se_Intercept","se","p_Intercept","p_value","gene_id","gene","Gene")
full_results$Variable <- oro

#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results$fdr <- p.adjust(full_results[,6],method="fdr")  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results[,6], 1e-10))
# results_total <- rbind(results_total,full_results)
write.csv(full_results,fs::path(dir.results,paste0("NEBULA_Hepatocytes_",oro,"_unadjusted_2000_hvg.csv")))
# }

oro_vars <- c("lipidpe_pcmgs","lipid_ep","lipidrcr_state2","lipidrcr_state4_c1","lipidrcr_state4_c12","lipid_le")
# results_total <- data.frame()
for (oro in oro_vars) {
  oro <- c("lipid_ep")
  full_results <- read.csv(fs::path(dir.results,paste0("NEBULA_Hepatocytes_",oro,"_unadjusted_2000_hvg.csv")))
  full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$logFC > 0, "#990000",
                               ifelse(full_results$fdr < 0.05 & full_results$logFC < 0, "#003366", "lightgray"))
  
  # Identify significant points (fdr < 0.05)
  significant_df <- full_results[full_results$fdr < 0.05, ]
  
  Genes <- length(unique(full_results$gene))
  Nuclei <- ncol(so_celltype)
  Nonconvergence_Rate <- nebula_nonconverged_percent
  # full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$logFC3 > 0, "lightcoral",
  #                               ifelse(full_results$fdr3 < 0.2 & full_results$logFC3 < 0, "lightblue", "gray"))
  # 
  # # Identify significant points (fdr < 0.05)
  # significant_df3 <- full_results[full_results$fdr3 < 0.2, ]
  
  max <- max(full_results$logFC)
  # max <- 3.1
  min <- min(full_results$logFC)
  # Create the volcano plot using ggplot
  volcano_plot <- ggplot(full_results, aes(x = logFC, y = PValue10, color = color)) +
    geom_point(alpha = 0.7) +  # Plot points with transparency
    scale_color_identity() +  # Use the color column directly
    theme_minimal() +  # Minimal theme
    labs(
      title = paste0(str_to_title(str_replace_all(oro,"_"," "))),
      subtitle = "Hepatocytes, Unadjusted (REML,Log Normal,Pooled Offset)",
      x = "FC",
      y = "-log10(P-Value)",
      color = "FC Direction Direction",
      caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
    ) +
    theme(
      plot.title = element_text(hjust = 0),
      axis.text.x = element_text(angle = 0, hjust = 1)
    )+
    # xlim(min,max)+
    xlim(-3,3)+
    # # Add labels for significant points
    geom_text(data = significant_df, aes(label = gene),
              vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
  # Add labels for significant points with ggrepel
  # geom_text_repel(data = significant_df, aes(label = Gene),
  #                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)
  
  volcano_plot
  png(fs::path(dir.results, paste0("Plot_NEBULA_Hepatocytes_",oro,"_unadjusted_2000_pooled_offset.png")),
      width = 2500, height = 2100, res = 300)
  print(volcano_plot)
  dev.off()
  
  #Gsea
  # full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_pooled_offset.csv")) %>%
  # clean_names() %>% 
  # dplyr::select(-x) %>% 
  full_results <- full_results %>% 
    dplyr::select(-X) %>%
    dplyr::rename(LogFC=logFC)
  
  # Filter out the gmt files for KEGG, Reactome and GOBP
  list.files(bg_path)
  gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
  gmt_files
  kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
  reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
  go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)
  
  # rank genes by t-stats in DiD
  rankings_dHep <- full_results$LogFC
  names(rankings_dHep) <- full_results$Gene
  rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
  plot(rankings_dHep)
  min(rankings_dHep)
  max(rankings_dHep)
  
  set.seed(1234)
  
  kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                                stats = rankings_dHep,
                                scoreType = 'std', 
                                minSize = 3,
                                maxSize = 500,
                                nproc = 1)
  
  reactome_res_dHep <- fgsea(pathways = reactome,
                             stats = rankings_dHep,
                             scoreType = 'std', 
                             minSize = 3,
                             maxSize = 500,
                             nproc = 1)
  go_res_dHep <- fgsea(pathways = go,
                       stats = rankings_dHep,
                       scoreType = 'std', 
                       minSize = 5,
                       maxSize = 500,
                       nproc = 1)
  
  dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                           "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                           "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
  rownames(dHep_fgsea) <- c("adj.pval", "p.val")
  
  ##### KEGG Legacy
  # a <- plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG (REML, Log Normal, Offset)", xnudge = 0.03)+ theme(legend.position.inside = c(0.2, 0.5))
  
  # ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted.jpeg"),
  #        width = 13, height = 10, scale = 1)
  
  ##### REACTOME
  b <- plot_fgsea_transpose(reactome_res_dHep, title = paste0(str_to_title(str_replace_all(oro,"_"," "))," in Hepatocytes Unadjusted Top 30 REACTOME (REML, Log Normal, Offset)"), xlimit = 5) + theme(legend.position.inside = c(0.15, 0.5))
  
  # ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted.jpeg"),
  #        width = 13, height = 10, scale = 1)
  ##### GO
  c <- plot_fgsea_transpose(go_res_dHep, title = paste0(str_to_title(str_replace_all(oro,"_"," "))," in Hepatocytes Unadjusted Top 30 GO (REML, Log Normal, Offset)"), xlimit = 8)+ theme(legend.position.inside = c(0.1, 0.5))
  
  # ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml_offset.jpeg"),
  #        width = 13, height = 10, scale = 1)
  
  combined_plot <- b + c + plot_layout(ncol = 1)
  print(combined_plot)
  ggsave(fs::path(dir.results,paste0(oro,"_Hepatocytes_top30_pathways_unadjusted.jpeg")),
         width = 15, height = 20, scale = 1)
  
  # results_total <- rbind(results_total,results)
}
# results_total <- results_total %>% 
#   dplyr::select(-X)

# write.csv(results_total,fs::path(dir.results,"NEBULA_Hepatocytes_Oroboros_unadjusted_2000_hvg.csv"))

# # Define significance stars
# results_total <- results_total %>%
#   mutate(signif = case_when(
#     fdr < 0.01 ~ "**",
#     fdr < 0.05 ~ "*",
#     TRUE ~ ""
#   ))
# 
# # Select only the needed columns and rename LogFC for clarity
# heatmap_data <- results_total %>%
#   dplyr::select(Gene, Variable, logFC, signif)
# 
# # custom_order <- oro_vars
# # heatmap_data$Variable <- factor(heatmap_data$Variable, levels = custom_order)
# # custom_labels <- str_to_title(str_replace_all(oro,"_"," "))
# 
# # Carb_cytoc
# # Carb_s
# # Carb_omy
# # Carb_fccp
# #  
# # Lipid_cytoc
# # Lipid_s
# # Lipid_omy
# # Lipid_fccp
# 
# 
# heat_map_p <- ggplot(heatmap_data, aes(x = Variable, y = Gene, fill = logFC)) +
#   geom_tile(color = "grey90") +
#   geom_text(aes(label = signif), size = 3, color = "black") +
#   scale_fill_gradient2(
#     low = "#264653", mid = "white", high = "darkred",
#     midpoint = 0,
#     name = "LogFC"
#   ) +
#   theme_minimal() +
#   labs(
#     # title = "TCA Cycle Genes vs. PET Variables (T2D)",
#     # subtitle = "Proximal Tubule Cells",
#     x = "Exposure",
#     y = "Gene") +
#   # scale_x_discrete(labels = setNames(custom_labels, custom_order))+
#   theme(
#     text = element_text(face="bold"),
#     axis.text.y = element_text(size = 6),
#     axis.text.x = element_text(angle = 45, hjust = 1),
#     axis.ticks.y = element_blank(),
#     axis.title.x = element_blank(),
#     panel.grid = element_blank()
#   )
# 
# # custom_colors <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51", "darkred")
# 
# png(fs::path(dir.results, "Heatmap_TCA_cycle_NEBULA_PT_PET_unadjusted_pooled_offset_T2D.png"), 
#     width = 1500, height = 2000, res = 300)
# print(heat_map_p)
# dev.off()

```

### i. NEBULA
```{r}
#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
length(which(is.na(so_celltype$carb_adp)))


counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)[20]
# # count_gene <- counts_hvg[g, , drop = FALSE]
# meta_gene_hvg <- .subset(so_celltype_hvg,features=g)@meta.data
# pred_gene_hvg <- model.matrix(~steatosis_cat, data = meta_gene_hvg)
# data_g_hvg = list(count = counts_hvg, id = id, pred = pred_gene_hvg)
# offset_hvg = Matrix::colSums(data_g_hvg$count) 

#Filter out NAs for oroboros variables
# so_celltype$carbfccp_fact <- factor(so_celltype$carbfccp)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~carbfccp_fact, data = meta_gene)
    # library <- meta_gene$pooled_offset
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

#Oroboros Variables
oro <- colnames(so_liver_sn@meta.data[,c(85:114)])
# Carb_cytoc
# Carb_s
# Carb_omy
# Carb_fccp
#  
# Lipid_cytoc
# Lipid_s
# Lipid_omy
# Lipid_fccp

# With parallelization
oro <- oro[1]
results_total <- data.frame()
for (exposure in oro) {
  cl <- makeCluster(20)
  registerDoParallel(cl)
  
  start_time <- Sys.time()
  nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
    tryCatch({
      count_gene <- counts_hvg[g, , drop = FALSE]
      # meta_gene <- subset(so_celltype,features=g)@meta.data
      # Get expression data for that gene
      gene_expr <- FetchData(so_celltype, vars = g)
      # Define which cells express the gene (e.g., expression > 0)
      cells_use <- rownames(gene_expr[gene_expr[[g]] > 0, ])
      so_subset <- subset(so_celltype, cells = cells_use)
      meta_gene <- so_subset@meta.data
      pred.formula <- as.formula(paste0("~",exposure))
      pred_gene <- model.matrix(pred.formula, data = meta_gene)
      # library <- meta_gene$library_size
      library <- meta_gene$pooled_offset
      
      if (ncol(data_g_gene$count) != nrow(data_g_gene$pred)) {
        data_g_gene$count <- t(data_g_gene$count)
      }
      
      data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
      
      if (is.null(data_g_gene)) {
        data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
      }
      
      #With offset
      result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset=data_g_gene$offset)
      
      list(gene = g, result = result)  # return both gene name and result
      
    }, error = function(e) {
      list(gene = g, summary = NA, overdispersion = NA, convergence = NA, algorithm = NA, covariance = NA, random_effect = NA)
    })
  }
  
  stopCluster(cl)
  end_time <- Sys.time()
  print(end_time - start_time)
  
  # set the names of results based on gene names
  nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
  names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
  nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results
  
  hep_nebula_converged <- map_dfr(
    names(nebula_results_list),
    function(gene_name) {
      converged <- nebula_results_list[[gene_name]]$convergence
      df <- data.frame(Gene = gene_name,
                       Convergence_Code = converged)
      return(df)
    }
  )
  
  nebula_summaries <- map_dfr(
    names(nebula_results_list),
    function(gene_name) {
      df <- nebula_results_list[[gene_name]]$summary
      df <- df %>% mutate(Gene = gene_name)
      return(df)
    }
  )
  nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
  ) 
  
  full_results <- as.data.frame(nebula_summaries)
  colnames(full_results) <- c("logFC_Intercept","logFC","se_Intercept","se","p_Intercept","p_value","gene_id","gene","Gene")
  full_results$Variable <- exposure
  
  #Calculate number of genes filtered out for low expression 
  low_exp <- length(genes_list)-length(full_results$gene)
  #Filter out non-converging genes
  full_results <- full_results %>% 
    filter(!gene %in%  nonconverge_genes)
  #Calculate nonconvergence rate
  nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
  # nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
  # print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
  full_results$fdr <- p.adjust(full_results[,6],method="fdr")  
  # mutate(fdr3=p.adjust(PValue3,method="fdr"))
  full_results$PValue10 <- -log10(pmax(full_results[,6], 1e-10))
  results_total <- rbind(results_total,full_results)
}

write.csv(results_total,fs::path(dir.results,"NEBULA_Hepatocytes_Oroboros_unadjusted_2000_hvg.csv"))

# Define significance stars
results_total <- results_total %>%
  mutate(signif = case_when(
    fdr < 0.01 ~ "**",
    fdr < 0.05 ~ "*",
    TRUE ~ ""
  ))

# Select only the needed columns and rename LogFC for clarity
heatmap_data <- results_total %>%
  dplyr::select(Gene, Variable, logFC, signif)

custom_order <- oro
heatmap_data$Variable <- factor(heatmap_data$Variable, levels = custom_order)
custom_labels <- str_to_title(str_replace_all(oro,"_"," "))

# Carb_cytoc
# Carb_s
# Carb_omy
# Carb_fccp
#  
# Lipid_cytoc
# Lipid_s
# Lipid_omy
# Lipid_fccp


heat_map_p <- ggplot(heatmap_data, aes(x = Variable, y = Gene, fill = logFC)) +
  geom_tile(color = "grey90") +
  geom_text(aes(label = signif), size = 3, color = "black") +
  scale_fill_gradient2(
    low = "#264653", mid = "white", high = "darkred",
    midpoint = 0,
    name = "LogFC"
  ) +
  theme_minimal() +
  labs(title = "TCA Cycle Genes vs. PET Variables (T2D)",
       subtitle = "Proximal Tubule Cells",
       x = "Exposure",
       y = "Gene") +
  scale_x_discrete(labels = setNames(custom_labels, custom_order))+
  theme(
    text = element_text(face="bold"),
    axis.text.y = element_text(size = 6),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.ticks.y = element_blank(),
    axis.title.x = element_blank(),
    panel.grid = element_blank()
  )

# custom_colors <- c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51", "darkred")

png(fs::path(dir.results, "Heatmap_TCA_cycle_NEBULA_PT_PET_unadjusted_pooled_offset_T2D.png"), 
    width = 1500, height = 2000, res = 300)
print(heat_map_p)
dev.off()

```

### ii. glmmTMB
```{r}
#Access the raw count matrix
raw_counts <- GetAssayData(so_celltype_hvg, layer = "counts")

# Step 2: Round manually for sparse matrix (dgCMatrix)
raw_counts_rounded <- raw_counts
raw_counts_rounded@x <- round(raw_counts@x)

# Step 3: Extract the full gene expression matrix (rounded counts)
expr_matrix <- as.matrix(raw_counts_rounded)

#View how rounded counts data looks
#1. Randomly select a few genes for visualization (e.g., top 50 or random)
selected_genes <- sample(rownames(raw_counts_rounded), 25)  # Random 50 genes

#2. Get the expression data for those genes (rounded counts)
selected_data <- as.matrix(raw_counts_rounded[selected_genes, ])

# Convert matrix to data.frame
df_selected_data <- as.data.frame(t(selected_data))  # Transpose so that genes are columns

# 3. **Boxplot** for raw counts of selected genes across cells (to see distribution)
df_selected_data_long <- reshape2::melt(df_selected_data)  # Convert to long format

ggplot(df_selected_data_long, aes(x = variable, y = value, fill = variable)) +
  geom_boxplot() +
  labs(title = "Boxplot of Selected Genes' Counts", x = "Genes", y = "Count") +
  theme_minimal()+
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
    # legend.position = "none"
  )

# 4. **Heatmap** of raw counts (scaled by genes) for a few cells (e.g., top 50 cells)
top_cells <- sample(colnames(raw_counts_rounded), 50)  # Random 50 cells
heatmap_data <- as.matrix(raw_counts_rounded[selected_genes, top_cells])

# Plot heatmap
pheatmap(heatmap_data, 
         cluster_rows = FALSE, 
         cluster_cols = FALSE, 
         show_rownames = TRUE, 
         show_colnames = FALSE,
         main = "Heatmap of Gene Expression (Selected Genes & Cells)") 

#Check rownames and colnames: Rownames = genes, Colnames = cells
#view(expr_matrix)
#Transpose gene expression dataset to merge with metadata
expr_matrix <- t(expr_matrix)
expr_matrix <- data.frame(expr_matrix) #Make a dataframe again after transposing

#Set gene list
gene_list_total <- hvgs

#Create cell variable to merge metadata in
expr_matrix$cellname <- rownames(expr_matrix) 
rownames(expr_matrix) <- NULL

# Extract the metadata
metadata <- .subset(so_celltype_hvg,features=g)@meta.data

#Create cell variable to merge metadata into gene expression data
metadata$cellname <- rownames(metadata)
rownames(metadata) <- NULL

# Combine the gene expression data and metadata
data <- tidylog::full_join(metadata,expr_matrix,by="cellname")
rm(metadata,expr_matrix)

#Make sure exposure/independent/x variable or group variable is a factor variable
data$steatosis_cat <- factor(data$steatosis_cat)
#Make sure to set reference level
data$steatosis_cat <- relevel(data$steatosis_cat,ref="Low Steatosis (0+1)")

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hepatocytes <- full_results$LogFC
names(rankings_Hepatocytes) <- full_results$Gene
rankings_Hepatocytes <- sort(rankings_Hepatocytes, decreasing = TRUE)
plot(rankings_Hepatocytes)
min(rankings_Hepatocytes)
max(rankings_Hepatocytes)

set.seed(1234)

kegg_legacy_res_Hepatocytes <- fgsea(pathways = kegg_legacy,
                                     stats = rankings_Hepatocytes,
                                     scoreType = 'std', 
                                     minSize = 3,
                                     maxSize = 500,
                                     nproc = 1)

reactome_res_Hepatocytes <- fgsea(pathways = reactome,
                                  stats = rankings_Hepatocytes,
                                  scoreType = 'std', 
                                  minSize = 3,
                                  maxSize = 500,
                                  nproc = 1)
go_res_Hepatocytes <- fgsea(pathways = go,
                            stats = rankings_Hepatocytes,
                            scoreType = 'std', 
                            minSize = 5,
                            maxSize = 500,
                            nproc = 1)

Hepatocytes_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hepatocytes[, padj < 0.05]), sum(kegg_legacy_res_Hepatocytes[, pval < 0.05])),
                                "REACTOME"=c(sum(reactome_res_Hepatocytes[, padj < 0.05]), sum(reactome_res_Hepatocytes[, pval < 0.05])),
                                "GO"=c(sum(go_res_Hepatocytes[, padj < 0.05]), sum(go_res_Hepatocytes[, pval < 0.05])))
rownames(Hepatocytes_fgsea) <- c("adj.pval", "p.val")
Hepatocytes_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hepatocytes_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hepatocytes_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
##### GO
plot_fgsea_transpose(go_res_Hepatocytes, title = "Hepatocytes Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hepatocytes_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```
### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)
# Use lapply to pull all summaries and add gene names
# nebula_summaries <- do.call(rbind, lapply(names(nebula_results_list), function(gene) {
#   if (!is.null(nebula_results_list[[gene]]$summary)) {
#     df <- nebula_results_list[[gene]]$summary
#     df$gene <- gene  # Add gene name as a new column
#     return(df)
#   } else {
#     NULL
#   }
# }))
nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hepatocytes_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hepatocytes, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hepatocytes_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## b. Hep-1
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 
so_celltype@assays
head(GetAssayData(so_celltype, layer = "counts")[, 1:5])  # Raw counts
head(GetAssayData(so_celltype, layer = "data")[, 1:5])  

nrow(so_celltype) #11300 genes
ncol(so_celltype) #24808 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# # Subset Seurat object to only HVGs
# so_celltype_hvg <- subset(so_celltype, features = hvgs)
# DefaultAssay(so_celltype_hvg) <- "RNA" 

```

### i. NEBULA
```{r}
#Round counts layer for nb
# counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
counts_hvg <- round(GetAssayData(so_celltype, layer = "counts")[hvgs, ]) # load counts and round

# List of genes
genes_list <- rownames(counts_hvg)

# With parallelization
cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype,features=g)@meta.data
    pred_gene <- model.matrix(~carbp_m, data = meta_gene)
    library <- meta_gene$pooled_offset
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene,offset=library)
    
    if (is.null(data_g_gene)) {
      data_g_gene <- list(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, offset = library)
    }
    
    #With offset
    result <- nebula(count = data_g_gene$count, id = data_g_gene$id, pred = data_g_gene$pred, ncore = 1, reml=T,model="NBLMM",output_re = T,covariance=T,offset = data_g_gene$offset)
    
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-1, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-1
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-1_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-1 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-1 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-1 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-1 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-1 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-1 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-1 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_1_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_1_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep1 <- full_results$LogFC
names(rankings_hep1) <- full_results$Gene
rankings_hep1 <- sort(rankings_hep1, decreasing = TRUE)
plot(rankings_hep1)
min(rankings_hep1)
max(rankings_hep1)

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep1, title = "Hep-1 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep1_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep1, title = "Hep-1 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep1_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep1, title = "Hep-1 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep1_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-1")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_1_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-1, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_1_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```


## c. Hep-2
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-2, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)

print(volcano_plot)

dev.off()
```

### ii. Pathway Enrichment
```{r echo = F}
#Hep-2
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::select(c("gene","log_fc_steatosis_cat_high_steatosis_2_3","fdr")) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)
sig_pos <- full_results %>% 
  filter(LogFC>0 & fdr<0.05)
sig_neg <- full_results %>% 
  filter(LogFC<0 & fdr<0.05)

#Run postive enricher analysis 
sig_pos_en <- enrichr(as.character(sig_pos$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[1]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_pos_en_df <- as.data.frame(sig_pos_en[[2]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_pos_en[[2]]), title = "Model 1 Positive Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Positive Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_pos_en[[3]]), title = "Model 1 Positive Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[3]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_pos_en_df <- as.data.frame(sig_pos_en[[5]])
# # Extract the number of overlapping genes
# sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Positive Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_pos_en_df <- as.data.frame(sig_pos_en[[4]])
# Extract the number of overlapping genes
sig_pos_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_pos_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_pos_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Positive Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_pos_en[[4]]), title = "Model 1 Positive Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# # Arrange 6 plots in 3 rows x 2 columns
# combined_plot <- (p1 | p2) / 
#   (p3 | p4) 
# 
# # Display or save the plot
# figure <- combined_plot + 
#   plot_annotation(
#     title = "Positive Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
#     theme = theme(
#       plot.title = element_text(
#         hjust = 0.5,         # Center title
#         size = 18,           # Bigger font
#         face = "bold"        # Bold text
#       )
#     )
#   )

# ggsave(fs::path(dir.results, "Positive_Hep-2_gsea_figure.jpeg"), figure, width = 18, height = 10)

#Run negative enricher analysis 
sig_neg_en <- enrichr(as.character(sig_neg$Gene), dbs)
#Plot results
# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[1]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - GO")
p1 <- plotEnrich(filtered_results, title = "GO")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_go.jpeg"),width=10,height=7)

# Convert enrichment results to a dataframe
sig_neg_en_df <- as.data.frame(sig_neg_en[[2]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
p2 <- plotEnrich(filtered_results, title = "KEGG")
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Kegg")
# plotEnrich(as.data.frame(sig_neg_en[[2]]), title = "Model 1 Negative Hep-2 - Kegg")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_kegg.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '22")
# p3 <- plotEnrich(filtered_results, title = "Negative Hep-2 - Reactome '22")
# # plotEnrich(as.data.frame(sig_neg_en[[3]]), title = "Model 1 Negative Hep-2 - Reactome '22")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome22.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[3]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - Reactome '24")
p3 <- plotEnrich(filtered_results, title = "Reactome '24")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

# sig_neg_en_df <- as.data.frame(sig_neg_en[[5]])
# # Extract the number of overlapping genes
# sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# # Filter pathways with at least 3 overlapping genes
# filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# # Plot the filtered results
# p5 <- plotEnrich(filtered_results, title = "Negative Hep-2 - MsigDB Computational")
# # plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Computational")
# # plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# # ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)

sig_neg_en_df <- as.data.frame(sig_neg_en[[4]])
# Extract the number of overlapping genes
sig_neg_en_df$Num_Genes <- as.numeric(sub("/.*", "", sig_neg_en_df$Overlap))
# Filter pathways with at least 3 overlapping genes
filtered_results <- subset(sig_neg_en_df, Num_Genes >= 3)
# Plot the filtered results
# plotEnrich(filtered_results, title = "Model 1 Negative Hep-2 - MsigDB Hallmark '20")
# plotEnrich(as.data.frame(sig_neg_en[[4]]), title = "Model 1 Negative Hep-2 - Reactome '24")
# ggsave(fs::path(dir.results,"All_Cell_Types_pos_gsea_reactome24.jpeg"),width=10,height=7)
p4 <- plotEnrich(filtered_results, title = "MsigDB '20")

# Arrange 6 plots in 3 rows x 2 columns
combined_plot <- (p1 | p2) / 
  (p3 | p4) 

# Display or save the plot
figure <- combined_plot + 
  plot_annotation(
    title = "Negative Gene Set Enrichment Analysis for Hep-2 vs. Steatosis (High/Low)",
    theme = theme(
      plot.title = element_text(
        hjust = 0.5,         # Center title
        size = 18,           # Bigger font
        face = "bold"        # Bold text
      )
    )
  )

ggsave(fs::path(dir.results, "NEBULA_Negative_Hep_2_gsea_figure.jpeg"), figure, width = 18, height = 10)

```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_2_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_hep2 <- full_results$LogFC
names(rankings_hep2) <- full_results$Gene
rankings_hep2 <- sort(rankings_hep2, decreasing = TRUE)
plot(rankings_hep2)
min(rankings_hep2)
max(rankings_hep2)


set.seed(1234)

kegg_legacy_res_hep2 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_hep2,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_hep2 <- fgsea(pathways = reactome,
                           stats = rankings_hep2,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_hep2 <- fgsea(pathways = go,
                     stats = rankings_hep2,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

hep2_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_hep2[, padj < 0.05]), sum(kegg_legacy_res_hep2[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_hep2[, padj < 0.05]), sum(reactome_res_hep2[, pval < 0.05])),
                         "GO"=c(sum(go_res_hep2[, padj < 0.05]), sum(go_res_hep2[, pval < 0.05])))
rownames(hep2_fgsea) <- c("adj.pval", "p.val")
hep2_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_hep2, title = "Hep-2 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"hep2_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_hep2, title = "Hep-2 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"hep2_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_hep2, title = "Hep-2 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"hep2_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-2")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_2_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-2, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_2_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## d. Hep-3
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-3, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_3_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep3 <- full_results$LogFC
names(rankings_Hep3) <- full_results$Gene
rankings_Hep3 <- sort(rankings_Hep3, decreasing = TRUE)
plot(rankings_Hep3)
min(rankings_Hep3)
max(rankings_Hep3)


set.seed(1234)

kegg_legacy_res_Hep3 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep3,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep3 <- fgsea(pathways = reactome,
                           stats = rankings_Hep3,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep3 <- fgsea(pathways = go,
                     stats = rankings_Hep3,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep3_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep3[, padj < 0.05]), sum(kegg_legacy_res_Hep3[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep3[, padj < 0.05]), sum(reactome_res_Hep3[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep3[, padj < 0.05]), sum(go_res_Hep3[, pval < 0.05])))
rownames(Hep3_fgsea) <- c("adj.pval", "p.val")
Hep3_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep3, title = "Hep-3 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep3_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep3, title = "Hep-3 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep3_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep3, title = "Hep-3 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep3_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-3")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #130124 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_3_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-3, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_3_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## e. Hep-4
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv"))

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-4, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_4_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep4 <- full_results$LogFC
names(rankings_Hep4) <- full_results$Gene
rankings_Hep4 <- sort(rankings_Hep4, decreasing = TRUE)
plot(rankings_Hep4)
min(rankings_Hep4)
max(rankings_Hep4)


set.seed(1234)

kegg_legacy_res_Hep4 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep4,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep4 <- fgsea(pathways = reactome,
                           stats = rankings_Hep4,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep4 <- fgsea(pathways = go,
                     stats = rankings_Hep4,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep4_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep4[, padj < 0.05]), sum(kegg_legacy_res_Hep4[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep4[, padj < 0.05]), sum(reactome_res_Hep4[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep4[, padj < 0.05]), sum(go_res_Hep4[, pval < 0.05])))
rownames(Hep4_fgsea) <- c("adj.pval", "p.val")
Hep4_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep4, title = "Hep-4 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep4_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep4, title = "Hep-4 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep4_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep4, title = "Hep-4 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep4_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-4")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_4_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-4, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_4_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## f. Hep-5
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "Hep-5, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_Hep_5_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_Hep5 <- full_results$LogFC
names(rankings_Hep5) <- full_results$Gene
rankings_Hep5 <- sort(rankings_Hep5, decreasing = TRUE)
plot(rankings_Hep5)
min(rankings_Hep5)
max(rankings_Hep5)


set.seed(1234)

kegg_legacy_res_Hep5 <- fgsea(pathways = kegg_legacy,
                              stats = rankings_Hep5,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_Hep5 <- fgsea(pathways = reactome,
                           stats = rankings_Hep5,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_Hep5 <- fgsea(pathways = go,
                     stats = rankings_Hep5,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

Hep5_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_Hep5[, padj < 0.05]), sum(kegg_legacy_res_Hep5[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_Hep5[, padj < 0.05]), sum(reactome_res_Hep5[, pval < 0.05])),
                         "GO"=c(sum(go_res_Hep5[, padj < 0.05]), sum(go_res_Hep5[, pval < 0.05])))
rownames(Hep5_fgsea) <- c("adj.pval", "p.val")
Hep5_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_Hep5, title = "Hep-5 Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"Hep5_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_Hep5, title = "Hep-5 Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"Hep5_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_Hep5, title = "Hep-5 Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"Hep5_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="Hep-5")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_Hep_5_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "Hep-5, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_Hep_5_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```

## g. dHep
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #11300 genes
ncol(so_celltype) #22983 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)

# Subset Seurat object to only HVGs
so_celltype_hvg <- subset(so_celltype, features = hvgs)
DefaultAssay(so_celltype_hvg) <- "RNA" 
```

### i. NEBULA
```{r}
counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(20)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)

end_time <- Sys.time()
print(end_time - start_time)

# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 
# end_time <- Sys.time()
# print(end_time - start_time)

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv"))
# full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
#   dplyr::select(-X) %>% 
#   clean_names() %>% 
#   dplyr::rename(`logFC_steatosis_catHigh Steatosis (2+3)`=log_fc_steatosis_cat_high_steatosis_2_3,
#                 PValue10=p_value10)

full_results$color <- ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` > 0, "lightcoral",
                             ifelse(full_results$fdr < 0.05 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)` < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- full_results[full_results$fdr < 0.05, ]

Genes <- length(unique(full_results$gene))
Nuclei <- ncol(so_celltype_hvg)
Nonconvergence_Rate <- nebula_nonconverged_percent
# full_results$color3 <- ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 > 0, "lightcoral",
#                               ifelse(full_results$fdr3 < 0.2 & full_results$`logFC_steatosis_catHigh Steatosis (2+3)`3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- full_results[full_results$fdr3 < 0.2, ]

max <- max(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# max <- 3.1
min <- min(full_results$`logFC_steatosis_catHigh Steatosis (2+3)`)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  theme_minimal() +  # Minimal theme
  labs(
    title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    subtitle = "dHep, Unadjusted (REML)",
    x = "logFC",
    y = "-log10(P-Value)",
    color = "LogFC Direction Direction",
    caption = paste0("FDR < 0.05, Genes = ",Genes,", Nuclei = ",Nuclei,", Non-Convergence Rate: ",Nonconvergence_Rate,", Genes Filtered out for Low Expression: ",low_exp)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  # xlim(min,max)+
  xlim(-5,5)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_NEBULA_dHep_Steatosis_unadjusted_2000_hvg.png"), 
    width = 3000, height = 2100, res = 300)
print(volcano_plot)
dev.off()
```

### iii. GSEA
```{r echo = F}
full_results <- read.csv(fs::path(dir.results,"NEBULA_dHep_Steatosis_unadjusted_2000_hvg_reml.csv")) %>%
  clean_names() %>% 
  dplyr::select(-x) %>% 
  dplyr::rename(Gene=gene,LogFC=log_fc_steatosis_cat_high_steatosis_2_3)

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(full_results$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(full_results$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(full_results$Gene), savefile = FALSE)

# rank genes by t-stats in DiD
rankings_dHep <- full_results$LogFC
names(rankings_dHep) <- full_results$Gene
rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
plot(rankings_dHep)
min(rankings_dHep)
max(rankings_dHep)


set.seed(1234)

kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                              stats = rankings_dHep,
                              scoreType = 'std', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_dHep <- fgsea(pathways = reactome,
                           stats = rankings_dHep,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_dHep <- fgsea(pathways = go,
                     stats = rankings_dHep,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                         "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
rownames(dHep_fgsea) <- c("adj.pval", "p.val")
dHep_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"dHep_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_dHep, title = "dHep Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"dHep_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_dHep, title = "dHep Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"dHep_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)
```

### iv. Targeted Genes
```{r}
#Filter to PT Cells
so_celltype <- subset(so_liver_sn,celltype=="dHep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
Nuclei <- ncol(so_celltype) #19319 nuclei

# # Subset Seurat object to targeted genes
target_genes <- c("PNPLA3","GCKR","TM6SF2","APOC3","MTOR")
so_celltype_hvg <- subset(so_celltype, features = target_genes)
DefaultAssay(so_celltype_hvg) <- "RNA"

counts_hvg <- round(GetAssayData(so_celltype_hvg, layer = "counts")) # load counts and round
# meta_hvg_tal <- attempt_so_hvg_tal@meta.data
# pred = model.matrix(~treatment*visit, data = meta_hvg_tal)
# data_g_hvg_tal = group_cell(count=counts_hvg_tal, id=meta_hvg_tal$subject, pred=pred)

#Make sure exposure/independent/x variable or group variable is a factor variable
so_celltype_hvg$steatosis_cat <- factor(so_celltype_hvg$steatosis_cat)
#Make sure to set reference level
so_celltype_hvg$steatosis_cat <- relevel(so_celltype_hvg$steatosis_cat,ref="Low Steatosis (0+1)")


# With parallelization
# List of genes
genes_list <- rownames(counts_hvg)

cl <- makeCluster(1)
registerDoParallel(cl)

start_time <- Sys.time()

nebula_results_list <- foreach(g = genes_list, .packages = c("nebula", "Matrix")) %dopar% {
  tryCatch({
    count_gene <- counts_hvg[g, , drop = FALSE]
    meta_gene <- subset(so_celltype_hvg, features = g)@meta.data
    pred_gene <- model.matrix(~steatosis_cat, data = meta_gene)
    data_g_gene <- group_cell(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene)
    
    result <- nebula(count = count_gene, id = meta_gene$Kit_Lot, pred = pred_gene, ncore = 1,output_re = T)
    list(gene = g, result = result)  # return both gene name and result
    
  }, error = function(e) {
    NULL
  })
}

stopCluster(cl)
end_time <- Sys.time()
print(end_time - start_time)


# set the names of results based on gene names
nebula_results_list <- Filter(Negate(is.null), nebula_results_list)  # remove NULLs first
names(nebula_results_list) <- sapply(nebula_results_list, function(x) x$gene)  # set names
nebula_results_list <- lapply(nebula_results_list, function(x) x$result)  # clean list back to just results

hep_nebula_converged <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    converged <- nebula_results_list[[gene_name]]$convergence
    df <- data.frame(Gene = gene_name,
                     Convergence_Code = converged)
    return(df)
  }
)

nebula_summaries <- map_dfr(
  names(nebula_results_list),
  function(gene_name) {
    df <- nebula_results_list[[gene_name]]$summary
    df <- df %>% mutate(Gene = gene_name)
    return(df)
  }
)
nonconverge_genes <- unique(hep_nebula_converged$Gene[which(hep_nebula_converged$Convergence_Code==-40)]
) 

full_results <- as.data.frame(nebula_summaries)
#Calculate number of genes filtered out for low expression 
low_exp <- length(genes_list)-length(full_results$gene)
#Filter out non-converging genes
full_results <- full_results %>% 
  filter(!gene %in%  nonconverge_genes)
#Calculate nonconvergence rate
nebula_nonconverged_percent <- paste0(round((1-(length(genes_list)-length(nonconverge_genes))/length(genes_list))*100,3),"%")
# nebula_nonconverged_percent <- (length(rownames(counts_hvg))-length(unique(full_results$gene)))/length(rownames(counts_hvg))
# print(paste0(nebula_nonconverged_percent*100, "% failed to converge"))
full_results <- full_results %>%
  mutate(fdr=p.adjust(`p_steatosis_catHigh Steatosis (2+3)`,method="fdr"))  
# mutate(fdr3=p.adjust(PValue3,method="fdr"))
full_results$PValue10 <- -log10(pmax(full_results$`p_steatosis_catHigh Steatosis (2+3)`, 1e-10))  # Avoid log(0)

write.csv(full_results,fs::path(dir.results,"Targeted_Genes_NEBULA_dHep_Steatosis_unadjusted_reml.csv"))

# Calculate Confidence Intervals (95% CI)
full_results <- full_results %>%
  mutate(
    CI_Lower = `logFC_steatosis_catHigh Steatosis (2+3)` - 1.96 * `se_steatosis_catHigh Steatosis (2+3)`,
    CI_Upper = `logFC_steatosis_catHigh Steatosis (2+3)` + 1.96 * `se_steatosis_catHigh Steatosis (2+3)`
  )
full_results <- full_results %>% 
  mutate(sig=ifelse(`p_steatosis_catHigh Steatosis (2+3)`<0.05,"*",""))

# Create the coefficient plot
dot_plot <- ggplot(full_results, aes(x = `logFC_steatosis_catHigh Steatosis (2+3)`, y = Gene)) +
  geom_errorbarh(aes(xmin = CI_Lower, xmax = CI_Upper), height = 0.2, color="orange") +  # Add horizontal error bars for CIs
  geom_point(size = 3,color="purple") +  # Add points for the Beta (Log2FC)
  # geom_point(size = 3,aes(color=sig)) +  # Add points for the Beta (Log2FC)
  labs(title = "Steatosis (High vs. Low) in All Cell Types",
       subtitle = "dHep, Unadjusted (REML)",
       x = "LogFC", y = "Gene",
       caption=paste0("Nuclei = ",Nuclei) )+
  geom_vline(xintercept = 0, linetype = "dotted", color = "black") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 10),
        axis.title.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        text =element_text(size=10)) 
dot_plot
ggsave(fs::path(dir.results, "NEBULA_Targeted_dHep_Steatosis_unadjusted_reml.jpeg"), dot_plot, width = 7, height = 6)

```


#5. Pseudotime Analysis 
##a. Prepare for Analysis 
```{r}
## Make SO into SCE object for pseudotime analysis
sce_HEP <- as.SingleCellExperiment(subset(so_liver_sn,celltype2=="Hep"), assay = "RNA")

#Barcharts by Steatosis Groups
n_low <- nrow(so_liver_sn@meta.data %>% filter(celltype2=="Hep" & steatosis_cat=="Low Steatosis (0+1)"))
n_high <- nrow(so_liver_sn@meta.data %>% filter(celltype2=="Hep" & steatosis_cat=="High Steatosis (2+3)"))

# By PT subtypes (by celltype)
cellcount<- so_liver_sn@meta.data %>% 
  filter(celltype2=="Hep") %>%
  dplyr::group_by(steatosis_cat, celltype) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(n = case_when(steatosis_cat == "Low Steatosis (0+1)" ~ n_low, steatosis_cat == "High Steatosis (2+3)" ~ n_high),
                ratio = count / n) %>% ungroup() %>%
  group_by(celltype) %>%
  dplyr::mutate(proportion = ratio/sum(ratio)) %>%
  dplyr::select(proportion, steatosis_cat, celltype) %>%
  ggplot(aes(x= celltype, y= proportion, fill = steatosis_cat)) + 
  geom_bar(stat = "identity", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion",
       fill = "Steatosis Grade") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  ggtitle("Proportion of cells") +
  scale_fill_manual(values = c("#264653", "#2a9d8f"))

cellcount
png(fs::path(dir.results, "Barchart_Hep_Cells_by_Steatosis_Cat.png"),
    width = 2500, height = 2000, res = 300)
print(cellcount)
dev.off()

n_hep1 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-1"))
n_hep2 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-2"))
n_hep3 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-3"))
n_hep4 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-4"))
n_hep5 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-5"))
n_dhep <- nrow(so_liver_sn@meta.data %>% filter(celltype == "dHep"))

# By PT subtypes (by celltype)
cellcount2 <- so_liver_sn@meta.data %>% 
  filter(!is.na(steatosis_cat)) %>%
  filter(celltype2 == "Hep") %>%
  dplyr::group_by(steatosis_cat, celltype) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(n = case_when(celltype == "Hep-1" ~ n_hep1, 
                              celltype == "Hep-2" ~ n_hep2,
                              celltype == "Hep-3" ~ n_hep3, 
                              celltype == "Hep-4" ~ n_hep4,
                              celltype == "Hep-5" ~ n_hep5,
                              celltype == "dHep" ~ n_dhep),
                ratio = count / n) %>% ungroup() %>%
  group_by(steatosis_cat) %>%
  dplyr::mutate(proportion = ratio/sum(ratio)) %>%
  dplyr::select(proportion, steatosis_cat, celltype) %>%
  ggplot(aes(x= steatosis_cat, y= proportion, fill = celltype)) + 
  geom_bar(stat = "identity", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = NULL,
       fill = "celltype") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 15),
        text = element_text(size = 20)) +
  ggtitle("Proportion of Cells by Steatosis Category") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

cellcount2
png(fs::path(dir.results, "Barchart_Hep_Cells_by_Steatosis.png"),
    width = 1500, height = 2500, res = 300)
print(cellcount2)
dev.off()

#By ALT
so_liver_sn$alt_cat <- factor(so_liver_sn$alt)
n_hep1 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-1"))
n_hep2 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-2"))
n_hep3 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-3"))
n_hep4 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-4"))
n_hep5 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-5"))
n_dhep <- nrow(so_liver_sn@meta.data %>% filter(celltype == "dHep"))

# By PT subtypes (by celltype)
cellcount_alt <- so_liver_sn@meta.data %>% 
  filter(!is.na(alt_cat)) %>%
  filter(celltype2 == "Hep") %>%
  dplyr::group_by(alt_cat, celltype) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(n = case_when(celltype == "Hep-1" ~ n_hep1, 
                              celltype == "Hep-2" ~ n_hep2,
                              celltype == "Hep-3" ~ n_hep3, 
                              celltype == "Hep-4" ~ n_hep4,
                              celltype == "Hep-5" ~ n_hep5,
                              celltype == "dHep" ~ n_dhep),
                ratio = count / n) %>% ungroup() %>%
  group_by(alt_cat) %>%
  dplyr::mutate(proportion = ratio/sum(ratio)) %>%
  dplyr::select(proportion, alt_cat, celltype) %>%
  ggplot(aes(x= alt_cat, y= proportion, fill = celltype)) + 
  geom_bar(stat = "identity", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = NULL,
       fill = "celltype") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 15),
        text = element_text(size = 20)) +
  ggtitle("ALT") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

cellcount_alt
png(fs::path(dir.results, "Barchart_Hep_Cells_by_ALT.png"),
    width = 2500, height = 2000, res = 300)
print(cellcount_alt)
dev.off()

#By AST
so_liver_sn$ast_cat <- factor(so_liver_sn$ast)
n_hep1 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-1"))
n_hep2 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-2"))
n_hep3 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-3"))
n_hep4 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-4"))
n_hep5 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-5"))
n_dhep <- nrow(so_liver_sn@meta.data %>% filter(celltype == "dHep"))

# By Hep subtypes (by celltype)
cellcount_ast <- so_liver_sn@meta.data %>% 
  filter(!is.na(ast_cat)) %>%
  filter(celltype2 == "Hep") %>%
  dplyr::group_by(ast_cat, celltype) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(n = case_when(celltype == "Hep-1" ~ n_hep1, 
                              celltype == "Hep-2" ~ n_hep2,
                              celltype == "Hep-3" ~ n_hep3, 
                              celltype == "Hep-4" ~ n_hep4,
                              celltype == "Hep-5" ~ n_hep5,
                              celltype == "dHep" ~ n_dhep),
                ratio = count / n) %>% ungroup() %>%
  group_by(ast_cat) %>%
  dplyr::mutate(proportion = ratio/sum(ratio)) %>%
  dplyr::select(proportion, ast_cat, celltype) %>%
  ggplot(aes(x= ast_cat, y= proportion, fill = celltype)) + 
  geom_bar(stat = "identity", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = NULL,
       fill = "celltype") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 15),
        text = element_text(size = 20)) +
  ggtitle("AST") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

png(fs::path(dir.results, "Barchart_Hep_Cells_by_AST.png"),
    width = 2500, height = 2000, res = 300)
print(cellcount_ast)
dev.off()

#By ggt
so_liver_sn$ggt_cat <- factor(so_liver_sn$ggt)
n_hep1 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-1"))
n_hep2 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-2"))
n_hep3 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-3"))
n_hep4 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-4"))
n_hep5 <- nrow(so_liver_sn@meta.data %>% filter(celltype == "Hep-5"))
n_dhep <- nrow(so_liver_sn@meta.data %>% filter(celltype == "dHep"))

# By PT subtypes (by celltype)
cellcount_ggt <- so_liver_sn@meta.data %>% 
  filter(!is.na(ggt_cat)) %>%
  filter(celltype2 == "Hep") %>%
  dplyr::group_by(ggt_cat, celltype) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(n = case_when(celltype == "Hep-1" ~ n_hep1, 
                              celltype == "Hep-2" ~ n_hep2,
                              celltype == "Hep-3" ~ n_hep3, 
                              celltype == "Hep-4" ~ n_hep4,
                              celltype == "Hep-5" ~ n_hep5,
                              celltype == "dHep" ~ n_dhep),
                ratio = count / n) %>% ungroup() %>%
  group_by(ggt_cat) %>%
  dplyr::mutate(proportion = ratio/sum(ratio)) %>%
  dplyr::select(proportion, ggt_cat, celltype) %>%
  ggplot(aes(x= ggt_cat, y= proportion, fill = celltype)) + 
  geom_bar(stat = "identity", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = NULL,
       fill = "celltype") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold", size = 15),
        text = element_text(size = 20)) +
  ggtitle("GGT") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))

png(fs::path(dir.results, "Barchart_Hep_Cells_by_ggt.png"),
    width = 2500, height = 2000, res = 300)
print(cellcount_ggt)
dev.off()

#Marker genes of healthy hepatocytes: 
healthy_hepatocyte_genes <- c("ALB", "CYP3A4", "ARG1", "IGF1", "HNF4A", "APOE", "SLC10A2", "FASN", "ABCB1", "GLUL")

#Marker genes of adaptive/mal adaptive hepatocytes: 
adaptive_maladaptive_genes <- c("COL1A1", "ACTA2", "TNF", "IL6", "FASN", "CASP3", "SOD1", "CAT", "HGF", "PCNA")

# Healthy markers
DotPlot(so_Hep, features = healthy_hepatocyte_genes, group.by = "celltype") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("Healthy Hepatocyte Marker Expression")

# Adaptive/maladaptive markers
DotPlot(so_Hep, features = adaptive_maladaptive_genes, group.by = "celltype") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("Adaptive/Maladaptive Hepatocyte Marker Expression")

```

##b. Visualize 
```{r, echo = F}
# so_liver_sn <- FindVariableFeatures(object = so_liver_sn)
# so_liver_sn <- RunPCA(so_liver_sn, features = VariableFeatures(object = so_liver_sn),assay="RNA")
# ElbowPlot(so_liver_sn)
# 
# # # Find neighbors and clusters (again using integrated data)
# so_liver_sn <- FindNeighbors(so_liver_sn, assay = "RNA", dims = 1:20)
# so_liver_sn <- FindClusters(so_liver_sn, resolution = 0.5)
# 
# # Perform UMAP 
# #Print plot
# png(fs::path(dir.results, "UMAP_by_celltype.png"),
#     width = 4500, height = 3000, res = 300)
# p <- DimPlot(so_liver_sn, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
#   ggtitle("UMAP Plot by Cell Type")
# print(p)
# dev.off()
# 
# 
# #Print plot
# png(fs::path(dir.results, "UMAP_by_Steatosis.png"),
#     width = 4500, height = 3000, res = 300)
# p <- DimPlot(
#   so_liver_sn,
#   reduction = "umap",
#   group.by = "steatosis_cat",
#   label = FALSE,
#   raster = FALSE
# ) +
#   scale_color_manual(values = c("darkblue","lightgray")) +
#   ggtitle("UMAP by Steatosis Status")
# print(p)
# dev.off()
# 
# #Barcharts of proportions
# cellcount <- so_liver_sn@meta.data
# prop_plot <- ggplot(data=cellcount,aes(StudyID, fill = celltype)) + 
#   geom_bar(stat = "count", position = "fill") +
#   theme_classic() +
#   labs(x = NULL,
#        y = "Proportion of Cells",
#        fill = "Cell type") +
#   theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
#   theme(plot.title = element_text(hjust=0.5, face="bold"),
#         text = element_text(size = 20)) 
# # ggtitle("PT Cells") +
# # scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
# png(fs::path(dir.results, "Barchart_All_Participants_All_Cells.png"),width = 4500, height = 3000, res = 300)
# print(prop_plot)
# dev.off()
# 
# prop_plot <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
#   geom_bar(stat = "count", position = "fill") +
#   theme_classic() +
#   labs(x = NULL,
#        y = "Proportion of Cells",
#        fill = "Cell Type") +
#   theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
#   theme(plot.title = element_text(hjust=0.5, face="bold"),
#         text = element_text(size = 20)) 
# png(fs::path(dir.results, "Barchart_Steatosis_Cat_All_Cells.png"),width = 3000, height = 3000, res = 300)
# print(prop_plot)
# dev.off()

#Hep cells (Hep-1 thru dHep)
so_liver_sn
so_Hep <- subset(so_liver_sn, celltype2 == "Hep")
so_Hep
DefaultAssay(so_Hep) <- "RNA"
so_Hep <- NormalizeData(so_Hep)
so_Hep <- ScaleData(so_Hep)
so_Hep <- FindVariableFeatures(object = so_Hep)
# so_Hep <- RunPCA(so_Hep, features = VariableFeatures(object = so_Hep),assay="RNA")
ElbowPlot(so_Hep)

# # Find neighbors and clusters (again using integrated data)
so_Hep <- FindNeighbors(so_Hep, assay = "RNA", dims = 1:20)
so_Hep@graphs
so_Hep <- FindClusters(so_Hep, resolution = 0.5)
DefaultAssay(so_Hep) <- "RNA"
# Perform UMAP 
#Print plot
png(fs::path(dir.results, "UMAP_by_Hep_Celltypes.png"),
    width = 4000, height = 3000, res = 300)
p <- DimPlot(so_Hep, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")
print(p)
dev.off()


#Print plot
png(fs::path(dir.results, "UMAP_by_Steatosis_Hep.png"),
    width = 4000, height = 3000, res = 300)
p <- DimPlot(
  so_Hep,
  reduction = "umap",
  group.by = "steatosis_cat",
  label = FALSE,
  raster = FALSE
) +
  scale_color_manual(values = c("darkblue","lightgray")) +
  ggtitle("UMAP by Steatosis Status")+
  theme(legend.position = "none")
print(p)
dev.off()

#Barcharts of proportions
cellcount <- so_Hep@meta.data
prop_plot <- ggplot(data=cellcount,aes(StudyID, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  # ggtitle("PT Cells") +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
png(fs::path(dir.results, "Barchart_All_Participants_Hep_Cells.png"),width = 4500, height = 3000, res = 300)
print(prop_plot)
dev.off()

prop_plot <- ggplot(data=cellcount,aes(steatosis_cat, fill = celltype)) + 
  geom_bar(stat = "count", position = "fill") +
  theme_classic() +
  labs(x = NULL,
       y = "Proportion of Cells",
       fill = "Cell Type") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold"),
        text = element_text(size = 20)) +
  scale_fill_manual(values = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51","darkred"))
png(fs::path(dir.results, "Barchart_Steatosis_Cat_All_Cells.png"),width = 3000, height = 3000, res = 300)
print(prop_plot)
dev.off()


# so_Hep <- NormalizeData(so_Hep)
# so_Hep <- ScaleData(so_Hep)
# 
# so_Hep <- FindVariableFeatures(object = so_Hep)
# so_Hep <- RunPCA(so_Hep, features = VariableFeatures(object = so_Hep),assay="RNA")
# ElbowPlot(so_Hep)
# 
# # # Find neighbors and clusters (again using integrated data)
# so_Hep <- FindNeighbors(so_Hep, assay = "RNA", dims = 1:20)
# so_Hep <- FindClusters(so_Hep, resolution = 0.5)
# 
# so_Hep <- RunUMAP(so_Hep, dims = 1:20)
# so_Hep <- RunTSNE(so_Hep, dimred = "UMAP")
# 
DimPlot(object = so_Hep, reduction = "umap")
DimPlot(object = so_Hep, reduction = "umap", group.by = "steatosis_grade")
DimPlot(object = so_Hep, reduction = "umap", group.by = "celltype")

DimPlot(so_Hep, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")

umap_df <- as.data.frame(so_Hep@reductions$umap@cell.embeddings)
umap_df <- umap_df %>% 
  filter(UMAP_1 < 5) %>% 
  filter(UMAP_2 > -8)
umap_keep <- rownames(umap_df)
length(umap_keep)  # confirm how many cells are kept - 103031

so_Hep_filtered <- subset(so_Hep, cells = umap_keep)
nrow(so_Hep_filtered)
ncol(so_Hep_filtered)
# png(fs::path(dir.results, "UMAP_by_Hep_Filtered_5_neg8.png"),
# width = 4000, height = 3000, res = 300)
p <- DimPlot(so_Hep_filtered, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")
print(p)
# dev.off()

# png(fs::path(dir.results, "UMAP_by_Hep_Unfiltered.png"),
# width = 4000, height = 3000, res = 300)
p <- DimPlot(so_Hep, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")
# print(p)
# dev.off()

umap_df <- as.data.frame(so_Hep@reductions$umap@cell.embeddings)
umap_df <- umap_df %>% 
  filter(UMAP_1 < 4) %>% 
  filter(UMAP_2 > -7)
umap_keep <- rownames(umap_df)
length(umap_keep)

so_Hep_filtered <- subset(so_Hep, cells = umap_keep)
nrow(so_Hep_filtered)
ncol(so_Hep_filtered)
# png(fs::path(dir.results, "UMAP_by_Hep_Filtered_5_neg8.png"),
# width = 4000, height = 3000, res = 300)
p <- DimPlot(so_Hep_filtered, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")
print(p)
# dev.off()

umap_df <- as.data.frame(so_Hep@reductions$umap@cell.embeddings)
umap_df <- umap_df %>% 
  filter(UMAP_1 < 4.6) %>% 
  filter(UMAP_2 > -7)
umap_keep <- rownames(umap_df)
length(umap_keep)

so_Hep_filtered <- subset(so_Hep, cells = umap_keep)
nrow(so_Hep_filtered)
ncol(so_Hep_filtered)
# png(fs::path(dir.results, "UMAP_by_Hep_Filtered_5_neg8.png"),
# width = 4000, height = 3000, res = 300)
p <- DimPlot(so_Hep_filtered, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
  ggtitle("UMAP Plot by Cell Type")+
  theme(legend.position = "none")
print(p)

# png(fs::path(dir.results, "UMAP_by_Hep_Celltypes2.png"),
# width = 4000, height = 3000, res = 300)
# 
# so_Hep <- so_Hep[, umap_keep]
# DimPlot(object = so_Hep, reduction = "umap")
# DimPlot(object = so_Hep, reduction = "umap", group.by = "steatosis_cat")
# DimPlot(object = so_Hep, reduction = "umap", group.by = "celltype")
# 
# #Print plot
# png(fs::path(dir.results, "UMAP_by_Hep_Celltypes2.png"),
# width = 4000, height = 3000, res = 300)
# p <- DimPlot(so_Hep, reduction = "umap", group.by = "celltype",raster = FALSE, label = TRUE, label.size = 4)+
#   ggtitle("UMAP Plot by Cell Type")+
#    theme(legend.position = "none")
# print(p)
# dev.off()
# 
# 
# #Print plot
# png(fs::path(dir.results, "UMAP_by_Steatosis_Hep2.png"),
#     width = 4000, height = 3000, res = 300)
# p <- DimPlot(
#   so_Hep,
#   reduction = "umap",
#   group.by = "steatosis_cat",
#   label = FALSE,
#   raster = FALSE
# ) +
#   scale_color_manual(values = c("darkblue","lightgray")) +
#   ggtitle("UMAP by Steatosis Status")+
#    theme(legend.position = "none")
# print(p)
# dev.off()
```

##c. Totem Clustering
```{r}
# # Totem clustering for trajectory analysis
so_Hep <- subset(so_liver_sn, celltype2 == "Hep")
DefaultAssay(so_Hep) <- "RNA"

so_Hep <- FindVariableFeatures(so_Hep, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_Hep)


so_Hep <- NormalizeData(so_Hep)
so_Hep <- ScaleData(so_Hep)
sce_HEP <- as.SingleCellExperiment(so_Hep)
gc()
sce_HEP <- PrepareTotem(sce_HEP)
sce_HEP <- RunDimRed(object = sce_HEP,
                     dim.red.method = "pca",
                     dim.red.features = hvgs,
                     dim.reduction.par.list = list(ndim=5))
# 
## where so@assays$RNA@counts is the normalized expression count
# gc()
# dim_red <- dimred_pca(t(subset(so, generaltype == "PT" & celltype != "PT_lowQuality")@assays$RNA@counts), ndim=2)
dim_red <- reducedDim(sce_HEP, type = "pca")
# 
sce_HEP <- RunClustering(sce_HEP,
                         k.range = 3:20,
                         min.cluster.size = 5,
                         N.clusterings=10000)
gc()
viz_cell <- VizCellConnectivity(sce_HEP,viz.dim.red = dim_red)

# 
pushover(message = "done w/ VizCellConnectivity")

sce_HEP <- SelectClusterings(sce_HEP,selection.method = 1,
                             selection.N.models = 10,
                             selection.stratified=FALSE,
                             prior.clustering = NULL)
VizMST(sce_HEP,clustering.names = ReturnTrajNames(sce_HEP),viz.dim.red = dim_red)
# VizMST(sce_HEP,clustering = so_Hep$celltype,viz.dim.red = dim_red)

ReturnMSTNetwork(sce_HEP,clustering.name = "3.367")
sce_HEP <- RunSmoothing(sce_HEP)

VizClustering(sce_HEP,clustering = so_Hep$celltype,viz.dim.red = dim_red)


# Step 1: Get cell types with names
celltypes <- so_Hep$celltype
names(celltypes) <- colnames(so_Hep)  # make sure they have cell names

# Step 2: Subset to match sce_HEP
common_cells <- intersect(colnames(sce_HEP), names(celltypes))
celltypes <- celltypes[common_cells]

# Step 3: Make sure the order matches sce_HEP
celltypes <- celltypes[colnames(sce_HEP)]
celltypes <- as.factor(celltypes)
all(names(celltypes) == colnames(sce_HEP))  # should be TRUE
# If not, fix it:
celltypes <- celltypes[colnames(sce_HEP)]

# Assume dim_red is already defined like this:
# dim_red <- reducedDims(sce_HEP)[["<your-dim-reduction>"]]  # already done

# Check cell names in both
head(rownames(dim_red))          # cell names from dim_red
head(names(celltypes))           # cell names from celltypes

# Ensure both are character vectors of matching cell IDs
common_cells <- intersect(rownames(dim_red), names(celltypes))

# Subset both to shared cells
dim_red <- dim_red[common_cells, , drop = FALSE]
celltypes <- celltypes[common_cells]

# Confirm order
celltypes <- celltypes[rownames(dim_red)]

# Confirm 1:1 matching
stopifnot(all(names(celltypes) == rownames(dim_red)))
table(is.na(celltypes))  # Should show no NAs
length(celltypes) == nrow(dim_red)  # Should be TRUE

length(celltypes)          # should equal
nrow(dim_red)              # the number of rows in dim_red
# Remove NA entries
valid <- !is.na(celltypes)
celltypes <- celltypes[valid]
dim_red <- dim_red[valid, , drop = FALSE]
sce_HEP <- sce_HEP[, rownames(dim_red)]

# Step 4: Visualize
VizClustering(sce_HEP, clustering = celltypes, viz.dim.red = dim_red)
VizMST(sce_HEP, clustering = celltypes, viz.dim.red = dim_red)

# Debug your slingshot clustering visualization

# Step 1: Check dimensions and data integrity
cat("=== Debugging Data Dimensions ===\n")
cat("Original celltypes length:", length(celltypes), "\n")
cat("sce_HEP columns:", ncol(sce_HEP), "\n")
cat("dim_red rows:", nrow(dim_red), "\n")

# Check for NA values
cat("NA values in celltypes:", sum(is.na(celltypes)), "\n")

# Step 2: Ensure all dimensions match
cat("\n=== Checking Dimension Alignment ===\n")

# Make sure all objects have the same number of cells
n_cells_sce <- ncol(sce_HEP)
n_cells_dimred <- nrow(dim_red)
n_celltypes <- length(celltypes)

cat("Cells in sce_HEP:", n_cells_sce, "\n")
cat("Cells in dim_red:", n_cells_dimred, "\n")
cat("Cells in celltypes:", n_celltypes, "\n")

# Check if dimensions match
if (n_cells_sce != n_cells_dimred || n_cells_sce != n_celltypes) {
    cat("WARNING: Dimension mismatch detected!\n")
}

# Step 3: Check clustering integrity
cat("\n=== Checking Clustering Integrity ===\n")
cat("Unique cell types:", length(unique(celltypes)), "\n")
cat("Cell type distribution:\n")
print(table(celltypes))

# Check for empty or problematic groups
empty_groups <- names(table(celltypes))[table(celltypes) == 0]
if (length(empty_groups) > 0) {
    cat("Empty groups found:", empty_groups, "\n")
}

# Step 4: Fix the data alignment
cat("\n=== Fixing Data Alignment ===\n")

# Method 1: If you have cell names/barcodes, use them for alignment
if (!is.null(colnames(sce_HEP)) && !is.null(rownames(dim_red))) {
    # Find common cells
    common_cells <- intersect(colnames(sce_HEP), rownames(dim_red))
    cat("Common cells found:", length(common_cells), "\n")
    
    if (length(common_cells) > 0) {
        # Subset to common cells
        sce_HEP_aligned <- sce_HEP[, common_cells]
        dim_red_aligned <- dim_red[common_cells, , drop = FALSE]
        
        # If celltypes has names, align by names
        if (!is.null(names(celltypes))) {
            celltypes_aligned <- celltypes[common_cells]
        } else {
            # If no names, you'll need to manually align
            cat("WARNING: celltypes has no names - manual alignment needed\n")
            celltypes_aligned <- celltypes[1:length(common_cells)]
        }
    }
} else {
    # Method 2: If no cell names, ensure same order and length
    cat("No cell names found - using positional alignment\n")
    
    # Find minimum length
    min_length <- min(ncol(sce_HEP), nrow(dim_red), length(celltypes))
    cat("Using first", min_length, "cells\n")
    
    sce_HEP_aligned <- sce_HEP[, 1:min_length]
    dim_red_aligned <- dim_red[1:min_length, , drop = FALSE]
    celltypes_aligned <- celltypes[1:min_length]
}

# Step 5: Remove any remaining NA values
valid_cells <- !is.na(celltypes_aligned)
sce_HEP_final <- sce_HEP_aligned[, valid_cells]
dim_red_final <- dim_red_aligned[valid_cells, , drop = FALSE]
celltypes_final <- celltypes_aligned[valid_cells]

cat("\n=== Final Data Check ===\n")
cat("Final sce_HEP columns:", ncol(sce_HEP_final), "\n")
cat("Final dim_red rows:", nrow(dim_red_final), "\n")
cat("Final celltypes length:", length(celltypes_final), "\n")
cat("Final cell type distribution:\n")
print(table(celltypes_final))

# Step 6: Try visualization with cleaned data
cat("\n=== Attempting Visualization ===\n")

# First try VizMST (usually more robust)
tryCatch({
    VizMST(sce_HEP_final, clustering = celltypes_final, viz.dim.red = dim_red_final)
    cat("VizMST successful!\n")
}, error = function(e) {
    cat("VizMST failed:", e$message, "\n")
})

# Then try VizClustering
tryCatch({
    VizClustering(sce_HEP_final, clustering = celltypes_final, viz.dim.red = dim_red_final)
    cat("VizClustering successful!\n")
}, error = function(e) {
    cat("VizClustering failed:", e$message, "\n")
})

# Alternative visualization approach
cat("\n=== Alternative Visualization ===\n")
library(ggplot2)

# Create a simple scatter plot
plot_df <- data.frame(
    dim_red_final,
    celltype = as.factor(celltypes_final)
)

p <- ggplot(plot_df, aes(x = X1, y = X2, color = celltype)) +
    geom_point(alpha = 0.7, size = 0.8) +
    theme_minimal() +
    labs(title = "Cell Types on Dimensionality Reduction",
         x = "Dimension 1", y = "Dimension 2",
         color = "Cell Type") +
    guides(color = guide_legend(override.aes = list(size = 2, alpha = 1)))

print(p)

# If you want to add trajectory information
if (exists("sds")) {  # sds is your SlingshotDataSet
    tryCatch({
        # Extract trajectory information
        curves <- slingCurves(sds)
        
        # Add curves to plot
        for (i in seq_along(curves)) {
            curve_coords <- curves[[i]]$s[curves[[i]]$ord, ]
            p <- p + geom_path(data = data.frame(X1 = curve_coords[,1], 
                                               X2 = curve_coords[,2]),
                             aes(x = X1, y = X2), 
                             color = "black", size = 1, inherit.aes = FALSE)
        }
        print(p)
    }, error = function(e) {
        cat("Could not add trajectories:", e$message, "\n")
    })
}
```

##d. Fit Slingshot
```{r echo = F}
# After visualizing UMAP by cell type clusters, we fit slingshot two ways: 
#   i) setting start cluster as PT-3, ending cluster as PT-4 (PT-3 having the most "healthy" cells, PT-4 having the most "diseased" cells),
#   ii) unsupervised.
#   
# The first method resulted in two trajectories (lineage 1: PT-3  PT-2  PT-5  PT-4; lineage 2: PT-3  PT-2  PT-1), and the second method also resulted in two trajectories (lineage 1: PT-4  PT-5  PT-2  PT-3; lineage 2: PT-4  PT-5  PT-2  PT-1).
# 
# Lineage 1 of the unsupervised method was the exact inverse of lineage 1 of the supervised method. Therefore, we moved forward with the initial method.

# End with Hep-5
sce_HEP <- as.SingleCellExperiment(so_Hep)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep, reduction = "umap")

sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, end.clus = 'Hep-5')
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_hep5.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)


# #start point hep-1
# sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = 'Hep-1')
# shuffle <- sample(ncol(sce_HEP))
# 
# jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_dHep.jpeg"), width = 1000, height =800, quality = 100)
# par(mar = c(5, 6, 1, 1))
# plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
#      xlab = "UMAP-1", ylab = "UMAP-2",
#      col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
#      cex.lab = 3, cex.axis = 2)
# lines(SlingshotDataSet(sce_HEP), type = "lineages")
# lines(SlingshotDataSet(sce_HEP))
# dev.off()
# 
# SlingshotDataSet(sce_HEP)
# 
# #start point hep-2
# sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = 'Hep-2')
# shuffle <- sample(ncol(sce_HEP))
# 
# jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_dHep.jpeg"), width = 1000, height =800, quality = 100)
# par(mar = c(5, 6, 1, 1))
# plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
#      xlab = "UMAP-1", ylab = "UMAP-2",
#      col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
#      cex.lab = 3, cex.axis = 2)
# lines(SlingshotDataSet(sce_HEP), type = "lineages")
# lines(SlingshotDataSet(sce_HEP))
# dev.off()
# 
# SlingshotDataSet(sce_HEP)

#Hep-5 after filtering
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")

sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, end.clus = 'Hep-5')
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_hep5_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

#dHep after filtering
# sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
# reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")

sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, end.clus = 'dHep')
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_dHep_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

#Try start and end point
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")
sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = "Hep-1", end.clus = 'dHep')
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_start_hep1_end_dHep_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

#Just hep-1 start point
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")
sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = "Hep-1")
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_start_hep1_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)


#Just hep2 start point
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")
sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = "Hep-2")
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_start_hep2_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

#Just hep2 start point to Hep-5
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")
sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = "Hep-2",end.clus="Hep-5")
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_start_hep2_end_hep5_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

#Just hep1 start point to Hep-5
sce_HEP <- as.SingleCellExperiment(so_Hep_filtered)
reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered, reduction = "umap")
sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, start.clus = "Hep-1",end.clus="Hep-5")
shuffle <- sample(ncol(sce_HEP))

jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_start_hep1_end_hep5_filtered.jpeg"), width = 1000, height =800, quality = 100)
par(mar = c(5, 6, 1, 1))
plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
     xlab = "UMAP-1", ylab = "UMAP-2",
     col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
     cex.lab = 3, cex.axis = 2)
lines(SlingshotDataSet(sce_HEP), type = "lineages")
lines(SlingshotDataSet(sce_HEP))
dev.off()

SlingshotDataSet(sce_HEP)

# #Remove dHep
# so_Hep_filtered2 <- subset(so_Hep_filtered,celltype!="dHep")
# sce_HEP <- as.SingleCellExperiment(so_Hep_filtered2)
# reducedDim(sce_HEP, "UMAP") <- Embeddings(so_Hep_filtered2, reduction = "umap")
# 
# sce_HEP <- slingshot(sce_HEP, reducedDim = 'UMAP', clusterLabels = colData(sce_HEP)$celltype, end.clus = 'Hep-5')
# shuffle <- sample(ncol(sce_HEP))

# jpeg(fs::path(dir.results,"Hep_slingshot_UMAP_end_hep5_filtered_no_dHep.jpeg"), width = 1000, height =800, quality = 100)
# par(mar = c(5, 6, 1, 1))
# plot(reducedDims(sce_HEP)$UMAP[shuffle, ], asp = 1, pch = 16,
#      xlab = "UMAP-1", ylab = "UMAP-2",
#      col = hcl.colors(100, alpha = .5, palette = "PinkYl")[cut(sce_HEP$slingPseudotime_1, breaks = 100)][shuffle],
#      cex.lab = 3, cex.axis = 2)
# lines(SlingshotDataSet(sce_HEP), type = "lineages")
# lines(SlingshotDataSet(sce_HEP))
# dev.off()
# 
# SlingshotDataSet(sce_HEP)

# 
# plot_df <- as.data.frame(slingPseudotime(sce_Hep_unsupervised)) %>%
#   dplyr::mutate(Group = colData(sce_Hep_unsupervised)$Group,
#                 gbm_thick_quartile = colData(sce_Hep_unsupervised)$gbm_thick_quartile,
#                 bmi_quartile = colData(sce_Hep_unsupervised)$bmi_quartile,
#                 bmi_manual_cat = colData(sce_Hep_unsupervised)$bmi_manual_cat,
#                 diabetes_duration_quartile = colData(sce_Hep_unsupervised)$diabetes_duration_quartile,
#                 diabetes_duration_manual_cat = colData(sce_Hep_unsupervised)$diabetes_duration_manual_cat,
#                 age_quartile = colData(sce_Hep_unsupervised)$age_quartile,
#                 raw_m_quartile = colData(sce_Hep_unsupervised)$raw_m_quartile)
# lines(SlingshotDataSet(sce_Hep_unsupervised))

# plot(reducedDims(sce_Hep_unsupervised)$UMAP[shuffle, ], asp = 1, pch = 16,
#      xlab = "UMAP-1", ylab = "UMAP-2",
#      col = sce_Hep_unsupervised$Group)
# 
# lines(SlingshotDataSet(sce_Hep_unsupervised), type = "lineages")
# legend("topright", pch = 16, col = alpha(c("HC" = "#ffcb77", "T1D" = "#fe6d73"), 0.4), bty = "n",
#        legend = levels(factor(colData(sce)$Group)))
# 
# plotUMAP(sce_Hep_unsupervised, colour_by="pseudotime", point_alpha=0.3) +
#     geom_segment(data=grid.df,
#         mapping=aes(x=start.1, y=start.2, xend=end.1, yend=end.2),
#         arrow=arrow(length=unit(0.05, "inches"), type="closed"))
```

```{r echo = F}
# Function to plot all smooth after fitGAM
plot_smooth <- function(sce, gene, title = gene) {
  non_missing_cells <- (colData(sce)$tradeSeq$conditions != "NA")
  non_missing_cells <- as.data.frame(non_missing_cells) %>% 
    dplyr::mutate(non_missing_cells = case_when(non_missing_cells == T ~ 1))
  cnt <- assays(sce)$counts[gene, , drop=F]
  cnt <- as.data.frame(t(cnt))
  cnt <- cnt*non_missing_cells
  colnames(cnt) <- "exp"
  cnt <- cnt %>% filter(!is.na(exp))
  pseudotime <- (colData(sce)$crv$pseudotime.Lineage1 * non_missing_cells %>%
                   filter(!is.na(non_missing_cells)))$non_missing_cells
  condition <- colData(sce)$tradeSeq$conditions
  
  if ("NA" %in% condition){
    condition <- condition[condition != "NA"]
  }
  
  ncondition <- length(unique(condition))
  
  if (ncondition == 2){
    colors = c("#007e5d",  "#ff8989")
  }
  if (ncondition == 3){
    colors = c("#ef767a", "#456990", "#49beaa")
  }
  if(ncondition == 4){
    colors = c("#003049", "#d62828", "#f77f00", "#fcbf49")
  }
  if(ncondition == 5){
    colors = c("#264653", "#2a9d8f", "#e9c46a", "#f4a261", "#e76f51")
  }
  
  cnt_smooth <- predictSmooth(sce, gene, nPoints=100) %>%
    filter(lineage == 1) %>% dplyr::select(-lineage)
  cnt_smooth <- cnt_smooth[, c("yhat", "time", "condition")]
  colnames(cnt_smooth) <- c("exp", "pseudotime", "condition")
  cnt_smooth <- cnt_smooth %>% filter(condition!="NA")
  
  p <- ggplot(NULL, aes(x=pseudotime, y=log1p(exp), color=condition)) + 
    geom_point(data=cnt, size=1, alpha = 0.3) +
    geom_line(data=cnt_smooth, linewidth=1.5) +
    labs(x = "Pseudotime",
         y = "Log(expression + 1)", 
         title = title,
         color = "Group") +
    theme_minimal() +
    theme(panel.border = element_blank(), panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) +
    scale_color_manual(values = colors)
  return(p)
}
```





#6. Proteomics Analysis
## a. TODAY 
###i. Proteomics Cleaning
```{r, include=FALSE}
# knitr::opts_knit$set(root.dir = "~/Volumes/Peds Endo-1/Petter Bjornstad/TODAY subaward")
# load somalogic data, with QC samples already excluded
# load("/Volumes/Peds Endo-1/Petter Bjornstad/TODAY subaward/Somalogic data raw/soma.Rdata")
load("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/TODAY subaward/Somalogic data raw/soma.Rdata")

# load analyte info
load("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/TODAY subaward/Somalogic data raw/analytes.Rdata")

# load comorbidity data
load("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/TODAY subaward/Clinical data/comorb.Rdata")

# load baseline risk factors
load("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/TODAY subaward/Clinical data/TODAY/baserisk.Rdata")

#Get key for peoteins and gene names 
key <- analytes %>% 
  dplyr::select(c("AptName","SeqId","EntrezGeneSymbol"))
saveRDS(key,"/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver project/Data/gene_prot.RDS")

# take only the baseline soma samples
# can we just take the earliest or is it possible someone would have a follow-up sample but not baseline?
# probably can take the first and then check the years to make sure compatible with TODAY
base <- soma %>% arrange(releaseid,Date.Drawn) %>% dplyr::group_by(releaseid) %>% dplyr::filter(row_number()==1)
# these 3 release IDs have first sample after end of recruitment, so they must be missing the baseline visit
base <- base %>% filter(!releaseid %in% c("65-85903","65-47984","65-25901"))

# the above section of code used to work but seems to have broken....this should accomplish the same thing?
#base <- soma %>% filter(visit==1)

# merge in complication data
base <- left_join(base, comorb, by="releaseid")
# this was previously:
# base <- merge(base, comorb, by="releaseid",all.x=T, all.y=F)

# merge in baseline risk factors
base <- left_join(base, baserisk, by="releaseid")
# this was previously:
#base <- merge(base, baserisk, by="releaseid",all.x=T, all.y=F)

# log transform UAlbCreat
base$log_UAlbCreat <- log(base$UAlbCreat + 0.0000001)

# identify columns corresponding to proteins
#is_seq <- function(.x) grepl("^seq\\.[0-9]{4}", .x) # regex for analytes
is_seq <- function(.x) grepl("seq", .x)
seq <- is_seq(names(base))

# convert to numeric
base[,seq] <- apply(base[,seq],2,as.numeric)

# are there proteins with low variability?
no_var = caret::nearZeroVar(base[,seq])
# none

# log transform
base_log <- base %>% modify_if(is_seq(names(.)), log)

# scale by SD
base_log_scale <- base_log
predictors <- colnames(base_log[seq])
for (i in 1:length(predictors)) {
  base_log_scale[,paste0(predictors[i])] <- base_log_scale[,paste0(predictors[i])]/sd(unlist(base_log[,paste0(predictors[i])]))
}

saveRDS(base_log,"/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Liver project/Data/TODAY_proteomics.rds")

```
###i. New
```{r}
#Load in AST and ALT in today 
today <- read.csv(fs::path(dir.dat,"TODAY subaward","Clinical data","TODAY","CBL.csv")) # length(unique(today$releaseid)) - 699 ids
today <- today %>%
  filter(mvisit!="R")
#dplyr::select(c("releaseid","ALT","AST"))
#Load in Proteomics in Today
prot <- readRDS(fs::path(dir.dat,"Liver project","Data","TODAY_proteomics.rds"))
# which(colnames(prot)=="AGEBASE") #7406
# which(colnames(prot)=="releaseid") #7326
# which(colnames(prot)=="sex") #7399
# proteins_keep <- which(grepl("seq.",colnames(prot)))
# prot <- prot %>% 
#   dplyr::select(all_of(c("releaseid","AGEBASE","sex",proteins_keep)))
prot <- prot[c(7326,7406,7399,which(grepl("seq.",colnames(prot))))] #length(unique(prot$releaseid)) - 376 unique ids
proteins <- colnames(prot)[which(grepl("seq.",colnames(prot)))]

#rm(today,today_M0,prot)

# #Determine which proteins correlate with genes associated with alt and ast
# #AST
# ast_genes <- read.xlsx(fs::path(dir.dat,"Liver project","Results","Bulk_Results_for_Ast Di (High vs. Low).xlsx"))
# ast_genes_pos <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC>0) %>%
#   arrange(desc(avg_log2FC)) %>%
#   dplyr::slice(1:200) 
# ast_genes_pos <- ast_genes_pos$Gene
# ast_genes_neg <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC<0) %>%
#   arrange(avg_log2FC) %>%
#   dplyr::slice(1:200) 
# ast_genes_neg <- ast_genes_neg$Gene
# ast_genes <- c(ast_genes_pos,ast_genes_neg)

#Key Protein Gene
key <- readRDS(fs::path(dir.dat,"Liver project","Data","gene_prot.RDS"))

so_celltype <- subset(so_liver_sn,celltype2=="Hep")
DefaultAssay(so_celltype) <- "RNA" 

nrow(so_celltype) #28964 genes
ncol(so_celltype) #130124 nuclei

## Select Highly Variable Genes (HVGs)
so_celltype <- FindVariableFeatures(so_celltype, selection.method = "vst", nfeatures = 2000)
hvgs <- VariableFeatures(so_celltype)
rm(so_celltype)

hvg_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% hvgs)]
prot <- prot[c("releaseid","AGEBASE","sex",hvg_proteins)]

#Calculate sensoring and time to event
baseline <- today %>% 
  filter(mvisit=="M00") %>% 
  mutate(baseline_alt=ALT,
         baseline_ast=AST) %>% 
  dplyr::select(all_of(c("releaseid","baseline_alt","baseline_ast"))) 
dat <- tidylog::left_join(today,baseline,by="releaseid")
dat <- dat %>% 
  mutate(event_alt=ifelse(ALT>=2*(baseline_alt),1,0)) %>% 
  mutate(event_ast=ifelse(AST>=2*(baseline_ast),1,0)) 
#Calculate time to event & censoring
dat_alt <- dat %>% 
  group_by(releaseid) %>%
  summarize(
    event_status_alt = ifelse(any(event_alt == 1, na.rm = TRUE), 1, 0), # 1 if event occurred, 0 if censored
    time_to_event_alt = ifelse(
      any(event_alt == 1, na.rm = TRUE), 
      min(days[event_alt == 1], na.rm = TRUE), # Time of first event
      max(days, na.rm = TRUE)                 # Time of last observation (censoring)
    )) %>% 
  ungroup()
dat_ast <- dat %>% 
  group_by(releaseid) %>%
  summarize(
    event_status_ast = ifelse(any(event_ast == 1, na.rm = TRUE), 1, 0), # 1 if event occurred, 0 if censored
    time_to_event_ast = ifelse(
      any(event_ast == 1, na.rm = TRUE), 
      min(days[event_ast == 1], na.rm = TRUE), # Time of first event
      max(days, na.rm = TRUE)                 # Time of last observation (censoring)
    )) %>% 
  ungroup()

dat_covs <- today %>% 
  filter(mvisit=="M00") %>% 
  dplyr::select(all_of(c("releaseid","HbA1c")))
#Merge proteins in 
dat <- tidylog::right_join(dat_covs,prot,by="releaseid")
dat <- tidylog::right_join(dat_alt,dat,by="releaseid")
dat <- tidylog::right_join(dat_ast,dat,by="releaseid")
rm(baseline,dat_alt,dat_ast,dat_covs,prot,today)

#Create loop for linear regression between proteins & liver enzymes
results <- data.frame()
for(y in c("alt","ast")) {
  for (x in hvg_proteins) {
    # M0 <- as.formula(paste0("Surv(time_to_event_alt,event_status_alt) ~ ",x,"+AGEBASE"))
    M0 <- as.formula(paste0("Surv(time_to_event_",y,",event_status_",y,") ~ ",x))
    M1 <- coxph(M0,data=dat)
    logHR <- summary(M1)$coef[1,1]
    pval <- summary(M1)$coef[1,5]
    results_full <- data.frame(Protein=x,Outcome=paste0("Time to doubling of ",str_to_upper(y)),LogHR=logHR,Pvalue=pval)
    results <- rbind(results,results_full)
  }
}
#ALT
results_alt <- results %>%
  filter(Outcome=="Time to doubling of ALT") %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr"))
results_alt <- results_alt %>%
  mutate(sig=ifelse(fdr<0.05,"*",""))
write.csv(results_alt,fs::path(dir.results,"TODAY_proteomics_ALT_hvg.csv"))

#AST
results_ast <- results %>%
  filter(Outcome=="Time to doubling of AST") %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr"))
results_ast <- results_ast %>%
  mutate(sig=ifelse(fdr<0.05,"*",""))
write.csv(results_alt,fs::path(dir.results,"TODAY_proteomics_AST_hvg.csv"))

results_total <- rbind(results_ast,results_alt)

#Match to gene name
key$Protein <- key$AptName
results_total <- tidylog::left_join(results_total,key,by="Protein")
results_ast <- tidylog::left_join(results_ast,key,by="Protein")
results_alt <- tidylog::left_join(results_alt,key,by="Protein")
write.csv(results_alt,fs::path(dir.results,"TODAY_proteomics_AST_hvg.csv"))
write.csv(results_alt,fs::path(dir.results,"TODAY_proteomics_ALT_hvg.csv"))

sig_results <- results_total %>% 
  filter(sig=="*")

write.csv(results_total,fs::path(dir.results,"TODAY_proteomics_alt_ast_hvg_adj_age.csv"))

# mutate(fdr3=p.adjust(PValue3,method="fdr"))
results_total$PValue10 <- -log10(pmax(results_total$Pvalue, 1e-10))  # Avoid log(0)


results_total <- results_total %>% 
  dplyr::rename(Gene=EntrezGeneSymbol)

results_total$color <- ifelse(results_total$fdr < 0.05 & results_total$LogHR > 0, "lightcoral",
                              ifelse(results_total$fdr < 0.05 & results_total$LogHR < 0, "lightblue", "gray"))

# Identify significant points (fdr < 0.05)
significant_df <- results_total[results_total$fdr < 0.05, ]

Genes <- length(unique(results_total$Gene))
# Nuclei <- ncol(so_celltype_hvg)
N <- length(unique(dat$releaseid))
# Nonconvergence_Rate <- nebula_nonconverged_percent
# results_total$color3 <- ifelse(results_total$fdr3 < 0.2 & results_total$LogHR3 > 0, "lightcoral",
#                               ifelse(results_total$fdr3 < 0.2 & results_total$LogHR3 < 0, "lightblue", "gray"))
# 
# # Identify significant points (fdr < 0.05)
# significant_df3 <- results_total[results_total$fdr3 < 0.2, ]

max <- max(results_total$LogHR)
# max <- 3.1
min <- min(results_total$LogHR)
# Create the volcano plot using ggplot
volcano_plot <- ggplot(results_total, aes(x = LogHR, y = PValue10, color = color)) +
  geom_point(alpha = 0.7) +  # Plot points with transparency
  scale_color_identity() +  # Use the color column directly
  # theme_minimal() +  # Minimal theme
  theme_bw()+
  labs(
    # title = "High Steatosis (2+3) vs. Low Steatosis (0+1)",
    # subtitle = "Hep-1, Unadjusted (REML)",
    x = "logHR",
    # y = "-log10(P-Value)",
    color = "LogHR Direction Direction",
    caption = paste0("FDR < 0.05, Proteins = ",Genes,", N = ",N)
  ) +
  theme(
    plot.title = element_text(hjust = 0),
    axis.text.x = element_text(angle = 0, hjust = 1)
  )+
  xlim(min,max)+
  # # Add labels for significant points
  geom_text(data = significant_df, aes(label = Gene),
            vjust = 1, hjust = 1, size = 3, check_overlap = TRUE, color = "black")+
  facet_wrap(~Outcome)
# Add labels for significant points with ggrepel
# geom_text_repel(data = significant_df, aes(label = Gene),
#                 size = 3, color = "black", box.padding = 0.5, max.overlaps = 20)

volcano_plot

# pdf(fs::path(dir.results,"Plot_NEBULA_Hepatocytes_Steatosis_unadjusted_2000_hvg_reml.pdf"),width=10,height=7)
# print(volcano_plot)
# dev.off()

png(fs::path(dir.results, "Plot_ALT_AST_Proteomics_adj_2000_hvg.png"), 
    width = 3000, height = 1500, res = 300)
print(volcano_plot)
dev.off()

#GSEA
full_results <- read.csv(fs::path(dir.results,"TODAY_proteomics_AST_hvg.csv")) %>%
  # clean_names() %>% 
  dplyr::select(-X) %>%
  dplyr::rename(Gene=EntrezGeneSymbol)

gene_direction_summary <- full_results %>%
  # mutate(Direction=ifelse(LogHR))
  group_by(Gene) %>%
  summarise(
    n_aptamers = n(),
    n_positive = sum(LogHR > 0),
    n_negative = sum(LogHR < 0),
    n_significant = sum(Pvalue < 0.05),
    direction_consistent = n_distinct(sign(LogHR)) == 1,
    LogHR_range = paste0(round(min(LogHR), 3), " to ", round(max(LogHR), 3)),
    .groups = "drop"
  ) %>%
  filter(n_aptamers > 1)  


gene_stats <- full_results %>%
  filter(!is.na(logHR)) %>%  # remove NAs
  group_by(Gene) %>%
  arrange(desc(abs(logHR))) %>%  # sort by magnitude
  slice(1) %>%  # take the top one
  ungroup()

# Filter out the gmt files for KEGG, Reactome and GOBP
list.files(bg_path)
gmt_files <- list.files(path = bg_path, pattern = '.gmt', full.names = TRUE)
gmt_files
kegg_legacy <- prepare_gmt(gmt_files[1], unique(gene_logHRs$Gene), savefile = FALSE)
reactome <- prepare_gmt(gmt_files[3], unique(gene_logHRs$Gene), savefile = FALSE)
go <- prepare_gmt(gmt_files[4], unique(gene_logHRs$Gene), savefile = FALSE)

# rank genes by t-logHRs in DiD
rankings_dHep <- full_results$LogHR
names(rankings_dHep) <- full_results$Gene
rankings_dHep <- sort(rankings_dHep, decreasing = TRUE)
plot(rankings_dHep)
min(rankings_dHep)
max(rankings_dHep)


set.seed(1234)

kegg_legacy_res_dHep <- fgsea(pathways = kegg_legacy,
                              stats = rankings_dHep,
                              # scoreType = 'std', 
                              scoreType = 'pos', 
                              minSize = 3,
                              maxSize = 500,
                              nproc = 1)

reactome_res_dHep <- fgsea(pathways = reactome,
                           stats = rankings_dHep,
                           scoreType = 'std', 
                           minSize = 3,
                           maxSize = 500,
                           nproc = 1)
go_res_dHep <- fgsea(pathways = go,
                     stats = rankings_dHep,
                     scoreType = 'std', 
                     minSize = 5,
                     maxSize = 500,
                     nproc = 1)

dHep_fgsea <- data.frame("KEGG Legacy"=c(sum(kegg_legacy_res_dHep[, padj < 0.05]), sum(kegg_legacy_res_dHep[, pval < 0.05])),
                         "REACTOME"=c(sum(reactome_res_dHep[, padj < 0.05]), sum(reactome_res_dHep[, pval < 0.05])),
                         "GO"=c(sum(go_res_dHep[, padj < 0.05]), sum(go_res_dHep[, pval < 0.05])))
rownames(dHep_fgsea) <- c("adj.pval", "p.val")
dHep_fgsea

##### KEGG Legacy
plot_fgsea_transpose(kegg_legacy_res_dHep, title = "dHep Unadjusted Top 30 KEGG", xnudge = 0.03)

ggsave(fs::path(dir.results,"dHep_top30_kegg_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### REACTOME
plot_fgsea_transpose(reactome_res_dHep, title = "dHep Unadjusted Top 30 REACTOME", xlimit = 5)

ggsave(fs::path(dir.results,"dHep_top30_reactome_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)

##### GO
plot_fgsea_transpose(go_res_dHep, title = "dHep Unadjusted Top 30 GO", xlimit = 8)

ggsave(fs::path(dir.results,"dHep_top30_go_pathways_unadjusted_reml.jpeg"),
       width = 13, height = 10, scale = 1)


#Visualize pathway results
```
###ii. Old
```{r}
#Load in AST and ALT in today 
today <- read.csv(fs::path(dir.dat,"TODAY subaward","Clinical data","TODAY","CBL.csv"))
# length(unique(today$releaseid)) - 699 ids
# today_M0 <- today %>%
#   filter(mvisit=="M00")
#dplyr::select(c("releaseid","ALT","AST"))
#Load in Proteomics in Today
prot <- readRDS(fs::path(dir.dat,"Liver project","TODAY_proteomics.rds"))
prot <- prot[c(7326,7404,7397,which(grepl("seq.",colnames(prot))))] #length(unique(prot$releaseid)) - 376 unique ids
proteins <- colnames(prot)[which(grepl("seq.",colnames(prot)))]

#rm(today,today_M0,prot)

# #Determine which proteins correlate with genes associated with alt and ast
# #AST
# ast_genes <- read.xlsx(fs::path(dir.dat,"Liver project","Results","Bulk_Results_for_Ast Di (High vs. Low).xlsx"))
# ast_genes_pos <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC>0) %>%
#   arrange(desc(avg_log2FC)) %>%
#   dplyr::slice(1:200) 
# ast_genes_pos <- ast_genes_pos$Gene
# ast_genes_neg <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC<0) %>%
#   arrange(avg_log2FC) %>%
#   dplyr::slice(1:200) 
# ast_genes_neg <- ast_genes_neg$Gene
# ast_genes <- c(ast_genes_pos,ast_genes_neg)

#Key Protein Gene
key <- readRDS(fs::path(dir.dat,"Liver project","gene_prot.RDS"))
sens_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% sens_genes)]
prot_sens <- prot[c("releaseid","AGEBASE","sex",sens_proteins)]
data <- tidylog::right_join(today,prot_sens,by="releaseid")
data <- data %>% #filter out recruitment visits
  filter(mvisit!="R")

# Compute baseline ALT for each participant
data <- data %>%
  group_by(releaseid) %>%
  mutate(baseline_alt = first(ALT)) %>% 
  mutate(baseline_ast = first(AST)) %>% 
  ungroup()

#Check if ALT has doubled at each visit
data <- data %>%
  mutate(event_alt = ifelse(ALT >= 2 * baseline_alt, 1, 0)) %>% 
  mutate(event_ast = ifelse(AST >= 2 * baseline_ast, 1, 0))

# Ensure the event persists after it occurs
data <- data %>%
  group_by(releaseid) %>%
  mutate(event_alt = cummax(event_alt)) %>%  # Marks 1 for all visits after doubling occurs
  mutate(event_ast = cummax(event_ast)) %>% 
  ungroup() %>% 
  mutate(days_to_alt = ifelse(event_alt==1,days,))

#Censor individuals who didnt develop event


#Create loop for cox proportional hazards
#ALT
results_alt <- data.frame()
for (protein in sens_proteins){
  # formula <- as.formula(paste0("Surv(days, event_alt) ~ ", protein))
  formula <- as.formula(paste0("Surv(days, event_alt) ~ ", protein,"+AGEBASE+sex+HbA1c"))
  # Fit the Cox model
  cox_model <- coxph(formula, data = data)
  
  # Store the summary or model output
  LogHR <- summary(cox_model)$coef[1,1]
  Risk <- ifelse(LogHR>0,"Increased",
                 ifelse(LogHR<0,"Decreased","No Effect"))
  Pvalue <- summary(cox_model)$coef[1,5]
  results_data_alt <- data.frame(Protein=protein,Outcome="ALT",LogHazardRatio=LogHR,Risk=Risk,Pvalue=Pvalue)
  results_alt <- rbind(results_alt,results_data_alt)
}
rm(results_data_alt)
results_alt <- results_alt %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr")) %>% 
  mutate(sig=case_when(fdr<0.05 & Risk=="Increased" ~ "Significant Positive",
                       fdr<0.05 & Risk=="Decreased" ~"Significant Negative",
                       fdr>0.05~"Not Significant"))

#AST
results_ast <- data.frame()
for (protein in sens_proteins){
  # formula <- as.formula(paste0("Surv(days, event_ast) ~ ", protein))
  formula <- as.formula(paste0("Surv(days, event_ast) ~ ", protein,"+AGEBASE+sex+HbA1c"))
  # Fit the Cox model
  cox_model <- coxph(formula, data = data)
  
  # Store the summary or model output
  LogHR <- summary(cox_model)$coef[1,1]
  Risk <- ifelse(LogHR>0,"Increased",
                 ifelse(LogHR<0,"Decreased","No Effect"))
  Pvalue <- summary(cox_model)$coef[1,5]
  results_data_ast <- data.frame(Protein=protein,Outcome="AST",LogHazardRatio=LogHR,Risk=Risk,Pvalue=Pvalue)
  results_ast <- rbind(results_ast,results_data_ast)
}
rm(results_data_ast)
results_ast <- results_ast %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr")) %>% 
  mutate(sig=case_when(fdr<0.05 & Risk=="Increased" ~ "Significant Positive",
                       fdr<0.05 & Risk=="Decreased" ~"Significant Negative",
                       fdr>0.05~"Not Significant"))

#Combine Results
results <- rbind(results_ast,results_alt)
results <- results %>% 
  dplyr::rename(AptName=Protein)
results <- tidylog::left_join(results,key,by="AptName")
results <- results %>% 
  mutate(Gene_Name=paste0(EntrezGeneSymbol,"_",SeqId))
results_alt <- results %>% 
  filter(Outcome=="ALT") 
results_ast <- results %>% 
  filter(Outcome=="AST") 

#Save results
# write.csv(results_alt,fs::path(dir.results,"ALT_cox_results_unadjusted.csv"))
# write.csv(results_ast,fs::path(dir.results,"AST_cox_results_unadjusted.csv"))
write.csv(results_alt,fs::path(dir.results,"ALT_cox_results_adjusted.csv"))
write.csv(results_ast,fs::path(dir.results,"AST_cox_results_adjusted.csv"))

# Create the volcano plot
volcano_plot_alt <- ggplot(results_alt, aes(x = LogHazardRatio, y = -log10(fdr), color = sig)) +
  geom_point(alpha = 0.8, size = 2) +  # Points with transparency
  # scale_color_manual(values = c("grey", "red")) +  # Custom colors for significance
  scale_color_manual(values = c("Not Significant" = "grey", 
                                "Significant Positive" = "red", 
                                "Significant Negative" = "blue")) +
  theme_minimal() +  # Clean theme
  labs(
    title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of ALT (U/L), adjusted for Age, Sex, and HbA1c",
    # title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of ALT (U/L)",
    x = "Log Hazard Ratio",
    y = "-log10(FDR)"
  ) +
  theme(
    legend.position = "top",
    legend.title = element_blank()
  )

# Highlight top significant genes (optional)
top_genes_alt <- subset(results_alt, sig== "Significant Positive" | sig=="Significant Negative")
if (nrow(top_genes_alt) > 0) {
  volcano_plot_alt <- volcano_plot_alt +
    # geom_text(data = top_genes_alt, aes(label = EntrezGeneSymbol), 
    #           vjust = 1.5, size = 5, color = "black")
    geom_text_repel(data = top_genes_alt, aes(label = Gene_Name), 
                    box.padding = 0.35, 
                    point.padding = 0.5, 
                    max.overlaps = Inf,  # Prevent overlap
                    size = 3, 
                    color = "black", 
                    segment.size = 0.5)
}
plot(volcano_plot_alt)

# Create the volcano plot
volcano_plot_ast <- ggplot(results_ast, aes(x = LogHazardRatio, y = -log10(fdr), color = sig)) +
  geom_point(alpha = 0.8, size = 2) +  # Points with transparency
  # scale_color_manual(values = c("grey", "red")) +  # Custom colors for significance
  scale_color_manual(values = c("Not Significant" = "grey", 
                                "Significant Positive" = "red", 
                                "Significant Negative" = "blue")) +
  theme_minimal() +  # Clean theme
  labs(
    title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of AST (U/L), adjusted for Age, Sex, and HbA1c",
    # title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of AST (U/L)",
    x = "Log Hazard Ratio",
    y = "-log10(FDR)"
  ) +
  theme(
    legend.position = "top",
    legend.title = element_blank()
  )

# Highlight top significant genes (optional)
top_genes_ast <- subset(results_ast, sig== "Significant Positive" | sig=="Significant Negative")
if (nrow(top_genes_ast) > 0) {
  volcano_plot_ast <- volcano_plot_ast +
    # geom_text(data = top_genes_ast, aes(label = EntrezGeneSymbol), 
    #           vjust = 1.5, size = 5, color = "black")
    geom_text_repel(data = top_genes_ast, aes(label = Gene_Name), 
                    box.padding = 0.35, 
                    point.padding = 0.5, 
                    max.overlaps = Inf,  # Prevent overlap
                    size = 3, 
                    color = "black", 
                    segment.size = 0.5)
}
plot(volcano_plot_ast)

#Make barchart of significant with alt/ast & proteins


#Compare significant results from alt/ast to transcriptomics alt/ast 


```

#b. TEEN-LABs
###i. New
```{r}
#Proteomics teen labs 
load(fs::path(dir.dat,"Teen Labs","Data_Cleaned","analysis_dataset.RData"))
dat_baseline <- df %>% 
  filter(visit=="Month 1")
dat_followup <- df %>% 
  filter(visit=="Month 6") 
dat <- df %>% 
  filter(visit=="Month 1" | visit=="Month 6")

#Protein key
key <- analyte_info %>% 
  dplyr::select(all_of(c("AptName","SeqId","EntrezGeneSymbol")))

#Senescence Proteins
hvg_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% hvgs)]
prot <- df[c("ID","visit",hvg_proteins)]
protein <- colnames(prot)[-c(1,2)]
prot_baseline <- prot %>% 
  filter(visit=="Month 1")
prot_followup <- prot %>% 
  filter(visit=="Month 6")
#Name proteins baseline or followup 
colnames(prot_baseline)[-c(1,2)] <- paste0(colnames(prot_baseline)[-c(1,2)],"_baseline")
colnames(prot_followup)[-c(1,2)] <- paste0(colnames(prot_followup)[-c(1,2)],"_followup")
prot_baseline <- prot_baseline %>% 
  dplyr::select(-visit)
prot_baseline_dat <- prot_baseline
prot_followup <- prot_followup %>% 
  dplyr::select(-visit)

prot <- tidylog::left_join(prot_baseline,prot_followup,by="ID")
# Find columns with "baseline" and "followup"
baseline_cols <- grep("_baseline$", colnames(prot), value = TRUE)
followup_cols <- grep("_followup$", colnames(prot), value = TRUE)

# Sort to ensure matching order
baseline_cols <- sort(baseline_cols)
followup_cols <- sort(followup_cols)

# Create new columns for differences
for (i in seq_along(baseline_cols)) {
  protein_name <- sub("_baseline$", "", baseline_cols[i])  # Get protein name (e.g., "protein1")
  diff_col <- paste0(protein_name, "_diff")               # Create a difference column name
  prot[[diff_col]] <- prot[[followup_cols[i]]] - prot[[baseline_cols[i]]]
}

#Covariates at baseline
dat_cov <- dat %>% 
  filter(visit=="Month 1") %>% 
  dplyr::select(all_of(c("ID","SEX","HBA1C","age"))) 

rm(df,analyte_info)

#ALT/AST
liv <- read.csv("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Teen Labs/Data_Cleaned/Teen-LABS_AST_ALT.csv")
liv <- liv %>% 
  dplyr::rename(visit=Visit) %>% 
  filter(visit==1 | visit==6) %>% 
  mutate(visit=case_when(visit==1~"Month 1",
                         visit==6~"Month 6")) %>% 
  mutate(across(everything(),~ifelse(.==-5,NA,.))) %>% 
  mutate(across(everything(),~ifelse(.==-7,1,.)))
liv_baseline <- liv %>% 
  filter(visit=="Month 1") %>% 
  dplyr::select(-visit) %>% 
  dplyr::rename(ALT_baseline=ALT,
                AST_baseline=AST)
liv_followup <- liv %>% 
  filter(visit=="Month 6") %>% 
  dplyr::select(-visit) %>% 
  dplyr::rename(ALT_followup=ALT,
                AST_followup=AST)
liv <- tidylog::left_join(liv_baseline,liv_followup,by="ID")
liv <- liv %>% 
  mutate(ALT_diff = ALT_followup-ALT_baseline,
         AST_diff = AST_followup-AST_baseline) 


dat_cov$ID <- as.character(dat_cov$ID)
prot$ID<- as.character(prot$ID)

#Baseline proteins & baseline covariates and baseline and followup alt/ast
dat_all <- tidylog::right_join(dat_cov,liv,by=c("ID"))
dat_all <- tidylog::left_join(dat_all,prot,by=c("ID"))

#Run linear mixed effect model & difference model 
protein_diff <- paste0(protein,"_diff")
results <- data.frame()
covs <- c("age","SEX","HBA1C")
for (x in protein_diff) {
  for (y in c("AST_diff","ALT_diff")) {
    M0_crude <- as.formula(paste0(y,"~",x))
    M0_adj <- as.formula(paste0(y,"~",x,"+",paste(covs, collapse = "+")))
    M1_crude <- lm(M0_crude,data=dat_all)
    M1_adj <- lm(M0_adj,data=dat_all)
    coef_crude <- summary(M1_crude)$coef[2,1]
    coef_adj <- summary(M1_adj)$coef[2,1]
    p_crude <- summary(M1_crude)$coef[2,4]
    p_adj <- summary(M1_adj)$coef[2,4]
    diff_results <- data.frame(Protein=x,Enzyme=y,Coef=coef_adj,PValue=p_adj)
    results <- rbind(results,diff_results)
  }
}
r_alt <- results %>% 
  filter(Enzyme=="ALT_diff") %>% 
  mutate(fdr=p.adjust(PValue,method="fdr"))
r_ast <- results %>% 
  filter(Enzyme=="AST_diff") %>% 
  mutate(fdr=p.adjust(PValue,method="fdr"))
results <- rbind(r_alt,r_ast)
results <- results %>% 
  mutate(sig=ifelse(fdr<0.05,"*",""))
write.csv(results,fs::path(dir.results,"Diff_Results_Liver_Teenlabs_Imputed.csv"))

#Baseline protein with change in ALT/AST pre and post surgery
liv <- read.csv("/Users/hhampson/Library/CloudStorage/OneDrive-UW/Biostatistics Core Shared Drive/Teen Labs/Data_Cleaned/Teen-LABS_AST_ALT.csv")
liv <- liv %>% 
  dplyr::rename(visit=Visit) %>% 
  filter(visit==1 | visit==6) %>% 
  mutate(visit=case_when(visit==1~"Month 1",
                         visit==6~"Month 6")) %>% 
  mutate(across(everything(),~ifelse(.==-5,NA,.))) %>% 
  mutate(across(everything(),~ifelse(.==-7,NA,.)))
prot_baseline_dat$ID <- as.character(prot_baseline_dat$ID)
dat_baseline <- tidylog::left_join(liv,prot_baseline_dat,by=c("ID"))
dat_baseline <- tidylog::left_join(dat_cov,dat_baseline,by="ID")
dat_baseline <- dat_baseline %>% 
  mutate(time=case_when(visit=="Month 1"~1,
                        visit=="Month 6"~6))

#Linear Mixed Effects Model
protein_baseline <- paste0(protein,"_baseline")
covs <- c("age","SEX","HBA1C")

results <- data.frame()
for (x in c("AST","ALT")) {
  for (y in protein_baseline) {
    M0_crude <- as.formula(paste0(y,"~",x,"+time+(1|ID)"))
    M0_adj <- as.formula(paste0(y,"~",x,"*time+",paste(covs, collapse = "+"),"+(1|ID)"))
    M1_crude <- lmer(M0_crude,data=dat_baseline)
    M1_adj <- lmer(M0_adj,data=dat_baseline)
    Main <- summary(M1_adj)$coef[2,1]
    Main_p <- summary(M1_adj)$coef[2,5]
    Time <- summary(M1_adj)$coef[3,1]
    Time_p <- summary(M1_adj)$coef[3,5]
    Int <- summary(M1_adj)$coef[7,1]
    Int_p <- summary(M1_adj)$coef[7,5]
    results_int <- data.frame(Protein=x,Enzyme=y,Main,Main_p,Time,Time_p,Int,Int_p)
    results <- rbind(results,results_int)
  }
}
r_alt <- results %>% 
  filter(Enzyme=="ALT") %>% 
  mutate(Main_fdr=p.adjust(Main_p,method="fdr")) %>% 
  mutate(Int_fdr=p.adjust(Int_p,method="fdr"))
r_ast <- results %>% 
  filter(Enzyme=="AST") %>% 
  mutate(Main_fdr=p.adjust(Main_p,method="fdr")) %>% 
  mutate(Int_fdr=p.adjust(Int_p,method="fdr"))
results <- rbind(r_alt,r_ast)
results <- results %>% 
  mutate(Main_sig=ifelse(Main_fdr<0.05,"*","")) %>% 
  mutate(Int_sig=ifelse(Int_fdr<0.05,"*",""))
results$AptName <- str_remove(results$Protein,"_baseline")
results <- tidylog::left_join(results,key,by="AptName")
write.csv(results,fs::path(dir.results,"LMEM_TeenLabs_Liver_Results_Imputed.csv"))
```
###ii. Old
```{r}
#Proteomics teen labs 
load(fs::path(dir.dat,"Teen Labs","Data_Cleaned","analysis_dataset.RData"))
#Senescence Proteins
sens_proteins <- analyte_info$AptName[which(analyte_info$EntrezGeneSymbol %in% sens_genes)]

alt_ast <- read.csv(fs::path(dir.dat,"Liver Project","Teen-LABS_AST_ALT.csv")) %>% 
  filter(Visit==1|Visit==6) %>% 
  mutate(Visit=ifelse(Visit==1,"PreSurgery","PostSurgery"))
df_baseline <- df %>% 
  filter(visit=="Month 1") %>% 
  dplyr::rename(Visit=visit) %>% 
  mutate(Visit="PreSurgery") %>% 
  mutate(ID=as.character(ID)) %>% 
  dplyr::select(all_of(c("ID",sens_proteins)))
# df_followup <- df %>% 
#   filter(visit=="Month 6") %>% 
#   dplyr::rename(Visit=visit) %>% 
#   mutate(Visit="PostSurgery") %>% 
#   mutate(ID=as.character(ID)) %>% 
#   dplyr::select(all_of(c("ID","Visit",sens_proteins)))

#Prepare data
dat_baseline <- tidylog::left_join(alt_ast,df_baseline,by=c("ID"))
dat <- dat_baseline
rm(dat_baseline)
# dat_followup <- tidylog::left_join(dat,df_followup,by=c("ID"))
# dat_all <- tidylog::left_join(dat,df,by=c("ID","Visit"))

hist(dat$AST)
hist(dat$ALT)
#Log transform alt and ast
dat$ALT_log <- log(dat$ALT)
dat$AST_log <- log(dat$AST)
dat$Visit <- factor(dat$Visit)
dat$Visit <- relevel(dat$Visit, ref = "PreSurgery")

# #Make sure everyone has baseline proteins
# prot <- df %>% 
#   dplyr::select(c("ID","Visit",sens_proteins))

#Linear Mixed Effects Model
#ALT
plot_list <- list()
results <- data.frame()
for (protein in sens_proteins) {
  gene_name <- paste0(analyte_info$EntrezGeneSymbol[which(analyte_info$AptName %in% protein)],"_",analyte_info$SeqId[which(analyte_info$AptName %in% protein)])
  M0 <- as.formula(paste0("ALT_log ~ ",protein,"*Visit + (1 | ID)"))
  M1 <- lmer(M0, data = dat)
  main_effect <- summary(M1)$coef[2,1]
  main_p <- summary(M1)$coef[2,5]
  time_effect <- summary(M1)$coef[3,1]
  time_p <- summary(M1)$coef[3,5]
  int_effect <- summary(M1)$coef[4,1]
  int_p <- summary(M1)$coef[4,5]
  results_dat <- data.frame(Apt=gene_name,Outcome="ALT_log",MainEffect=main_effect,MainPVal=main_p,TimeEffect=time_effect,TimePVal=time_p,Interaction=int_effect,IntPVal=int_p)
  results <- rbind(results,results_dat)
  
  int_plot <- ggplot(dat, aes(x = !!sym(protein), y = ALT_log, color = Visit)) +
    geom_point() +
    geom_smooth(method = "lm", se = TRUE) +
    labs(title = paste0("Interaction between ",gene_name," and Visit on ALT_log"))+
    theme_minimal()
  
  # Create the spaghetti plot
  spaghetti_plot <- ggplot(dat, aes(x = Visit, y = ALT_log, group = ID, color = !!sym(protein))) +
    geom_line() +  # Connect ALT levels for each individual
    geom_point() +  # Add points for ALT levels at each time point
    # scale_color_viridis_d(guide = "none") +  # Optional: Viridis color palette, hide legend
    # scale_color_viridis_c(guide = "colorbar", name = "Protein") +
    scale_color_gradient(low = "blue", high = "red", name = gene_name) +
    # scale_color_gradient2(low = "blue", mid = "white", high = "red", midpoint = median(dat[[protein]],na.rm=T), name = "Protein") +
    labs(
      title = paste0("Spaghetti Plot of ALT Levels Pre- and Post-Surgery by ",gene_name),
      x = "Visit",
      y = "ALT Levels"
    ) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "gray85")) 
  
  # Combine the two plots
  combined_plot <- plot_grid(int_plot, spaghetti_plot, ncol = 1, labels = c("A", "B"))
  
  # Add to the list
  plot_list[[protein]] <- combined_plot
}

write.csv(results,fs::path(dir.results,paste0("Interaction between ",gene_name," and Visit on ALT_log.csv")))
# # Save all combined plots to a single PDF
# pdf("combined_interaction_and_spaghetti_plots.pdf", width = 10, height = 14)
# for (combined_plot in plot_list) {
#   plot(combined_plot)
# }
# dev.off()

#AST
plot_list <- list()
results <- data.frame()
for (protein in sens_proteins) {
  gene_name <- paste0(analyte_info$EntrezGeneSymbol[which(analyte_info$AptName %in% protein)],"_",analyte_info$SeqId[which(analyte_info$AptName %in% protein)])
  M0 <- as.formula(paste0("AST_log ~ ",protein,"*Visit + (1 | ID)"))
  M1 <- lmer(M0, data = dat)
  main_effect <- summary(M1)$coef[2,1]
  main_p <- summary(M1)$coef[2,5]
  time_effect <- summary(M1)$coef[3,1]
  time_p <- summary(M1)$coef[3,5]
  int_effect <- summary(M1)$coef[4,1]
  int_p <- summary(M1)$coef[4,5]
  results_dat <- data.frame(Apt=gene_name,Outcome="AST_log",MainEffect=main_effect,MainPVal=main_p,TimeEffect=time_effect,TimePVal=time_p,Interaction=int_effect,IntPVal=int_p)
  results <- rbind(results,results_dat)
  
  int_plot <- ggplot(dat, aes(x = !!sym(protein), y = ALT_log, color = Visit)) +
    geom_point() +
    geom_smooth(method = "lm", se = TRUE) +
    labs(title = paste0("Interaction between ",gene_name," and Visit on AST_log"))+
    theme_minimal()
  
  # Create the spaghetti plot
  spaghetti_plot <- ggplot(dat, aes(x = Visit, y = ALT_log, group = ID, color = !!sym(protein))) +
    geom_line() +  # Connect ALT levels for each individual
    geom_point() +  # Add points for ALT levels at each time point
    # scale_color_viridis_d(guide = "none") +  # Optional: Viridis color palette, hide legend
    # scale_color_viridis_c(guide = "colorbar", name = "Protein") +
    scale_color_gradient(low = "blue", high = "red", name = gene_name) +
    # scale_color_gradient2(low = "blue", mid = "white", high = "red", midpoint = median(dat[[protein]],na.rm=T), name = "Protein") +
    labs(
      title = paste0("Spaghetti Plot of AST Levels Pre- and Post-Surgery by ",gene_name),
      x = "Visit",
      y = "ALT Levels"
    ) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "gray85")) 
  
  # Combine the two plots
  combined_plot <- plot_grid(int_plot, spaghetti_plot, ncol = 1, labels = c("A", "B"))
  
  # Add to the list
  plot_list[[protein]] <- combined_plot
}

write.csv(results,fs::path(dir.results,paste0("Interaction between ",gene_name," and Visit on AST_log.csv")))
# # Save all combined plots to a single PDF
# pdf("combined_interaction_and_spaghetti_plots.pdf", width = 10, height = 14)
# for (combined_plot in plot_list) {
#   plot(combined_plot)
# }
# dev.off()
```


#7. Overlap Analysis & Interpretation
```{r}

Hep1 <- readRDS(fs::path(dir.results,"Hep_1_Steatosis_Reactome.rds"))
Hep2 <- readRDS(fs::path(dir.results,"Hep_2_Steatosis_Reactome.rds"))
Hep3 <- readRDS(fs::path(dir.results,"Hep_3_Steatosis_Reactome.rds"))
Hep4 <- readRDS(fs::path(dir.results,"Hep_4_Steatosis_Reactome.rds"))
Hep5 <- readRDS(fs::path(dir.results,"Hep_5_Steatosis_Reactome.rds"))
dHep <- readRDS(fs::path(dir.results,"dHep_Steatosis_Reactome.rds"))

# Create a named list of the 6 cell type data frames
hep_list <- list(
  dHep = dHep,
  Hep1 = Hep1,
  Hep2 = Hep2,
  Hep3 = Hep3,
  Hep4 = Hep4,
  Hep5 = Hep5
)

# Run comparison
overlap_results <- compare_pathways(ast, hep_list)

# View the result
print(overlap_results)


```

