---
title: "Liver Analysis"
author: "Hailey Hampson"
date: "2024-09-24"
output: html_document
--- 

#1. Set up Libraries & Directores
```{r libraries, echo=F, include = F}
library(reprex)
library(tidyverse)
library(BiocManager)        
library(arsenal)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(future)
library(colorspace)
library(patchwork)
library(ggdendro)
library(cowplot)
library(ggpubr)
library(venn)
library(rstatix)
library(table1)
library(Biobase)
library(ReactomeGSA)
library(GSEABase)
library(msigdbr)
library(kableExtra)
library(knitr)
library(SingleCellExperiment)
library(fgsea)
library(EnhancedVolcano)
library(openxlsx)
library(BiocManager)
library(MAST)
library(ggrepel)
# library(qpcR)
library(ggpubr)
library(openxlsx)
library(ggplot2)
library(GGally)
library(GSEABase)
library(limma)
library(reshape2)
library(data.table)
library(knitr)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(stringr)
library(NMF)
library(rsvd)
library(RColorBrewer)
library(MAST)
library(devtools)
# install_github("Sun-lab/ideas",force=T)
library(ideas)
library(foreach)
library(doRNG)
library(doParallel)
library(fs)
registerDoParallel(cores = 6)
library(VennDiagram)
#install.packages("survival")
library(survival)
#install.packages("lme4")  # If not already installed
library(lme4)
#install.packages("lmerTest")
library(lmerTest)



#Local file path
# dir.dat <- c("/Volumes/Peds Endo/Petter Bjornstad")
# dir.dat2 <- c("/Volumes/Peds Endo/Petter Bjornstad/scRNA/data_clean")
# dir.code <- c("/Users/hhampson/Documents/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
# dir.results <- c("/Users/hhampson/Documents/UW/1_Ongoing Projects/Liver scRNAseq/2_Results")

#Lambda file path
dir.dat <- c("/run/user/1026/gvfs/smb-share:server=ucdenver.pvt,share=som/PEDS/RI Biostatistics Core/Shared/Shared Projects/Laura/Peds Endo/Petter Bjornstad")
dir.code <- c("/home/Github_Repo/CHCO-Code/Petter Bjornstad/Liver analysis/Liver scRNAseq")
dir.results <- c(fs::path(dir.dat,"Liver project/Results"))

#Load functions
source("Liver_functions.R")

```

#2. Load Full Data & Clean: snRNA & MetaData
```{r echo = F}
# Liver snRNA data processing
gc()
#Local
# so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))
#Lambda
so_liver_sn <- readRDS(fs::path(dir.dat,"scRNA","data_raw","NoRef_PetterLiver_ClinData_Labels_Char_041924.RDS"))

gc()
#Local
# meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
#Lambda
meta_liver_raw <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
unique(meta_liver_raw$Kit_Lot)
meta_liver_raw <- meta_liver_raw %>%
  filter(StudyID!="IT2D-18") %>%
  filter(StudyID!="IT2D-16")
meta_liver_sn <-  so_liver_sn@meta.data[,1:11] %>%
  dplyr::mutate(RNAlater_ID = SampleID) %>%
  left_join(meta_liver_raw)

rownames(meta_liver_sn) <- rownames(so_liver_sn@meta.data)
so_liver_sn <- AddMetaData(so_liver_sn, meta_liver_sn)
rm(meta_liver_sn,meta_liver_raw)

#Switch default assay in seurat object to RNA
DefaultAssay(so_liver_sn) <- "RNA"
gc()

#Create liver disease and drug groups
so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
  mutate(both=ifelse(diagnosis_of_diabetes=="Yes" & glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(diagnosis_of_diabetes=="Yes" & sglt2=="No" & glp1agonist=="No","Yes","No")) %>%
  mutate(hc=ifelse(diagnosis_of_diabetes=="No","Yes","No"))

# Create a single grouping variable
so_liver_sn@meta.data <- so_liver_sn@meta.data %>% 
  mutate(group2=case_when(both == "Yes" ~ "both",
                          sglt2_exclusive == "Yes" ~ "sglt2_exclusive",
                          glp1_exclusive == "Yes" ~ "glp1_exclusive",
                          neither == "Yes" ~ "neither",
                          hc == "Yes" ~ "obese_control"))

so_liver_sn@meta.data$group2 <- factor(so_liver_sn@meta.data$group2, levels = c("obese_control","neither", "sglt2_exclusive", "glp1_exclusive","both"))

# #Create Diabetes only seurat object
# so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")

#Create senesence so in all cell types
sens_gene_in_data <- intersect(sens_genes, rownames(so_liver_sn))
# Subset the Seurat object to include only genes in sens_gene_in_data
so_sens_all <- subset(so_liver_sn, features = sens_gene_in_data)

# #Filter to hepatocytes only
# so_liver_sn$Hepatocyte <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Yes","No")
# so_hep <- subset(so_liver_sn, Hepatocyte=="Yes")
# rm(so_liver_sn)
# DefaultAssay(so_hep) <- "RNA"  

```

#3. Descriptive Statistics
```{r echo = F}
#Define disease and drug groups
dat <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
dat <- dat %>% 
  mutate(both=ifelse(glp1agonist=="Yes" & sglt2=="Yes","Yes","No")) %>% 
  mutate(sglt2_exclusive=ifelse(sglt2=="Yes" & glp1agonist=="No","Yes","No")) %>% 
  mutate(glp1_exclusive=ifelse(sglt2=="No" & glp1agonist=="Yes","Yes","No")) %>% 
  mutate(neither=ifelse(sglt2=="No" & glp1agonist=="No","Yes","No")) 

table(dat2$sglt2_exclusive) #-/+ #5 No, 0 yes
table(dat2$glp1_exclusive) #+/- # 2 Yes, 3 No
table(dat2$both) #+/+ #1 Both, 4 not both
table(dat2$neither) #-/- #3 neither, 2 on someting

#Create table 1
dat <- read.csv(fs::path(dir.dat,"scRNA","data_clean","liver_biopsy_metadata_PN.csv"))
dat <- dat %>% 
  filter(diagnosis_of_diabetes=="Yes")
cor(dat$ast,dat$alt,method="pearson")
cor(dat$ast,dat$alt,method="spearman")
cor(dat$alt,dat$tg,method="spearman")

dat$diagnosis_of_diabetes <- factor(dat$diagnosis_of_diabetes, levels=c("Yes","No"), labels=c("Type 2 Diabetes", "Obese Controls"))
dat$nih_sex     <- factor(dat$nih_sex, levels=c("Male", "Female"), labels=c("Male", "Female"))
dat$nih_ethnicity   <- factor(dat$nih_ethnicity, levels=c("Hispanic_Or_Latino","NonHispanic"), labels=c("Hispanic or Latino","Non-Hispanic/Non-Latino"))
dat$nih_race   <- factor(dat$nih_race, levels=c("White","BlackAfAm","Multiracial","Other"),
                         labels=c("White","Black","Multirace","Other"))
dat$steatosis_grade <- as.factor(dat$steatosis_grade)
dat$fibrosis_stage <- as.factor(dat$fibrosis_stage)
dat$lobular_inflammation_percent <- as.factor(dat$lobular_inflammation_percent)

label(dat$age)      <- "Age (y)"
label(dat$sglt2_exclusive) <- "SGLT2 Inhibitors (Yes/No)"
label(dat$glp1agonist) <- "GLP-1 Receptor Agonists (Yes/No)"
label(dat$nih_sex)      <- "Sex"
label(dat$nih_race)    <- "Race"
label(dat$nih_ethnicity)    <- "Ethnicity"
label(dat$diagnosis_of_diabetes)  <- "Diabetes Status"
label(dat$bmi)   <- "Body Mass Index (kg/m2)"
label(dat$diagnosis_of_MASLD)  <- "MASLD Status"
label(dat$sbp)     <- "Systolic Blood Pressure (mmHg)"
label(dat$dbp) <- "Diastolic Blood Pressure (mmHg)"
label(dat$tg) <-  "Triglycerides (mg/dL)"
label(dat$creatinine) <-  "Creatinine (mg/dL)"
label(dat$steatosis_percent) <- "Steatosis Percent (%)"
label(dat$fibrosis_stage) <- "Fibrosis Stage"
label(dat$lobular_inflammation_percent) <- "Lobular Inflammation Percent (%)"
label(dat$sglt2_exclusive) <- "Exclusive SGLT2 Inhibitors (Yes/No)"
label(dat$glp1_exclusive) <- "Exclusive GLP-1 Receptor Agonists (Yes/No)"
label(dat$both) <- "SLGT2i + GLP-1ra (Yes/No)"
label(dat$neither) <- "No Medication (Yes/No)"

table1(~ age + nih_sex + nih_race + nih_ethnicity + bmi + diagnosis_of_MASLD + sbp + dbp + tg + creatinine + sglt2_exclusive + glp1_exclusive + both | diagnosis_of_diabetes, data=dat)

table1(~ age + nih_sex + nih_race + nih_ethnicity + bmi + diagnosis_of_MASLD + sbp + dbp + tg + creatinine + sglt2_exclusive + glp1_exclusive + both, data=dat)

table1(~ age + nih_sex + nih_race + diagnosis_of_MASLD + bmi+ tg + ast + alt + ggt + steatosis_percent + lobular_inflammation_percent + fibrosis_stage | diagnosis_of_diabetes, data = dat)

table1(~ diagnosis_of_MASLD + diagnosis_of_diabetes + a1c  + bmi+ tg + ast + alt + ggt + steatosis_percent + lobular_inflammation_percent + fibrosis_stage + sglt2 | glp1agonist, data = dat)


pvalue <- function(x,...) {
  # Construct vectors of data y, and groups (strata) g
  y <- unlist(x)
  g <- factor(rep(1:length(x), times=sapply(x, length)))
  if (is.numeric(y)) {
    # For numeric variables, perform a standard 2-sample t-test
    p <- t.test(y ~ g)$p.value
  } else {
    # For categorical variables, perform a chi-squared test of independence
    p <- chisq.test(table(y, g))$p.value
  }
  # Format the p-value, using an HTML entity for the less-than sign.
  # The initial empty string places the output on the line below the variable label.
  c("", sub("<", "&lt;", format.pval(p, digits=3, eps=0.001)))
}

mean_sd <- function(x) {
  # Handle missing values explicitly
  s <- stats::sd(x, na.rm=TRUE)
  m <- mean(x, na.rm=TRUE)
  sprintf("%.2f (%.2f)", m, s)
}

table1(~ age + nih_sex + nih_race + bmi + diagnosis_of_MASLD + sbp + dbp + tg | diagnosis_of_diabetes, data=dat,overall=F, extra.col=list(`P-value`=pvalue), render.continuous = mean_sd)

table1(~ age + nih_sex + nih_race + bmi + diagnosis_of_MASLD + sbp + dbp + tg + diagnosis_of_diabetes | glp1agonist, data=dat,overall=F, extra.col=list(`P-value`=pvalue), render.continuous = mean_sd)

table1(~ age + nih_sex + nih_race + bmi + diagnosis_of_MASLD + sbp + dbp + tg | diagnosis_of_diabetes, data=dat,render.continuous = render.continuous.custom)

table1(~ age + nih_sex + nih_race + diagnosis_of_MASLD + bmi+ tg + ast + alt + ggt + steatosis_percent + lobular_inflammation_percent + fibrosis_stage | diagnosis_of_diabetes, data = dat,render.continuous = render.continuous.custom)

# Generate the table
table1(~ age + nih_sex + nih_race + diagnosis_of_MASLD + bmi+ tg + ast + alt + ggt + steatosis_percent + lobular_inflammation_percent + fibrosis_stage | diagnosis_of_diabetes, 
       data = dat, 
       overall = , 
       extra.col = list(`P-value` = pvalue), 
       render.continuous = render.continuous.custom)
#bmi,masld,TGs,AST/ALT (Liver enzymes),ggt, steatosis grade and or percentages,lobular inflam grade, fibrosis stage

# Descriptive stats
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_diabetes", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

## MASLD Y/N
liver_meta_raw_sc <- meta_liver_raw %>%
  filter(Cryostor_ID!="")
form <- paste("diagnosis_of_MASLD", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

## GLP-1RA Y/N
liver_meta_raw_sc <- liver_meta_raw %>%
  filter(Cryostor_ID!="")
form <- paste("glp1agonist", paste(colnames(liver_meta_raw_sc)[6:36], collapse = " + "), sep = " ~ ")

summary(arsenal::tableby(formula = as.formula(form), data = liver_meta_raw_sc, test = F))

```

#4. DEGS & GSEA
##a. All Cell Types
```{r}
#Among Type 2: No meds vs. GLP-1 Exclusive
##SGLT2i only -/+ #5 No, 0 yes
#GLP-1 only +/- # 2 Yes, 3 No
#Both +/+ #1 Both, 4 not both
#Neither-/- #3 neither, 2 on someting

all_genes <- rownames(so_liver_sn)
#Diabetes yes vs. no
degs_fxn(so=so_liver_sn,cell=NULL,gene_set=all_genes,exposure="diagnosis_of_diabetes",exp_group="Yes",ref_group="No",enrichment="Yes",top_gsea = 30)

#Create diab so
so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")
#GLP-1 vs. None
so_sub <- subset(so_diab,group2!="both")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="neither",enrichment="Yes",top_gsea = 30)
#GLP-1 vs. Both
so_sub <- subset(so_diab,group2!="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="glp1_exclusive",enrichment="Yes",top_gsea = 30)
#Both vs. None
so_sub <- subset(so_diab,group2!="glp1_exclusive")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="neither",enrichment="Yes",top_gsea = 30)

```

##b. Specific Cell Types
```{r}
#Major Cell Types
#unique(so_liver_sn$celltype)
so_liver_sn$celltype <- as.character(so_liver_sn$celltype)
so_liver_sn$celltype2 <- ifelse(grepl("Hep-",so_liver_sn$celltype),"Hepatocyte",
                                ifelse(grepl("Stellate-",so_liver_sn$celltype),"Stellate",
                                       ifelse(grepl("EC-",so_liver_sn$celltype),"EC",so_liver_sn$celltype)))
all_cell_types <- unique(so_liver_sn$celltype2)
all_genes <- rownames(so_liver_sn)
#Diabetes yes vs. no
for (cell_type in all_cell_types) {
  degs_fxn(so=so_liver_sn,cell=cell_type,gene_set=all_genes,exposure="diagnosis_of_diabetes",exp_group="Yes",ref_group="No",enrichment="Yes",top_gsea = 30)
}

#Create diab so 
so_diab <- subset(so_liver_sn,diagnosis_of_diabetes=="Yes")

#GLP-1 vs. None
so_sub <- subset(so_diab,group2!="both")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-6]
for (cell_type in all_cell_types) {
  degs_fxn(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="neither",enrichment="Yes",top_gsea = 30)
}
#GLP-1 vs. Both
so_sub <- subset(so_diab,group2!="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-6]
for (cell_type in all_cell_types) {
  degs_fxn(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="glp1_exclusive",enrichment="Yes",top_gsea = 30)
}
#Both vs. None
so_sub <- subset(so_diab,group2!="glp1_exclusive")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-11]
for (cell_type in all_cell_types) {
  degs_fxn(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="neither",enrichment="Yes",top_gsea = 30)
}
```
##c. Senescence Pseudobulk
```{r}
#Among Type 2: No meds vs. GLP-1 Exclusive
##SGLT2i only -/+ #5 No, 0 yes
#GLP-1 only +/- # 2 Yes, 3 No
#Both +/+ #1 Both, 4 not both
#Neither-/- #3 neither, 2 on someting

all_genes <- rownames(so_sens_all)
#Diabetes yes vs. no
degs_fxn_sens(so=so_sens_all,cell=NULL,gene_set=all_genes,exposure="diagnosis_of_diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)

#Create diab so
#so_diab <- subset(so_sens_all,diagnosis_of_diabetes=="Yes")
#None vs. OC 
so_sub <- subset(so_sens_all,group2=="neither" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="neither",ref_group="obese_control",enrichment="No",top_gsea = 30)

#GLP-1 vs. OC
so_sub <- subset(so_sens_all,group2=="glp1_exclusive" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="obese_control",enrichment="No",top_gsea = 30)

#Both vs. OC
so_sub <- subset(so_sens_all,group2=="both" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="obese_control",enrichment="No",top_gsea = 30)

#GLP-1 vs. None
so_sub <- subset(so_sens_all,group2=="glp1_exclusive" | group2=="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="neither",enrichment="No",top_gsea = 30)
#GLP-1 vs. Both
so_sub <- subset(so_sens_all,group2=="both" | group2=="glp1_exclusive")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="glp1_exclusive",enrichment="No",top_gsea = 30)
#Both vs. None
so_sub <- subset(so_sens_all,group2=="both" | group2=="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
degs_fxn_sens(so=so_sub,cell=NULL,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="neither",enrichment="No",top_gsea = 30)

```
##d. Senescence Specific Cells
```{r}
#Major Cell Types
#unique(so_sens_all$celltype)
so_sens_all$celltype <- as.character(so_sens_all$celltype)
so_sens_all$celltype2 <- ifelse(grepl("Hep-",so_sens_all$celltype),"Hepatocyte",
                                ifelse(grepl("Stellate-",so_sens_all$celltype),"Stellate",
                                       ifelse(grepl("EC-",so_sens_all$celltype),"EC",so_sens_all$celltype)))
all_cell_types <- unique(so_sens_all$celltype2)
all_genes <- rownames(so_sens_all)
#Diabetes yes vs. no
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sens_all,cell=cell_type,gene_set=all_genes,exposure="diagnosis_of_diabetes",exp_group="Yes",ref_group="No",enrichment="No",top_gsea = 30)
}

#Create diab so 
#so_diab <- subset(so_sens_all,diagnosis_of_diabetes=="Yes")
#None vs. OC
so_sub <- subset(so_sens_all,group2=="neither" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-11]
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="neither",ref_group="obese_control",enrichment="No",top_gsea = 30)
}

#GLP-1 vs. OC
so_sub <- subset(so_sens_all,group2=="glp1_exclusive" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="obese_control",enrichment="No",top_gsea = 30)
}
#Both vs. OC
so_sub <- subset(so_sens_all,group2=="both" | group2 == "obese_control")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-11]
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="obese_control",enrichment="No",top_gsea = 30)
}

#GLP-1 vs. None
so_sub <- subset(so_sens_all,group2=="glp1_exclusive" | group2=="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-6]
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="glp1_exclusive",ref_group="neither",enrichment="No",top_gsea = 30)
}
#GLP-1 vs. Both
so_sub <- subset(so_sens_all,group2=="both" | group2=="glp1_exclusive")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-6]
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="glp1_exclusive",enrichment="No",top_gsea = 30)
}
#Both vs. None
so_sub <- subset(so_sens_all,group2=="both" | group2=="neither")
so_sub$group2 <- factor(so_sub$group2)
#Define genes
all_genes <- rownames(so_sub)
all_cell_types <- unique(so_sub$celltype2)[-c(10,11)]
for (cell_type in all_cell_types) {
  degs_fxn_sens(so=so_sub,cell=cell_type,gene_set=all_genes,exposure="group2",exp_group="both",ref_group="neither",enrichment="No",top_gsea = 30)
}
```

#5. MAST Associations
##a. All Cell Types
```{r}
#Diabetes
mast_fxn_sens(so=so_sens_all,cell=NULL,exposure="diagnosis_of_diabetes",gene_set=rownames(so_sens_all),exp_group="Yes",ref_group="No")

#AST/ALT
so_diab <- subset(so_sens_all, diagnosis_of_diabetes=="Yes")
gene_set <- rownames(so_diab)

#AST
mast_fxn_sens(so=so_diab,cell=NULL,exposure="ast",covariate=NULL,gene_set=gene_set,exp_group=NULL,ref_group=NULL)

#ALT
mast_fxn_sens(so=so_diab,cell=NULL,exposure="alt",covariate=NULL,gene_set=gene_set,exp_group=NULL,ref_group=NULL)

#Medication Groups
#Neither vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="neither")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,exposure="Medication",covariate="alt",gene_set=gene_set,exp_group="neither",ref_group="obese_control")

#GLP-1 vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="glp1_exclusive")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,exposure="Medication",covariate="alt",gene_set=gene_set,exp_group="glp1_exclusive",ref_group="obese_control")

#Both vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="obese_control")

#GLP-1 vs. Neither
so_diab <- subset(so_sens_all,diagnosis_of_diabetes=="Yes")
so_subset <- subset(so_diab, group2=="neither"|group2=="glp1_exclusive")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="glp1_exclusive",ref_group="neither")

#Both vs. Neither
so_subset <- subset(so_diab, group2=="neither"|group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="neither")

#Both vs. GLP-1
so_subset <- subset(so_diab, group2=="glp1_exclusive"|group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
mast_fxn_sens(so=so_subset,cell=NULL,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="glp1_exclusive")

# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_diab, layer = "data"))

# Extract metadata
cell_metadata <- so_diab@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~ast+group2
exp <- paste0("ast")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle1 <- merge(summary_dt[contrast==exp & component=='H',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle1[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle1$component <- "FC_H"
# fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle2$component <- "C"
# fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle3$component <- "D"

# fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle1
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

# #Only plot C component if D component is significant - aka significant expression of gene
# sig_results <- result1 %>% 
#   filter(component=="D") %>% 
#   filter(fdr<0.05) %>% 
#   pull(Gene)
# result1 <- result1 %>% 
#   filter(component=="C") %>% 
#   filter(Gene %in% sig_results) %>% 
#   filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Mast_AST_LogFC_all_participants.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# if (!is.null(cell)) {
#   title <- paste0("Continuous Estimate for ",condition," among", cell," cells")
# } else {
#   title <- paste0("Continuous Estimate for ",condition)
# }
# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "LogFC", y = "Gene",title="Pairwise LogFC for AST among All Participants") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_LogFC_Ast_All_Participants.pdf"),width=10,height=10)
plot(p_C)
dev.off()

#ALT
so_diab <- subset(so_sens_all, diagnosis_of_diabetes=="Yes")
gene_set <- rownames(so_diab)
# exposure_vars <- c("ast","alt")
# for (exp in exposure_vars){
#   mast_fxn(so=so_sens_all,cell=NULL,exposure=exp,covariate=NULL,gene_set=rownames(so_sens_all),batch_size=length(rownames(so_sens_all)),exp_group=NULL,ref_group=NULL)
# }

# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_diab, layer = "data"))

# Extract metadata
cell_metadata <- so_diab@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~alt
exp <- paste0("alt")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle1 <- merge(summary_dt[contrast==exp & component=='H',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle1[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle1$component <- "FC_H"
# fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle2$component <- "C"
# fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle3$component <- "D"

# fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle1
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

# #Only plot C component if D component is significant - aka significant expression of gene
# sig_results <- result1 %>% 
#   filter(component=="D") %>% 
#   filter(fdr<0.05) %>% 
#   pull(Gene)
# result1 <- result1 %>% 
#   filter(component=="C") %>% 
#   filter(Gene %in% sig_results) %>% 
#   filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Mast_ALT_LogFC.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# if (!is.null(cell)) {
#   title <- paste0("Continuous Estimate for ",condition," among", cell," cells")
# } else {
#   title <- paste0("Continuous Estimate for ",condition)
# }
# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "LogFC", y = "Gene",title="Pairwise LogFC for ALT among YT2D") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_LogFC_ALT.pdf"),width=10,height=10)
plot(p_C)
dev.off()

so_diab <- subset(so_sens_all, diagnosis_of_diabetes=="Yes")
gene_set <- rownames(so_diab)
# exposure_vars <- c("ast","alt")
# for (exp in exposure_vars){
#   mast_fxn(so=so_sens_all,cell=NULL,exposure=exp,covariate=NULL,gene_set=rownames(so_sens_all),batch_size=length(rownames(so_sens_all)),exp_group=NULL,ref_group=NULL)
# }

# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_diab, layer = "data"))

# Extract metadata
cell_metadata <- so_diab@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~alt
exp <- paste0("alt")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Mast_ALT.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# if (!is.null(cell)) {
#   title <- paste0("Continuous Estimate for ",condition," among", cell," cells")
# } else {
#   title <- paste0("Continuous Estimate for ",condition)
# }
# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for ALT among YT2D") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_Alt.pdf"),width=10,height=10)
plot(p_C)
dev.off()

#AST
ast_prot_pos <- read.csv(fs::path(dir.results,"AST_cox_results_adjusted.csv")) %>% 
  filter(sig=="Significant Positive") %>% 
  pull(EntrezGeneSymbol)
ast_prot_neg <- read.csv(fs::path(dir.results,"AST_cox_results_adjusted.csv")) %>% 
  filter(sig=="Significant Negative")%>% 
  pull(EntrezGeneSymbol)

ast_mast_pos <- read.csv(fs::path(dir.results,"Ast.csv")) %>% 
  filter(component=="C") %>% 
  filter(LogFoldChange>0) %>% 
  pull(Gene)
ast_prot_pos[which(ast_prot_pos %in% ast_mast_pos)] #"ACVR1B" "NRG1"
ast_mast_pos[which(ast_mast_pos %in% ast_prot_pos)]

ast_mast_neg <- read.csv(fs::path(dir.results,"Ast.csv")) %>% 
  filter(component=="C") %>% 
  filter(LogFoldChange<0) %>% 
  pull(Gene)
ast_prot_neg[which(ast_prot_neg %in% ast_mast_neg)] "ANG"
ast_mast_neg[which(ast_mast_neg %in% ast_prot_neg)]

#ALT
alt_prot_pos <- read.csv(fs::path(dir.results,"ALT_cox_results_adjusted.csv")) %>% 
  filter(sig=="Significant Positive") %>% 
  pull(EntrezGeneSymbol)
alt_prot_neg <- read.csv(fs::path(dir.results,"ALT_cox_results_adjusted.csv")) %>% 
  filter(sig=="Significant Negative")%>% 
  pull(EntrezGeneSymbol)

alt_mast_pos <- read.csv(fs::path(dir.results,"Alt.csv")) %>% 
  filter(component=="C") %>% 
  filter(LogFoldChange>0) %>% 
  pull(Gene)
alt_prot_pos[which(alt_prot_pos %in% alt_mast_pos)] #"IL7"     "ACVR1B"  "ANGPTL4" "C3"      "NRG1" 
alt_mast_pos[which(alt_mast_pos %in% alt_prot_pos)]

alt_mast_neg <- read.csv(fs::path(dir.results,"Alt.csv")) %>% 
  filter(component=="C") %>% 
  filter(LogFoldChange<0) %>% 
  pull(Gene)
alt_prot_neg[which(alt_prot_neg %in% alt_mast_neg)] #"ANG"
alt_mast_neg[which(alt_mast_neg %in% alt_prot_neg)]#"ANG"  "LCP1"

#Medication
#Neither vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="neither")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set=rownames(so_subset)
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate="alt",gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")

# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication + ast
exp <- paste0("Medicationneither")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_Neither_OC_adjusted_ast.csv")))

#GLP-1 vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="glp1_exclusive")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate=NULL,gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")
# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication + ast
exp <- paste0("Medicationglp1_exclusive")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_GLP1_OC_adjusted_ast.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# if (!is.null(cell)) {
#   title <- paste0("Continuous Estimate for ",condition," among", cell," cells")
# } else {
#   title <- paste0("Continuous Estimate for ",condition)
# }
# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for GLP-1 vs. Obese Control") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_GLP1_OC_Adj_Ast.pdf"),width=10,height=10)
plot(p_C)
dev.off()

#Both vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate=NULL,gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")
# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication + ast
exp <- paste0("Medicationboth")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_Both_OC_adjusted_ast.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for GLP-1 + SLGT2 vs. Obese Control") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_Both_OC_Adj_Ast.pdf"),width=10,height=10)
plot(p_C)
dev.off()


#GLP-1 vs. None
so_diab <- subset(so_sens_all, diagnosis_of_diabetes=="Yes")
so_subset <- subset(so_diab, group2=="glp1_exclusive" | group2=="neither")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
# exposure_vars <- c("ast","alt")
# for (exp in exposure_vars){
#   mast_fxn(so=so_sens_all,cell=NULL,exposure=exp,covariate=NULL,gene_set=rownames(so_sens_all),batch_size=length(rownames(so_sens_all)),exp_group=NULL,ref_group=NULL)
# }

# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication + ast
exp <- paste0("Medicationglp1_exclusive")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle1 <- merge(summary_dt[contrast==exp & component=='H',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='logFC', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle1[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle1$component <- "FC_H"
# fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle2$component <- "C"
# fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
#                    summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
# fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
# fcHurdle3$component <- "D"

# fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle1
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

# #Only plot C component if D component is significant - aka significant expression of gene
# sig_results <- result1 %>% 
#   filter(component=="D") %>% 
#   filter(fdr<0.05) %>% 
#   pull(Gene)
# result1 <- result1 %>% 
#   filter(component=="C") %>% 
#   filter(Gene %in% sig_results) %>% 
#   filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Mast_GLP1_neither_LogFC.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  # filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# if (!is.null(cell)) {
#   title <- paste0("Continuous Estimate for ",condition," among", cell," cells")
# } else {
#   title <- paste0("Continuous Estimate for ",condition)
# }
# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "LogFC", y = "Gene",title="Pairwise LogFC for GLP-1 vs. No Medication among YT2D") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_LogFC_GLP1_Neither.pdf"),width=10,height=10)
plot(p_C)
dev.off()

so_subset <- subset(so_sens_all,group2=="glp1_exclusive" | group2=="neither")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
so_subset$ast <- factor(as.character(so_subset$ast))
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate=NULL,gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")
# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication 
exp <- paste0("Medicationglp1_exclusive")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_GLP1_neither_crude.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for GLP-1 vs. No Medication") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_GLP1_Neither.pdf"),width=10,height=10)
plot(p_C)
dev.off()

#Both vs. GLP-1
so_subset <- subset(so_sens_all,group2=="glp1_exclusive" | group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
so_subset$ast <- factor(as.character(so_subset$ast))
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate=NULL,gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")
# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication 
exp <- paste0("Medicationboth")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_both_GLP1_crude.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for GLP-1 + SGLT2 vs. GLP1") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_Both_GLP1.pdf"),width=10,height=10)
plot(p_C)
dev.off()

#Both vs. Neither
so_subset <- subset(so_sens_all,group2=="neither" | group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
so_subset$ast <- factor(as.character(so_subset$ast))
# mast_fxn(so=so_subset,cell=NULL,exposure="Medication",covariate=NULL,gene_set=rownames(so_subset),batch_size=length(rownames(so_subset)),exp_group="neither",ref_group="obese_control")
# Extract the expression data matrix from so (e.g., normalized counts)
expression_matrix <- as.matrix(GetAssayData(so_subset, layer = "data"))

# Extract metadata
cell_metadata <- so_subset@meta.data
#cell_metadata[[exposure]] <- factor(cell_metadata[[exposure]])

# Create SingleCellAssay object
sca <- FromMatrix(exprsArray = expression_matrix, cData = cell_metadata)

# Assuming gene_set is a vector of gene names
sca_gene_set <- sca[rownames(sca) %in% gene_set, ]

#Define the formula
# model_formula <- as.formula(paste0("~", exposure,"+",covariate))
model_formula <- ~Medication 
exp <- paste0("Medicationboth")

# Run the linear model with zlm
zlm_results <- zlm(formula = model_formula, sca = sca_gene_set)

# Summarize results and perform likelihood ratio test (LRT) for outcome
summary_zlm <- summary(zlm_results, doLRT = exp)

# Convert the summary to a data table for easy manipulation
summary_dt <- summary_zlm$datatable

#Overall with Hurdle and logFC
fcHurdle2 <- merge(summary_dt[contrast==exp & component=='C',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='C', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle2[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle2$component <- "C"
fcHurdle3 <- merge(summary_dt[contrast==exp & component=='D',.(primerid, `Pr(>Chisq)`)], #hurdle P values
                   summary_dt[contrast==exp & component=='D', .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients
fcHurdle3[,fdr:=p.adjust(`Pr(>Chisq)`, 'fdr')]
fcHurdle3$component <- "D"

fcHurdle <- rbind(fcHurdle2,fcHurdle3)

result1 <- fcHurdle
result1 <- result1 %>%
  filter(fdr<0.05) %>%
  dplyr::rename(LogFoldChange=coef) %>% 
  dplyr::rename(Gene=primerid)

#Only plot C component if D component is significant - aka significant expression of gene
sig_results <- result1 %>% 
  filter(component=="D") %>% 
  filter(fdr<0.05) %>% 
  pull(Gene)
result1 <- result1 %>% 
  filter(component=="C") %>% 
  filter(Gene %in% sig_results) %>% 
  filter(fdr<0.05)

write.csv(result1,fs::path(dir.results,paste0("Medication_both_neither_crude.csv")))

# Filter top 10 positive and negative log2FoldChange
top_pos_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange>0) %>% 
  arrange(desc(LogFoldChange)) 
# dplyr::slice(1:30) 
# dplyr::rename(Gene=primerid,
#               LogFC=coef)

top_neg_C <- as.data.frame(result1) %>%
  filter(component=="C") %>% 
  filter(fdr<0.05) %>%
  filter(LogFoldChange<0)  %>% 
  arrange(LogFoldChange)
# dplyr::slice(1:30) 

top_pos_C$Direction <- "Positive"
top_neg_C$Direction <- "Negative"

# Combine and prepare for plotting
top_genes_C <- bind_rows(top_pos_C, top_neg_C)

# Set sorting order: all negative genes first, then all positive genes
top_genes_C <- top_genes_C %>%
  mutate(Gene = factor(Gene, levels = Gene[order(Direction, LogFoldChange)]))  # Order by Direction and log2FC

# # Bar chart with flipped x and y axes
p_C <- ggplot(top_genes_C, aes(x = LogFoldChange, y = Gene, fill = Direction)) +
  geom_bar(stat = "identity") +
  labs(x = "Effect Estimate", y = "Gene",title="Continuous Estimate for GLP-1 + SGLT2 vs. No Medication") +
  scale_fill_manual(values = c("blue", "red"), labels = c("Lower Expression", "Higher Expression")) +
  theme_minimal()+
  theme(element_text(family="Times",size=20))+
  theme(legend.position="none")

pdf(fs::path(dir.results,"Mast_Barchart_Continuous_Estimate_Both_Neither.pdf"),width=10,height=10)
plot(p_C)
dev.off()
```

##b. Single Cell
```{r}
#Major Cell Types
#unique(so_sens_all$celltype)
so_sens_all$celltype <- as.character(so_sens_all$celltype)
so_sens_all$celltype2 <- ifelse(grepl("Hep-",so_sens_all$celltype),"Hepatocyte",
                                ifelse(grepl("Stellate-",so_sens_all$celltype),"Stellate",
                                       ifelse(grepl("EC-",so_sens_all$celltype),"EC",so_sens_all$celltype)))
all_cell_types <- unique(so_sens_all$celltype2)
gene_set <- rownames(so_sens_all)
#Diabetes
for (cell in all_cell_types){
  mast_fxn_sens(so=so_sens_all,cell=cell,exposure="diagnosis_of_diabetes",gene_set=gene_set,exp_group="Yes",ref_group="No")
}


#AST/ALT
so_diab <- subset(so_sens_all, diagnosis_of_diabetes=="Yes")
gene_set <- rownames(so_diab)

#AST
for (cell in all_cell_types){
  mast_fxn_sens(so=so_diab,cell=cell,exposure="ast",covariate=NULL,gene_set=gene_set,exp_group=NULL,ref_group=NULL)
} 

#ALT
for (cell in all_cell_types){
  mast_fxn_sens(so=so_diab,cell=cell,exposure="alt",covariate=NULL,gene_set=gene_set,exp_group=NULL,ref_group=NULL)
}

#Medication Groups
#Neither vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="neither")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
all_cell_types <- unique(so_sens_all$celltype2)[-11]
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="neither",ref_group="obese_control")
} 
#GLP-1 vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="glp1_exclusive")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="glp1_exclusive",ref_group="obese_control")
} 

#Both vs. OC
so_subset <- subset(so_sens_all,group2=="obese_control" | group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="obese_control")
}

#GLP-1 vs. Neither
# so_diab <- subset(so_sens_all,diagnosis_of_diabetes=="Yes")
# so_subset <- subset(so_diab, group2=="neither"|group2=="glp1_exclusive")
so_subset <- subset(so_sens_all, group2=="neither"|group2=="glp1_exclusive")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="glp1_exclusive",ref_group="neither")
}

#Both vs. Neither
# so_subset <- subset(so_diab, group2=="neither"|group2=="both")
so_subset <- subset(so_sens_all, group2=="neither"|group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="neither")
} 

#Both vs. GLP-1
# so_subset <- subset(so_diab, group2=="glp1_exclusive"|group2=="both")
so_subset <- subset(so_sens_all, group2=="glp1_exclusive"|group2=="both")
so_subset$Medication <- so_subset$group2
so_subset$Medication <- factor(so_subset$Medication)
gene_set <- rownames(so_subset)
for (cell in all_cell_types){
  mast_fxn_sens(so=so_subset,cell=cell,covariate="alt",exposure="Medication",gene_set=gene_set,exp_group="both",ref_group="glp1_exclusive")
}
```



#6. GSEA
```{r}
Idents(so_liver_sn) <- so_liver_sn$celltype
de.markers(so_liver_sn, genes, "diagnosis_of_diabetes", id2 = "No", id1 = "Yes", NULL, "_top")

#Gene set enrichment analysis
sce_sn_hep <- as.SingleCellExperiment(so_liver_sn_hep)
rm(so_liver_sn_hep)
## C2 category is according to canonical pathways: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4707969/pdf/nihms-743907.pdf
geneSets <- msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG")
### filter background to only include genes that we assessed
geneSets$gene_symbol <- toupper(geneSets$gene_symbol)
geneSets <- geneSets[geneSets$gene_symbol %in% names(sce_sn_hep),]
m_list <- geneSets %>% split(x = .$gene_symbol, f = .$gs_name)
stats <- m$p_val_adj
names(stats) <- rownames(m)
eaRes <- fgsea(pathways = m_list, stats = na.omit(stats))
#ooEA <- order(eaRes$pval, decreasing = FALSE)
#kable(head(eaRes[ooEA, 1:7], n = 20))
# Convert the leadingEdge column to comma-separated strings
eaRes$leadingEdge <- sapply(eaRes$leadingEdge, function(x) paste(x, collapse = ", "))
gc()
# Save to CSV
write.csv(eaRes, fs::path(dir.results,"GSEA_Diabetes_Hepatocytes.csv"))
gc()

#IPA Pathways 
test_ipa <- readxl::read_xls("/Users/choiyej/Dropbox/PANTHER/IPA results/tanner_stage_12_345_rh_de_table_export.xls", skip = 1)
colnames(test_ipa) <- c("pathway", "neglog_p", "ratio", "zscore", "molecules")

# Volcano plot
sig_cutoff = -log(0.05, base = 10)

test_ipa <- test_ipa %>%
  dplyr::mutate(direction = case_when(zscore < 0 & neglog_p > sig_cutoff ~ "Negative", 
                                      zscore > 0 & neglog_p > sig_cutoff ~ "Positive",
                                      zscore == 0 & neglog_p > sig_cutoff ~ "Unknown",
                                      T ~ "NS")) %>%
  group_by(direction) %>%
  dplyr::mutate(count = row_number(),
                label = case_when(direction != "NS" ~ paste0(str_sub(direction, 1,1), count))) %>%
  ungroup()

test_ipa$direction <- factor(test_ipa$direction, levels = c("Negative", "Positive", "Unknown", "NS"))

ipa.plot <- test_ipa %>% filter(neglog_p > 0) %>%
  ggplot(aes(x = zscore, y = neglog_p)) + 
  geom_point(aes(color = direction), alpha = 0.5, size = 2) +
  geom_hline(aes(yintercept = sig_cutoff), color = "red", linetype = "dashed") +
  labs(x = "Z-Score",
       y = "-log(p-value)",
       color = "Direction") + 
  scale_color_manual(values = c("#bf0603", "#457b9d",
                                "#f6bd60", "#dad7cd")) +
  ggrepel::geom_label_repel(aes(label = label, color = direction),
                            label.size = 0.15,
                            max.overlaps = 100, 
                            min.segment.length = 0.01) +
  theme_bw()

# Table
ipa.table.neg <- test_ipa %>%
  dplyr::filter(direction == "Negative") %>%
  dplyr::select(label, pathway) 
ipa.table.pos <- test_ipa %>%
  dplyr::filter(direction == "Positive") %>%
  dplyr::select(label, pathway) 
ipa.table.unk <- test_ipa %>%
  dplyr::filter(direction == "Unknown") %>%
  dplyr::select(label, pathway) 

cbind.fill <- function(...) {
  nm <- list(...) 
  nm <- lapply(nm, as.matrix)
  n <- max(sapply(nm, nrow)) 
  do.call(cbind, lapply(nm, function(x) {
    filled_matrix <- ifelse(is.na(x), "", x)
    rbind(filled_matrix, matrix("", n - nrow(filled_matrix), ncol(filled_matrix)))
  }))
}

ipa.table.comb <- cbind.fill(ipa.table.neg,ipa.table.pos,ipa.table.unk)
gg.ipa.table <- ggtexttable(ipa.table.comb, rows = NULL,
                            cols = c("Label", "Pathway", "Label", "Pathway","Label", "Pathway"),
                            theme = ttheme("blank",
                                           tbody.style = tbody_style(fill = "white", 
                                                                     size = 8.5, 
                                                                     hjust = 0,
                                                                     x = 0.1))) %>%
  tab_add_hline(at.row = 1:2, row.side = "top", linewidth = 0.8)



ggarrange(ipa.plot, gg.ipa.table,
          ncol = 2, heights = c(1,1), widths = c(0.5,1),
          legend = "top")


```


#7. Visualize Results
##a. All Cell Types
```{r}
exposure <- "Group2"
pdf(fs::path(dir.results,paste0("venn_results_bulk_",exposure,".pdf")),width=20,height=15)
visualize_function_bulk(exposure)
dev.off()
```

##b. Specific Cell Types
```{r}
cell_types <- c("Hepatocyte","EC","Stellate","Cholang","NKC/NKT","Kup/MON","dHep","Kup/MAC","B/Plasma","Hep/Immune")
cell_types <- str_replace_all(cell_types,"/","_")

#GLP-1 vs. Neither
visualize_cell(exposure = "Group2",exp_group = "glp1_exclusive",ref_group = "neither")

#GLP-1 + SGLT2 vs. Neither
visualize_cell(exposure = "Group2",exp_group = "both",ref_group = "neither")

#GLP-1 vs. Both
visualize_cell(exposure = "Group2",exp_group = "both",ref_group = "glp1_exclusive")





exposure <- "Group2"
cell_type <- c("Hepatocyte","EC","Stellate","Cholang","NKC/NKT","Kup/MON","dHep","Kup/MAC","B/Plasma","Hep/Immune")
cell_type <- str_replace_all(cell_type,"/","_")
#cFIB/cImmune
for (cell in cell_type){
  pdf(fs::path(dir.results,paste0("venn_results_",cell,"_",exposure,".pdf")),width=20,height=15)
  visualize_function(exposure,cell)
  dev.off()
}
```

#8. TODAY 
##a. New
```{r}
#Load in AST and ALT in today 
today <- read.csv(fs::path(dir.dat,"TODAY subaward","Clinical data","TODAY","CBL.csv")) # length(unique(today$releaseid)) - 699 ids
today <- today %>%
  filter(mvisit!="R")
#dplyr::select(c("releaseid","ALT","AST"))
#Load in Proteomics in Today
prot <- readRDS(fs::path(dir.dat,"Liver project","TODAY_proteomics.rds"))
prot <- prot[c(7326,7397,7404,which(grepl("seq.",colnames(prot))))] #length(unique(prot$releaseid)) - 376 unique ids
proteins <- colnames(prot)[which(grepl("seq.",colnames(prot)))]

#rm(today,today_M0,prot)

# #Determine which proteins correlate with genes associated with alt and ast
# #AST
# ast_genes <- read.xlsx(fs::path(dir.dat,"Liver project","Results","Bulk_Results_for_Ast Di (High vs. Low).xlsx"))
# ast_genes_pos <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC>0) %>%
#   arrange(desc(avg_log2FC)) %>%
#   dplyr::slice(1:200) 
# ast_genes_pos <- ast_genes_pos$Gene
# ast_genes_neg <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC<0) %>%
#   arrange(avg_log2FC) %>%
#   dplyr::slice(1:200) 
# ast_genes_neg <- ast_genes_neg$Gene
# ast_genes <- c(ast_genes_pos,ast_genes_neg)

#Key Protein Gene
key <- readRDS(fs::path(dir.dat,"Liver project","gene_prot.RDS"))
sens_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% sens_genes)]
prot <- prot[c("releaseid","AGEBASE","sex",sens_proteins)]

#Calculate sensoring and time to event
baseline <- today %>% 
  filter(mvisit=="M00") %>% 
  mutate(baseline_alt=ALT,
         baseline_ast=AST) %>% 
  dplyr::select(all_of(c("releaseid","baseline_alt","baseline_ast"))) 
dat <- tidylog::left_join(today,baseline,by="releaseid")
dat <- dat %>% 
  mutate(event_alt=ifelse(ALT>=2*(baseline_alt),1,0)) %>% 
  mutate(event_ast=ifelse(AST>=2*(baseline_ast),1,0)) 
#Calculate time to event & censoring
dat_alt <- dat %>% 
  group_by(releaseid) %>%
  summarize(
    event_status_alt = ifelse(any(event_alt == 1, na.rm = TRUE), 1, 0), # 1 if event occurred, 0 if censored
    time_to_event_alt = ifelse(
      any(event_alt == 1, na.rm = TRUE), 
      min(days[event_alt == 1], na.rm = TRUE), # Time of first event
      max(days, na.rm = TRUE)                 # Time of last observation (censoring)
    )) %>% 
  ungroup()
dat_ast <- dat %>% 
  group_by(releaseid) %>%
  summarize(
    event_status_ast = ifelse(any(event_ast == 1, na.rm = TRUE), 1, 0), # 1 if event occurred, 0 if censored
    time_to_event_ast = ifelse(
      any(event_ast == 1, na.rm = TRUE), 
      min(days[event_ast == 1], na.rm = TRUE), # Time of first event
      max(days, na.rm = TRUE)                 # Time of last observation (censoring)
    )) %>% 
  ungroup()

dat_covs <- today %>% 
  filter(mvisit=="M00") %>% 
  dplyr::select(all_of(c("releaseid","HbA1c")))
#Merge proteins in 
dat <- tidylog::right_join(dat_covs,prot,by="releaseid")
dat <- tidylog::right_join(dat_alt,dat,by="releaseid")
dat <- tidylog::right_join(dat_ast,dat,by="releaseid")
rm(baseline,dat_alt,dat_ast,dat_covs,prot,today)

#Create loop for linear regression between proteins & liver enzymes
results <- data.frame()
for(y in c("alt","ast")) {
  for (x in sens_proteins) {
    M0 <- as.formula(paste0("Surv(time_to_event_alt,event_status_alt) ~ ",x,"+HbA1c+sex+AGEBASE"))
    M0 <- as.formula(paste0("Surv(time_to_event_",y,",event_status_",y,") ~ ",x))
    M1 <- coxph(M0,data=dat)
    logHR <- summary(M1)$coef[1,1]
    pval <- summary(M1)$coef[1,5]
    results_full <- data.frame(Protein=x,Outcome=paste0("Time to doubling of ",str_to_upper(y)),LogHR=logHR,Pvalue=pval)
    results <- rbind(results,results_full)
  }
}
#ALT
results_alt <- results %>%
  filter(Outcome=="Time to doubling of ALT") %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr"))
results_alt <- results_alt %>%
  mutate(sig=ifelse(fdr<0.05,"*",""))
#AST
results_ast <- results %>%
  filter(Outcome=="Time to doubling of AST") %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr"))
results_ast <- results_ast %>%
  mutate(sig=ifelse(fdr<0.05,"*",""))
results_total <- rbind(results_ast,results_alt)

#Match to gene name
key$Protein <- key$AptName
results_total <- tidylog::left_join(results_total,key,by="Protein")

sig_results <- results_total %>% 
  filter(sig=="*")
```
##b. Old
```{r}
#Load in AST and ALT in today 
today <- read.csv(fs::path(dir.dat,"TODAY subaward","Clinical data","TODAY","CBL.csv"))
# length(unique(today$releaseid)) - 699 ids
# today_M0 <- today %>%
#   filter(mvisit=="M00")
#dplyr::select(c("releaseid","ALT","AST"))
#Load in Proteomics in Today
prot <- readRDS(fs::path(dir.dat,"Liver project","TODAY_proteomics.rds"))
prot <- prot[c(7326,7404,7397,which(grepl("seq.",colnames(prot))))] #length(unique(prot$releaseid)) - 376 unique ids
proteins <- colnames(prot)[which(grepl("seq.",colnames(prot)))]

#rm(today,today_M0,prot)

# #Determine which proteins correlate with genes associated with alt and ast
# #AST
# ast_genes <- read.xlsx(fs::path(dir.dat,"Liver project","Results","Bulk_Results_for_Ast Di (High vs. Low).xlsx"))
# ast_genes_pos <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC>0) %>%
#   arrange(desc(avg_log2FC)) %>%
#   dplyr::slice(1:200) 
# ast_genes_pos <- ast_genes_pos$Gene
# ast_genes_neg <- ast_genes %>%
#   filter(p_val_adj<0.05) %>%
#   filter(avg_log2FC<0) %>%
#   arrange(avg_log2FC) %>%
#   dplyr::slice(1:200) 
# ast_genes_neg <- ast_genes_neg$Gene
# ast_genes <- c(ast_genes_pos,ast_genes_neg)

#Key Protein Gene
key <- readRDS(fs::path(dir.dat,"Liver project","gene_prot.RDS"))
sens_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% sens_genes)]
prot_sens <- prot[c("releaseid","AGEBASE","sex",sens_proteins)]
data <- tidylog::right_join(today,prot_sens,by="releaseid")
data <- data %>% #filter out recruitment visits
  filter(mvisit!="R")

# Compute baseline ALT for each participant
data <- data %>%
  group_by(releaseid) %>%
  mutate(baseline_alt = first(ALT)) %>% 
  mutate(baseline_ast = first(AST)) %>% 
  ungroup()

#Check if ALT has doubled at each visit
data <- data %>%
  mutate(event_alt = ifelse(ALT >= 2 * baseline_alt, 1, 0)) %>% 
  mutate(event_ast = ifelse(AST >= 2 * baseline_ast, 1, 0))

# Ensure the event persists after it occurs
data <- data %>%
  group_by(releaseid) %>%
  mutate(event_alt = cummax(event_alt)) %>%  # Marks 1 for all visits after doubling occurs
  mutate(event_ast = cummax(event_ast)) %>% 
  ungroup() %>% 
  mutate(days_to_alt = ifelse(event_alt==1,days,))

#Censor individuals who didnt develop event


#Create loop for cox proportional hazards
#ALT
results_alt <- data.frame()
for (protein in sens_proteins){
  # formula <- as.formula(paste0("Surv(days, event_alt) ~ ", protein))
  formula <- as.formula(paste0("Surv(days, event_alt) ~ ", protein,"+AGEBASE+sex+HbA1c"))
  # Fit the Cox model
  cox_model <- coxph(formula, data = data)
  
  # Store the summary or model output
  LogHR <- summary(cox_model)$coef[1,1]
  Risk <- ifelse(LogHR>0,"Increased",
                 ifelse(LogHR<0,"Decreased","No Effect"))
  Pvalue <- summary(cox_model)$coef[1,5]
  results_data_alt <- data.frame(Protein=protein,Outcome="ALT",LogHazardRatio=LogHR,Risk=Risk,Pvalue=Pvalue)
  results_alt <- rbind(results_alt,results_data_alt)
}
rm(results_data_alt)
results_alt <- results_alt %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr")) %>% 
  mutate(sig=case_when(fdr<0.05 & Risk=="Increased" ~ "Significant Positive",
                       fdr<0.05 & Risk=="Decreased" ~"Significant Negative",
                       fdr>0.05~"Not Significant"))

#AST
results_ast <- data.frame()
for (protein in sens_proteins){
  # formula <- as.formula(paste0("Surv(days, event_ast) ~ ", protein))
  formula <- as.formula(paste0("Surv(days, event_ast) ~ ", protein,"+AGEBASE+sex+HbA1c"))
  # Fit the Cox model
  cox_model <- coxph(formula, data = data)
  
  # Store the summary or model output
  LogHR <- summary(cox_model)$coef[1,1]
  Risk <- ifelse(LogHR>0,"Increased",
                 ifelse(LogHR<0,"Decreased","No Effect"))
  Pvalue <- summary(cox_model)$coef[1,5]
  results_data_ast <- data.frame(Protein=protein,Outcome="AST",LogHazardRatio=LogHR,Risk=Risk,Pvalue=Pvalue)
  results_ast <- rbind(results_ast,results_data_ast)
}
rm(results_data_ast)
results_ast <- results_ast %>% 
  mutate(fdr = p.adjust(Pvalue,method="fdr")) %>% 
  mutate(sig=case_when(fdr<0.05 & Risk=="Increased" ~ "Significant Positive",
                       fdr<0.05 & Risk=="Decreased" ~"Significant Negative",
                       fdr>0.05~"Not Significant"))

#Combine Results
results <- rbind(results_ast,results_alt)
results <- results %>% 
  dplyr::rename(AptName=Protein)
results <- tidylog::left_join(results,key,by="AptName")
results <- results %>% 
  mutate(Gene_Name=paste0(EntrezGeneSymbol,"_",SeqId))
results_alt <- results %>% 
  filter(Outcome=="ALT") 
results_ast <- results %>% 
  filter(Outcome=="AST") 

#Save results
# write.csv(results_alt,fs::path(dir.results,"ALT_cox_results_unadjusted.csv"))
# write.csv(results_ast,fs::path(dir.results,"AST_cox_results_unadjusted.csv"))
write.csv(results_alt,fs::path(dir.results,"ALT_cox_results_adjusted.csv"))
write.csv(results_ast,fs::path(dir.results,"AST_cox_results_adjusted.csv"))

# Create the volcano plot
volcano_plot_alt <- ggplot(results_alt, aes(x = LogHazardRatio, y = -log10(fdr), color = sig)) +
  geom_point(alpha = 0.8, size = 2) +  # Points with transparency
  # scale_color_manual(values = c("grey", "red")) +  # Custom colors for significance
  scale_color_manual(values = c("Not Significant" = "grey", 
                                "Significant Positive" = "red", 
                                "Significant Negative" = "blue")) +
  theme_minimal() +  # Clean theme
  labs(
    title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of ALT (U/L), adjusted for Age, Sex, and HbA1c",
    # title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of ALT (U/L)",
    x = "Log Hazard Ratio",
    y = "-log10(FDR)"
  ) +
  theme(
    legend.position = "top",
    legend.title = element_blank()
  )

# Highlight top significant genes (optional)
top_genes_alt <- subset(results_alt, sig== "Significant Positive" | sig=="Significant Negative")
if (nrow(top_genes_alt) > 0) {
  volcano_plot_alt <- volcano_plot_alt +
    # geom_text(data = top_genes_alt, aes(label = EntrezGeneSymbol), 
    #           vjust = 1.5, size = 5, color = "black")
    geom_text_repel(data = top_genes_alt, aes(label = Gene_Name), 
                    box.padding = 0.35, 
                    point.padding = 0.5, 
                    max.overlaps = Inf,  # Prevent overlap
                    size = 3, 
                    color = "black", 
                    segment.size = 0.5)
}
plot(volcano_plot_alt)

# Create the volcano plot
volcano_plot_ast <- ggplot(results_ast, aes(x = LogHazardRatio, y = -log10(fdr), color = sig)) +
  geom_point(alpha = 0.8, size = 2) +  # Points with transparency
  # scale_color_manual(values = c("grey", "red")) +  # Custom colors for significance
  scale_color_manual(values = c("Not Significant" = "grey", 
                                "Significant Positive" = "red", 
                                "Significant Negative" = "blue")) +
  theme_minimal() +  # Clean theme
  labs(
    title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of AST (U/L), adjusted for Age, Sex, and HbA1c",
    # title = "Cox Proportional Hazards Model for Senescence Proteins and Doubling of AST (U/L)",
    x = "Log Hazard Ratio",
    y = "-log10(FDR)"
  ) +
  theme(
    legend.position = "top",
    legend.title = element_blank()
  )

# Highlight top significant genes (optional)
top_genes_ast <- subset(results_ast, sig== "Significant Positive" | sig=="Significant Negative")
if (nrow(top_genes_ast) > 0) {
  volcano_plot_ast <- volcano_plot_ast +
    # geom_text(data = top_genes_ast, aes(label = EntrezGeneSymbol), 
    #           vjust = 1.5, size = 5, color = "black")
    geom_text_repel(data = top_genes_ast, aes(label = Gene_Name), 
                    box.padding = 0.35, 
                    point.padding = 0.5, 
                    max.overlaps = Inf,  # Prevent overlap
                    size = 3, 
                    color = "black", 
                    segment.size = 0.5)
}
plot(volcano_plot_ast)



```

#9. TEEN-LABs
##a. New
```{r}
#Proteomics teen labs 
load(fs::path(dir.dat,"Teen Labs","Data_Cleaned","analysis_dataset.RData"))
dat_baseline <- df %>% 
  filter(visit=="Month 1")
dat_followup <- df %>% 
  filter(visit=="Month 6") 
dat <- df %>% 
  filter(visit=="Month 1" | visit=="Month 6")

#Protein key
key <- analyte_info %>% 
  dplyr::select(all_of(c("AptName","SeqId","EntrezGeneSymbol")))

#Senescence Proteins
sens_proteins <- key$AptName[which(key$EntrezGeneSymbol %in% sens_genes)]
prot <- df[c("ID","visit",sens_proteins)]
protein <- colnames(prot)[-c(1,2)]
prot_baseline <- prot %>% 
  filter(visit=="Month 1")
prot_followup <- prot %>% 
  filter(visit=="Month 6")
#Name proteins baseline or followup 
colnames(prot_baseline)[-c(1,2)] <- paste0(colnames(prot_baseline)[-c(1,2)],"_baseline")
colnames(prot_followup)[-c(1,2)] <- paste0(colnames(prot_followup)[-c(1,2)],"_followup")
prot_baseline <- prot_baseline %>% 
  dplyr::select(-visit)
prot_baseline_dat <- prot_baseline
prot_followup <- prot_followup %>% 
  dplyr::select(-visit)

prot <- tidylog::left_join(prot_baseline,prot_followup,by="ID")
# Find columns with "baseline" and "followup"
baseline_cols <- grep("_baseline$", colnames(prot), value = TRUE)
followup_cols <- grep("_followup$", colnames(prot), value = TRUE)

# Sort to ensure matching order
baseline_cols <- sort(baseline_cols)
followup_cols <- sort(followup_cols)

# Create new columns for differences
for (i in seq_along(baseline_cols)) {
  protein_name <- sub("_baseline$", "", baseline_cols[i])  # Get protein name (e.g., "protein1")
  diff_col <- paste0(protein_name, "_diff")               # Create a difference column name
  prot[[diff_col]] <- prot[[followup_cols[i]]] - prot[[baseline_cols[i]]]
}

#Covariates at baseline
dat_cov <- dat %>% 
  filter(visit=="Month 1") %>% 
  dplyr::select(all_of(c("ID","SEX","HBA1C","age"))) 

rm(df,analyte_info)

#ALT/AST
liv <- read.csv(fs::path(dir.dat,"Liver Project","Teen-LABS_AST_ALT.csv"))
liv <- liv %>% 
  dplyr::rename(visit=Visit) %>% 
  filter(visit==1 | visit==6) %>% 
  mutate(visit=case_when(visit==1~"Month 1",
                         visit==6~"Month 6")) %>% 
  mutate(across(everything(),~ifelse(.==-5,NA,.))) %>% 
  mutate(across(everything(),~ifelse(.==-7,1,.)))
liv_baseline <- liv %>% 
  filter(visit=="Month 1") %>% 
  dplyr::select(-visit) %>% 
  dplyr::rename(ALT_baseline=ALT,
                AST_baseline=AST)
liv_followup <- liv %>% 
  filter(visit=="Month 6") %>% 
  dplyr::select(-visit) %>% 
  dplyr::rename(ALT_followup=ALT,
                AST_followup=AST)
liv <- tidylog::left_join(liv_baseline,liv_followup,by="ID")
liv <- liv %>% 
  mutate(ALT_diff = ALT_followup-ALT_baseline,
         AST_diff = AST_followup-AST_baseline) 


dat_cov$ID <- as.character(dat_cov$ID)
prot$ID<- as.character(prot$ID)

#Baseline proteins & baseline covariates and baseline and followup alt/ast
dat_all <- tidylog::right_join(dat_cov,liv,by=c("ID"))
dat_all <- tidylog::left_join(dat_all,prot,by=c("ID"))

#Run linear mixed effect model & difference model 
protein_diff <- paste0(protein,"_diff")
results <- data.frame()
covs <- c("age","SEX","HBA1C")
for (x in protein_diff) {
  for (y in c("AST_diff","ALT_diff")) {
    M0_crude <- as.formula(paste0(y,"~",x))
    M0_adj <- as.formula(paste0(y,"~",x,"+",paste(covs, collapse = "+")))
    M1_crude <- lm(M0_crude,data=dat_all)
    M1_adj <- lm(M0_adj,data=dat_all)
    coef_crude <- summary(M1_crude)$coef[2,1]
    coef_adj <- summary(M1_adj)$coef[2,1]
    p_crude <- summary(M1_crude)$coef[2,4]
    p_adj <- summary(M1_adj)$coef[2,4]
    diff_results <- data.frame(Protein=x,Enzyme=y,Coef=coef_adj,PValue=p_adj)
    results <- rbind(results,diff_results)
  }
}
r_alt <- results %>% 
  filter(Enzyme=="ALT_diff") %>% 
  mutate(fdr=p.adjust(PValue,method="fdr"))
r_ast <- results %>% 
  filter(Enzyme=="AST_diff") %>% 
  mutate(fdr=p.adjust(PValue,method="fdr"))
results <- rbind(r_alt,r_ast)
results <- results %>% 
  mutate(sig=ifelse(fdr<0.05,"*",""))
write.csv(results,fs::path(dir.results,"Diff_Results_Liver_Teenlabs_Imputed.csv"))

#Baseline protein with change in ALT/AST pre and post surgery
liv <- read.csv(fs::path(dir.dat,"Liver Project","Teen-LABS_AST_ALT.csv"))
liv <- liv %>% 
  dplyr::rename(visit=Visit) %>% 
  filter(visit==1 | visit==6) %>% 
  mutate(visit=case_when(visit==1~"Month 1",
                         visit==6~"Month 6")) %>% 
  mutate(across(everything(),~ifelse(.==-5,NA,.))) %>% 
  mutate(across(everything(),~ifelse(.==-7,NA,.)))
prot_baseline_dat$ID <- as.character(prot_baseline_dat$ID)
dat_baseline <- tidylog::left_join(liv,prot_baseline_dat,by=c("ID"))
dat_baseline <- tidylog::left_join(dat_cov,dat_baseline,by="ID")
dat_baseline <- dat_baseline %>% 
  mutate(time=case_when(visit=="Month 1"~1,
                        visit=="Month 6"~6))

#Linear Mixed Effects Model
protein_baseline <- paste0(protein,"_baseline")
covs <- c("age","SEX","HBA1C")

results <- data.frame()
for (x in protein_baseline) {
  for (y in c("AST","ALT")) {
    M0_crude <- as.formula(paste0(y,"~",x,"*time+(1|ID)"))
    M0_adj <- as.formula(paste0(y,"~",x,"*time+",paste(covs, collapse = "+"),"+(1|ID)"))
    M1_crude <- lmer(M0_crude,data=dat_baseline)
    M1_adj <- lmer(M0_adj,data=dat_baseline)
    Main <- summary(M1_adj)$coef[2,1]
    Main_p <- summary(M1_adj)$coef[2,5]
    Time <- summary(M1_adj)$coef[3,1]
    Time_p <- summary(M1_adj)$coef[3,5]
    Int <- summary(M1_adj)$coef[7,1]
    Int_p <- summary(M1_adj)$coef[7,5]
    results_int <- data.frame(Protein=x,Enzyme=y,Main,Main_p,Time,Time_p,Int,Int_p)
    results <- rbind(results,results_int)
  }
}
r_alt <- results %>% 
  filter(Enzyme=="ALT") %>% 
  mutate(Main_fdr=p.adjust(Main_p,method="fdr")) %>% 
  mutate(Int_fdr=p.adjust(Int_p,method="fdr"))
r_ast <- results %>% 
  filter(Enzyme=="AST") %>% 
  mutate(Main_fdr=p.adjust(Main_p,method="fdr")) %>% 
  mutate(Int_fdr=p.adjust(Int_p,method="fdr"))
results <- rbind(r_alt,r_ast)
results <- results %>% 
  mutate(Main_sig=ifelse(Main_fdr<0.05,"*","")) %>% 
  mutate(Int_sig=ifelse(Int_fdr<0.05,"*",""))
results$AptName <- str_remove(results$Protein,"_baseline")
results <- tidylog::left_join(results,key,by="AptName")
write.csv(results,fs::path(dir.results,"LMEM_TeenLabs_Liver_Results_Imputed.csv"))
```
##b. Old
```{r}
#Proteomics teen labs 
load(fs::path(dir.dat,"Teen Labs","Data_Cleaned","analysis_dataset.RData"))
#Senescence Proteins
sens_proteins <- analyte_info$AptName[which(analyte_info$EntrezGeneSymbol %in% sens_genes)]

alt_ast <- read.csv(fs::path(dir.dat,"Liver Project","Teen-LABS_AST_ALT.csv")) %>% 
  filter(Visit==1|Visit==6) %>% 
  mutate(Visit=ifelse(Visit==1,"PreSurgery","PostSurgery"))
df_baseline <- df %>% 
  filter(visit=="Month 1") %>% 
  dplyr::rename(Visit=visit) %>% 
  mutate(Visit="PreSurgery") %>% 
  mutate(ID=as.character(ID)) %>% 
  dplyr::select(all_of(c("ID",sens_proteins)))
# df_followup <- df %>% 
#   filter(visit=="Month 6") %>% 
#   dplyr::rename(Visit=visit) %>% 
#   mutate(Visit="PostSurgery") %>% 
#   mutate(ID=as.character(ID)) %>% 
#   dplyr::select(all_of(c("ID","Visit",sens_proteins)))

#Prepare data
dat_baseline <- tidylog::left_join(alt_ast,df_baseline,by=c("ID"))
dat <- dat_baseline
rm(dat_baseline)
# dat_followup <- tidylog::left_join(dat,df_followup,by=c("ID"))
# dat_all <- tidylog::left_join(dat,df,by=c("ID","Visit"))

hist(dat$AST)
hist(dat$ALT)
#Log transform alt and ast
dat$ALT_log <- log(dat$ALT)
dat$AST_log <- log(dat$AST)
dat$Visit <- factor(dat$Visit)
dat$Visit <- relevel(dat$Visit, ref = "PreSurgery")

# #Make sure everyone has baseline proteins
# prot <- df %>% 
#   dplyr::select(c("ID","Visit",sens_proteins))

#Linear Mixed Effects Model
#ALT
plot_list <- list()
results <- data.frame()
for (protein in sens_proteins) {
  gene_name <- paste0(analyte_info$EntrezGeneSymbol[which(analyte_info$AptName %in% protein)],"_",analyte_info$SeqId[which(analyte_info$AptName %in% protein)])
  M0 <- as.formula(paste0("ALT_log ~ ",protein,"*Visit + (1 | ID)"))
  M1 <- lmer(M0, data = dat)
  main_effect <- summary(M1)$coef[2,1]
  main_p <- summary(M1)$coef[2,5]
  time_effect <- summary(M1)$coef[3,1]
  time_p <- summary(M1)$coef[3,5]
  int_effect <- summary(M1)$coef[4,1]
  int_p <- summary(M1)$coef[4,5]
  results_dat <- data.frame(Apt=gene_name,Outcome="ALT_log",MainEffect=main_effect,MainPVal=main_p,TimeEffect=time_effect,TimePVal=time_p,Interaction=int_effect,IntPVal=int_p)
  results <- rbind(results,results_dat)
  
  int_plot <- ggplot(dat, aes(x = !!sym(protein), y = ALT_log, color = Visit)) +
    geom_point() +
    geom_smooth(method = "lm", se = TRUE) +
    labs(title = paste0("Interaction between ",gene_name," and Visit on ALT_log"))+
    theme_minimal()
  
  # Create the spaghetti plot
  spaghetti_plot <- ggplot(dat, aes(x = Visit, y = ALT_log, group = ID, color = !!sym(protein))) +
    geom_line() +  # Connect ALT levels for each individual
    geom_point() +  # Add points for ALT levels at each time point
    # scale_color_viridis_d(guide = "none") +  # Optional: Viridis color palette, hide legend
    # scale_color_viridis_c(guide = "colorbar", name = "Protein") +
    scale_color_gradient(low = "blue", high = "red", name = gene_name) +
    # scale_color_gradient2(low = "blue", mid = "white", high = "red", midpoint = median(dat[[protein]],na.rm=T), name = "Protein") +
    labs(
      title = paste0("Spaghetti Plot of ALT Levels Pre- and Post-Surgery by ",gene_name),
      x = "Visit",
      y = "ALT Levels"
    ) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "gray85")) 
  
  # Combine the two plots
  combined_plot <- plot_grid(int_plot, spaghetti_plot, ncol = 1, labels = c("A", "B"))
  
  # Add to the list
  plot_list[[protein]] <- combined_plot
}

write.csv(results,fs::path(dir.results,paste0("Interaction between ",gene_name," and Visit on ALT_log.csv")))
# # Save all combined plots to a single PDF
# pdf("combined_interaction_and_spaghetti_plots.pdf", width = 10, height = 14)
# for (combined_plot in plot_list) {
#   plot(combined_plot)
# }
# dev.off()

#AST
plot_list <- list()
results <- data.frame()
for (protein in sens_proteins) {
  gene_name <- paste0(analyte_info$EntrezGeneSymbol[which(analyte_info$AptName %in% protein)],"_",analyte_info$SeqId[which(analyte_info$AptName %in% protein)])
  M0 <- as.formula(paste0("AST_log ~ ",protein,"*Visit + (1 | ID)"))
  M1 <- lmer(M0, data = dat)
  main_effect <- summary(M1)$coef[2,1]
  main_p <- summary(M1)$coef[2,5]
  time_effect <- summary(M1)$coef[3,1]
  time_p <- summary(M1)$coef[3,5]
  int_effect <- summary(M1)$coef[4,1]
  int_p <- summary(M1)$coef[4,5]
  results_dat <- data.frame(Apt=gene_name,Outcome="AST_log",MainEffect=main_effect,MainPVal=main_p,TimeEffect=time_effect,TimePVal=time_p,Interaction=int_effect,IntPVal=int_p)
  results <- rbind(results,results_dat)
  
  int_plot <- ggplot(dat, aes(x = !!sym(protein), y = ALT_log, color = Visit)) +
    geom_point() +
    geom_smooth(method = "lm", se = TRUE) +
    labs(title = paste0("Interaction between ",gene_name," and Visit on AST_log"))+
    theme_minimal()
  
  # Create the spaghetti plot
  spaghetti_plot <- ggplot(dat, aes(x = Visit, y = ALT_log, group = ID, color = !!sym(protein))) +
    geom_line() +  # Connect ALT levels for each individual
    geom_point() +  # Add points for ALT levels at each time point
    # scale_color_viridis_d(guide = "none") +  # Optional: Viridis color palette, hide legend
    # scale_color_viridis_c(guide = "colorbar", name = "Protein") +
    scale_color_gradient(low = "blue", high = "red", name = gene_name) +
    # scale_color_gradient2(low = "blue", mid = "white", high = "red", midpoint = median(dat[[protein]],na.rm=T), name = "Protein") +
    labs(
      title = paste0("Spaghetti Plot of AST Levels Pre- and Post-Surgery by ",gene_name),
      x = "Visit",
      y = "ALT Levels"
    ) +
    theme_minimal() +
    theme(panel.grid.major = element_line(color = "gray85")) 
  
  # Combine the two plots
  combined_plot <- plot_grid(int_plot, spaghetti_plot, ncol = 1, labels = c("A", "B"))
  
  # Add to the list
  plot_list[[protein]] <- combined_plot
}

write.csv(results,fs::path(dir.results,paste0("Interaction between ",gene_name," and Visit on AST_log.csv")))
# # Save all combined plots to a single PDF
# pdf("combined_interaction_and_spaghetti_plots.pdf", width = 10, height = 14)
# for (combined_plot in plot_list) {
#   plot(combined_plot)
# }
# dev.off()


```

#10. Pseudotime Analysis - Hepatocytes
```{r}
#Marker genes of healthy hepatocytes: 
healthy_hepatocyte_genes <- c("ALB", "CYP3A4", "ARG1", "IGF1", "HNF4A", "APOE", "SLC10A2", "FASN", "ABCB1", "GLUL")

#Marker genes of adaptive/mal adaptive hepatocytes: 
adaptive_maladaptive_genes <- c("COL1A1", "ACTA2", "TNF", "IL6", "FASN", "CASP3", "SOD1", "CAT", "HGF", "PCNA")


```

#11. Senescence Score 
```{r}
sens_genes <- intersect(sens_genes, rownames(so_liver_sn))
oro_vars <- colnames(so_liver_sn@meta.data)[85:114]

so_liver_sn <- AddModuleScore(
  object =so_liver_sn,
  features = list(sens_genes),
  name = "SenescenceScore",
  replace = TRUE
)

# Aggregate senescence scores at the individual level (mean or median per individual)
individual_data <- so_liver_sn@meta.data %>%
  group_by(Kit_Lot) %>%
  summarise(
    avg_senescence = mean(SenescenceScore1, na.rm = TRUE),  # mean score per individual
    median_senescence = median(SenescenceScore1, na.rm = TRUE),  # or median score if preferred
    alt=first(alt),
    ast=first(ast),
    diagnosis_of_diabetes=first(diagnosis_of_diabetes),
    group2=first(group2),
    across(all_of(oro_vars), ~first(.), .names = "first_{.col}")
  )


# Boxplot or violin plot
#Type 2 Diabetes
ggplot(so_liver_sn@meta.data, aes(x = diagnosis_of_diabetes, y = SenescenceScore1, fill = diagnosis_of_diabetes)) +
  geom_violin(trim = FALSE) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  theme_minimal() +
  labs(title = "Senescence Score by Diabetes Status", x = "Diabetes (Yes vs. No)", y = "Senescence Score")

#Aggregated
ggplot(individual_data, aes(x = diagnosis_of_diabetes, y = avg_senescence, fill = diagnosis_of_diabetes)) +
  geom_violin(trim = FALSE) +
  geom_boxplot(width = 0.2, outlier.shape = NA) +
  theme_minimal() +
  labs(title = "Senescence Score by Diabetes Status", x = "Diabetes (Yes vs. No)", y = "Average Senescence Score")

#Difference in Senescence Score for Type 2 vs. No Type 2
t.test(SenescenceScore1 ~ diagnosis_of_diabetes, data = so_liver_sn@meta.data)

#Medication
#Difference in Senescence Score for Medicaiton Type
ggplot(so_liver_sn@meta.data, aes(x = group2, y = SenescenceScore1, fill = group2)) +
  geom_jitter(width = 0.2, alpha = 0.5, size = 0.8) +  # Add jitter for individual data points
  geom_boxplot(outlier.shape = NA, width = 0.6) +  # Boxplot without outliers
  #geom_jitter(width = 0.2, alpha = 0.5, size = 0.8) +  # Add jitter for individual data points
  theme_minimal() +
  labs(title = "Senescence Score by Medication Status", x = "Medication", y = "Senescence Score")

#Aggregated
ggplot(individual_data, aes(x = group2, y = avg_senescence, fill = group2)) +
  #geom_jitter(width = 0.2, alpha = 0.5, size = 0.8) +  # Add jitter for individual data points
  geom_boxplot(outlier.shape = NA, width = 0.6) +  # Boxplot without outliers
  geom_jitter(width = 0.2, alpha = 0.5, size = 0.8) +  # Add jitter for individual data points
  theme_minimal() +
  labs(title = "Average Senescence Score by Medication Status", x = "Medication", y = "Average Senescence Score")

#Add random effect for individual - cells not independent
#LMEM

#Anova
#singel cell
mod <- aov(SenescenceScore1 ~ group2, data = so_liver_sn@meta.data)
summary(mod)
# Post-hoc test: Tukey's Honest Significant Difference (HSD)
TukeyHSD(mod)

#Bulk
mod <- aov(avg_senescence ~ group2, data = individual_data)
summary(mod)
# Post-hoc test: Tukey's Honest Significant Difference (HSD)
TukeyHSD(mod)

#Correlate with oroborus data
oro_vars <- colnames(so_liver_sn@meta.data)[85:114]
for (oro in oro_vars){
  p <- ggplot(so_liver_sn@meta.data, aes(x = !!sym(oro), y = SenescenceScore1)) +
  geom_point(alpha = 0.6) +  # Scatter points
  geom_smooth(method = "lm", color = "blue", se = FALSE) +  # Linear regression line
  theme_minimal() +
  labs(title = paste0("Senescence Score vs. ",str_to_title(oro)),
       x = paste0(str_to_title(oro)),
       y = "Senescence Score") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
  print(p)
}

#Correlate with alt/ast
#AST
# Scatter plot with regression line
ggplot(so_liver_sn@meta.data, aes(x = ast, y = SenescenceScore1)) +
  geom_point(alpha = 0.6) +  # Scatter points
  geom_smooth(method = "lm", color = "blue", se = FALSE) +  # Linear regression line
  theme_minimal() +
  labs(title = "Senescence Score vs. AST (U/L)",
       x = "Continuous Exposure Variable",
       y = "Senescence Score") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))  # Rotate x-axis labels if necessary

cor.test(so_liver_sn@meta.data$SenescenceScore1, so_liver_sn@meta.data$ast)
lm_model <- lm(SenescenceScore1 ~ ast, data = so_liver_sn@meta.data)
summary(lm_model)
#ALT
cor.test(so_liver_sn@meta.data$SenescenceScore1, so_liver_sn@meta.data$ast)
lm_model <- lm(SenescenceScore1 ~ ast, data = so_liver_sn@meta.data)
summary(lm_model)

#Medication
ggplot(so_liver_sn@meta.data, aes(x = group2, y = SenescenceScore1)) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Group", y = "Senescence Score")

# Perform ANOVA
anova_result <- aov(SenescenceScore1 ~ group2, data = so_liver_sn@meta.data)

# Summary of the ANOVA
summary(anova_result)
shapiro.test(so_liver_sn@meta.data$SenescenceScore1) # For the whole dataset


```

#12. Oroborus Analysis
```{r}
oro_vars <- colnames(so_liver_sn@meta.data)[85:114]
#meta.data <- so_liver_sn@meta.data

## Function to summarize each variable
#agg_meta.data <- meta.data %>%
#  group_by(Kit_Lot) %>%
#  summarise(across(all_of(oro_vars), ~ mean(.x, na.rm = TRUE), .names = "mean_{.col}"))

#MAST between oro vars and senescence genes
genes <- rownames(so_sens_all)
for (oro in oro_vars){
mast_sens_fxn(so=so_sens_all,cell=NULL,exposure=oro,gene_set=genes,exp_group=NULL,ref_group=NULL)
}

#Association between oro vars and senescence score

```


#13. Targeted DEGs 
```{r}


```


