---
title: "ATTEMPT scRNA report (subset for outline)"
author: "Ye Ji Choi"
format:
  revealjs:
    embed-resources: true
    scrollable: true
    transition: slide
    controls-layout: bottom-right
    menu: true
    toc: true
    toc-depth: 1
    fontsize: 15
    number-sections: true
---
```{r include = F}
library(jsonlite)
library(aws.s3)
library(knitr)
library(dplyr)
library(kableExtra)
library(tidyverse)
library(lmerTest)
library(emmeans)
library(broom)
library(ggbeeswarm)
library(lemon)
library(quantreg)

source("~/GitHub/CHCO-Code/Petter Bjornstad/Resources/YC/R Functions/correlation_function.R")
source("~/GitHub/CHCO-Code/Petter Bjornstad/ATTEMPT/attempt_functions.R")
```

```{r include = F}
## Create an S3 client
keys <- fromJSON("/Users/choiyej/Library/CloudStorage/OneDrive-TheUniversityofColoradoDenver/Bjornstad Pyle Lab/keys.json")

Sys.setenv(
  "AWS_ACCESS_KEY_ID" = keys$MY_ACCESS_KEY,
  "AWS_SECRET_ACCESS_KEY" = keys$MY_SECRET_KEY,
  "AWS_DEFAULT_REGION" = "",
  "AWS_REGION" = "",
  "AWS_S3_ENDPOINT" = "s3.kopah.uw.edu"
)
```

```{r echo = F}
harm_dat <- read.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/Data Harmonization/Data Clean/harmonized_dataset.csv", na.strings = "")

attempt_redcap <- harm_dat %>% filter(study == "ATTEMPT") %>%
  dplyr::summarise(across(where(negate(is.numeric)), ~ ifelse(all(is.na(.x)), NA_character_, last(na.omit(.x)))),
                   across(where(is.numeric), ~ ifelse(all(is.na(.x)), NA_real_, mean(.x, na.rm = TRUE))),
                   .by = c(record_id, visit)) %>%
  dplyr::mutate(race_ethnicity_condensed = case_when(race == "White" &
                                                       ethnicity == "Not Hispanic or Latino" ~ "Not Hispanic or Latino White",
                                                     race == "Black or African American" &
                                                       ethnicity == "Not Hispanic or Latino" ~ "Not Hispanic or Latino Black",
                                                     ethnicity == "Hispanic or Latino" ~ "Hispanic or Latino",
                                                     T ~ "Not Hispanic or Latino Other")) %>%
  arrange(record_id) %>%
  dplyr::select(-hba1c)

attempt_redcap_denver <- attempt_redcap %>%
  filter(grepl("^3", record_id)) %>%
  mutate(visit = case_when(visit == "baseline" ~ 0,
                           visit == "4_weeks_post" ~ 4,
                           visit == "4_months_post" ~ 16,
                           visit == "screening"  ~ -4,
                           visit == "follow_up" ~ 18),
         record_id = as.integer(record_id)) %>%
  dplyr::select(-c(ethnicity, bmi, gfr_raw_plasma, gfr_bsa_plasma, date, sex, mri_date, PWV,
                   creatinine_s, cystatin_c_s, temp, pulse, height, avg_c_r2, avg_k_r2, 
                   sbp, dbp, hct, age, weight)) %>%
  rowwise() %>%
  mutate(avg_c_t1 = mean(c(t1_r_cortex, t1_l_cortex), na.rm = T),
         avg_k_t1 = mean(c(t1_r_kidney, t1_l_kidney), na.rm = T),
         tkv = (volume_left + volume_right)) %>%
  ungroup() %>%
  dplyr::rename(subject_id = record_id)

# clinical dataset from Antoine (merged version)
s3load("Clinical Data/ATTEMPT_AC.RData", "attempt", region = "")
attempt_dat <- merged_data
attempt_dat[attempt_dat == ""] <- NA
attempt_dat <- attempt_dat %>%
  filter(visit %in% c("baseline", "4_weeks_post", "4_months_post", "screening", "follow_up")) %>%
  mutate(visit = case_when(visit == "baseline" ~ 0,
                           visit == "4_weeks_post" ~ 4,
                           visit == "4_months_post" ~ 16,
                           visit == "screening"  ~ -4,
                           visit == "follow_up" ~ 18),
                record_id = as.integer(record_id)) %>%
  dplyr::rename(subject_id = record_id,
                treatment = treatment_arm)

# extract IDs with biopsy
attempt_clean_so_metadata <- s3readRDS("cleaned_data/attempt_clean_so_metadata.rds", "attempt", region = "")
attempt_biopsy_yn <- attempt_clean_so_metadata %>%
  dplyr::select(subject_id, visit) %>%
  distinct(subject_id, visit, .keep_all = T) %>%
  mutate(visit = case_when(visit == "PRE" ~ 0,
                           visit == "POST" ~ 16),
         biopsy_yn = "Yes")
attempt_biopsy_yn <- attempt_biopsy_yn %>%
  rbind(attempt_biopsy_yn %>%
          filter(visit == 0) %>%
          mutate(visit = -4))

# merge REDCap version (for Denver) and dataset from Antoine
attempt_dat <- left_join(attempt_dat, attempt_redcap_denver) %>%
  left_join(attempt_biopsy_yn) %>%
  mutate(biopsy_yn = case_when(biopsy_yn == "Yes" ~ "Yes",
                               T ~ "No"))
attempt_dat$treatment <- factor(attempt_dat$treatment, levels = c("Placebo", "Dapagliflozin 5mg"))

attempt_dat %>%
  filter(biopsy_yn == "Yes") %>%
  filter(visit %in% c(0)) %>%
  filter(subject_id != 30001) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id) %>%
  write.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Clean/ATTEMPT_BL_LN2.csv", na = "",
            row.names = F)

attempt_dat %>%
  filter(biopsy_yn == "Yes") %>%
  filter(visit %in% c(16)) %>%
  filter(subject_id != 30013) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id) %>%
  rbind(attempt_dat[attempt_dat$subject_id == 30472,] %>%
  filter(visit %in% c(16)) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id)) %>%
  write.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Clean/ATTEMPT_4M_LN2.csv", na = "",
            row.names = F)
```

```{r echo = F}
delta_df <- attempt_dat %>%
  filter(!is.na(visit)) %>%
  pivot_wider(id_cols = c(subject_id, treatment, site, sex),
              names_from = visit,
              values_from = c(hba1c, cgm_tir, weight, mgfr_jodal_bsa, mgfr_jodal, avg_c_r2, avg_k_r2,
                              avg_c_t1, avg_k_t1, avg_pcascl, avg_c_adc)) %>%
  mutate(
    delta_hba1c_4 = hba1c_4 - hba1c_0,
    delta_hba1c_16 = hba1c_16 - hba1c_0,
    delta_tir_4 = cgm_tir_4 - cgm_tir_0,
    delta_tir_16 = cgm_tir_16 - cgm_tir_0,
    delta_weight_4 = weight_4 - weight_0,
    delta_weight_16 = weight_16 - weight_0,
    delta_mgfr_jodal_bsa_4 = mgfr_jodal_bsa_4 - `mgfr_jodal_bsa_-4`,
    delta_mgfr_jodal_bsa_16 = mgfr_jodal_bsa_16 - `mgfr_jodal_bsa_-4`,
    delta_mgfr_jodal_4 = mgfr_jodal_4 - `mgfr_jodal_-4`,
    delta_mgfr_jodal_16 = mgfr_jodal_16 - `mgfr_jodal_-4`,
    delta_avg_c_r2_4 = avg_c_r2_4 - `avg_c_r2_-4`,
    delta_avg_c_r2_16 = avg_c_r2_16 - `avg_c_r2_-4`,
    delta_avg_k_r2_4 = avg_k_r2_4 - `avg_k_r2_-4`,
    delta_avg_k_r2_16 = avg_k_r2_16 - `avg_k_r2_-4`,
    delta_avg_c_t1_4 = avg_c_t1_4 - `avg_c_t1_0`,
    delta_avg_c_t1_16 = avg_c_t1_16 - `avg_c_t1_0`,
    delta_avg_k_t1_4 = avg_k_t1_4 - `avg_k_t1_0`,
    delta_avg_k_t1_16 = avg_k_t1_16 - `avg_k_t1_0`,
    delta_avg_pcascl_4 = avg_pcascl_4 - `avg_pcascl_0`,
    delta_avg_pcascl_16 = avg_pcascl_16 - `avg_pcascl_0`,
    delta_avg_c_adc_4 = avg_c_adc_4 - `avg_c_adc_0`,
    delta_avg_c_adc_16 = avg_c_adc_16 - `avg_c_adc_0`,    
    delta_hba1c_0 = 0,
    delta_tir_0 = 0,
    delta_weight_0 = 0,
    `delta_mgfr_jodal_bsa_-4` = 0,
    `delta_mgfr_jodal_-4` = 0,
    `delta_avg_c_r2_-4` = 0,
    `delta_avg_k_r2_-4` = 0,
    `delta_avg_pcascl_0` = 0,
    `delta_avg_c_adc_0` = 0,
    `delta_avg_c_t1_0` = 0,
    `delta_avg_k_t1_0` = 0
  ) %>%
  dplyr::select(subject_id, treatment, site, sex, matches("delta_.*_(-?4|16|0)$")) %>%
  pivot_longer(
    cols = -c(subject_id, treatment, sex, site),
    names_to = c("variable", "visit"),
    names_pattern = "delta_(.*)_(-?\\d+)"
  ) %>% 
  pivot_wider(names_from = variable, values_from = value) %>%
  mutate(visit = as.numeric(visit))

delta_df$treatment <- factor(delta_df$treatment, levels = c("Placebo", "Dapagliflozin 5mg"))

delta_df <- delta_df %>% left_join(
attempt_dat %>%
  group_by(subject_id) %>%
  filter(visit == 0) %>%
  dplyr::select(subject_id, age, diabetes_dx_duration, hba1c, cgm_tir, weight) %>%
  dplyr::rename(hba1c_bl = hba1c,
                tir_bl = cgm_tir,
                weight_bl = weight)) %>% left_join(
attempt_dat %>%
  group_by(subject_id) %>%
  filter(visit == -4) %>%
  dplyr::select(subject_id, mgfr_jodal_bsa, mgfr_jodal,
                avg_c_r2, avg_k_r2) %>%
  dplyr::rename(mgfr_jodal_bsa_bl = mgfr_jodal_bsa,
                mgfr_jodal_bl = mgfr_jodal,
                avg_c_r2_bl = avg_c_r2,
                avg_k_r2_bl = avg_k_r2))
```

# Title

# Abstract

(look at REMODEL)

# Introduction

# Methods

- Mixed models for MRI and lipid profiles.

- Described change over time on mGFR, HbA1c, TIR, weight, and R2*

- Described change over time on mGFR, HbA1c, TIR, weight, and R2* based on baseline characteristics

- Data cleaning: filtered cells with less than X% expression, gathered gene scores using known PT markers to exclude cells with scores above X. 

- scRNA using NEBULA limited it to top 2000 HVG's.

- Gene ~ treatment*visit + (1|subject)

- Not limited to top 2000 HVGs for downstream analysis with CROCODILE cohort.

- Looked for overlaps between ATTEMPT and CROCODILE cohort to identify "reversal" genes.

- Slingshot using both KPMP and non-KPMP cell annotations, with known cell types as start/end cluster.

- Quantile regression on modes of pseudotime to associate with clinical vars (change in weight, hba1c, mGFR, TIR)

- Cell-cell-interaction analysis using LIANA.

# Results

## Describe cohort

```{r echo = F, results='asis'}
# hist(subset(attempt_dat2, visit == "PRE")$hba1c)
# hist(subset(attempt_dat2, visit == "PRE")$emu_urine_acr_mean)
# hist(subset(attempt_dat2, visit == "PRE")$cgm_tir)
# hist(attempt_dat$gfr_bsa_plasma)

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + ldl_serum_mmoll + hdl_serum_mmoll, data = subset(attempt_dat, visit == 0)), digits = 1))

kable(summary(arsenal::tableby(treatment ~ mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == -4)), digits = 1))
```

### MRI

### Proteomics

### Kidney biopsies

```{r echo = F, results='asis'}
kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse  + ldl_serum_mmoll + hdl_serum_mmoll + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, visit == 0  & biopsy_yn == "Yes")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == -4 & biopsy_yn == "Yes")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, visit == 16  & biopsy_yn == "Yes")), digits = 1))
```



## Clinical changes

- ∆ mGFR in whole

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_mgfr_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime.jpeg){width=70%}

- ∆ mGFR in Denver

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_mgfr_overtime_denver.jpeg){width=70%}

---

- ∆ mGFR by tertiles

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_jodal_bsa_category_bars.jpeg){width=70%}

---

- Summarize in text HbA1c, TIR, weight, mGFR and refer to PO paper

PO paper outcomes

|        | Placebo | Dapagliflozin |
|--------|---------|---------------|
| HbA1c  |   `+`   |      `-`      |
| TIR    |   `-`   |      `+`      |
| Weight |   `+`   |      `-`      |
| mGFR   |   `-`   |   `- (more)`  |

---

- Kidney MRI changes in whole 

| R2* | Oxygen availability |
|-----|---------------------|
| `-` | `+`                 |
| `+` | `-`                 |

In CROCODILE T1D vs. HC (ref): Lower medulla R2* in T1D compared to HC

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime.jpeg){width=70%}


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime.jpeg){width=70%}


- Kidney MRI changes in Denver

| R2* | Oxygen availability |
|-----|---------------------|
| `-` | `+`                 |
| `+` | `-`                 |

In CROCODILE T1D vs. HC (ref): Lower medulla R2* in T1D compared to HC

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime_denver.jpeg){width=70%}


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime_denver.jpeg){width=70%}

---

- Kidney R2* per tertile of R2* and mGFR

| R2* | Oxygen availability |
|-----|---------------------|
| `-` | `+`                 |
| `+` | `-`                 |


In CROCODILE T1D vs. HC (ref): Lower medulla R2* in T1D compared to HC

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_mgfr_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_mgfr_category_bars.jpeg){width=70%}

---

- Describe quantitative morphometrics (and safety?)


## Transcriptomics

### Describe # of cells and type of cells annotated; UMAP; # DEGs; upregulated plots


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_celltype.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_kpmpcelltype.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_treat.jpeg){width=50%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_visit.jpeg){width=50%}

### DEGs degs vs. placebo: refer to figures, summarize cells findings across kidney segments, and summary figure. Briefly summary pathology findings (decide on reference)

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("PT", "TAL", "PC", "Immune", "Immune_Myeloid", "Immune_Lymphoid", "IC", "EC", "FIBVSMCP", "POD")

for (cells in celltype_list) {
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Dot Plots/", cells, "_hvg_subtypes_nebula_scores_volcano_heat.jpeg)\n\n"))
}
```

### DEGs association w/ ∆ clinical variables refer to figures, summarize findings across all types, and refer to pathways

#### Change in mGFR

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_mgfr_jodal_bsa_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

#### Change in HbA1c

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_hba1c_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

#### Change in weight

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_weight_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

#### Change in TIR

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_tir_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

### Proportion of all cell types and pseudotime trajectories. Epithelial cell density plot w/ + and - ∆ in clinical variables

| General type      | Sub type                      |
|-------------------|-------------------------------|
| PT                | PT-1, PT-2, PT-3, PT-4, PT-5  |
| TAL               | TAL-1, TAL-2, TAL-3           |
| PC                | tPC-IC, PC-1, PC-2            |
| IC                | IC-A, IC-B                    |
| EC                | EC-PTC, EC-GC, EC-AEA, EC-LYM |
| Immune (Myeloid)  | MAC, MON                      |
| Immune (Lymphoid) | NKT/NKC, T, B                 |
| FIB/VSMC/P        | FIB, VSMC/P                   |

---

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_cellcounts_plot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_cellcounts_proportions_plot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_general_cellcounts_plot_clean.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_general_cellcounts_proportions_plot_clean.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_pt_treatment_visit_proportions_plot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_pt_visit_treatment_proportions_plot_kpmp.jpeg){width=70%}

---



### Reversal to HC state

#### TID vs. HC (CRC)

```{r echo = F, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "ic", "ec", "fibvsmcp")

for (cells in celltype_list) {
  
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/Volcano Plots/", cells ,"_targeted_volcano_attempt_directions_labeled.jpeg){width=70%}\n\n"))
    
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/", cells, "_croc_attempt_dot.jpeg){width=70%}\n\n"))
  
}
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/Volcano Plots/pod_targeted_volcano_attempt_directions_labeled.jpeg){width=70%}

#### General reversal - briefly summarize/describe 2 pathway analysis






## Proteomics

### Describe DAP Plasma

### Describe DAP urine

### Associate DAP w/ DEGs

#### similar to 3c

#### Perhaps limit to epithelial cells for urine and endothelial cells for plasma





