---
title: "ATTEMPT scRNA report"
author: "Ye Ji Choi"
format:
  revealjs:
    embed-resources: true
    scrollable: true
    transition: slide
    controls-layout: bottom-right
    menu: true
    toc: true
    toc-depth: 1
    fontsize: 15
    number-sections: true
    filters:
      - figure-path.lua
---

```{r include = F}
library(jsonlite)
library(aws.s3)
library(knitr)
library(dplyr)
library(kableExtra)
library(tidyverse)
library(lmerTest)
library(emmeans)
library(broom)
library(ggbeeswarm)
library(lemon)
library(quantreg)
library(readxl)

source("~/GitHub/CHCO-Code/Petter Bjornstad/Resources/YC/R Functions/correlation_function.R")
source("~/GitHub/CHCO-Code/Petter Bjornstad/ATTEMPT/attempt_functions.R")
```

```{r include = F}
## Create an S3 client
keys <- fromJSON("/Users/choiyej/Library/CloudStorage/OneDrive-TheUniversityofColoradoDenver/Bjornstad Pyle Lab/keys.json")

Sys.setenv(
  "AWS_ACCESS_KEY_ID" = keys$MY_ACCESS_KEY,
  "AWS_SECRET_ACCESS_KEY" = keys$MY_SECRET_KEY,
  "AWS_DEFAULT_REGION" = "",
  "AWS_REGION" = "",
  "AWS_S3_ENDPOINT" = "s3.kopah.uw.edu"
)
```

# Introduction {data-name="Introduction"}

This is a report compiling results from the ATTEMPT scRNA analysis. 

Primary outcome paper: https://rdcu.be/epIA2

Next steps:

- Complete the report

- Add scrolling menu?

- plot showing the relationship between the top gene and change in clinical variable

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura%20Pyle%20-%20Bjornstad/Biostatistics%20Core%20Shared%20Drive/ATTEMPT/Results/Figures/ATTEMPT_FU_Consort_diagram.png){width="100%"}


# Baseline descriptive tables

```{r echo = T}
harm_dat <- read.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/Data Harmonization/Data Clean/harmonized_dataset.csv", na.strings = "")

attempt_redcap <- harm_dat %>% filter(study == "ATTEMPT") %>%
  dplyr::summarise(across(where(negate(is.numeric)), ~ ifelse(all(is.na(.x)), NA_character_, last(na.omit(.x)))),
                   across(where(is.numeric), ~ ifelse(all(is.na(.x)), NA_real_, mean(.x, na.rm = TRUE))),
                   .by = c(record_id, visit)) %>%
  dplyr::mutate(race_ethnicity_condensed = case_when(race == "White" &
                                                       ethnicity == "Not Hispanic or Latino" ~ "Not Hispanic or Latino White",
                                                     race == "Black or African American" &
                                                       ethnicity == "Not Hispanic or Latino" ~ "Not Hispanic or Latino Black",
                                                     ethnicity == "Hispanic or Latino" ~ "Hispanic or Latino",
                                                     T ~ "Not Hispanic or Latino Other")) %>%
  arrange(record_id) %>%
  dplyr::select(-hba1c)

attempt_redcap_denver <- attempt_redcap %>%
  filter(grepl("^3", record_id)) %>%
  mutate(visit = case_when(visit == "baseline" ~ 0,
                           visit == "4_weeks_post" ~ 4,
                           visit == "4_months_post" ~ 16,
                           visit == "screening"  ~ -4,
                           visit == "follow_up" ~ 18),
         record_id = as.integer(record_id)) %>%
  dplyr::select(-c(ethnicity, bmi, gfr_raw_plasma, gfr_bsa_plasma, date, sex, mri_date,
                   creatinine_s, cystatin_c_s, temp, pulse, height, avg_c_r2, avg_k_r2, 
                   sbp, dbp, hct, age, weight)) %>%
  rowwise() %>%
  mutate(avg_c_t1 = mean(c(t1_r_cortex, t1_l_cortex), na.rm = T),
         avg_k_t1 = mean(c(t1_r_kidney, t1_l_kidney), na.rm = T),
         tkv = (volume_left + volume_right)) %>%
  ungroup() %>%
  dplyr::rename(subject_id = record_id)

# clinical dataset from Antoine (merged version)
attempt_dat <- s3readRDS("Clinical Data/ATTEMPT_AC.RDS", "attempt", region = "")
attempt_dat[attempt_dat == ""] <- NA
attempt_dat <- attempt_dat %>%
  filter(visit %in% c("baseline", "4_weeks_post", "4_months_post", "screening", "follow_up")) %>%
  mutate(visit = case_when(visit == "baseline" ~ 0,
                           visit == "4_weeks_post" ~ 4,
                           visit == "4_months_post" ~ 16,
                           visit == "screening"  ~ -4,
                           visit == "follow_up" ~ 18),
                record_id = as.integer(record_id)) %>%
  dplyr::rename(subject_id = record_id,
                treatment = treatment_arm)

# extract IDs with biopsy
attempt_clean_so_metadata <- s3readRDS("cleaned_data/attempt_clean_so_metadata.rds", "attempt", region = "")
attempt_biopsy_yn <- attempt_clean_so_metadata %>%
  dplyr::select(subject_id, visit) %>%
  distinct(subject_id, visit, .keep_all = T) %>%
  mutate(visit = case_when(visit == "PRE" ~ 0,
                           visit == "POST" ~ 16),
         biopsy_yn = "Yes")
attempt_biopsy_yn <- attempt_biopsy_yn %>%
  rbind(attempt_biopsy_yn %>%
          filter(visit == 0) %>%
          mutate(visit = -4))

# merge REDCap version (for Denver) and dataset from Antoine
attempt_dat <- left_join(attempt_dat, attempt_redcap_denver) %>%
  left_join(attempt_biopsy_yn) %>%
  group_by(subject_id, visit) %>%
  mutate(biopsy_yn = case_when(biopsy_yn == "Yes" ~ "Yes",
                               T ~ "No"),
         avg_ketones = mean(c(mgfr_ketones_pre, mgfr_ketone_iv), na.rm = T)) %>%
  ungroup()
attempt_dat$treatment <- factor(attempt_dat$treatment, levels = c("Placebo", "Dapagliflozin 5mg"))

attempt_dat %>%
  filter(biopsy_yn == "Yes") %>%
  filter(visit %in% c(0)) %>%
  filter(subject_id != 30001) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id) %>%
  write.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Clean/ATTEMPT_BL_LN2.csv", na = "",
            row.names = F)

attempt_dat %>%
  filter(biopsy_yn == "Yes") %>%
  filter(visit %in% c(16)) %>%
  filter(subject_id != 30013) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id) %>%
  rbind(attempt_dat[attempt_dat$subject_id == 30472,] %>%
  filter(visit %in% c(16)) %>%
  dplyr::select(subject_id, visit, treatment) %>%
  arrange(subject_id)) %>%
  write.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Clean/ATTEMPT_4M_LN2.csv", na = "",
            row.names = F)
```
```{r echo = F}
# morphometric report
morph_report <- read_excel(file.path(root_path, "ATTEMPT/Results/Morphometry_qualitative-Aug20-2025.xlsx")) %>%
  filter(!is.na(Average)) %>%
  dplyr::select(-Average) %>%  
  pivot_longer(
    cols = starts_with("SU-"),
    names_to = "Pathology_report_ID",
    values_to = "value"
  ) %>%
  pivot_wider(
    names_from = IDs,
    values_from = value
  ) %>%
  dplyr::mutate(across(3:ncol(.), ~ case_when(
    . == 0 ~ "No",
    . == 1 ~ "Yes"
  )))

biopsy_master <- read_excel("/Users/choiyej/Library/CloudStorage/OneDrive-TheUniversityofColoradoDenver/Bjornstad Pyle Lab/Biopsy master tracker.xlsx", skip = 2) 
colnames(biopsy_master) <- gsub(" ", "_", colnames(biopsy_master))  # Replace spaces with underscores
colnames(biopsy_master) <- gsub("/", "_", colnames(biopsy_master))  # Replace slashes with underscores
colnames(biopsy_master) <- gsub("\\(", "_", colnames(biopsy_master))  # Replace opening parentheses with underscores
colnames(biopsy_master) <- gsub("\\)", "", colnames(biopsy_master))  # Remove closing parentheses
morph_report_merged <- biopsy_master %>%
  right_join(morph_report) %>%
  dplyr::mutate(visit = case_when(Visit_ID == "BL" ~ 0,
                                  Visit_ID == "4M" ~ 16),
                subject_id = case_when(Study == "PANDA" ~ Coenroll_ID__Same_visit_only, 
                                       T ~ Study_ID),
                subject_id = as.numeric(subject_id),
                kit_id = Kit_ID) %>%
  dplyr::select(subject_id, visit, Pathology_report_ID, 32:43)

attempt_dat <- attempt_dat %>% left_join(morph_report_merged)

# subset(attempt_dat, biopsy_yn == "Yes" & is.na(Pathology_report_ID) & visit %in% c(0, 16))

s3saveRDS(attempt_dat, bucket = "attempt", 
          object = "cleaned_data/attempt_dat.rds",
          region = "")
```

## Baseline Table (Full cohort, Dapa vs. Placebo)

```{r echo = T, results='asis'}
# hist(subset(attempt_dat2, visit == "PRE")$hba1c)
# hist(subset(attempt_dat2, visit == "PRE")$emu_urine_acr_mean)
# hist(subset(attempt_dat2, visit == "PRE")$cgm_tir)
# hist(attempt_dat$gfr_bsa_plasma)

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + ldl_serum_mmoll + hdl_serum_mmoll, data = subset(attempt_dat, visit == 0)), digits = 1))

kable(summary(arsenal::tableby(treatment ~ mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == -4)), digits = 1))

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + ldl_serum_mmoll + hdl_serum_mmoll + mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == 16)), digits = 1))

# kable(summary(arsenal::tableby(visit ~  hba1c + mgfr_jodal_bsa + mgfr_jodal, data = attempt_dat), digits = 1))

# kable(summary(arsenal::tableby(treatment ~ mgfr_ketone_iv + mgfr_ketones_pre + mgfr_ketones_t240, data = subset(attempt_dat, visit == -4)), digits = 1))
# kable(summary(arsenal::tableby(treatment ~ mgfr_ketone_iv + mgfr_ketones_pre + mgfr_ketones_t240, data = subset(attempt_dat, visit == 16)), digits = 1))
```

```{r echo = T}
# kable(summary(arsenal::tableby(treatment ~ mgfr_si + mgfr_si_adjusted + mgfr_bm_adult + mgfr_bm_adult_adjusted + mgfr_bm_adult_adjusted_cl1 + mgfr_bm_combined + mgfr_bm_combined_adjusted + mgfr_bm_combined_adjusted_cl1 + mgfr_bm_child + mgfr_bm_child_adjusted + mgfr_bm_child_adjusted_cl1 + mgfr_jodal + mgfr_jodal  + mgfr_jodal_bsa , data = subset(attempt_dat, visit == 0)), digits = 1))
```


```{r echo = T, include = F}
# N's for consort diagram
bl_attempt_dat <- subset(attempt_dat, visit == -4) # R2* collected at -4
bl_attempt_dat_denver <- subset(bl_attempt_dat, site == "Denver")
bl_attempt_dat_london <- subset(bl_attempt_dat, site == "London")
bl_attempt_dat_toronto <- subset(bl_attempt_dat, site == "Toronto")

# Number of each treatment group at each site
table(bl_attempt_dat$site, bl_attempt_dat$treatment)

# Number of subjects with R2* at each site/treatment group
table(!is.na(bl_attempt_dat_denver$avg_c_r2), bl_attempt_dat_denver$treatment)
table(!is.na(bl_attempt_dat_london$avg_c_r2), bl_attempt_dat_london$treatment)
table(!is.na(bl_attempt_dat_toronto$avg_c_r2), bl_attempt_dat_toronto$treatment)

# Number of subjects with R2* AND biopsy in Denver
table(bl_attempt_dat_denver$biopsy_yn == "Yes" & !is.na(bl_attempt_dat_denver$avg_c_r2), 
      bl_attempt_dat_denver$treatment)

# Number of subjects with T1 mapping in Denver (regardless of biopsy)
table(!is.na(subset(attempt_dat, visit == 0 & site == "Denver")$avg_c_t1), 
      subset(attempt_dat, visit == 0 & site == "Denver")$treatment)

# Number of subjects with biopsy in Denver (regardless of MRI)
table(bl_attempt_dat_denver$biopsy_yn, bl_attempt_dat_denver$treatment)

# Number of subjects with T1 AND biopsy in Denver
table(bl_attempt_dat_denver$biopsy_yn == "Yes" & 
        !is.na(subset(attempt_dat, visit == 0 & site == "Denver")$avg_c_t1), 
      bl_attempt_dat_denver$treatment)
```

## Baseline Table (Denver only cohort, Dapa vs. Placebo)

```{r echo = T, results='asis'}
kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + ldl_serum_mmoll + hdl_serum_mmoll + avg_c_t1 + avg_k_t1 + avg_pcascl + avg_c_adc, data = subset(attempt_dat, visit == 0 & site == "Denver")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + ldl_serum_mmoll + hdl_serum_mmoll + avg_c_t1 + avg_k_t1 + avg_pcascl + avg_c_adc, data = subset(attempt_dat, visit == 16 & site == "Denver")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == -4 & site == "Denver")), digits = 1))

# denver_mri_ids <- attempt_dat %>%
#   filter(site == "Denver") %>%
#   filter(!is.na(avg_c_r2)|!is.na(avg_k_r2)) %>%
#   dplyr::select(subject_id, visit) %>%
#   mutate(visit = case_when(visit == -4 ~ "PRE",
#                            visit == 16 ~ "POST"))
# 
# write.csv(denver_mri_ids, "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Clean/Denver_MRI_IDs.csv", row.names = F)
```


## Baseline Table (Biopsy only cohort, Dapa vs. Placebo)

```{r echo = T, results='asis'}
kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse  + ldl_serum_mmoll + hdl_serum_mmoll + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, visit == 0  & biopsy_yn == "Yes")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ mgfr_jodal_bsa + mgfr_jodal + mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2, data = subset(attempt_dat, visit == -4 & biopsy_yn == "Yes")), digits = 1))

kable(summary(arsenal::tableby(treatment ~ age + sex + weight+ height + bmi + hba1c + kwt(emu_urine_acr_mean, "Nmiss","medianq1q3", "range") + creatinine_s + cystatin_c_s + albumin_serum_gl + cgm_tir + diabetes_dx_duration  + creatinine_s + cystatin_c_s + sbp + dbp + cholesterol_serum_mmoll + triglycerides_serum_mmoll + pulse + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, visit == 16  & biopsy_yn == "Yes")), digits = 1))
```

# Clinical variable plots

## Plots over time (Full cohort)

```{r echo = T}
delta_df <- attempt_dat %>%
  filter(!is.na(visit)) %>%
  pivot_wider(
    id_cols = c(subject_id, treatment, site, sex),
    names_from = visit,
    values_from = c(
      hba1c, cgm_tir, weight, mgfr_jodal_bsa, mgfr_jodal, avg_c_r2, avg_k_r2,
      avg_c_t1, avg_k_t1, avg_pcascl, avg_c_adc, avg_ketones,
      cholesterol_serum_mmoll, triglycerides_serum_mmoll, ldl_serum_mmoll, hdl_serum_mmoll
    )
  ) %>%
  mutate(
    # Existing deltas
    delta_hba1c_4 = hba1c_4 - hba1c_0,
    delta_hba1c_16 = hba1c_16 - hba1c_0,
    delta_tir_4 = cgm_tir_4 - cgm_tir_0,
    delta_tir_16 = cgm_tir_16 - cgm_tir_0,
    delta_weight_4 = weight_4 - weight_0,
    delta_weight_16 = weight_16 - weight_0,
    delta_mgfr_jodal_bsa_4 = mgfr_jodal_bsa_4 - `mgfr_jodal_bsa_-4`,
    delta_mgfr_jodal_bsa_16 = mgfr_jodal_bsa_16 - `mgfr_jodal_bsa_-4`,
    delta_mgfr_jodal_4 = mgfr_jodal_4 - `mgfr_jodal_-4`,
    delta_mgfr_jodal_16 = mgfr_jodal_16 - `mgfr_jodal_-4`,
    delta_avg_ketones_4 = avg_ketones_4 - `avg_ketones_-4`,
    delta_avg_ketones_16 = avg_ketones_16 - `avg_ketones_-4`,
    delta_avg_c_r2_4 = avg_c_r2_4 - `avg_c_r2_-4`,
    delta_avg_c_r2_16 = avg_c_r2_16 - `avg_c_r2_-4`,
    delta_avg_k_r2_4 = avg_k_r2_4 - `avg_k_r2_-4`,
    delta_avg_k_r2_16 = avg_k_r2_16 - `avg_k_r2_-4`,
    delta_avg_c_t1_4 = avg_c_t1_4 - `avg_c_t1_0`,
    delta_avg_c_t1_16 = avg_c_t1_16 - `avg_c_t1_0`,
    delta_avg_k_t1_4 = avg_k_t1_4 - `avg_k_t1_0`,
    delta_avg_k_t1_16 = avg_k_t1_16 - `avg_k_t1_0`,
    delta_avg_pcascl_4 = avg_pcascl_4 - `avg_pcascl_0`,
    delta_avg_pcascl_16 = avg_pcascl_16 - `avg_pcascl_0`,
    delta_avg_c_adc_4 = avg_c_adc_4 - `avg_c_adc_0`,
    delta_avg_c_adc_16 = avg_c_adc_16 - `avg_c_adc_0`,
    
    # New lipid deltas (baseline = visit 0)
    delta_cholesterol_4 = cholesterol_serum_mmoll_4 - cholesterol_serum_mmoll_0,
    delta_cholesterol_16 = cholesterol_serum_mmoll_16 - cholesterol_serum_mmoll_0,
    delta_triglycerides_4 = triglycerides_serum_mmoll_4 - triglycerides_serum_mmoll_0,
    delta_triglycerides_16 = triglycerides_serum_mmoll_16 - triglycerides_serum_mmoll_0,
    delta_ldl_4 = ldl_serum_mmoll_4 - ldl_serum_mmoll_0,
    delta_ldl_16 = ldl_serum_mmoll_16 - ldl_serum_mmoll_0,
    delta_hdl_4 = hdl_serum_mmoll_4 - hdl_serum_mmoll_0,
    delta_hdl_16 = hdl_serum_mmoll_16 - hdl_serum_mmoll_0,
    
    # Zero deltas for baseline
    delta_hba1c_0 = 0,
    delta_tir_0 = 0,
    delta_weight_0 = 0,
    `delta_avg_ketones_-4` = 0,
    `delta_mgfr_jodal_bsa_-4` = 0,
    `delta_mgfr_jodal_-4` = 0,
    `delta_avg_c_r2_-4` = 0,
    `delta_avg_k_r2_-4` = 0,
    `delta_avg_pcascl_0` = 0,
    `delta_avg_c_adc_0` = 0,
    `delta_avg_c_t1_0` = 0,
    `delta_avg_k_t1_0` = 0,
    delta_cholesterol_0 = 0,
    delta_triglycerides_0 = 0,
    delta_ldl_0 = 0,
    delta_hdl_0 = 0
  ) %>%
  dplyr::select(subject_id, treatment, site, sex, matches("delta_.*_(-?4|16|0)$")) %>%
  pivot_longer(
    cols = -c(subject_id, treatment, sex, site),
    names_to = c("variable", "visit"),
    names_pattern = "delta_(.*)_(-?\\d+)"
  ) %>% 
  pivot_wider(names_from = variable, values_from = value) %>%
  mutate(visit = as.numeric(visit))

delta_df$treatment <- factor(delta_df$treatment, levels = c("Placebo", "Dapagliflozin 5mg"))

# Add baseline values for covariate adjustment (includes lipid baselines)
delta_df <- delta_df %>% 
  left_join(
    attempt_dat %>%
      group_by(subject_id) %>%
      filter(visit == 0) %>%
      dplyr::select(
        subject_id, age, diabetes_dx_duration, 
        hba1c, cgm_tir, weight,
        cholesterol_serum_mmoll, triglycerides_serum_mmoll, 
        ldl_serum_mmoll, hdl_serum_mmoll
      ) %>%
      dplyr::rename(
        hba1c_bl = hba1c,
        tir_bl = cgm_tir,
        weight_bl = weight,
        cholesterol_bl = cholesterol_serum_mmoll,
        triglycerides_bl = triglycerides_serum_mmoll,
        ldl_bl = ldl_serum_mmoll,
        hdl_bl = hdl_serum_mmoll
      )
  ) %>% 
  left_join(
    attempt_dat %>%
      group_by(subject_id) %>%
      filter(visit == -4) %>%
      dplyr::select(subject_id, mgfr_jodal_bsa, mgfr_jodal,
                    avg_c_r2, avg_k_r2, avg_ketones) %>%
      dplyr::rename(
        mgfr_jodal_bsa_bl = mgfr_jodal_bsa,
        mgfr_jodal_bl = mgfr_jodal,
        avg_c_r2_bl = avg_c_r2,
        avg_k_r2_bl = avg_k_r2,
        avg_ketones_bl = avg_ketones
      )
  )

s3saveRDS(delta_df, bucket = "attempt", 
          object = "cleaned_data/attempt_delta_df.rds",
          region = "")
```



### mGFR (BSA)

```{r echo = T, include = F, eval = T}
plot_mean_ci_stars(attempt_dat,
                   y_var = mgfr_jodal_bsa, 
                   y_axis_title = expression("Measured GFR (ml min"^{-1}~"1.73 m"^{-2}~")"),
                   visits_to_plot = c(-4, 16),
                   baseline_visit = -4,
                   legend_position = c(0.3,0.2),
                   covariates = "site")
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_mgfr_jodal_bsa <- lmer(
  mgfr_jodal_bsa ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = subset(delta_df, visit %in% c(-4, 16))
)

# summary(model_mgfr_jodal_bsa)

emm_mgfr_jodal_bsa <- emmeans(model_mgfr_jodal_bsa, 
                              ~ treatment | visit, data = subset(delta_df, visit %in% c(-4, 16)))
summary(emm_mgfr_jodal_bsa)

# Pairwise comparisons at each week
contrast(emm_mgfr_jodal_bsa, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "mgfr_jodal_bsa", baseline_var = "mgfr_jodal_bsa_bl", visit_week = 16)
```


#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "mgfr_jodal_bsa",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  proper_name =  "mGFR (BSA)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_jodal_bsa_category_bars.jpeg"
)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_jodal_bsa_category_bars.jpeg){width=70%}

### mGFR (Raw)

```{r echo = T, include = F, eval = T}
plot_mean_ci_stars(attempt_dat,
                   y_var = mgfr_jodal, 
                   y_axis_title = expression("Measured GFR (ml min"^{-1}~"1.73 m"^{-2}~")"),
                   visits_to_plot = c(-4, 16),
                   baseline_visit = -4,
                   legend_position = c(0.3,0.2),
                   covariates = "site")
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_raw_overtime.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_raw_overtime.jpeg){width=70%}


```{r echo = T, include = T, eval = T}
model_mgfr_jodal <- lmer(
  mgfr_jodal ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = subset(delta_df, visit %in% c(-4, 16))
)

# summary(model_mgfr_jodal)

emm_mgfr_jodal <- emmeans(model_mgfr_jodal, 
                              ~ treatment | visit, data = subset(delta_df, visit %in% c(-4, 16)))
summary(emm_mgfr_jodal)

# Pairwise comparisons at each week
contrast(emm_mgfr_jodal, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "mgfr_jodal", baseline_var = "mgfr_jodal_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "mgfr_jodal",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  proper_name =  "mGFR (Raw)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_jodal_raw_category_bars.jpeg"
)
```
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_jodal_raw_category_bars.jpeg){width=70%}

### HbA1c

```{r echo = T}
plot_mean_ci_stars2 <- function(data, y_var, y_axis_title, legend_position = c(0.6, 0.8)) {
  
  dodge_val <- 0.08
  
  # Dynamically evaluate y_var
  y_sym <- rlang::ensym(y_var)
  
  # Keep only visits 0, 4, 16
  visits_to_plot <- c(-4,0,4,16,18)
  data <- data %>% filter(visit %in% visits_to_plot)
  
  # 1. Calculate mean and 95% CI (t-based)
  mean_dat <- data %>%
    group_by(treatment_arm, visit) %>%
    dplyr::summarise(
      n = sum(!is.na(!!y_sym)),
      mean_y = mean(!!y_sym, na.rm = TRUE),
      sd_y = sd(!!y_sym, na.rm = TRUE),
      t_crit = qt(0.975, df = n - 1),
      lower = mean_y - (t_crit * sd_y / sqrt(n)),
      upper = mean_y + (t_crit * sd_y / sqrt(n)),
      .groups = "drop"
    )
  
  # 2. Run t-tests between treatments at each visit
  ttest_results <- data %>%
    group_by(visit) %>%
    dplyr::summarise(
      p_value = tryCatch(
        t.test(!!y_sym ~ treatment_arm)$p.value,
        error = function(e) NA_real_
      ),
      .groups = "drop"
    ) %>%
    mutate(
      stars = case_when(
        p_value < 0.001 ~ "***",
        p_value < 0.01  ~ "**",
        p_value < 0.05  ~ "*",
        TRUE            ~ ""
      )
    )
  
  # 3. Set y position for stars (above upper CI)
  ttest_results <- ttest_results %>%
    left_join(mean_dat %>% group_by(visit) %>%
                dplyr::summarise(max_upper = max(upper), .groups = "drop"),
              by = "visit") %>%
    mutate(y_position = max_upper + 0.05 * max_upper)  # 5% above max_upper
  
  # 4. Plot
  p <- data %>%
    ggplot(aes(x = factor(visit, levels = visits_to_plot), y = !!y_sym, color = treatment_arm)) +
      geom_errorbar(data = mean_dat, 
                    aes(x = factor(visit, levels = visits_to_plot), ymin = lower, ymax = upper,  group = treatment_arm), 
                    inherit.aes = FALSE,
                    width = 0.1, size = 0.3, position = position_dodge(width = dodge_val),
                    color = "black") +
      geom_line(data = mean_dat, aes(x = factor(visit, levels = visits_to_plot), y = mean_y, group = treatment_arm), 
                position = position_dodge(width = dodge_val),
                size = 1.2) +
      geom_point(data = mean_dat, aes(x = factor(visit, levels = visits_to_plot), y = mean_y, shape = treatment_arm), 
                 size = 11, position = position_dodge(width = dodge_val), color = "white") +
      geom_point(data = mean_dat, aes(x = factor(visit, levels = visits_to_plot), y = mean_y, shape = treatment_arm), 
                 size = 7, position = position_dodge(width = dodge_val)) +
      geom_text(data = ttest_results, 
                aes(x = factor(visit, levels = visits_to_plot), y = y_position, label = stars), 
                inherit.aes = FALSE,
                size = 6, vjust = 0) +
      scale_shape_manual(values = c("A" = 15,   # Square
                                    "B" = 18)) +   # Diamond
      scale_color_manual(values = c("A" = "#f8ae9d", 
                                    "B" = "#a7b298")) +
      scale_x_discrete(expand = expansion(mult = c(0.3, 0.3))) +
      scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
      theme(panel.background = element_blank(),
            legend.position = legend_position,
            legend.justification = c(1, 1),
            legend.background = element_blank()) +
      labs(x = "Visit", 
           y = y_axis_title,
           color = NULL, shape = NULL) +
    guides(color = guide_legend(override.aes = list(size=5)))
  
  return(p)
}
```


```{r echo = T, include = F}
attempt_031425_raw <- read.csv("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Data Raw/ATTEMPT_DenverDataRequest_20250314.csv")

attempt_031425_raw <- attempt_031425_raw %>%
  mutate(visit = case_when(visit == "V1 Screening" ~ -4,
                           visit == "V2 Randomization" ~ 0,
                           visit == "V3 Safety" ~ 4,
                           visit == "V4 Final" ~ 16,
                           visit == "V5 Follow-Up" ~ 18))

attempt_031425_raw$hba1c_percent <- as.numeric(attempt_031425_raw$hba1c_percent)
plot_mean_ci_stars2(attempt_031425_raw, 
                   y_var = hba1c_percent, y_axis_title = "HbA1c (%)", legend_position = c(0.3,0.2))

plot_mean_ci_stars(subset(attempt_dat), 
                   y_var = hba1c, y_axis_title = "HbA1c (%)", legend_position = c(0.3,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_overtime_2.jpeg", width = 7, height = 5)

plot_mean_ci_stars(data = subset(delta_df), y_var = hba1c, 
                   y_axis_title = "\u0394 HbA1c (%)",
                  legend_position = c(0.3,0.2),
                  visits_to_plot = c(0, 4, 16)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_hba1c_overtime_2.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_overtime_2.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_hba1c_overtime_2.jpeg){width=70%}



```{r echo = T, include = T, eval = T}
model_hba1c <- lmer(
  hba1c ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = delta_df
)

# summary(model_hba1c)

emm_hba1c <- emmeans(model_hba1c, ~ treatment | visit, 
                     data = delta_df)
summary(emm_hba1c)

# Pairwise comparisons at each week
contrast(emm_hba1c, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "hba1c", baseline_var = "hba1c_bl", visit_week = 4)
run_ancova(delta_df, outcome_var = "hba1c", baseline_var = "hba1c_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "hba1c",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = 0,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "HbA1c (%)",
    bin_breaks = c(-Inf, 7.5, 8, Inf),
  bin_labels = c("<7.5", "7.5-8", "\u22658"),
  y_limits = c(-2, 2),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_category_bars.jpeg"
)

plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "mgfr_jodal_bsa",
  value_var = "hba1c",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = 0,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "HbA1c (%)",
  x_label = "\nBaseline mGFR (BSA)",
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_by_mgfr_category_bars.jpeg"
)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_by_mgfr_category_bars.jpeg){width=70%}


### TIR

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(0, 16)), 
                  visits_to_plot = c(0, 16),
                   y_var = cgm_tir, y_axis_title = "Time in Range (%)", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/tir_overtime.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, visit %in% c(0,16)), 
                   visits_to_plot = c(0, 16),
                   y_var = tir, y_axis_title = "\u0394 Time in Range (%)",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_tir_overtime.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/tir_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_tir_overtime.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_tir <- lmer(
  tir ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = delta_df
)

# summary(model_tir)

emm_tir <- emmeans(model_tir, ~ treatment | visit)
summary(emm_tir)

# Pairwise comparisons at each week
contrast(emm_tir, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "tir", baseline_var = "tir_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "cgm_tir",
  visit_var = "visit",
  visit_baseline = 0,
  visit_pre = 0,
  visit_post = 16,
  bin_breaks = c(-Inf, 45, 60, Inf),
  bin_labels = c("<45", "45-60", "\u226560"),
  proper_name =  "TIR (%)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/cgm_tir_category_bars.jpeg"
)


attempt_dat %>%
  mutate(visit = case_when(visit <= 0 ~ "baseline", T ~ "followup")) %>%
  ggplot(aes(x = mgfr_jodal_bsa, y = cgm_tir, color = treatment, shape = visit)) +
  geom_point()

plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "mgfr_jodal_bsa",
  value_var = "cgm_tir",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = 0,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "TIR (%)",
  x_label = "\nBaseline mGFR (BSA)",
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/cgm_tir_by_mgfr_category_bars.jpeg"
)
```
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/cgm_tir_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/cgm_tir_by_mgfr_category_bars.jpeg){width=70%}

### Weight

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(0, 4, 16)), 
                   visits_to_plot = c(0, 4, 16),
                   y_var = weight, y_axis_title = "Weight", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_overtime.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = delta_df, 
                   visits_to_plot = c(0, 4, 16),
                   y_var = weight, y_axis_title = "\u0394 Weight",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_weight_overtime.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_weight_overtime.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_weight <- lmer(
  weight ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = delta_df
)

# summary(model_weight)

emm_weight <- emmeans(model_weight, ~ treatment | visit)
summary(emm_weight)

# Pairwise comparisons at each week
contrast(emm_weight, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "weight", baseline_var = "weight_bl", visit_week = 4)
run_ancova(delta_df, outcome_var = "weight", baseline_var = "weight_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "weight",
  visit_var = "visit",
  visit_baseline = 0,
  visit_pre = 0,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Weight (kg)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_category_bars.jpeg"
)

plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "mgfr_jodal_bsa",
  value_var = "weight",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = 0,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "Weight (kg)",
  x_label = "\nBaseline mGFR (BSA)",
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_by_mgfr_category_bars.jpeg"
)
```
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_by_mgfr_category_bars.jpeg){width=70%}

### Avg cortex R2*

```{r echo = T, include = F}
plot_mean_ci_stars(attempt_dat, baseline_visit = -4,
                   visits_to_plot = c(-4, 16),
                   y_var = avg_c_r2, y_axis_title = "Avg cortical R2*", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime.jpeg", width = 7, height = 5)

plot_mean_ci_stars(data = delta_df, baseline_visit = -4,
                   visits_to_plot = c(-4, 16),
                   y_var = avg_c_r2, y_axis_title = "\u0394 Avg cortical R2*",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_avg_c_r2 <- lmer(
  avg_c_r2 ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = delta_df
)

# summary(model_avg_c_r2)

emm_avg_c_r2 <- emmeans(model_avg_c_r2, ~ treatment | visit)
summary(emm_avg_c_r2)

# Pairwise comparisons at each week
contrast(emm_avg_c_r2, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "avg_c_r2", baseline_var = "avg_c_r2_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "avg_c_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Cortical R2*",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_category_bars.jpeg"
)

# by baseline mGFR
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "mgfr_jodal_bsa",
  value_var = "avg_c_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "Cortical R2*",
  x_label = "\nBaseline mGFR (BSA)",
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_mgfr_category_bars.jpeg"
)

# by baseline HbA1c
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "hba1c",
  value_var = "avg_c_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "Cortical R2*",
  x_label = "\nBaseline HbA1c (%)",
  bin_breaks = c(-Inf, 7.5, 8, Inf),
  bin_labels = c("<7.5", "7.5-8", "\u22658"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_hba1c_category_bars.jpeg"
)

# by baseline weight
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "weight",
  value_var = "avg_c_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Cortical R2*",
  x_label = "\nBaseline Weight (kg)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_weight_category_bars.jpeg"
)

# by baseline tir
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "cgm_tir",
  value_var = "avg_c_r2",
  visit_var = "visit",
  visit_baseline = 0,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Cortical R2*",
  x_label = "\nBaseline TIR (%)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_cgm_tir_category_bars.jpeg"
)
```
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_mgfr_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_hba1c_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_weight_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_by_cgm_tir_category_bars.jpeg){width=70%}

### Avg kidney R2*

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(-4, 16)), 
                   visits_to_plot = c(-4, 16), baseline_visit = -4,
                   y_var = avg_k_r2, y_axis_title = "Avg kidney R2*", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = delta_df, 
                   visits_to_plot = c(-4, 16),baseline_visit = -4,
                   y_var = avg_k_r2, y_axis_title = "\u0394 Avg kidney R2*",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime.jpeg){width=70%}

Difference in kidney R2* correlation with baseline kidney R2*

```{r echo = T, include = T, eval = T}
cor.test(subset(delta_df, visit == 16)$avg_k_r2, 
          subset(delta_df, visit == 16)$avg_k_r2_bl)
```


```{r echo = T, include = T, eval = T}
model_avg_k_r2 <- lmer(
  avg_k_r2 ~ treatment * factor(visit) +
    age + sex + diabetes_dx_duration + 
    (1 | subject_id),
  data = delta_df
)

# summary(model_avg_k_r2)

emm_avg_k_r2 <- emmeans(model_avg_k_r2, ~ treatment | visit)
summary(emm_avg_k_r2)

# Pairwise comparisons at each week
contrast(emm_avg_k_r2, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(delta_df, outcome_var = "avg_k_r2", baseline_var = "avg_k_r2_bl", visit_week = 16)
```

#### Tertiles/bins

```{r echo = T, include = F, eval = T}
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  value_var = "avg_k_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  top_padding = 0.5,
  proper_name =  "Kidney R2*",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_category_bars.jpeg"
)

plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "mgfr_jodal_bsa",
  value_var = "avg_k_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "Kidney R2*",
  x_label = "\nBaseline mGFR (BSA)",
  bin_breaks = c(-Inf, 90, 110, Inf),
  bin_labels = c("<90", "90-110", "\u2265110"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_mgfr_category_bars.jpeg"
)

# by baseline HbA1c
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "hba1c",
  value_var = "avg_k_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = F,
  proper_name =  "Kidney R2*",
  x_label = "\nBaseline HbA1c (%)",
  bin_breaks = c(-Inf, 7.5, 8, Inf),
  bin_labels = c("<7.5", "7.5-8", "\u22658"),
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_hba1c_category_bars.jpeg"
)

# by baseline weight
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "weight",
  value_var = "avg_k_r2",
  visit_var = "visit",
  visit_baseline = -4,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Kidney R2*",
  x_label = "\nBaseline Weight (kg)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_weight_category_bars.jpeg"
)

# by baseline tir
plot_delta_by_category(
  data = attempt_dat,
  id_var = "subject_id",
  treatment_var = "treatment",
  x_var = "cgm_tir",
  value_var = "avg_k_r2",
  visit_var = "visit",
  visit_baseline = 0,
  visit_pre = -4,
  visit_post = 16,
  bin_by_tertiles = T,
  proper_name =  "Kidney R2*",
  x_label = "\nBaseline TIR (%)",
  output_path = "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_cgm_tir_category_bars.jpeg"
)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_mgfr_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_hba1c_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_weight_category_bars.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_by_cgm_tir_category_bars.jpeg){width=70%}

### Correlations

*Pooled*

```{r echo = T}
vars_to_replicate <- c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2", 
                         "avg_c_t1", "avg_k_t1", "avg_pcascl", "avg_c_adc")
for (var in vars_to_replicate) {
  delta_df[[paste0(var, "_1")]] <- delta_df[[var]]
}
dict = data.frame(hba1c = "\u0394 HbA1c",
                  tir = "\u0394 TIR",
                  weight = "\u0394 Weight",
                  mgfr_jodal = "\u0394 mGFR",
                  mgfr_jodal_bsa = "\u0394 mGFR (BSA)",
                  avg_c_r2 = "\u0394 Cortical R2*",
                  avg_k_r2 = "\u0394 Kidney R2*",
                  avg_c_t1 = "\u0394 Cortical T1",
                  avg_k_t1 = "\u0394 Kidney T1",
                  avg_pcascl = "\u0394 ASL",
                  avg_c_adc = "\u0394 Cortical ADC")
for (var in vars_to_replicate) {
  dict[[paste0(var, "_1")]] <- dict[[var]]
}

corr_plot_modified(data = subset(delta_df, visit == 16),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1"),
                   dict = dict,
                   adj_var = "treatment")

```

*Placebo*

```{r echo = T}
corr_plot_modified(data = subset(delta_df, treatment == "Placebo" & visit == 16),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1"),
                   dict = dict)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16)$mgfr_jodal)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16)$avg_c_r2)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16)$avg_k_r2)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16)$avg_k_r2,
         subset(delta_df, treatment == "Placebo" & visit == 16)$mgfr_jodal)
```

*Dapagliflozin*

```{r echo = T}
corr_plot_modified(data = subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1"),
                   dict = dict)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$mgfr_jodal)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$avg_c_r2)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$avg_k_r2)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$avg_k_r2,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16)$mgfr_jodal)
```

## Plots over time (Denver cohort)

### mGFR (BSA)

```{r echo = T, include = F}
# Primary outcome paper figure 2A (Denver only)
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(-4,16) & site == "Denver"),
                   y_var = mgfr_jodal_bsa, 
                   y_axis_title = expression("Measured GFR (ml min"^{-1}~"1.73 m"^{-2}~")"),
                   legend_position = c(0.3,0.2),
                   visits_to_plot = c(-4, 16),
                   baseline_visit = -4,
                   covariates = "site")
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime_denver.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_overtime_denver.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_mgfr_jodal_bsa <- lmer(
  mgfr_jodal_bsa ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_mgfr_jodal_bsa)

emm_mgfr_jodal_bsa <- emmeans(model_mgfr_jodal_bsa, ~ treatment | visit)
summary(emm_mgfr_jodal_bsa)

# Pairwise comparisons at each week
contrast(emm_mgfr_jodal_bsa, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

### mGFR (Raw)

```{r echo = T, include = F}
# Primary outcome paper figure 2A (Denver only)
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(-4,16) & site == "Denver"),
                   y_var = mgfr_jodal, 
                   y_axis_title = expression("Measured GFR (ml min"^{-1}~"1.73 m"^{-2}~")"),
                   legend_position = c(0.3,0.2),
                   visits_to_plot = c(-4, 16),
                   baseline_visit = -4,
                   covariates = "site")
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_raw_overtime_denver.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/mgfr_raw_overtime_denver.jpeg){width=70%}


```{r echo = T, include = T, eval = T}
model_mgfr_jodal <- lmer(
  mgfr_jodal ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df,site == "Denver")
)

# summary(model_mgfr_jodal)

emm_mgfr_jodal <- emmeans(model_mgfr_jodal, ~ treatment | visit)
summary(emm_mgfr_jodal)

# Pairwise comparisons at each week
contrast(emm_mgfr_jodal, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

### HbA1c

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(0, 4, 16) & site == "Denver"), 
                   y_var = hba1c, y_axis_title = "HbA1c (%)", legend_position = c(0.3,0.2),
                   visits_to_plot = c(0, 4, 16))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_overtime_2_denver.jpeg", width = 7, height = 5)

plot_mean_ci_stars(data = subset(delta_df, site == "Denver"), y_var = hba1c, 
                   y_axis_title = "\u0394 HbA1c (%)",
                  legend_position = c(0.3,0.2),
                  visits_to_plot = c(0, 4, 16)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_hba1c_overtime_2_denver.jpeg", width = 7, height = 5)
```

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/hba1c_overtime_2_denver.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_hba1c_overtime_2_denver.jpeg){width=70%}



```{r echo = T, include = T, eval = T}
model_hba1c <- lmer(
  hba1c ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_hba1c)

emm_hba1c <- emmeans(model_hba1c, ~ treatment | visit)
summary(emm_hba1c)

# Pairwise comparisons at each week
contrast(emm_hba1c, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

### TIR

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(0, 16) & site == "Denver"), 
                   y_var = cgm_tir, y_axis_title = "Time in Range (%)", legend_position = c(0.6,0.2),
                   visits_to_plot = c(0, 16))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/tir_overtime_denver.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, visit %in% c(0,16) & site == "Denver"), 
                   y_var = tir, y_axis_title = "\u0394 Time in Range (%)",
                   legend_position = c(0.3,0.8),
                    visits_to_plot = c(0, 16)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_tir_overtime_denver.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/tir_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_tir_overtime_denver.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_tir <- lmer(
  tir ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_tir)

emm_tir <- emmeans(model_tir, ~ treatment | visit)
summary(emm_tir)

# Pairwise comparisons at each week
contrast(emm_tir, method = "revpairwise", by = "visit", adjust = "bonferroni")
```


### Weight

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, visit %in% c(0, 4,16) & site == "Denver"), 
                   y_var = weight, y_axis_title = "Time in Range (%)", legend_position = c(0.6,0.2),
                    visits_to_plot = c(0, 4, 16))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_overtime_denver.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, site == "Denver"), 
                   y_var = weight, y_axis_title = "\u0394 Time in Range (%)",
                   legend_position = c(0.3,0.8),
                    visits_to_plot = c(0, 4, 16)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_weight_overtime_denver.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/weight_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_weight_overtime_denver.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_weight <- lmer(
  weight ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_weight)

emm_weight <- emmeans(model_weight, ~ treatment | visit)
summary(emm_weight)

# Pairwise comparisons at each week
contrast(emm_weight, method = "revpairwise", by = "visit", adjust = "bonferroni")
```


### Avg cortex R2*

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, site == "Denver" & visit %in% c(-4, 16)), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_c_r2, y_axis_title = "Avg cortical R2*", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime_denver.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, site == "Denver"), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_c_r2, y_axis_title = "\u0394 Avg cortical R2*",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime_denver.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_c_r2_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_c_r2_overtime_denver.jpeg){width=70%}

```{r echo = T, include = T, eval = T}
model_avg_c_r2 <- lmer(
  avg_c_r2 ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_avg_c_r2)

emm_avg_c_r2 <- emmeans(model_avg_c_r2, ~ treatment | visit)
summary(emm_avg_c_r2)

# Pairwise comparisons at each week
contrast(emm_avg_c_r2, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "avg_c_r2", baseline_var = "avg_c_r2_bl", visit_week = 16)
```

### Avg kidney R2*

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, site == "Denver" & visit %in% c(-4, 16)), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_k_r2, y_axis_title = "Avg kidney R2*", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime_denver.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, site == "Denver"), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_k_r2, y_axis_title = "\u0394 Avg kidney R2*",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime_denver.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_k_r2_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_k_r2_overtime_denver.jpeg){width=70%}


Difference in kidney R2* correlation with baseline kidney R2*

```{r echo = T, include = T, eval = T}
cor.test(subset(delta_df, site == "Denver" & visit == 16)$avg_k_r2, 
          subset(delta_df, site == "Denver" & visit == 16)$avg_k_r2_bl)
```


```{r echo = T, include = T, eval = T}
model_avg_k_r2 <- lmer(
  avg_k_r2 ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_avg_k_r2)

emm_avg_k_r2 <- emmeans(model_avg_k_r2, ~ treatment | visit)
summary(emm_avg_k_r2)

# Pairwise comparisons at each week
contrast(emm_avg_k_r2, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "avg_k_r2", baseline_var = "avg_k_r2_bl", visit_week = 16)
```


### Ketones (avg of mgfr_ketone_iv and mgfr_ketones_pre)

```{r echo = T, include = F}
plot_mean_ci_stars(subset(attempt_dat, site == "Denver" & visit %in% c(-4, 16)), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_ketones, y_axis_title = "Ketones", legend_position = c(0.6,0.2))
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_ketones_overtime_denver.jpeg", width = 7, height = 5)


plot_mean_ci_stars(data = subset(delta_df, site == "Denver"), 
                   visits_to_plot = c(-4, 16),
                   y_var = avg_ketones, y_axis_title = "\u0394 Ketones",
                   legend_position = c(0.3,0.8)) +
  geom_hline(yintercept = 0, linetype = "dashed", linewidth = 0.1)
ggsave("/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_ketones_overtime_denver.jpeg", width = 7, height = 5)
```


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/avg_ketones_overtime_denver.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Plot over time/delta_avg_ketones_overtime_denver.jpeg){width=70%}


Difference in ketones correlation with baseline ketones

```{r echo = T, include = T, eval = T}
cor.test(subset(delta_df, site == "Denver" & visit == 16)$avg_ketones, 
          subset(delta_df, site == "Denver" & visit == 16)$avg_ketones_bl)
```


```{r echo = T, include = T, eval = T}
model_avg_ketones <- lmer(
  avg_ketones ~ treatment * factor(visit) +
    (1 | subject_id),
  data = subset(delta_df, site == "Denver")
)

# summary(model_avg_ketones)

emm_avg_ketones <- emmeans(model_avg_ketones, ~ treatment | visit)
summary(emm_avg_ketones)

# Pairwise comparisons at each week
contrast(emm_avg_ketones, method = "revpairwise", by = "visit", adjust = "bonferroni")
```

```{r echo = T}
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "avg_ketones", baseline_var = "avg_ketones_bl", visit_week = 16)
```

### Correlations

*Pooled*

```{r echo = T}
corr_plot_modified(data = subset(delta_df, visit == 16 & site == "Denver"),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2", 
                         "avg_c_t1", "avg_k_t1", "avg_pcascl", "avg_c_adc"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1", 
                         "avg_c_t1_1", "avg_k_t1_1", "avg_pcascl_1", "avg_c_adc_1"),
                   dict = dict,
                   adj_var = "treatment")
```

*Placebo*

```{r echo = T}
corr_plot_modified(data = subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver"),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2", 
                         "avg_c_t1", "avg_k_t1", "avg_pcascl", "avg_c_adc"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1", 
                         "avg_c_t1_1", "avg_k_t1_1", "avg_pcascl_1", "avg_c_adc_1"),
                   dict = dict)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$mgfr_jodal)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$avg_c_r2)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$avg_k_r2)

cor.test(subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$avg_k_r2,
         subset(delta_df, treatment == "Placebo" & visit == 16 & site == "Denver")$mgfr_jodal)
```

*Dapagliflozin*

Not enough obs for ASL in dapagliflozin group

```{r echo = T}
corr_plot_modified(data = subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver"),
                   X = c("hba1c", "tir", "weight", "mgfr_jodal", "mgfr_jodal_bsa", "avg_c_r2", "avg_k_r2", 
                         "avg_c_t1", "avg_k_t1", "avg_c_adc"),
                   Y = c("hba1c_1", "tir_1", "weight_1", "mgfr_jodal_1", "mgfr_jodal_bsa_1", "avg_c_r2_1", "avg_k_r2_1", 
                         "avg_c_t1_1", "avg_k_t1_1", "avg_c_adc_1"),
                   dict = dict)


cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$mgfr_jodal)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$avg_c_r2)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$hba1c,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$avg_k_r2)

cor.test(subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$avg_k_r2,
         subset(delta_df, treatment == "Dapagliflozin 5mg" & visit == 16 & site == "Denver")$mgfr_jodal)
```


# Mixed models for clinical vars

## MRI Mixed model (Full cohort, Dapa vs. Placebo over time, Unadjusted)

```{r echo = T}
outcomes <- c(
  "mri_r2_cortex_l", "mri_r2_cortex_r", "avg_c_r2",
  "mri_r2_kidney_l", "mri_r2_kidney_r", "avg_k_r2"
)

# Run models and store summaries in a list
model_summaries <- lapply(outcomes, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = attempt_dat)
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- outcomes

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)


```

## MRI Mixed model (Full cohort, Dapa vs. Placebo over time, Adjusted for age, sex, diabetes duration, site, and BMI)

```{r echo = T}
outcomes <- c(
  "mri_r2_cortex_l", "mri_r2_cortex_r", "avg_c_r2",
  "mri_r2_kidney_l", "mri_r2_kidney_r", "avg_k_r2"
)

# Run models and store summaries in a list
model_summaries <- lapply(outcomes, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + age + sex + diabetes_dx_duration + site + bmi + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = attempt_dat)
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- outcomes

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)
```

## MRI Mixed model (Denver only cohort, Dapa vs. Placebo over time)

```{r echo = T}
# kable(summary(arsenal::tableby(visit ~ mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2 + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, site == "Denver"), strata = treatment), digits = 1))

outcomes <- c(
  "mri_r2_cortex_l", "mri_r2_cortex_r", "avg_c_r2",
  "mri_r2_kidney_l", "mri_r2_kidney_r", "avg_k_r2",
  "t1_l_cortex", "t1_r_cortex", "avg_c_t1",
  "t1_l_kidney", "t1_r_kidney", "avg_k_t1",
  "adc_left", "adc_right", "avg_c_adc",
  "pcasl3d_left", "pcasl3d_right", "avg_pcascl",
  "volume_left", "volume_right", "tkv"
)

# Run models and store summaries in a list
model_summaries <- lapply(outcomes, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = subset(attempt_dat, site == "Denver"))
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- outcomes

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)
```

## MRI Mixed model (Biopsy only cohort, Dapa vs. Placebo over time)

```{r echo = T}
# kable(summary(arsenal::tableby(visit ~ mri_r2_cortex_l + mri_r2_cortex_r + avg_c_r2 + mri_r2_kidney_l + mri_r2_kidney_r + avg_k_r2 + t1_l_cortex + t1_r_cortex + avg_c_t1 + t1_l_kidney + t1_r_kidney + avg_k_t1 + adc_left + adc_right + avg_c_adc + pcasl3d_left + pcasl3d_right + avg_pcascl + volume_left + volume_right + tkv, data = subset(attempt_dat, site == "Denver"), strata = treatment), digits = 1))

outcomes <- c(
  "mri_r2_cortex_l", "mri_r2_cortex_r", "avg_c_r2",
  "mri_r2_kidney_l", "mri_r2_kidney_r", "avg_k_r2",
  "t1_l_cortex", "t1_r_cortex", "avg_c_t1",
  "t1_l_kidney", "t1_r_kidney", "avg_k_t1",
  "adc_left", "adc_right", "avg_c_adc",
  "pcasl3d_left", "pcasl3d_right", "avg_pcascl",
  "volume_left", "volume_right", "tkv"
)

# Run models and store summaries in a list
model_summaries <- lapply(outcomes, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = subset(attempt_dat, biopsy_yn == "Yes"))
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- outcomes

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)
```


# Lipid profiles
## Lipid profile Mixed model (Full cohort, Dapa vs. Placebo over time, Unadjusted)

```{r echo = T}
lipid_profiles <- c(
  "cholesterol_serum_mmoll", "triglycerides_serum_mmoll", 
  "ldl_serum_mmoll","hdl_serum_mmoll"
)

# Run models and store summaries in a list
model_summaries <- lapply(lipid_profiles, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = subset(attempt_dat, visit %in% c(0, 16)))
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- lipid_profiles

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)

# hist(attempt_dat$cholesterol_serum_mmoll)
# hist(attempt_dat$triglycerides_serum_mmoll)
# hist(attempt_dat$ldl_serum_mmoll)
# hist(attempt_dat$hdl_serum_mmoll)

# subset(attempt_dat, visit %in% c(18))$cholesterol_serum_mmoll

```

```{r echo = T}
delta_df$treatment <- relevel(delta_df$treatment, ref = "Dapagliflozin 5mg")
run_ancova(delta_df, outcome_var = "cholesterol", baseline_var = "cholesterol_bl", visit_week = 16)$contrasts
run_ancova(delta_df, outcome_var = "triglycerides", baseline_var = "triglycerides_bl", visit_week = 16)$contrasts
run_ancova(delta_df, outcome_var = "ldl", baseline_var = "ldl_bl", visit_week = 16)$contrasts
run_ancova(delta_df, outcome_var = "hdl", baseline_var = "hdl_bl", visit_week = 16)$contrasts
```


## Lipid profiles Mixed model (Denver only cohort, Dapa vs. Placebo over time)

```{r echo = T}
lipid_profiles <- c(
  "cholesterol_serum_mmoll", "triglycerides_serum_mmoll", 
  "ldl_serum_mmoll","hdl_serum_mmoll"
)

# Run models and store summaries in a list
model_summaries <- lapply(lipid_profiles, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = subset(attempt_dat, site == "Denver" & visit %in% c(0, 16)))
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- lipid_profiles

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)
# 
# hist(subset(attempt_dat, site == "Denver")$cholesterol_serum_mmoll)
# hist(subset(attempt_dat, site == "Denver")$triglycerides_serum_mmoll)
# hist(subset(attempt_dat, site == "Denver")$ldl_serum_mmoll)
# hist(subset(attempt_dat, site == "Denver")$hdl_serum_mmoll)
```


```{r echo = T}
# Reverse reference level (make Placebo the comparison)
delta_df$treatment <- relevel(delta_df$treatment, ref = "Dapagliflozin 5mg")
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "cholesterol", baseline_var = "cholesterol_bl", visit_week = 16)$contrasts
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "triglycerides", baseline_var = "triglycerides_bl", visit_week = 16)$contrasts
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "ldl", baseline_var = "ldl_bl", visit_week = 16)$contrasts
run_ancova(subset(delta_df, site == "Denver"), outcome_var = "hdl", baseline_var = "hdl_bl", visit_week = 16)$contrasts

```

## Lipid profiles Mixed model (Biopsy only cohort, Dapa vs. Placebo over time)

```{r echo = T}
lipid_profiles <- c(
  "cholesterol_serum_mmoll", "triglycerides_serum_mmoll", 
  "ldl_serum_mmoll","hdl_serum_mmoll"
)

# Run models and store summaries in a list
model_summaries <- lapply(lipid_profiles, function(var) {
  cat("\n=====================\n")
  cat("Outcome:", var, "\n")
  cat("=====================\n")
  
  formula <- as.formula(paste(var, "~ visit * treatment + (1 | subject_id)"))
  model <- lmerTest::lmer(formula, data = subset(attempt_dat, biopsy_yn == "Yes"))
  print(summary(model))
  
  return(summary(model))
})
names(model_summaries) <- lipid_profiles

fixed_effects_df <- do.call(rbind, lapply(names(model_summaries), function(var) {
  model_summary <- model_summaries[[var]]
  fe <- coef(summary(model_summary)) %>%
    as.data.frame() %>%
    tibble::rownames_to_column("term") %>%
    mutate(outcome = var, .before = 1)
}))

clean_results <- fixed_effects_df %>%
  dplyr::rename(
    estimate = Estimate,
    std_error = `Std. Error`,
    df = df,
    t_value = `t value`,
    p_value = `Pr(>|t|)`
  ) %>%
  dplyr::select(outcome, term, estimate, std_error, df, t_value, p_value) %>%
  filter(term == "visit:treatmentDapagliflozin 5mg")

kable(clean_results, digits = 3)
```


# Transcriptomics

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_celltype.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_kpmpcelltype.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_treat.jpeg){width=50%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/UMAP/attempt_umap_visit.jpeg){width=50%}


---

## DEGs (Dapagliflozin vs. Placebo) - DiD using NEBULA across all celltypes

### KPMP Cell types

#### Top 2000 HVG

```{r echo = T, results = 'asis', eval = T}
celltype_groups <- list(
PT = c("PT-S1/S2", "PT-S3", "aPT"),
TAL = c("C-TAL-1", "C-TAL-2", "aTAL", "dTAL"),
PC = c("CCD-PC", "CNT-PC", "dCCD-PC", "M-PC", "tPC-IC"),
IC = c("IC-A", "IC-B", "aIC"),
EC = c("EC-AVR", "EC-GC", "EC-PTC", "EC-AEA", "EC-LYM", "EC/VSMC"),
Immune = c("MAC", "MON", "cDC", "pDC", "CD4+ T", "CD8+ T", "B", "NK"),
Immune_myeloid = c("MAC", "MON", "cDC", "pDC"),
Immune_lymphoid = c("CD4+ T", "CD8+ T", "B", "NK"),
VSMC_P_FIB = c("VSMC/P", "FIB"),
POD = "POD"
)


celltype_paragraphs <- list(
PT = "In PT cells, dapagliflozin caused downregulation of key metabolic and oxidative-stress-related
genes, aligning with improved tubular metabolic health (Fig. #). Notably, the gluconeogenic enzyme PCK1
(phosphoenolpyruvate carboxykinase) was markedly decreased (log2FC  2.1), potentially reversing the diabetes driven increase in renal gluconeogenesis. Multiple metallothionein (MT) isoforms (MT1X, MT1G, MT2A, etc.) were sharply downregulated (log2FC ~1.3 to 1.8), consistent with reduced oxidative stress since MTs are usually induced in hyperglycemia as antioxidants. MIOX (myo-inositol oxygenase), a PT-specific enzyme that generates ROS under high glucose, was also significantly lower, indicating less oxidative burden. APOE (apolipoprotein E), implicated in tubular lipid handling, was reduced; in diabetes, elevated PT ApoE promotes lipid accumulation, so its decrease suggests alleviated lipotoxic stress. Overall, these changes are consistent with dapagliflozin ameliorating the hyperglycemic milieu in PT cells (lower ROS, less aberrant gluconeogenesis and lipid uptake). A few transcripts were upregulated in PT (e.g. COL4A2, a collagen IV isoform, and structural proteins like AHNAK), which could reflect remodeling of the tubular basement membrane or cytoskeleton. Since there was no evidence of fibrosis on MRI and histopathology, this mild ECM upturn may represent a transient normalization of basement
membrane composition or acute repair response rather than progressive fibrosis.",
TAL = "TAL cells showed a mixed response: predominantly downregulation of stress/injury genes, but upregulation of metabolic genes, consistent with increased workload from enhanced distal Na+ delivery under SGLT2 inhibition. Notably, EGF (epidermal growth factor) expression rose (log2FC +0.39) consistent with SGLT2i fostering a healthier tubular epithelium. Many antioxidant and injury-response genes were suppressed: MT1G and other MT isoforms decreased (less oxidative stress need, similar to PT), and GSTA2 (glutathione S-transferase) fell, indicating reduced oxidative load. UCHL1 (a neuronally expressed deubiquitinase aberrantly expressed in injured kidneys) was lower, implying less cellular injury. Interestingly, PCK1 in TAL was upregulated (log2FC  +0.94), in contrast to its suppression in PT, consistent with prior data from our group. Aside from this, TAL cells overall exhibit changes consistent with reduced inflammatory signaling and oxidative stress under dapagliflozin. The net picture is TAL tubules working harder in ion transport (hence metabolic upregulation) but experiencing less diabetes-related stress.",
PC = "Parietal epithelial cells in the same glomerular cluster co up regulated PAPP A and its endogenous inhibitor STC1, implying finely balanced IGF signaling that supports controlled remodeling rather than unchecked growth.",
IC = "Intercalated cells in the cortical collecting duct showed a predominance of upregulated transcripts (all top 100 were increased by dapagliflozin vs. placebo). This pattern reflects a concerted activation of acid-base transport mechanisms, likely in response to the mild metabolic acidosis/ketosis induced by SGLT2 inhibition. Consistently, IC cells upregulated multiple genes involved in proton secretion and bicarbonate reabsorption: CA2 (carbonic anhydrase II) was up (log2FC +0.78), facilitating intracellular HCO3- generation for secretion; SLC26A7 (Cl-/HCO3 exchanger on basolateral membrane of type A intercalated cells) was strongly up (log2FC +1.02), which is crucial for bicarbonate reabsorption in acidic conditions. ATP6AP2 (proton pump accessory unit) rose (log2FC +0.95), indicating increased H+ pumping capacity. These changes are highly consistent with the expected physiologic response: the kidney ramping up acid secretion to counter the SGLT2i-induced acid load. Additionally, IGFBP7 (Insulin-like Growth Factor Binding Protein 7) was up (log2FC +1.00), which could indicate cellular stress in response to high metabolic state to proton pumps, but could also be temporarily elevated as cells adapt to a new workload. Intercalated cells also showed increased expression of various transporters: e.g. SLC4A9 (AE4) up (log2FC +0.42), another bicarbonate transporter; V-ATPase subunits and co-factors (like ATP6V1G3, perhaps). ",
EC = "Endothelial cells responded to dapagliflozin with a clear antifibrotic and anti-inflammatory expression profile. A prominent change was the downregulation of extracellular matrix (ECM) and adhesion molecules. Fibronectin-1 (FN1), a key fibrogenic ECM protein that accumulates in DKD (driven by TGF-), was significantly decreased (log2FC -1.14). This is strongly consistent with SGLT2is antifibrotic effect, as reduced fibronectin implies less ongoing remodeling in glomerular basement membranes and interstitium. Likewise, VCAM1 fell (log2FC -1.37); hyperglycemia and inflammation normally upregulate endothelial VCAM-1, promoting leukocyte adhesion and interstitial inflammation. Its suppression by dapagliflozin indicates reduced endothelial inflammation, likely contributing to less immune cell recruitment to the kidney. Endothelial injury markers improved: Clusterin (CLU), which is elevated in stressed glomeruli and serves as a protective chaperone, was downregulated (log2FC -0.9 in EC cluster) which is consistent with less cellular stress/injury to endothelium. Moreover, EMCN (endomucin), a glycocalyx protein, was up (logFC +0.42), suggesting restoration of the endothelial surface layer integrity (T1D often sheds glycocalyx, impairing endothelium). Netrin-4 (NTN4), an extracellular matrix protein with anti-angiogenic and anti-inflammatory properties in vasculature, was upregulated (log2FC +0.72); netrin-4 is known to provide an anti-inflammatory, endothelial-protective milieu. This aligns with improved endothelial function under SGLT2 inhibition. Another matricellular protein, SPARCL1 (hevin), was down (log2FC -1.33); its role in kidney is not well defined, but in other organs it can modulate angiogenesis and anti-fibrosis. Additionally, SLC14A1 (UT-B urea transporter), expressed in endothelial cells of the vasa recta, decreased (log2FC -0.55), potentially reflecting altered medullary urea handling due to osmotic diuresis. Additionally, multiple inflammation-related transcripts (adhesion molecules, chemokines like CXCL12 [log2FC 0.5], etc.) were suppressed. ",
Immune = "Lymphoid cells were broadly transcriptionally silent, indicating reduced intrarenal T- , B- , and NK cell activity, whereas myeloid cells suppressed interferon stimulated genes and increased oxidative metabolism markers such as PDK4 and SLC38A2hallmarks of a shift from inflammatory M1 toward reparative M2 phenotypes.",
VSMC_P_FIB = " Stromal fibroblast/smooth muscle/mesangial cells displayed a strongly anti fibrotic pattern, with higher TIMP3, A2M, and BMP responsive ID1 and no induction of TGF  targets; the modest CTNNB1 rise appeared isolated and non pathogenic. ",
POD = "Podocytes exhibited a distinct AMPK centered re programming: dapagliflozin dropped the interferon stress marker IFI6, and simultaneously raised ARRDC4, an  arrestin that retunes GLUT trafficking during energy stress, and thymosin 4 (TMSB4X), which fortifies the actin cytoskeleton. This triad signals a shift toward fatty acid/ketone oxidation, enhanced autophagy, and slit diaphragm stability, providing a plausible molecular basis for the drugs albuminuria lowering effect."
)

for (cells in names(celltype_groups)) {

cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Dot Plots/kpmp_", cells, "_hvg_subtypes_nebula_scores_volcano_heat.jpeg)\n\n"))

cat(paste0(celltype_paragraphs[cells], "\n\n"))

}
```


---

### Non-KPMP Cell types

#### Top 2000 HVG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("PT", "TAL", "PC", "Immune", "Immune_Myeloid", "Immune_Lymphoid", "IC", "EC", "FIBVSMCP", "POD")

for (cells in celltype_list) {

  # cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", tolower(cells), "_hvg_placebo_pvalue_nebula_reml_pooled.jpeg)\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Dot Plots/", cells, "_hvg_subtypes_nebula_scores_volcano_heat.jpeg)\n\n"))
}
```

---

#### Not limited to top 2000 HVG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("PT", "TAL", "PC", "Immune", "Immune_Myeloid", "Immune_Lymphoid", "IC", "EC", "FIBVSMCP", "POD")

for (cells in celltype_list) {

  # cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", tolower(cells), "_placebo_pvalue_nebula_reml_pooled.jpeg)\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Dot Plots/", cells, "_subtypes_nebula_scores_volcano_heat.jpeg)\n\n"))
}
```


## GSEA


### Positive PT KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_top30_kegg_pathways.jpeg){width=70%}


```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative PT KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive PT Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative PT Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive PT GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative PT GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PT GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pt_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))}
```

---

### Positive TAL KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative TAL KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive TAL Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative TAL Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive TAL GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative TAL GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### TAL GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/tal_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive PC KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative PC KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive PC Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative PC Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive PC GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative PC GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### PC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/pc_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative Immune GO

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---


### Positive Immune Myeloid KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune Myeloid KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune Myeloid Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune Myeloid Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune Myeloid GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative Immune Myeloid GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Myeloid GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_myeloid_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---


### Positive Immune Lymphoid KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune Lymphoid KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune Lymphoid Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative Immune Lymphoid Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive Immune Lymphoid GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative Immune Lymphoid GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### Immune Lymphoid GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/immune_lymphoid_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---


### Positive IC KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative IC KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive IC Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative IC Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive IC GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative IC GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### IC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ic_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))}
```

---


### Positive EC KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative EC KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive EC Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative EC Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive EC GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))

}
```

---

### Negative EC GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### EC GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/ec_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---


### Positive FIB/VSMC/P KEGG

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_top30_kegg_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("FIB/VSMC/P KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative FIB/VSMC/P KEGG
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### FIB/VSMC/P KEGG Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_kegg_legacy.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_kegg_legacy_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive FIB/VSMC/P Reactome

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_top30_reactome_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### FIB/VSMC/P Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_reactome.jpeg){width=70%}\n\n"))

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative FIB/VSMC/P Reactome
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### FIB/VSMC/P Reactome Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_reactome.jpeg){width=70%}\n\n"))
  
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_reactome_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Positive FIB/VSMC/P GO

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_top30_go_pathways.jpeg){width=70%}

```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### FIB/VSMC/P GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_go.jpeg){width=70%}\n\n"))
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_pos", i, "_go_heatmap.jpeg){width=70%}\n\n"))
}
```

---

### Negative FIB/VSMC/P GO
```{r echo=FALSE, results='asis'}
for (i in 1:10) {
  cat("#### FIB/VSMC/P GO Pathway Top ", i, "\n\n")  # Adds a section header

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_go.jpeg){width=70%}\n\n"))
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/fibvsmc_nebula_neg", i, "_go_heatmap.jpeg){width=70%}\n\n"))
  }
```

---


## IPA

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "ic", "ec", "fibvsmcp")

for (cells in celltype_list) {
  cat("### Activated IPA Pathway for", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Z score bar plots/nebula/", cells, "_top10_activated_pathway_zscores.jpeg)\n\n"))
}
```

---

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "ic", "ec", "fibvsmcp")

for (cells in celltype_list) {
  cat("### Inhibited IPA Pathway for", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Z score bar plots/nebula/", cells, "_top10_inhibited_pathway_zscores.jpeg){width=70%}\n\n"))
}
```


# T1D vs. HC

---


## T1D (CRC T1D vs. HC, same approach as #2, informed approach)


```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "ic", "ec", "fibvsmcp")

for (cells in celltype_list) {

    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_targeted_t1dvhc_", cells, "_deg.jpeg){width=50%}\n\n"))
  
    cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/Volcano Plots/", cells ,"_targeted_volcano_attempt_directions_labeled.jpeg){width=50%}\n\n"))
    
  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/", cells, "_croc_attempt_dot.jpeg){width=50%}\n\n"))
  
}

cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_targeted_t1dvhc_", "pod", "_deg.jpeg){width=50%}\n\n"))
  
cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/CROCODILE comparison/Volcano Plots/", "pod" ,"_targeted_volcano_attempt_directions_labeled.jpeg){width=50%}\n\n"))
```

---

## T1D (CRC T1D vs. HC, same approach as #2, non-informed approach)

![PT](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_pt_deg.jpeg){width=70%}
![TAL](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_tal_deg.jpeg){width=70%}
![PC](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_pc_deg.jpeg){width=70%}
![Immune](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_immune_deg.jpeg){width=70%}
![Immune Myeloid](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_immune_myeloid_deg.jpeg){width=70%}
![Immune Lymphoid](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_immune_lymphoid_deg.jpeg){width=70%}
![EC](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_ec_deg.jpeg){width=70%}
![IC](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_ic_deg.jpeg){width=70%}
![FIB/VSMC/P](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_fibvsmcp_deg.jpeg){width=70%}
![POD](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/CROCODILE/Results/Figures/Volcano/nebula/croc_pod_deg.jpeg){width=70%}

---

## T1D (CRC T1D + BL ATTEMPT vs. HC, same approach as #2, informed approach)

- Work in progress...

---

## T1D (CRC T1D + BL ATTEMPT vs. HC, same approach as #2, non-informed approach)

- Work in progress...

---

## Evaluate the effect of Dapagliflozin vs. Placebo on these DEGs characteristic of T1D

- Work in progress...


# Cell type proportions

## Subtype to general type clusters

::: columns

::: column

| General type      | Sub type                      |
|-------------------|-------------------------------|
| PT                | PT-1, PT-2, PT-3, PT-4, PT-5  |
| TAL               | TAL-1, TAL-2, TAL-3           |
| PC                | tPC-IC, PC-1, PC-2            |
| IC                | IC-A, IC-B                    |
| EC                | EC-PTC, EC-GC, EC-AEA, EC-LYM |
| Immune (Myeloid)  | MAC, MON                      |
| Immune (Lymphoid) | NKT/NKC, T, B                 |
| FIB/VSMC/P        | FIB, VSMC/P                   |
| POD.              | POD                           |


:::

::: column

| General type        | KPMP Sub type                                  |
|---------------------|------------------------------------------------|
| PT                  | PT-S1/S2, PT-S3, aPT                           |
| TAL                 | C-TAL-1, C-TAL-2, aTAL, dTAL                   |
| PC                  | CCD-PC, CNT-PC, dCCD-PC, M-PC, tPC-IC          |
| IC                  | IC-A, IC-B, aIC                                |
| EC                  | EC-AVR, EC-GC, EC-PTC, EC-AEA, EC-LYM, EC/VSMC |
| Immune (Myeloid)    | MAC, MON, cDC, pDC                             |
| Immune (Lymphoid)   | CD4+ T, CD8+ T, B, NK                          |
| FIB/VSMC/P          | FIB, VSMC/P                                    |
| POD                 | POD                                            |

:::
:::

## Proportion figures

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_cellcounts_plot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_cellcounts_proportions_plot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_general_cellcounts_plot_clean.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_general_cellcounts_proportions_plot_clean.jpeg){width=70%}


![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_pt_treatment_visit_proportions_plot.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/attempt_pt_visit_treatment_proportions_plot_kpmp.jpeg){width=70%}


## Upset and circular plot

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Upset/celltype_deg_upset.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/celltype_deg_chordplot.png){width=70%}

---

## T1D PRE vs. POST in Placebo (ATTEMPT only)

### Propeller

```{r echo = T, results='asis'}
propeller_res <- s3readRDS("PT/pt_propeller_combined.rds", "attempt", region ="")
propeller_res_placebo <- propeller_res %>% 
  filter(treatment == "Placebo") %>%
  arrange(celltype) %>%
  dplyr::select(PropMean.PRE, PropMean.POST, PropRatio, P.Value, FDR)

kable(propeller_res_placebo,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

**No significant change** in PT cell type proportion was observed in the placebo group post treatment. Numeric increase in PT-1, PT-2, and PT-3 cells were observed, and numeric decrease in PT-4 and PT-5 cells were observed.

```{r echo = T, results='asis'}
propeller_res_kpmp <- s3readRDS("PT/pt_propeller_combined_kpmp.rds", "attempt", region ="")
propeller_res_kpmp_placebo <- propeller_res_kpmp %>% 
  filter(treatment == "Placebo") %>%
  arrange(KPMP_celltype) %>%
  dplyr::select(PropMean.PRE, PropMean.POST, PropRatio, P.Value, FDR)

kable(propeller_res_kpmp_placebo,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

**No significant change** in PT cell type proportion was observed in the placebo group post treatment. Numeric increase in PT-S1/S2 cells were observed, and numeric decrease in PT-S3 and aPT cells were observed.



### Dirichlet Regression

```{r echo = T, results='asis'}
dirichlet_res_placebo <- s3readRDS("PT/pt_dirichlet_placebo.rds", "attempt", region ="")
dirichlet_res_placebo <- data.frame(dirichlet_res_placebo$coef.mat) %>%
  rownames_to_column(var = "coefficient") %>%
  mutate(coefficient = case_when(coefficient == "visitPOST" ~ "PT-1 (POST)",
                                 coefficient == "visitPOST.1" ~ "PT-2 (POST)",
                                 coefficient == "visitPOST.2" ~ "PT-3 (POST)",
                                 coefficient == "visitPOST.3" ~ "PT-4 (POST)",
                                 coefficient == "visitPOST.4" ~ "PT-5 (POST)")) %>%
  filter(!is.na(coefficient))

kable(dirichlet_res_placebo,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```


```{r echo = T, results='asis'}
dirichlet_res_kpmp_placebo <- s3readRDS("PT/pt_dirichlet_placebo_kpmp.rds", "attempt", region ="")
dirichlet_res_kpmp_placebo <- data.frame(dirichlet_res_kpmp_placebo$coef.mat) %>%
  rownames_to_column(var = "coefficient") %>%
  mutate(coefficient = case_when(coefficient == "visitPOST" ~ "PT-S1/S2 (POST)",
                                 coefficient == "visitPOST.1" ~ "PT-S3 (POST)",
                                 coefficient == "visitPOST.2" ~ "aPT (POST)")) %>%
  filter(!is.na(coefficient))

kable(dirichlet_res_kpmp_placebo,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

---

## T1D PRE vs. POST in Dapagliflozin (ATTEMPT only)

### Propeller

```{r echo = T, results='asis'}
propeller_res_dapa <- propeller_res %>% 
  filter(treatment == "Dapagliflozin") %>%
  arrange(celltype) %>%
  remove_rownames() %>%
  column_to_rownames("celltype") %>%
  dplyr::select(PropMean.PRE, PropMean.POST, PropRatio, P.Value, FDR)

kable(propeller_res_dapa,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

**No significant change** in PT cell type proportion was observed in the dapagliflozin group post treatment. Numeric increase in PT-1, and PT-2 cells were observed, and numeric decrease in PT-3, PT-4, and PT-5 cells were observed.

```{r echo = T, results='asis'}
propeller_res_kpmp <- s3readRDS("PT/pt_propeller_combined_kpmp.rds", "attempt", region ="")
propeller_res_kpmp_dapa <- propeller_res_kpmp %>% 
  filter(treatment == "Dapagliflozin") %>%
  arrange(KPMP_celltype) %>%
  dplyr::select(PropMean.PRE, PropMean.POST, PropRatio, P.Value, FDR)

kable(propeller_res_kpmp_dapa,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

**No significant change** in PT cell type proportion was observed in the dapa group post treatment. Numeric increase in PT-S1/S2 cells were observed, and numeric decrease in PT-S3 and aPT cells were observed.


### Dirichlet Regression

```{r echo = T, results='asis'}
dirichlet_res_dapa <- s3readRDS("PT/pt_dirichlet_dapa.rds", "attempt", region ="")
dirichlet_res_dapa <- data.frame(dirichlet_res_dapa$coef.mat) %>%
  rownames_to_column(var = "coefficient") %>%
  mutate(coefficient = case_when(coefficient == "visitPOST" ~ "PT-1 (POST)",
                                 coefficient == "visitPOST.1" ~ "PT-2 (POST)",
                                 coefficient == "visitPOST.2" ~ "PT-3 (POST)",
                                 coefficient == "visitPOST.3" ~ "PT-4 (POST)",
                                 coefficient == "visitPOST.4" ~ "PT-5 (POST)")) %>%
  filter(!is.na(coefficient))

kable(dirichlet_res_dapa,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```


```{r echo = T, results='asis'}
dirichlet_res_kpmp_dapa <- s3readRDS("PT/pt_dirichlet_dapa_kpmp.rds", "attempt", region ="")
dirichlet_res_kpmp_dapa <- data.frame(dirichlet_res_kpmp_dapa$coef.mat) %>%
  rownames_to_column(var = "coefficient") %>%
  mutate(coefficient = case_when(coefficient == "visitPOST" ~ "PT-S1/S2 (POST)",
                                 coefficient == "visitPOST.1" ~ "PT-S3 (POST)",
                                 coefficient == "visitPOST.2" ~ "aPT (POST)")) %>%
  filter(!is.na(coefficient))

kable(dirichlet_res_kpmp_dapa,format = "html", 
      digits = 3)  %>%
  kable_styling(font_size = 20,
                bootstrap_options = c("striped", "hover", "condensed"),
                full_width = FALSE)
```

---

## T1D vs. HC (CROCODILE + ATTEMPT)

---

## T1D vs. HC (CROCODILE only)

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/croc_pt_group_proportions_plot.jpeg)

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proportions/croc_pt_group_proportions_plot_kpmp.jpeg)

# Pseudotime trajectories (KPMP Cell types)

---

## T1D vs. HC (Cross sectional)

---

## Dapagliflozin vs. Placebo in T1D (serial data)

### PT

Lineage: PT-S1/S2, PT-S3, aPT

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pca_pt_kpmp_slingshot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_kpmp_slingshot_density_trtvisit.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_slingshot_violin.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_kpmp_slingshot_density_trt.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_kpmp_slingshot_percentile_heatmap.jpeg){width=70%}

#### Regression
```{r echo = T, results = 'asis'}
pt_kpmp_rq <- s3readRDS("Results/pt_kpmp_rq_fit.rds", "attempt", region ="")
pt_kpmp_rq_summary <- summary(pt_kpmp_rq, se = "nid")

kable(bind_rows(
  lapply(seq_along(pt_kpmp_rq_summary), function(i) {
    coef_df <- as.data.frame(pt_kpmp_rq_summary[[i]]$coefficients)
    coef_df$term <- rownames(coef_df)
    coef_df$tau <- pt_kpmp_rq$tau[i]
    coef_df
  }),
  .id = "quantile"
), row.names = F)
```

#### Associations with clinical variables

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa")

for (v in vars) {
  cat("Density plot for ", v, "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_density_pt_kpmp_", v, "_deltadelta_slingshot.jpeg){width=70%}\n\n"))
}
```

---

##### Regression

```{r echo = T, results='asis'}
# need to add TIR
vars <- c("hba1c", "weight",  "mgfr_jodal", "mgfr_jodal_bsa", "visit", "trt")

for (v in vars) {
  cat("Quantile Regression Summary for ", v, "\n\n")
  
  # Load the model
  model_path <- paste0("Results/pt_kpmp_rq_fit_", v, ".rds")
  pt_kpmp_rq_fit <- s3readRDS(model_path, bucket = "attempt", region = "")
  
  # Force correct summary method
  pt_kpmp_rq_summary <- quantreg:::summary.rqs(pt_kpmp_rq_fit, se = "nid")
  
  # Extract coefficient table
  coef_table <- bind_rows(
    lapply(seq_along(pt_kpmp_rq_summary), function(i) {
      coef_df <- as.data.frame(pt_kpmp_rq_summary[[i]]$coefficients)
      coef_df$term <- rownames(coef_df)
      coef_df$tau <- pt_kpmp_rq_fit$tau[i]
      coef_df
    }),
    .id = "quantile"
  )
  
  # Display nicely
  print(kable(coef_table, row.names = F))
  cat("\n\n")
}
```

### TAL

Lineage: aTAL, C-TAL-1, C-TAL-2, dTAL

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pca_tal_kpmp_slingshot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_kpmp_slingshot_density_trtvisit.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_slingshot_violin.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_kpmp_slingshot_density_trt.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_kpmp_slingshot_percentile_heatmap.jpeg){width=70%}

#### Regression
```{r echo = T, results = 'asis'}
tal_kpmp_rq <- s3readRDS("Results/tal_kpmp_rq_fit.rds", "attempt", region ="")
tal_kpmp_rq_summary <- summary(tal_kpmp_rq, se = "nid")

kable(bind_rows(
  lapply(seq_along(tal_kpmp_rq_summary), function(i) {
    coef_df <- as.data.frame(tal_kpmp_rq_summary[[i]]$coefficients)
    coef_df$term <- rownames(coef_df)
    coef_df$tau <- tal_kpmp_rq$tau[i]
    coef_df
  }),
  .id = "quantile"
), row.names = F)
```

#### Associations with clinical variables

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa")

for (v in vars) {
  cat("Density plot for ", v, "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_density_tal_kpmp_", v, "_deltadelta_slingshot.jpeg){width=70%}\n\n"))
}
```

---

##### Regression

```{r echo = T, results='asis', eval = F}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa", "visit", "trt")

for (v in vars) {
  cat("Quantile Regression Summary for ", v, "\n\n")
  
  # Load the model
  model_path <- paste0("Results/tal_kpmp_rq_fit_", v, ".rds")
  tal_kpmp_rq_fit <- s3readRDS(model_path, bucket = "attempt", region = "")
  
  # Force correct summary method
  tal_kpmp_rq_summary <- quantreg:::summary.rqs(tal_kpmp_rq_fit, se = "nid")
  
  # Extract coefficient table
  coef_table <- bind_rows(
    lapply(seq_along(tal_kpmp_rq_summary), function(i) {
      coef_df <- as.data.frame(tal_kpmp_rq_summary[[i]]$coefficients)
      coef_df$term <- rownames(coef_df)
      coef_df$tau <- tal_kpmp_rq_fit$tau[i]
      coef_df
    }),
    .id = "quantile"
  )
  
  # Display nicely
  print(kable(coef_table, row.names = F))
  cat("\n\n")
}
```

# Pseudotime trajectories (Non-KPMP Cell types)

---

## T1D vs. HC (Cross sectional)

---

## Dapagliflozin vs. Placebo in T1D (serial data)

### PT

Lineage: PT-2, PT-1, PT-3, PT-5, PT-4

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pca_pt_slingshot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_slingshot_density_trtvisit.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_slingshot_violin.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_slingshot_density_trt.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pt_slingshot_percentile_heatmap.jpeg){width=70%}

#### Regression
```{r echo = T, results = 'asis'}
pt_rq <- s3readRDS("Results/pt_rq_fit.rds", "attempt", region ="")
pt_rq_summary <- summary(pt_rq, se = "nid")

kable(bind_rows(
  lapply(seq_along(pt_rq_summary), function(i) {
    coef_df <- as.data.frame(pt_rq_summary[[i]]$coefficients)
    coef_df$term <- rownames(coef_df)
    coef_df$tau <- pt_rq$tau[i]
    coef_df
  }),
  .id = "quantile"
), row.names = F)
```

#### Associations with clinical variables

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa")

for (v in vars) {
  cat("Density plot for ", v, "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_density_pt_", v, "_deltadelta_slingshot.jpeg){width=70%}\n\n"))
}
```

---

##### Regression

```{r echo = T, results='asis'}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa", "visit", "trt")

for (v in vars) {
  cat("Quantile Regression Summary for ", v, "\n\n")
  
  # Load the model
  model_path <- paste0("Results/pt_rq_fit_", v, ".rds")
  pt_rq_fit <- s3readRDS(model_path, bucket = "attempt", region = "")
  
  # Force correct summary method
  pt_rq_summary <- quantreg:::summary.rqs(pt_rq_fit, se = "nid")
  
  # Extract coefficient table
  coef_table <- bind_rows(
    lapply(seq_along(pt_rq_summary), function(i) {
      coef_df <- as.data.frame(pt_rq_summary[[i]]$coefficients)
      coef_df$term <- rownames(coef_df)
      coef_df$tau <- pt_rq_fit$tau[i]
      coef_df
    }),
    .id = "quantile"
  )
  
  # Display nicely
  print(kable(coef_table, row.names = F))
  cat("\n\n")
}
```


### TAL

Lineage: TAL-2, TAL-1, TAL-3

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_pca_tal_slingshot.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_slingshot_density_trtvisit.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_slingshot_violin.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_slingshot_density_trt.jpeg){width=70%}

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_tal_slingshot_percentile_heatmap.jpeg){width=70%}

#### Regression
```{r echo = T, results = 'asis'}
tal_rq <- s3readRDS("Results/tal_rq_fit.rds", "attempt", region ="")
tal_rq_summary <- summary(tal_rq, se = "nid")

kable(bind_rows(
  lapply(seq_along(tal_rq_summary), function(i) {
    coef_df <- as.data.frame(tal_rq_summary[[i]]$coefficients)
    coef_df$term <- rownames(coef_df)
    coef_df$tau <- tal_rq$tau[i]
    coef_df
  }),
  .id = "quantile"
), row.names = F)
```

#### Associations with clinical variables

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa")

for (v in vars) {
  cat("Density plot for ", v, "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Slingshot/attempt_density_tal_", v, "_deltadelta_slingshot.jpeg){width=70%}\n\n"))
}
```

---

##### Regression

```{r echo = T, results='asis', eval = F}
vars <- c("hba1c", "weight", "tir", "mgfr_jodal", "mgfr_jodal_bsa", "visit", "trt")

for (v in vars) {
  cat("Quantile Regression Summary for ", v, "\n\n")
  
  # Load the model
  model_path <- paste0("Results/tal_rq_fit_", v, ".rds")
  tal_rq_fit <- s3readRDS(model_path, bucket = "attempt", region = "")
  
  # Force correct summary method
  tal_rq_summary <- quantreg:::summary.rqs(tal_rq_fit, se = "nid")
  
  # Extract coefficient table
  coef_table <- bind_rows(
    lapply(seq_along(tal_rq_summary), function(i) {
      coef_df <- as.data.frame(tal_rq_summary[[i]]$coefficients)
      coef_df$term <- rownames(coef_df)
      coef_df$tau <- tal_rq_fit$tau[i]
      coef_df
    }),
    .id = "quantile"
  )
  
  # Display nicely
  print(kable(coef_table, row.names = F))
  cat("\n\n")
}
```


# Associate DEGs with clinical markers adjusting for treatment group

---

## KPMP Change in mGFR (raw)

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_mgfr_jodal_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in mGFR (raw)

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_mgfr_jodal_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```



---

### IPA


---


### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_raw_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_raw_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_raw_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

---

## KPMP Change in mGFR (BSA)

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_mgfr_jodal_bsa_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in mGFR (BSA)

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_mgfr_jodal_bsa_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---

### IPA


---


### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_bsa_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_bsa_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_mgfr_bsa_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

---


## KPMP Change in TIR

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_tir_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in TIR

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n")

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_tir_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---

## KPMP Change in HbA1c

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_hba1c_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in HbA1c

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n")

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_hba1c_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---


### IPA


---


### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_hba1c_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_hba1c_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_hba1c_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

---

## KPMP Change in weight

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_weight_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in weight

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_weight_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---



### IPA


---


### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_weight_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_weight_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_weight_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

---

## KPMP Change in avg cortical R2*

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_avg_c_r2_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in avg cortical R2*

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_avg_c_r2_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---


### IPA


---


### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_c_r2_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_c_r2_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_c_r2_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

---

## KPMP Change in avg kidney R2*

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_avg_k_r2_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in avg kidney R2*

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_avg_k_r2_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

---

### IPA

---

### GSEA

#### KEGG

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_k_r2_top30_kegg_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### REACTOME

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_k_r2_top30_reactome_pathways.jpeg){width=70%}\n\n"))
}
```

---

#### GO

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "fibvsmc")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Pathways/nebula/", cells, "_avg_k_r2_top30_go_pathways.jpeg){width=70%}\n\n"))
}
```

## KPMP Change in avg ketones

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune",  "immune_myeloid", "immune_lymphoid", "ic", "ec", "vsmc_p_fib")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/kpmp_", cells, "_avg_ketones_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

## Change in avg ketones

```{r echo = T, results = 'asis', eval = T}
celltype_list <- c("pt", "tal", "pc", "immune", "ic", "ec", "fibvsmcp")

for (cells in celltype_list) {
  cat("#### ", toupper(cells), "\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Volcano Plots/nebula/", cells, "_avg_ketones_trtcolors_concordance.jpeg){width=70%}\n\n"))
}
```

# Cell-to-cell interaction

---

## T1D vs. HC

---

## Dapagliflozin vs. Placebo

![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/LIANA/liana_heatmap_2.jpeg){width=70%}
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/LIANA/liana_heatmap.jpeg){width=70%}

```{r echo = T, results = 'asis', eval = T}
# Define the folder path
folder_path <- "/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/LIANA/"

# List all files starting with "circle_"
file_list <- list.files(folder_path, pattern = "^circle_.*\\.jpeg$", full.names = FALSE)

# Loop through each file
for (file_name in file_list) {
  # Extract a label from the filename (optional)
  label <- tools::file_path_sans_ext(gsub("^circle_", "", file_name))

  cat("#### ", toupper(label), "\n\n") 

  cat(paste0("![](", file.path(folder_path, file_name), "){width=70%}\n\n"))
}
```


# Proteomics DiD

## Plasma Proteomics 

::: columns
::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_plasma_proteomics_mixed.jpeg)
:::

::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_plasma_proteomics_adjp_mixed.jpeg)
:::
:::

### Clinical associations

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "tir", "weight", "mgfr", "mgfr_bsa", "avg_c_r2", "avg_k_r2", "PWV")

for (var in vars) {
  cat("#### ", var, " in plasma\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Correlations/volcano_mixed_spearman_plasma_",  var, ".jpeg)\n\n"))
}
```


## Urine Proteomics 

#### Normalized by creatinine

::: columns
::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_urine_proteomics_mixed.jpeg)
:::

::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_urine_proteomics_adjp_mixed.jpeg)
:::
:::

#### Creatinine in model

::: columns
::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_urine_proteomics_mixed.jpeg)
:::

::: column
![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Volcano_DiD_urine_proteomics_adjp_mixed.jpeg)
:::
:::


### Clinical associations (Correlations with creatinine normalized urine, DiD ref model: creatinine normalized proteomics)

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "tir", "weight", "mgfr", "mgfr_bsa", "avg_c_r2", "avg_k_r2", "PWV")

for (var in vars) {
  cat("#### ", var, " in urine\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Correlations/volcano_mixed_spearman_urine_",  var, ".jpeg)\n\n"))
}
```

### Clinical associations (Correlations with creatinine normalized urine, DiD ref model: non-normalized proteomics with creatinine in model)

```{r echo = T, results = 'asis', eval = T}
vars <- c("hba1c", "tir", "weight", "mgfr", "mgfr_bsa", "avg_c_r2", "avg_k_r2", "PWV")

for (var in vars) {
  cat("#### ", var, " in urine\n\n") 

  cat(paste0("![](/Users/choiyej/Library/CloudStorage/OneDrive-SharedLibraries-UW/Laura Pyle - Bjornstad/Biostatistics Core Shared Drive/ATTEMPT/Results/Figures/Proteomics/Volcano Plots/Correlations/volcano_mixed_cr_spearman_urine_",  var, ".jpeg)\n\n"))
}
```

# Associate DEGs w/ proteomics

---

## EC DEGs with change in plasma proteomics

---

## PT DEGs with change in urine proteomics

---

## TAL DEGs with change in urine proteomics
